<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="A Python package for simulating kidney cancer progression with synthetic data generation and machine learning."><meta name=author content="Guillermo Prol-Castelo (Barcelona Supercomputing Center), for the EVENFLOW European Project"><link href=https://github.com/gprolcastelo/renalprog/ENRICHMENT_ANALYSIS/ rel=canonical><link href=../CLASSIFICATION/ rel=prev><link href=../ENRICHMENT_QUICK_REF/ rel=next><link rel=icon href=../assets/images/kidneys.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>Enrichment Analysis Details - renalprog</title><link rel=stylesheet href=../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../css/timeago.css><link rel=stylesheet href=../assets/_mkdocstrings.css><link rel=stylesheet href=../stylesheets/extra.css><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#dynamic-enrichment-analysis class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title=renalprog class="md-header__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../assets/images/kidneys.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> renalprog </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Enrichment Analysis Details </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../getting-started/ class=md-tabs__link> Getting Started </a> </li> <li class=md-tabs__item> <a href=../tutorials/ class=md-tabs__link> Tutorials </a> </li> <li class=md-tabs__item> <a href=../reproducibility/ class=md-tabs__link> Reproducibility </a> </li> <li class=md-tabs__item> <a href=../api/ class=md-tabs__link> API Reference </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=../advanced/architecture/ class=md-tabs__link> Advanced Topics </a> </li> <li class=md-tabs__item> <a href=../contributing/guidelines/ class=md-tabs__link> Contributing </a> </li> <li class=md-tabs__item> <a href=../citation/ class=md-tabs__link> Citation & License </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title=renalprog class="md-nav__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../assets/images/kidneys.png alt=logo> </a> renalprog </label> <div class=md-nav__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> <span class=md-ellipsis> Getting Started </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Getting Started </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../getting-started/ class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/data-requirements/ class=md-nav__link> <span class=md-ellipsis> Data Requirements </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> <span class=md-ellipsis> Tutorials </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tutorials/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/pretrained-models/ class=md-nav__link> <span class=md-ellipsis> Using Pretrained Models </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/complete-pipeline/ class=md-nav__link> <span class=md-ellipsis> Complete Pipeline </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step1-data-processing/ class=md-nav__link> <span class=md-ellipsis> Step 1 - Data Processing </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step2-vae-training/ class=md-nav__link> <span class=md-ellipsis> Step 2 - VAE Training </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step3-reconstruction/ class=md-nav__link> <span class=md-ellipsis> Step 3 - Reconstruction </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step4-trajectories/ class=md-nav__link> <span class=md-ellipsis> Step 4 - Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step5-classification/ class=md-nav__link> <span class=md-ellipsis> Step 5 - Classification </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step6-enrichment/ class=md-nav__link> <span class=md-ellipsis> Step 6 - Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/visualization/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> Reproducibility </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Reproducibility </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../reproducibility/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../reproducibility/requirements/ class=md-nav__link> <span class=md-ellipsis> System Requirements </span> </a> </li> <li class=md-nav__item> <a href=../reproducibility/data-preparation/ class=md-nav__link> <span class=md-ellipsis> Data Preparation </span> </a> </li> <li class=md-nav__item> <a href=../reproducibility/pipeline/ class=md-nav__link> <span class=md-ellipsis> Running the Pipeline </span> </a> </li> <li class=md-nav__item> <a href=../reproducibility/results/ class=md-nav__link> <span class=md-ellipsis> Expected Results </span> </a> </li> <li class=md-nav__item> <a href=../reproducibility/troubleshooting/ class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../api/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../api/config/ class=md-nav__link> <span class=md-ellipsis> Configuration </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5_3> <label class=md-nav__link for=__nav_5_3 id=__nav_5_3_label tabindex=0> <span class=md-ellipsis> Data Processing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_3_label aria-expanded=false> <label class=md-nav__title for=__nav_5_3> <span class="md-nav__icon md-icon"></span> Data Processing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../api/dataset/ class=md-nav__link> <span class=md-ellipsis> Dataset </span> </a> </li> <li class=md-nav__item> <a href=../api/features/ class=md-nav__link> <span class=md-ellipsis> Features </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5_4> <label class=md-nav__link for=__nav_5_4 id=__nav_5_4_label tabindex=0> <span class=md-ellipsis> Modeling </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_4_label aria-expanded=false> <label class=md-nav__title for=__nav_5_4> <span class="md-nav__icon md-icon"></span> Modeling </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../api/models/ class=md-nav__link> <span class=md-ellipsis> VAE Models </span> </a> </li> <li class=md-nav__item> <a href=../api/training/ class=md-nav__link> <span class=md-ellipsis> Training </span> </a> </li> <li class=md-nav__item> <a href=../api/prediction/ class=md-nav__link> <span class=md-ellipsis> Prediction </span> </a> </li> <li class=md-nav__item> <a href=../api/trajectories/ class=md-nav__link> <span class=md-ellipsis> Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../api/classification/ class=md-nav__link> <span class=md-ellipsis> Classification </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../api/enrichment/ class=md-nav__link> <span class=md-ellipsis> Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../api/plots/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> <li class=md-nav__item> <a href=../api/utils/ class=md-nav__link> <span class=md-ellipsis> Utilities </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_6 checked> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex> <span class=md-ellipsis> Advanced Topics </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=true> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> Advanced Topics </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../advanced/architecture/ class=md-nav__link> <span class=md-ellipsis> Architecture </span> </a> </li> <li class=md-nav__item> <a href=../TRAJECTORIES/ class=md-nav__link> <span class=md-ellipsis> Trajectory Generation Details </span> </a> </li> <li class=md-nav__item> <a href=../CLASSIFICATION/ class=md-nav__link> <span class=md-ellipsis> Classification Details </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Enrichment Analysis Details </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Enrichment Analysis Details </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#pipeline-steps class=md-nav__link> <span class=md-ellipsis> Pipeline Steps </span> </a> <nav class=md-nav aria-label="Pipeline Steps"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-deseq2-differential-expression-analysis class=md-nav__link> <span class=md-ellipsis> 1. DESeq2 Differential Expression Analysis </span> </a> </li> <li class=md-nav__item> <a href=#2-gsea-analysis class=md-nav__link> <span class=md-ellipsis> 2. GSEA Analysis </span> </a> </li> <li class=md-nav__item> <a href=#3-results-combination class=md-nav__link> <span class=md-ellipsis> 3. Results Combination </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#installation-requirements class=md-nav__link> <span class=md-ellipsis> Installation Requirements </span> </a> <nav class=md-nav aria-label="Installation Requirements"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#python-dependencies class=md-nav__link> <span class=md-ellipsis> Python Dependencies </span> </a> </li> <li class=md-nav__item> <a href=#gsea-cli-tool class=md-nav__link> <span class=md-ellipsis> GSEA CLI Tool </span> </a> </li> <li class=md-nav__item> <a href=#pathway-database class=md-nav__link> <span class=md-ellipsis> Pathway Database </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#usage class=md-nav__link> <span class=md-ellipsis> Usage </span> </a> <nav class=md-nav aria-label=Usage> <ul class=md-nav__list> <li class=md-nav__item> <a href=#basic-usage class=md-nav__link> <span class=md-ellipsis> Basic Usage </span> </a> </li> <li class=md-nav__item> <a href=#parameter-tuning-for-parallel-processing class=md-nav__link> <span class=md-ellipsis> Parameter Tuning for Parallel Processing </span> </a> </li> <li class=md-nav__item> <a href=#command-line-usage class=md-nav__link> <span class=md-ellipsis> Command Line Usage </span> </a> </li> <li class=md-nav__item> <a href=#advanced-usage class=md-nav__link> <span class=md-ellipsis> Advanced Usage </span> </a> </li> <li class=md-nav__item> <a href=#resume-from-checkpoint class=md-nav__link> <span class=md-ellipsis> Resume from Checkpoint </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#configuration class=md-nav__link> <span class=md-ellipsis> Configuration </span> </a> <nav class=md-nav aria-label=Configuration> <ul class=md-nav__list> <li class=md-nav__item> <a href=#thread-count class=md-nav__link> <span class=md-ellipsis> Thread Count </span> </a> </li> <li class=md-nav__item> <a href=#gsea-parameters class=md-nav__link> <span class=md-ellipsis> GSEA Parameters </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#output-format class=md-nav__link> <span class=md-ellipsis> Output Format </span> </a> <nav class=md-nav aria-label="Output Format"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#directory-structure class=md-nav__link> <span class=md-ellipsis> Directory Structure </span> </a> </li> <li class=md-nav__item> <a href=#final-results-format class=md-nav__link> <span class=md-ellipsis> Final Results Format </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#performance class=md-nav__link> <span class=md-ellipsis> Performance </span> </a> <nav class=md-nav aria-label=Performance> <ul class=md-nav__list> <li class=md-nav__item> <a href=#computational-requirements class=md-nav__link> <span class=md-ellipsis> Computational Requirements </span> </a> </li> <li class=md-nav__item> <a href=#benchmarks class=md-nav__link> <span class=md-ellipsis> Benchmarks </span> </a> </li> <li class=md-nav__item> <a href=#optimization-tips class=md-nav__link> <span class=md-ellipsis> Optimization Tips </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#cpu-over-subscription-node_fail-on-hpc class=md-nav__link> <span class=md-ellipsis> CPU Over-Subscription / NODE_FAIL on HPC </span> </a> </li> <li class=md-nav__item> <a href=#gsea-not-found class=md-nav__link> <span class=md-ellipsis> GSEA Not Found </span> </a> </li> <li class=md-nav__item> <a href=#memory-errors class=md-nav__link> <span class=md-ellipsis> Memory Errors </span> </a> </li> <li class=md-nav__item> <a href=#gsea-command-failures class=md-nav__link> <span class=md-ellipsis> GSEA Command Failures </span> </a> </li> <li class=md-nav__item> <a href=#missing-pathways class=md-nav__link> <span class=md-ellipsis> Missing Pathways </span> </a> </li> <li class=md-nav__item> <a href=#windows-specific-issues class=md-nav__link> <span class=md-ellipsis> Windows-Specific Issues </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#references class=md-nav__link> <span class=md-ellipsis> References </span> </a> </li> <li class=md-nav__item> <a href=#see-also class=md-nav__link> <span class=md-ellipsis> See Also </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../ENRICHMENT_QUICK_REF/ class=md-nav__link> <span class=md-ellipsis> Enrichment Quick Reference </span> </a> </li> <li class=md-nav__item> <a href=../PATHWAY_HEATMAP/ class=md-nav__link> <span class=md-ellipsis> Pathway Heatmap Analysis </span> </a> </li> <li class=md-nav__item> <a href=../GSEA_INSTALLATION/ class=md-nav__link> <span class=md-ellipsis> GSEA Installation </span> </a> </li> <li class=md-nav__item> <a href=../R_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> R Analysis Scripts </span> </a> </li> <li class=md-nav__item> <a href=../advanced/r-integration/ class=md-nav__link> <span class=md-ellipsis> R Integration </span> </a> </li> <li class=md-nav__item> <a href=../advanced/custom-models/ class=md-nav__link> <span class=md-ellipsis> Custom Models </span> </a> </li> <li class=md-nav__item> <a href=../advanced/performance/ class=md-nav__link> <span class=md-ellipsis> Performance Tuning </span> </a> </li> <li class=md-nav__item> <a href=../advanced/gpu/ class=md-nav__link> <span class=md-ellipsis> GPU Acceleration </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex=0> <span class=md-ellipsis> Contributing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Contributing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../contributing/guidelines/ class=md-nav__link> <span class=md-ellipsis> Guidelines </span> </a> </li> <li class=md-nav__item> <a href=../contributing/development/ class=md-nav__link> <span class=md-ellipsis> Development Setup </span> </a> </li> <li class=md-nav__item> <a href=../contributing/code-style/ class=md-nav__link> <span class=md-ellipsis> Code Style </span> </a> </li> <li class=md-nav__item> <a href=../contributing/testing/ class=md-nav__link> <span class=md-ellipsis> Testing </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex=0> <span class=md-ellipsis> Citation & License </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> Citation & License </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../citation/ class=md-nav__link> <span class=md-ellipsis> How to Cite </span> </a> </li> <li class=md-nav__item> <a href=../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> <li class=md-nav__item> <a href=../acknowledgments/ class=md-nav__link> <span class=md-ellipsis> Acknowledgments </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#pipeline-steps class=md-nav__link> <span class=md-ellipsis> Pipeline Steps </span> </a> <nav class=md-nav aria-label="Pipeline Steps"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-deseq2-differential-expression-analysis class=md-nav__link> <span class=md-ellipsis> 1. DESeq2 Differential Expression Analysis </span> </a> </li> <li class=md-nav__item> <a href=#2-gsea-analysis class=md-nav__link> <span class=md-ellipsis> 2. GSEA Analysis </span> </a> </li> <li class=md-nav__item> <a href=#3-results-combination class=md-nav__link> <span class=md-ellipsis> 3. Results Combination </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#installation-requirements class=md-nav__link> <span class=md-ellipsis> Installation Requirements </span> </a> <nav class=md-nav aria-label="Installation Requirements"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#python-dependencies class=md-nav__link> <span class=md-ellipsis> Python Dependencies </span> </a> </li> <li class=md-nav__item> <a href=#gsea-cli-tool class=md-nav__link> <span class=md-ellipsis> GSEA CLI Tool </span> </a> </li> <li class=md-nav__item> <a href=#pathway-database class=md-nav__link> <span class=md-ellipsis> Pathway Database </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#usage class=md-nav__link> <span class=md-ellipsis> Usage </span> </a> <nav class=md-nav aria-label=Usage> <ul class=md-nav__list> <li class=md-nav__item> <a href=#basic-usage class=md-nav__link> <span class=md-ellipsis> Basic Usage </span> </a> </li> <li class=md-nav__item> <a href=#parameter-tuning-for-parallel-processing class=md-nav__link> <span class=md-ellipsis> Parameter Tuning for Parallel Processing </span> </a> </li> <li class=md-nav__item> <a href=#command-line-usage class=md-nav__link> <span class=md-ellipsis> Command Line Usage </span> </a> </li> <li class=md-nav__item> <a href=#advanced-usage class=md-nav__link> <span class=md-ellipsis> Advanced Usage </span> </a> </li> <li class=md-nav__item> <a href=#resume-from-checkpoint class=md-nav__link> <span class=md-ellipsis> Resume from Checkpoint </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#configuration class=md-nav__link> <span class=md-ellipsis> Configuration </span> </a> <nav class=md-nav aria-label=Configuration> <ul class=md-nav__list> <li class=md-nav__item> <a href=#thread-count class=md-nav__link> <span class=md-ellipsis> Thread Count </span> </a> </li> <li class=md-nav__item> <a href=#gsea-parameters class=md-nav__link> <span class=md-ellipsis> GSEA Parameters </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#output-format class=md-nav__link> <span class=md-ellipsis> Output Format </span> </a> <nav class=md-nav aria-label="Output Format"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#directory-structure class=md-nav__link> <span class=md-ellipsis> Directory Structure </span> </a> </li> <li class=md-nav__item> <a href=#final-results-format class=md-nav__link> <span class=md-ellipsis> Final Results Format </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#performance class=md-nav__link> <span class=md-ellipsis> Performance </span> </a> <nav class=md-nav aria-label=Performance> <ul class=md-nav__list> <li class=md-nav__item> <a href=#computational-requirements class=md-nav__link> <span class=md-ellipsis> Computational Requirements </span> </a> </li> <li class=md-nav__item> <a href=#benchmarks class=md-nav__link> <span class=md-ellipsis> Benchmarks </span> </a> </li> <li class=md-nav__item> <a href=#optimization-tips class=md-nav__link> <span class=md-ellipsis> Optimization Tips </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#cpu-over-subscription-node_fail-on-hpc class=md-nav__link> <span class=md-ellipsis> CPU Over-Subscription / NODE_FAIL on HPC </span> </a> </li> <li class=md-nav__item> <a href=#gsea-not-found class=md-nav__link> <span class=md-ellipsis> GSEA Not Found </span> </a> </li> <li class=md-nav__item> <a href=#memory-errors class=md-nav__link> <span class=md-ellipsis> Memory Errors </span> </a> </li> <li class=md-nav__item> <a href=#gsea-command-failures class=md-nav__link> <span class=md-ellipsis> GSEA Command Failures </span> </a> </li> <li class=md-nav__item> <a href=#missing-pathways class=md-nav__link> <span class=md-ellipsis> Missing Pathways </span> </a> </li> <li class=md-nav__item> <a href=#windows-specific-issues class=md-nav__link> <span class=md-ellipsis> Windows-Specific Issues </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#references class=md-nav__link> <span class=md-ellipsis> References </span> </a> </li> <li class=md-nav__item> <a href=#see-also class=md-nav__link> <span class=md-ellipsis> See Also </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=dynamic-enrichment-analysis>Dynamic Enrichment Analysis<a class=headerlink href=#dynamic-enrichment-analysis title="Permanent link">&para;</a></h1> <p>This document describes the dynamic enrichment analysis pipeline for renalprog.</p> <h2 id=overview>Overview<a class=headerlink href=#overview title="Permanent link">&para;</a></h2> <p>The enrichment analysis pipeline performs Gene Set Enrichment Analysis (GSEA) on synthetic cancer progression trajectories. This allows us to identify biological pathways that are enriched at different stages of progression.</p> <h2 id=pipeline-steps>Pipeline Steps<a class=headerlink href=#pipeline-steps title="Permanent link">&para;</a></h2> <h3 id=1-deseq2-differential-expression-analysis>1. DESeq2 Differential Expression Analysis<a class=headerlink href=#1-deseq2-differential-expression-analysis title="Permanent link">&para;</a></h3> <p>For each synthetic trajectory timepoint: 1. Load trajectory gene expression data (reverse log-transform from RSEM) 2. Load healthy control samples (reverse log-transform from RSEM) 3. Run PyDESeq2 differential expression analysis comparing trajectory vs controls 4. Extract log2 fold-change and adjusted p-values for each gene 5. Rank genes by log2 fold-change 6. Save ranked gene list (<code>.rnk</code> file) for GSEA</p> <p><strong>Note</strong>: The pipeline uses PyDESeq2 for proper differential expression analysis, not simple fold-change calculations. This ensures statistical rigor and proper handling of count data variance.</p> <h3 id=2-gsea-analysis>2. GSEA Analysis<a class=headerlink href=#2-gsea-analysis title="Permanent link">&para;</a></h3> <p>For each ranked gene list: 1. Run GSEA using preranked mode 2. Test against pathway database (ReactomePathways.gmt) 3. Calculate enrichment scores and FDR q-values 4. Generate positive and negative enrichment reports</p> <h3 id=3-results-combination>3. Results Combination<a class=headerlink href=#3-results-combination title="Permanent link">&para;</a></h3> <p>Combine all GSEA results into a single dataset: - One row per (patient, timepoint, pathway) - Includes enrichment score (ES), normalized ES (NES), and FDR q-value - Missing pathways filled with NaN values</p> <h2 id=installation-requirements>Installation Requirements<a class=headerlink href=#installation-requirements title="Permanent link">&para;</a></h2> <h3 id=python-dependencies>Python Dependencies<a class=headerlink href=#python-dependencies title="Permanent link">&para;</a></h3> <p>The enrichment pipeline requires PyDESeq2 for differential expression analysis:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>pip<span class=w> </span>install<span class=w> </span>pydeseq2
</span></code></pre></div> <p>PyDESeq2 is a Python implementation of the DESeq2 method for differential expression analysis of count data.</p> <p><strong>Citation:</strong> <div class="language-text highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a>Muzellec, B., Telenczuk, M., &amp; Cabeli, V. (2022).
</span><span id=__span-1-2><a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>PyDESeq2: a python package for bulk RNA-seq differential expression analysis.
</span><span id=__span-1-3><a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>bioRxiv, 2022-12.
</span></code></pre></div></p> <h3 id=gsea-cli-tool>GSEA CLI Tool<a class=headerlink href=#gsea-cli-tool title="Permanent link">&para;</a></h3> <ol> <li>Download GSEA from: https://www.gsea-msigdb.org/gsea/downloads.jsp</li> <li>Extract to project root (creates <code>GSEA_4.3.2/</code> directory)</li> <li>Ensure <code>gsea-cli.sh</code> (Unix) or <code>gsea-cli.bat</code> (Windows) is executable</li> </ol> <p><strong>Citation:</strong> <div class="language-text highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a>Subramanian, A., Tamayo, P., Mootha, V. K., Mukherjee, S., Ebert, B. L., Gillette, M. A., ... &amp; Mesirov, J. P. (2005).
</span><span id=__span-2-2><a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a>Gene set enrichment analysis: a knowledge-based approach for interpreting genome-wide expression profiles.
</span><span id=__span-2-3><a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a>Proceedings of the National Academy of Sciences, 102(43), 15545-15550.
</span></code></pre></div></p> <h3 id=pathway-database>Pathway Database<a class=headerlink href=#pathway-database title="Permanent link">&para;</a></h3> <p>The ReactomePathways.gmt file is included in <code>data/external/ReactomePathways.gmt</code>.</p> <p><strong>Citation:</strong> <div class="language-text highlight"><pre><span></span><code><span id=__span-3-1><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a>Jassal, B., Matthews, L., Viteri, G., Gong, C., Lorente, P., Fabregat, A., ... &amp; D&#39;Eustachio, P. (2020).
</span><span id=__span-3-2><a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a>The reactome pathway knowledgebase.
</span><span id=__span-3-3><a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a>Nucleic acids research, 48(D1), D498-D503.
</span></code></pre></div></p> <h2 id=usage>Usage<a class=headerlink href=#usage title="Permanent link">&para;</a></h2> <h3 id=basic-usage>Basic Usage<a class=headerlink href=#basic-usage title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-4-1><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=kn>from</span><span class=w> </span><span class=nn>renalprog.enrichment</span><span class=w> </span><span class=kn>import</span> <span class=n>EnrichmentPipeline</span>
</span><span id=__span-4-2><a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>
</span><span id=__span-4-3><a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a><span class=c1># Initialize pipeline</span>
</span><span id=__span-4-4><a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a><span class=n>pipeline</span> <span class=o>=</span> <span class=n>EnrichmentPipeline</span><span class=p>(</span>
</span><span id=__span-4-5><a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a>    <span class=n>trajectory_dir</span><span class=o>=</span><span class=s1>&#39;data/interim/20251216_synthetic_data/kirc/early_to_late&#39;</span><span class=p>,</span>
</span><span id=__span-4-6><a id=__codelineno-4-6 name=__codelineno-4-6 href=#__codelineno-4-6></a>    <span class=n>output_dir</span><span class=o>=</span><span class=s1>&#39;data/processed/20251217_enrichment&#39;</span><span class=p>,</span>
</span><span id=__span-4-7><a id=__codelineno-4-7 name=__codelineno-4-7 href=#__codelineno-4-7></a>    <span class=n>cancer_type</span><span class=o>=</span><span class=s1>&#39;kirc&#39;</span><span class=p>,</span>
</span><span id=__span-4-8><a id=__codelineno-4-8 name=__codelineno-4-8 href=#__codelineno-4-8></a>    <span class=n>n_threads</span><span class=o>=</span><span class=mi>8</span><span class=p>,</span>
</span><span id=__span-4-9><a id=__codelineno-4-9 name=__codelineno-4-9 href=#__codelineno-4-9></a>    <span class=n>n_threads_per_deseq</span><span class=o>=</span><span class=mi>8</span><span class=p>,</span>  <span class=c1># Threads per DESeq2 analysis to prevent CPU over-subscription</span>
</span><span id=__span-4-10><a id=__codelineno-4-10 name=__codelineno-4-10 href=#__codelineno-4-10></a>    <span class=n>gsea_path</span><span class=o>=</span><span class=s1>&#39;./GSEA_4.3.2/gsea-cli.sh&#39;</span><span class=p>,</span>
</span><span id=__span-4-11><a id=__codelineno-4-11 name=__codelineno-4-11 href=#__codelineno-4-11></a>    <span class=n>pathways_file</span><span class=o>=</span><span class=s1>&#39;data/external/ReactomePathways.gmt&#39;</span>
</span><span id=__span-4-12><a id=__codelineno-4-12 name=__codelineno-4-12 href=#__codelineno-4-12></a><span class=p>)</span>
</span><span id=__span-4-13><a id=__codelineno-4-13 name=__codelineno-4-13 href=#__codelineno-4-13></a>
</span><span id=__span-4-14><a id=__codelineno-4-14 name=__codelineno-4-14 href=#__codelineno-4-14></a><span class=c1># Run complete pipeline</span>
</span><span id=__span-4-15><a id=__codelineno-4-15 name=__codelineno-4-15 href=#__codelineno-4-15></a><span class=n>results</span> <span class=o>=</span> <span class=n>pipeline</span><span class=o>.</span><span class=n>run</span><span class=p>()</span>
</span><span id=__span-4-16><a id=__codelineno-4-16 name=__codelineno-4-16 href=#__codelineno-4-16></a>
</span><span id=__span-4-17><a id=__codelineno-4-17 name=__codelineno-4-17 href=#__codelineno-4-17></a><span class=c1># Results are saved to: data/processed/20251217_enrichment/trajectory_enrichment.csv</span>
</span></code></pre></div> <h3 id=parameter-tuning-for-parallel-processing>Parameter Tuning for Parallel Processing<a class=headerlink href=#parameter-tuning-for-parallel-processing title="Permanent link">&para;</a></h3> <p>The pipeline supports parallel processing with careful control over CPU usage:</p> <ul> <li><code>n_threads</code>: Number of trajectory files to process in parallel (default: 4)</li> <li><code>n_threads_per_deseq</code>: Number of threads for each PyDESeq2 analysis (default: 8)</li> </ul> <p><strong>Important</strong>: Total CPU usage = <code>n_threads × n_threads_per_deseq</code>. Choose values based on your system:</p> <ul> <li><strong>Local PC (8 cores)</strong>: <code>n_threads=2, n_threads_per_deseq=4</code> → 8 threads total</li> <li><strong>Workstation (32 cores)</strong>: <code>n_threads=4, n_threads_per_deseq=8</code> → 32 threads total </li> <li><strong>HPC cluster (112 cores)</strong>: <code>n_threads=14, n_threads_per_deseq=8</code> → 112 threads total</li> </ul> <p>This prevents CPU over-subscription which can cause NODE_FAIL errors on SLURM clusters.</p> <h3 id=command-line-usage>Command Line Usage<a class=headerlink href=#command-line-usage title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-5-1><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=c1># Run complete pipeline</span>
</span><span id=__span-5-2><a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a>python<span class=w> </span>scripts/pipeline_steps/6_enrichment_analysis.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-5-3><a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a><span class=w>    </span>--trajectory_dir<span class=w> </span>data/interim/20251216_synthetic_data/kirc/early_to_late<span class=w> </span><span class=se>\</span>
</span><span id=__span-5-4><a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=w>    </span>--output_dir<span class=w> </span>data/processed/20251217_enrichment<span class=w> </span><span class=se>\</span>
</span><span id=__span-5-5><a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a><span class=w>    </span>--cancer_type<span class=w> </span>kirc<span class=w> </span><span class=se>\</span>
</span><span id=__span-5-6><a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a><span class=w>    </span>--n_threads<span class=w> </span><span class=m>8</span><span class=w> </span><span class=se>\</span>
</span><span id=__span-5-7><a id=__codelineno-5-7 name=__codelineno-5-7 href=#__codelineno-5-7></a><span class=w>    </span>--gsea_path<span class=w> </span>./GSEA_4.3.2/gsea-cli.sh<span class=w> </span><span class=se>\</span>
</span><span id=__span-5-8><a id=__codelineno-5-8 name=__codelineno-5-8 href=#__codelineno-5-8></a><span class=w>    </span>--pathways_file<span class=w> </span>data/external/ReactomePathways.gmt
</span></code></pre></div> <h3 id=advanced-usage>Advanced Usage<a class=headerlink href=#advanced-usage title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-6-1><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=c1># Run steps separately</span>
</span><span id=__span-6-2><a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=kn>from</span><span class=w> </span><span class=nn>renalprog.enrichment</span><span class=w> </span><span class=kn>import</span> <span class=p>(</span>
</span><span id=__span-6-3><a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a>    <span class=n>process_trajectories_for_deseq</span><span class=p>,</span>
</span><span id=__span-6-4><a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a>    <span class=n>run_gsea_parallel</span><span class=p>,</span>
</span><span id=__span-6-5><a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a>    <span class=n>combine_gsea_results</span>
</span><span id=__span-6-6><a id=__codelineno-6-6 name=__codelineno-6-6 href=#__codelineno-6-6></a><span class=p>)</span>
</span><span id=__span-6-7><a id=__codelineno-6-7 name=__codelineno-6-7 href=#__codelineno-6-7></a>
</span><span id=__span-6-8><a id=__codelineno-6-8 name=__codelineno-6-8 href=#__codelineno-6-8></a><span class=c1># Step 1: DESeq2 processing (reverse log-transform + differential expression)</span>
</span><span id=__span-6-9><a id=__codelineno-6-9 name=__codelineno-6-9 href=#__codelineno-6-9></a><span class=n>process_trajectories_for_deseq</span><span class=p>(</span>
</span><span id=__span-6-10><a id=__codelineno-6-10 name=__codelineno-6-10 href=#__codelineno-6-10></a>    <span class=n>trajectory_dir</span><span class=o>=</span><span class=s1>&#39;data/interim/20251216_synthetic_data/kirc/early_to_late&#39;</span><span class=p>,</span>
</span><span id=__span-6-11><a id=__codelineno-6-11 name=__codelineno-6-11 href=#__codelineno-6-11></a>    <span class=n>output_dir</span><span class=o>=</span><span class=s1>&#39;data/processed/20251217_enrichment&#39;</span><span class=p>,</span>
</span><span id=__span-6-12><a id=__codelineno-6-12 name=__codelineno-6-12 href=#__codelineno-6-12></a>    <span class=n>cancer_type</span><span class=o>=</span><span class=s1>&#39;kirc&#39;</span><span class=p>,</span>
</span><span id=__span-6-13><a id=__codelineno-6-13 name=__codelineno-6-13 href=#__codelineno-6-13></a>    <span class=n>n_threads</span><span class=o>=</span><span class=mi>8</span>
</span><span id=__span-6-14><a id=__codelineno-6-14 name=__codelineno-6-14 href=#__codelineno-6-14></a><span class=p>)</span>
</span><span id=__span-6-15><a id=__codelineno-6-15 name=__codelineno-6-15 href=#__codelineno-6-15></a>
</span><span id=__span-6-16><a id=__codelineno-6-16 name=__codelineno-6-16 href=#__codelineno-6-16></a><span class=c1># Step 2: Run GSEA</span>
</span><span id=__span-6-17><a id=__codelineno-6-17 name=__codelineno-6-17 href=#__codelineno-6-17></a><span class=n>run_gsea_parallel</span><span class=p>(</span>
</span><span id=__span-6-18><a id=__codelineno-6-18 name=__codelineno-6-18 href=#__codelineno-6-18></a>    <span class=n>deseq_dir</span><span class=o>=</span><span class=s1>&#39;data/processed/20251217_enrichment/deseq&#39;</span><span class=p>,</span>
</span><span id=__span-6-19><a id=__codelineno-6-19 name=__codelineno-6-19 href=#__codelineno-6-19></a>    <span class=n>n_threads</span><span class=o>=</span><span class=mi>8</span>
</span><span id=__span-6-20><a id=__codelineno-6-20 name=__codelineno-6-20 href=#__codelineno-6-20></a><span class=p>)</span>
</span><span id=__span-6-21><a id=__codelineno-6-21 name=__codelineno-6-21 href=#__codelineno-6-21></a>
</span><span id=__span-6-22><a id=__codelineno-6-22 name=__codelineno-6-22 href=#__codelineno-6-22></a><span class=c1># Step 3: Combine results</span>
</span><span id=__span-6-23><a id=__codelineno-6-23 name=__codelineno-6-23 href=#__codelineno-6-23></a><span class=n>results</span> <span class=o>=</span> <span class=n>combine_gsea_results</span><span class=p>(</span>
</span><span id=__span-6-24><a id=__codelineno-6-24 name=__codelineno-6-24 href=#__codelineno-6-24></a>    <span class=n>deseq_dir</span><span class=o>=</span><span class=s1>&#39;data/processed/20251217_enrichment/deseq&#39;</span><span class=p>,</span>
</span><span id=__span-6-25><a id=__codelineno-6-25 name=__codelineno-6-25 href=#__codelineno-6-25></a>    <span class=n>pathways_file</span><span class=o>=</span><span class=s1>&#39;data/external/ReactomePathways.gmt&#39;</span>
</span><span id=__span-6-26><a id=__codelineno-6-26 name=__codelineno-6-26 href=#__codelineno-6-26></a><span class=p>)</span>
</span></code></pre></div> <p><strong>Technical Details:</strong></p> <p>The DESeq2 processing involves: 1. <strong>Reverse log-transformation</strong>: Input data is log-transformed RSEM values, which are converted back to counts 2. <strong>PyDESeq2 analysis</strong>: Proper variance modeling and statistical testing 3. <strong>Rank file generation</strong>: Genes ranked by log2 fold-change for GSEA preranked mode</p> <h3 id=resume-from-checkpoint>Resume from Checkpoint<a class=headerlink href=#resume-from-checkpoint title="Permanent link">&para;</a></h3> <p>If the pipeline fails or is interrupted, you can resume:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-7-1><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># Skip DESeq if already completed</span>
</span><span id=__span-7-2><a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a>python<span class=w> </span>scripts/pipeline_steps/6_enrichment_analysis.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-7-3><a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=w>    </span>--trajectory_dir<span class=w> </span>data/interim/20251216_synthetic_data/kirc/early_to_late<span class=w> </span><span class=se>\</span>
</span><span id=__span-7-4><a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a><span class=w>    </span>--output_dir<span class=w> </span>data/processed/20251217_enrichment<span class=w> </span><span class=se>\</span>
</span><span id=__span-7-5><a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a><span class=w>    </span>--skip_deseq
</span><span id=__span-7-6><a id=__codelineno-7-6 name=__codelineno-7-6 href=#__codelineno-7-6></a>
</span><span id=__span-7-7><a id=__codelineno-7-7 name=__codelineno-7-7 href=#__codelineno-7-7></a><span class=c1># Skip GSEA if already completed</span>
</span><span id=__span-7-8><a id=__codelineno-7-8 name=__codelineno-7-8 href=#__codelineno-7-8></a>python<span class=w> </span>scripts/pipeline_steps/6_enrichment_analysis.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-7-9><a id=__codelineno-7-9 name=__codelineno-7-9 href=#__codelineno-7-9></a><span class=w>    </span>--trajectory_dir<span class=w> </span>data/interim/20251216_synthetic_data/kirc/early_to_late<span class=w> </span><span class=se>\</span>
</span><span id=__span-7-10><a id=__codelineno-7-10 name=__codelineno-7-10 href=#__codelineno-7-10></a><span class=w>    </span>--output_dir<span class=w> </span>data/processed/20251217_enrichment<span class=w> </span><span class=se>\</span>
</span><span id=__span-7-11><a id=__codelineno-7-11 name=__codelineno-7-11 href=#__codelineno-7-11></a><span class=w>    </span>--skip_deseq<span class=w> </span><span class=se>\</span>
</span><span id=__span-7-12><a id=__codelineno-7-12 name=__codelineno-7-12 href=#__codelineno-7-12></a><span class=w>    </span>--skip_gsea
</span></code></pre></div> <h2 id=configuration>Configuration<a class=headerlink href=#configuration title="Permanent link">&para;</a></h2> <h3 id=thread-count>Thread Count<a class=headerlink href=#thread-count title="Permanent link">&para;</a></h3> <p>The <code>n_threads</code> parameter controls parallelization: - <strong>Recommended</strong>: Number of CPU cores - 1 - <strong>Minimum</strong>: 1 (sequential processing, very slow) - <strong>Maximum</strong>: Number of CPU cores - <strong>Default</strong>: 4</p> <p>Example for different systems: <div class="language-python highlight"><pre><span></span><code><span id=__span-8-1><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># Modern PC (8+ cores)</span>
</span><span id=__span-8-2><a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=n>pipeline</span> <span class=o>=</span> <span class=n>EnrichmentPipeline</span><span class=p>(</span><span class=o>...</span><span class=p>,</span> <span class=n>n_threads</span><span class=o>=</span><span class=mi>6</span><span class=p>)</span>
</span><span id=__span-8-3><a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a>
</span><span id=__span-8-4><a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=c1># Older PC (4 cores)</span>
</span><span id=__span-8-5><a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a><span class=n>pipeline</span> <span class=o>=</span> <span class=n>EnrichmentPipeline</span><span class=p>(</span><span class=o>...</span><span class=p>,</span> <span class=n>n_threads</span><span class=o>=</span><span class=mi>2</span><span class=p>)</span>
</span><span id=__span-8-6><a id=__codelineno-8-6 name=__codelineno-8-6 href=#__codelineno-8-6></a>
</span><span id=__span-8-7><a id=__codelineno-8-7 name=__codelineno-8-7 href=#__codelineno-8-7></a><span class=c1># High-performance cluster</span>
</span><span id=__span-8-8><a id=__codelineno-8-8 name=__codelineno-8-8 href=#__codelineno-8-8></a><span class=n>pipeline</span> <span class=o>=</span> <span class=n>EnrichmentPipeline</span><span class=p>(</span><span class=o>...</span><span class=p>,</span> <span class=n>n_threads</span><span class=o>=</span><span class=mi>32</span><span class=p>)</span>
</span></code></pre></div></p> <h3 id=gsea-parameters>GSEA Parameters<a class=headerlink href=#gsea-parameters title="Permanent link">&para;</a></h3> <p>Default GSEA parameters in <code>generate_gsea_command()</code>: - <code>collapse</code>: false (use all genes) - <code>nperm</code>: 1000 permutations - <code>set_max</code>: 500 (maximum pathway size) - <code>set_min</code>: 15 (minimum pathway size)</p> <p>To modify, edit <code>renalprog/modeling/enrichment.py</code>: <div class="language-python highlight"><pre><span></span><code><span id=__span-9-1><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=k>def</span><span class=w> </span><span class=nf>generate_gsea_command</span><span class=p>(</span><span class=o>...</span><span class=p>):</span>
</span><span id=__span-9-2><a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>    <span class=n>cmd</span> <span class=o>=</span> <span class=p>(</span>
</span><span id=__span-9-3><a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a>        <span class=sa>f</span><span class=s1>&#39;&quot;</span><span class=si>{</span><span class=n>gsea_path</span><span class=si>}</span><span class=s1>&quot; GSEAPreranked &#39;</span>
</span><span id=__span-9-4><a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a>        <span class=sa>f</span><span class=s1>&#39;-gmx &quot;</span><span class=si>{</span><span class=n>gmt_file</span><span class=si>}</span><span class=s1>&quot; &#39;</span>
</span><span id=__span-9-5><a id=__codelineno-9-5 name=__codelineno-9-5 href=#__codelineno-9-5></a>        <span class=sa>f</span><span class=s1>&#39;-rnk &quot;</span><span class=si>{</span><span class=n>rnk_file</span><span class=si>}</span><span class=s1>&quot; &#39;</span>
</span><span id=__span-9-6><a id=__codelineno-9-6 name=__codelineno-9-6 href=#__codelineno-9-6></a>        <span class=sa>f</span><span class=s1>&#39;-out &quot;</span><span class=si>{</span><span class=n>output_dir</span><span class=si>}</span><span class=s1>&quot; &#39;</span>
</span><span id=__span-9-7><a id=__codelineno-9-7 name=__codelineno-9-7 href=#__codelineno-9-7></a>        <span class=sa>f</span><span class=s1>&#39;-collapse false &#39;</span>
</span><span id=__span-9-8><a id=__codelineno-9-8 name=__codelineno-9-8 href=#__codelineno-9-8></a>        <span class=sa>f</span><span class=s1>&#39;-nperm 2000 &#39;</span>  <span class=c1># Increase permutations</span>
</span><span id=__span-9-9><a id=__codelineno-9-9 name=__codelineno-9-9 href=#__codelineno-9-9></a>        <span class=sa>f</span><span class=s1>&#39;-set_max 1000 &#39;</span>  <span class=c1># Allow larger pathways</span>
</span><span id=__span-9-10><a id=__codelineno-9-10 name=__codelineno-9-10 href=#__codelineno-9-10></a>        <span class=sa>f</span><span class=s1>&#39;-set_min 10&#39;</span>  <span class=c1># Allow smaller pathways</span>
</span><span id=__span-9-11><a id=__codelineno-9-11 name=__codelineno-9-11 href=#__codelineno-9-11></a>    <span class=p>)</span>
</span></code></pre></div></p> <h2 id=output-format>Output Format<a class=headerlink href=#output-format title="Permanent link">&para;</a></h2> <h3 id=directory-structure>Directory Structure<a class=headerlink href=#directory-structure title="Permanent link">&para;</a></h3> <div class="language-text highlight"><pre><span></span><code><span id=__span-10-1><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a>output_dir/
</span><span id=__span-10-2><a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a>├── deseq/                          # DESeq results
</span><span id=__span-10-3><a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a>│   ├── early_to_late/             # Transition type
</span><span id=__span-10-4><a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a>│   │   ├── patient1/              # Patient trajectory
</span><span id=__span-10-5><a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a>│   │   │   ├── patient1_tp0_foldchange.rnk
</span><span id=__span-10-6><a id=__codelineno-10-6 name=__codelineno-10-6 href=#__codelineno-10-6></a>│   │   │   ├── patient1_tp1_foldchange.rnk
</span><span id=__span-10-7><a id=__codelineno-10-7 name=__codelineno-10-7 href=#__codelineno-10-7></a>│   │   │   ├── ...
</span><span id=__span-10-8><a id=__codelineno-10-8 name=__codelineno-10-8 href=#__codelineno-10-8></a>│   │   │   ├── gsea_tp0/          # GSEA output for timepoint 0
</span><span id=__span-10-9><a id=__codelineno-10-9 name=__codelineno-10-9 href=#__codelineno-10-9></a>│   │   │   │   ├── gsea_report_for_na_pos_*.tsv
</span><span id=__span-10-10><a id=__codelineno-10-10 name=__codelineno-10-10 href=#__codelineno-10-10></a>│   │   │   │   └── gsea_report_for_na_neg_*.tsv
</span><span id=__span-10-11><a id=__codelineno-10-11 name=__codelineno-10-11 href=#__codelineno-10-11></a>│   │   │   └── gsea_tp1/
</span><span id=__span-10-12><a id=__codelineno-10-12 name=__codelineno-10-12 href=#__codelineno-10-12></a>│   │   └── patient2/
</span><span id=__span-10-13><a id=__codelineno-10-13 name=__codelineno-10-13 href=#__codelineno-10-13></a>│   └── gsea_commands_*.cmd        # GSEA command files
</span><span id=__span-10-14><a id=__codelineno-10-14 name=__codelineno-10-14 href=#__codelineno-10-14></a>├── gsea/                           # GSEA working directory
</span><span id=__span-10-15><a id=__codelineno-10-15 name=__codelineno-10-15 href=#__codelineno-10-15></a>└── trajectory_enrichment.csv       # Final combined results
</span></code></pre></div> <h3 id=final-results-format>Final Results Format<a class=headerlink href=#final-results-format title="Permanent link">&para;</a></h3> <p><code>trajectory_enrichment.csv</code> columns: - <code>Patient</code>: Patient identifier (e.g., "TCGA-3Z-A93Z-01_to_TCGA-A3-A8OW-01") - <code>Idx</code>: Timepoint index (0 to n_samples-1) - <code>Transition</code>: Transition type (e.g., "early_to_late") - <code>NAME</code>: Pathway name (from ReactomePathways.gmt) - <code>ES</code>: Enrichment score - <code>NES</code>: Normalized enrichment score - <code>FDR q-val</code>: False discovery rate q-value</p> <p>Example: <div class="language-text highlight"><pre><span></span><code><span id=__span-11-1><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a>Patient,Idx,Transition,NAME,ES,NES,FDR q-val
</span><span id=__span-11-2><a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a>TCGA-3Z-A93Z-01_to_TCGA-A3-A8OW-01,0,early_to_late,Cell Cycle,0.65,2.13,0.001
</span><span id=__span-11-3><a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a>TCGA-3Z-A93Z-01_to_TCGA-A3-A8OW-01,0,early_to_late,DNA Repair,0.52,1.87,0.012
</span><span id=__span-11-4><a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a>TCGA-3Z-A93Z-01_to_TCGA-A3-A8OW-01,1,early_to_late,Cell Cycle,0.71,2.31,0.000
</span></code></pre></div></p> <h2 id=performance>Performance<a class=headerlink href=#performance title="Permanent link">&para;</a></h2> <h3 id=computational-requirements>Computational Requirements<a class=headerlink href=#computational-requirements title="Permanent link">&para;</a></h3> <ul> <li><strong>Memory</strong>: ~4 GB per thread</li> <li><strong>Disk Space</strong>: ~10 GB for KIRC dataset (500+ trajectories)</li> <li><strong>Time</strong>: ~2-6 hours for KIRC dataset (depends on threads)</li> </ul> <h3 id=benchmarks>Benchmarks<a class=headerlink href=#benchmarks title="Permanent link">&para;</a></h3> <p>Testing on different hardware configurations:</p> <table> <thead> <tr> <th>System</th> <th>Cores</th> <th>RAM</th> <th>Trajectories</th> <th>Time</th> <th>Threads Used</th> </tr> </thead> <tbody> <tr> <td>Modern Desktop (2023)</td> <td>16</td> <td>32 GB</td> <td>500</td> <td>2 hrs</td> <td>12</td> </tr> <tr> <td>Laptop (2020)</td> <td>8</td> <td>16 GB</td> <td>500</td> <td>4 hrs</td> <td>6</td> </tr> <tr> <td>Workstation (2018)</td> <td>4</td> <td>8 GB</td> <td>500</td> <td>6 hrs</td> <td>3</td> </tr> <tr> <td>HPC Cluster</td> <td>64</td> <td>256 GB</td> <td>500</td> <td>0.5 hrs</td> <td>48</td> </tr> </tbody> </table> <h3 id=optimization-tips>Optimization Tips<a class=headerlink href=#optimization-tips title="Permanent link">&para;</a></h3> <ol> <li><strong>Use SSD</strong>: GSEA creates many temporary files</li> <li><strong>Adjust threads</strong>: Set to CPU cores - 1 for best performance</li> <li><strong>Clean up</strong>: Use <code>--cleanup</code> flag to save disk space</li> <li><strong>Resume</strong>: Use <code>--skip_deseq</code> or <code>--skip_gsea</code> to resume interrupted runs</li> </ol> <h2 id=troubleshooting>Troubleshooting<a class=headerlink href=#troubleshooting title="Permanent link">&para;</a></h2> <h3 id=cpu-over-subscription-node_fail-on-hpc>CPU Over-Subscription / NODE_FAIL on HPC<a class=headerlink href=#cpu-over-subscription-node_fail-on-hpc title="Permanent link">&para;</a></h3> <p><strong>Error</strong>: <code>NODE_FAIL</code> on SLURM, or system reports &gt;500% CPU usage</p> <p><strong>Cause</strong>: Too many parallel processes using too many threads each</p> <p><strong>Solution</strong>: The pipeline uses two levels of parallelization: 1. <code>n_threads</code>: Number of trajectory files processed in parallel 2. <code>n_threads_per_deseq</code>: Number of threads for each PyDESeq2 analysis</p> <p>Total CPU usage = <code>n_threads × n_threads_per_deseq</code></p> <p><strong>Fix</strong>: Adjust these parameters based on available cores:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-12-1><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=c1># For 112-core HPC node</span>
</span><span id=__span-12-2><a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a>python<span class=w> </span>scripts/pipeline_steps/6_enrichment_analysis.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-12-3><a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a><span class=w>    </span>--n_threads<span class=w> </span><span class=m>14</span><span class=w> </span><span class=se>\</span>
</span><span id=__span-12-4><a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a><span class=w>    </span>--n_threads_per_deseq<span class=w> </span><span class=m>8</span><span class=w>  </span><span class=c1># Total: 112 threads</span>
</span><span id=__span-12-5><a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a>
</span><span id=__span-12-6><a id=__codelineno-12-6 name=__codelineno-12-6 href=#__codelineno-12-6></a><span class=c1># For 32-core workstation</span>
</span><span id=__span-12-7><a id=__codelineno-12-7 name=__codelineno-12-7 href=#__codelineno-12-7></a>python<span class=w> </span>scripts/pipeline_steps/6_enrichment_analysis.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-12-8><a id=__codelineno-12-8 name=__codelineno-12-8 href=#__codelineno-12-8></a><span class=w>    </span>--n_threads<span class=w> </span><span class=m>4</span><span class=w> </span><span class=se>\</span>
</span><span id=__span-12-9><a id=__codelineno-12-9 name=__codelineno-12-9 href=#__codelineno-12-9></a><span class=w>    </span>--n_threads_per_deseq<span class=w> </span><span class=m>8</span><span class=w>  </span><span class=c1># Total: 32 threads</span>
</span><span id=__span-12-10><a id=__codelineno-12-10 name=__codelineno-12-10 href=#__codelineno-12-10></a>
</span><span id=__span-12-11><a id=__codelineno-12-11 name=__codelineno-12-11 href=#__codelineno-12-11></a><span class=c1># For 8-core PC</span>
</span><span id=__span-12-12><a id=__codelineno-12-12 name=__codelineno-12-12 href=#__codelineno-12-12></a>python<span class=w> </span>scripts/pipeline_steps/6_enrichment_analysis.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-12-13><a id=__codelineno-12-13 name=__codelineno-12-13 href=#__codelineno-12-13></a><span class=w>    </span>--n_threads<span class=w> </span><span class=m>2</span><span class=w> </span><span class=se>\</span>
</span><span id=__span-12-14><a id=__codelineno-12-14 name=__codelineno-12-14 href=#__codelineno-12-14></a><span class=w>    </span>--n_threads_per_deseq<span class=w> </span><span class=m>4</span><span class=w>  </span><span class=c1># Total: 8 threads</span>
</span></code></pre></div> <p>The pipeline sets environment variables (OMP_NUM_THREADS, MKL_NUM_THREADS, etc.) to limit threading per process.</p> <h3 id=gsea-not-found>GSEA Not Found<a class=headerlink href=#gsea-not-found title="Permanent link">&para;</a></h3> <p><strong>Error</strong>: <code>GSEA CLI not found at ./GSEA_4.3.2/gsea-cli.sh</code></p> <p><strong>Solution</strong>: 1. Download GSEA from https://www.gsea-msigdb.org/gsea/downloads.jsp 2. Extract to project root 3. Or specify custom path with <code>--gsea_path</code></p> <h3 id=memory-errors>Memory Errors<a class=headerlink href=#memory-errors title="Permanent link">&para;</a></h3> <p><strong>Error</strong>: <code>OutOfMemoryError</code> or system freezes</p> <p><strong>Solution</strong>: 1. Reduce <code>n_threads</code> 2. Process trajectories in batches 3. Increase system swap space</p> <h3 id=gsea-command-failures>GSEA Command Failures<a class=headerlink href=#gsea-command-failures title="Permanent link">&para;</a></h3> <p><strong>Error</strong>: GSEA commands fail with non-zero exit code</p> <p><strong>Solution</strong>: 1. Check GSEA installation 2. Verify pathway file format (GMT) 3. Check file permissions 4. Review GSEA log files in output directories</p> <h3 id=missing-pathways>Missing Pathways<a class=headerlink href=#missing-pathways title="Permanent link">&para;</a></h3> <p><strong>Warning</strong>: Some pathways have all NaN values</p> <p><strong>Explanation</strong>: Normal - not all pathways are significant in every sample</p> <h3 id=windows-specific-issues>Windows-Specific Issues<a class=headerlink href=#windows-specific-issues title="Permanent link">&para;</a></h3> <p><strong>Error</strong>: Cannot run <code>gsea-cli.sh</code> on Windows</p> <p><strong>Solution</strong>: 1. Install Git Bash or WSL (Windows Subsystem for Linux) 2. Use <code>gsea-cli.bat</code> instead if available 3. Or run in WSL environment</p> <h2 id=references>References<a class=headerlink href=#references title="Permanent link">&para;</a></h2> <ol> <li><strong>GSEA Method</strong>:</li> <li>Subramanian et al. (2005). "Gene set enrichment analysis: a knowledge-based approach for interpreting genome-wide expression profiles." PNAS 102(43):15545-15550.</li> <li> <p>https://www.gsea-msigdb.org/</p> </li> <li> <p><strong>PyDESeq2</strong>:</p> </li> <li>Muzellec, B., Telenczuk, M., &amp; Cabeli, V. (2022). "PyDESeq2: a python package for bulk RNA-seq differential expression analysis." bioRxiv, 2022-12.</li> <li> <p>https://github.com/owkin/PyDESeq2</p> </li> <li> <p><strong>DESeq2 Original Method</strong>:</p> </li> <li> <p>Love, M.I., Huber, W., Anders, S. (2014). "Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2." Genome Biology 15(12):550.</p> </li> <li> <p><strong>Reactome Pathways</strong>:</p> </li> <li>Jassal et al. (2020). "The reactome pathway knowledgebase." Nucleic Acids Research 48(D1):D498-D503.</li> <li> <p>https://reactome.org/</p> </li> <li> <p><strong>Original Implementation</strong>:</p> </li> <li>Prol-Castelo, G. (2024). My_BRCA repository</li> <li>Files: <code>src_deseq_and_gsea_NCSR/py_deseq.py</code>, <code>trajectory_analysis.py</code></li> </ol> <h2 id=see-also>See Also<a class=headerlink href=#see-also title="Permanent link">&para;</a></h2> <ul> <li><a href=../CLASSIFICATION/ >Classification Pipeline</a></li> <li><a href=../TRAJECTORIES/ >Trajectory Generation</a></li> <li><a href=../R_ANALYSIS/ >R Analysis Scripts</a></li> </ul> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="December 19, 2025 16:49:24 UTC"><span class=timeago datetime=2025-12-19T16:49:24+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="December 19, 2025 16:49:24 UTC">2025-12-19</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="December 19, 2025 16:49:24 UTC"><span class=timeago datetime=2025-12-19T16:49:24+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="December 19, 2025 16:49:24 UTC">2025-12-19</span> </span> </aside> <form class=md-feedback name=feedback hidden> <fieldset> <legend class=md-feedback__title> Was this page helpful? </legend> <div class=md-feedback__inner> <div class=md-feedback__list> <button class="md-feedback__icon md-icon" type=submit title="This page was helpful" data-md-value=1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m7 0c0 .8-.7 1.5-1.5 1.5S14 10.3 14 9.5 14.7 8 15.5 8s1.5.7 1.5 1.5m-5 7.73c-1.75 0-3.29-.73-4.19-1.81L9.23 14c.45.72 1.52 1.23 2.77 1.23s2.32-.51 2.77-1.23l1.42 1.42c-.9 1.08-2.44 1.81-4.19 1.81"/></svg> </button> <button class="md-feedback__icon md-icon" type=submit title="This page could be improved" data-md-value=0> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10m-6.5-4c.8 0 1.5.7 1.5 1.5s-.7 1.5-1.5 1.5-1.5-.7-1.5-1.5.7-1.5 1.5-1.5M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m2 4.5c1.75 0 3.29.72 4.19 1.81l-1.42 1.42C14.32 16.5 13.25 16 12 16s-2.32.5-2.77 1.23l-1.42-1.42C8.71 14.72 10.25 14 12 14"/></svg> </button> </div> <div class=md-feedback__note> <div data-md-value=1 hidden> Thanks for your feedback! </div> <div data-md-value=0 hidden> Thanks for your feedback! Help us improve by opening an issue. </div> </div> </div> </fieldset> </form> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../CLASSIFICATION/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Classification Details"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Classification Details </div> </div> </a> <a href=../ENRICHMENT_QUICK_REF/ class="md-footer__link md-footer__link--next" aria-label="Next: Enrichment Quick Reference"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Enrichment Quick Reference </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2022-2025 EVENFLOW Project </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/gprolcastelo/renalprog target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://pypi.org/project/renalprog/ target=_blank rel=noopener title=pypi.org class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.8 200.5c-7.7-30.9-22.3-54.2-53.4-54.2h-40.1v47.4c0 36.8-31.2 67.8-66.8 67.8H172.7c-29.2 0-53.4 25-53.4 54.3v101.8c0 29 25.2 46 53.4 54.3 33.8 9.9 66.3 11.7 106.8 0 26.9-7.8 53.4-23.5 53.4-54.3v-40.7H226.2v-13.6h160.2c31.1 0 42.6-21.7 53.4-54.2 11.2-33.5 10.7-65.7 0-108.6M286.2 444.7a20.4 20.4 0 1 1 0-40.7 20.4 20.4 0 1 1 0 40.7M167.8 248.1h106.8c29.7 0 53.4-24.5 53.4-54.3V91.9c0-29-24.4-50.7-53.4-55.6-35.8-5.9-74.7-5.6-106.8.1-45.2 8-53.4 24.7-53.4 55.6v40.7h106.9v13.6h-147c-31.1 0-58.3 18.7-66.8 54.2-9.8 40.7-10.2 66.1 0 108.6 7.6 31.6 25.7 54.2 56.8 54.2H101v-48.8c0-35.3 30.5-66.4 66.8-66.4m-6.6-183.4a20.4 20.4 0 1 1 0 40.8 20.4 20.4 0 1 1 0-40.8"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "search.suggest", "search.highlight", "search.share", "content.code.copy", "content.code.annotate", "toc.follow"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../js/timeago.min.js></script> <script src=../js/timeago_mkdocs_material.js></script> <script src=../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>