<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="A Python package for simulating kidney cancer progression with synthetic data generation and machine learning."><meta name=author content="Guillermo Prol-Castelo (Barcelona Supercomputing Center), for the EVENFLOW European Project"><link href=https://github.com/gprolcastelo/renalprog/tutorials/data-requirements/ rel=canonical><link href=../quickstart/ rel=prev><link href=../ rel=next><link rel=icon href=../../assets/images/kidneys.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>Data Requirements - renalprog</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../css/timeago.css><link rel=stylesheet href=../../assets/_mkdocstrings.css><link rel=stylesheet href=../../stylesheets/extra.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#data-requirements class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=renalprog class="md-header__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../../assets/images/kidneys.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> renalprog </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Data Requirements </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> Home </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=../../getting-started/ class=md-tabs__link> Getting Started </a> </li> <li class=md-tabs__item> <a href=../ class=md-tabs__link> Tutorials </a> </li> <li class=md-tabs__item> <a href=../../reproducibility/ class=md-tabs__link> Reproducibility </a> </li> <li class=md-tabs__item> <a href=../../api/ class=md-tabs__link> API Reference </a> </li> <li class=md-tabs__item> <a href=../../advanced/architecture/ class=md-tabs__link> Advanced Topics </a> </li> <li class=md-tabs__item> <a href=../../contributing/guidelines/ class=md-tabs__link> Contributing </a> </li> <li class=md-tabs__item> <a href=../../citation/ class=md-tabs__link> Citation & License </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=renalprog class="md-nav__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../../assets/images/kidneys.png alt=logo> </a> renalprog </label> <div class=md-nav__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex> <span class=md-ellipsis> Getting Started </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Getting Started </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../getting-started/ class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=../quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Data Requirements </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Data Requirements </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#input-data-formats class=md-nav__link> <span class=md-ellipsis> Input Data Formats </span> </a> <nav class=md-nav aria-label="Input Data Formats"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-gene-expression-data class=md-nav__link> <span class=md-ellipsis> 1. Gene Expression Data </span> </a> </li> <li class=md-nav__item> <a href=#2-clinical-data class=md-nav__link> <span class=md-ellipsis> 2. Clinical Data </span> </a> </li> <li class=md-nav__item> <a href=#3-sample-metadata-optional class=md-nav__link> <span class=md-ellipsis> 3. Sample Metadata (Optional) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#tcga-data class=md-nav__link> <span class=md-ellipsis> TCGA Data </span> </a> <nav class=md-nav aria-label="TCGA Data"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#downloading-tcga-data class=md-nav__link> <span class=md-ellipsis> Downloading TCGA Data </span> </a> </li> <li class=md-nav__item> <a href=#tcga-data-structure class=md-nav__link> <span class=md-ellipsis> TCGA Data Structure </span> </a> </li> <li class=md-nav__item> <a href=#tcga-sample-ids class=md-nav__link> <span class=md-ellipsis> TCGA Sample IDs </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#custom-non-tcga-data class=md-nav__link> <span class=md-ellipsis> Custom (Non-TCGA) Data </span> </a> <nav class=md-nav aria-label="Custom (Non-TCGA) Data"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#preparing-your-own-data class=md-nav__link> <span class=md-ellipsis> Preparing Your Own Data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#data-preprocessing class=md-nav__link> <span class=md-ellipsis> Data Preprocessing </span> </a> <nav class=md-nav aria-label="Data Preprocessing"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#automatic-preprocessing class=md-nav__link> <span class=md-ellipsis> Automatic Preprocessing </span> </a> </li> <li class=md-nav__item> <a href=#preprocessing-parameters class=md-nav__link> <span class=md-ellipsis> Preprocessing Parameters </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#quality-control class=md-nav__link> <span class=md-ellipsis> Quality Control </span> </a> <nav class=md-nav aria-label="Quality Control"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#check-data-quality class=md-nav__link> <span class=md-ellipsis> Check Data Quality </span> </a> </li> <li class=md-nav__item> <a href=#sample-size-guidelines class=md-nav__link> <span class=md-ellipsis> Sample Size Guidelines </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#example-datasets class=md-nav__link> <span class=md-ellipsis> Example Datasets </span> </a> <nav class=md-nav aria-label="Example Datasets"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#kirc-kidney-renal-clear-cell-carcinoma class=md-nav__link> <span class=md-ellipsis> KIRC (Kidney Renal Clear Cell Carcinoma) </span> </a> </li> <li class=md-nav__item> <a href=#other-tcga-cancers class=md-nav__link> <span class=md-ellipsis> Other TCGA Cancers </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#sample-ids-dont-match class=md-nav__link> <span class=md-ellipsis> "Sample IDs don't match" </span> </a> </li> <li class=md-nav__item> <a href=#too-few-genes-after-filtering class=md-nav__link> <span class=md-ellipsis> "Too few genes after filtering" </span> </a> </li> <li class=md-nav__item> <a href=#data-looks-weird class=md-nav__link> <span class=md-ellipsis> "Data looks weird" </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> <span class=md-ellipsis> Tutorials </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../pretrained-models/ class=md-nav__link> <span class=md-ellipsis> Using Pretrained Models </span> </a> </li> <li class=md-nav__item> <a href=../complete-pipeline/ class=md-nav__link> <span class=md-ellipsis> Complete Pipeline </span> </a> </li> <li class=md-nav__item> <a href=../step1-data-processing/ class=md-nav__link> <span class=md-ellipsis> Step 1 - Data Processing </span> </a> </li> <li class=md-nav__item> <a href=../step2-vae-training/ class=md-nav__link> <span class=md-ellipsis> Step 2 - VAE Training </span> </a> </li> <li class=md-nav__item> <a href=../step3-reconstruction/ class=md-nav__link> <span class=md-ellipsis> Step 3 - Reconstruction </span> </a> </li> <li class=md-nav__item> <a href=../step4-trajectories/ class=md-nav__link> <span class=md-ellipsis> Step 4 - Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../step5-classification/ class=md-nav__link> <span class=md-ellipsis> Step 5 - Classification </span> </a> </li> <li class=md-nav__item> <a href=../step6-enrichment/ class=md-nav__link> <span class=md-ellipsis> Step 6 - Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../visualization/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> Reproducibility </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Reproducibility </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../reproducibility/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../reproducibility/requirements/ class=md-nav__link> <span class=md-ellipsis> System Requirements </span> </a> </li> <li class=md-nav__item> <a href=../../reproducibility/data-preparation/ class=md-nav__link> <span class=md-ellipsis> Data Preparation </span> </a> </li> <li class=md-nav__item> <a href=../../reproducibility/pipeline/ class=md-nav__link> <span class=md-ellipsis> Running the Pipeline </span> </a> </li> <li class=md-nav__item> <a href=../../reproducibility/results/ class=md-nav__link> <span class=md-ellipsis> Expected Results </span> </a> </li> <li class=md-nav__item> <a href=../../reproducibility/troubleshooting/ class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../api/config/ class=md-nav__link> <span class=md-ellipsis> Configuration </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5_3> <label class=md-nav__link for=__nav_5_3 id=__nav_5_3_label tabindex=0> <span class=md-ellipsis> Data Processing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_3_label aria-expanded=false> <label class=md-nav__title for=__nav_5_3> <span class="md-nav__icon md-icon"></span> Data Processing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/dataset/ class=md-nav__link> <span class=md-ellipsis> Dataset </span> </a> </li> <li class=md-nav__item> <a href=../../api/features/ class=md-nav__link> <span class=md-ellipsis> Features </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5_4> <label class=md-nav__link for=__nav_5_4 id=__nav_5_4_label tabindex=0> <span class=md-ellipsis> Modeling </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_4_label aria-expanded=false> <label class=md-nav__title for=__nav_5_4> <span class="md-nav__icon md-icon"></span> Modeling </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/models/ class=md-nav__link> <span class=md-ellipsis> VAE Models </span> </a> </li> <li class=md-nav__item> <a href=../../api/training/ class=md-nav__link> <span class=md-ellipsis> Training </span> </a> </li> <li class=md-nav__item> <a href=../../api/prediction/ class=md-nav__link> <span class=md-ellipsis> Prediction </span> </a> </li> <li class=md-nav__item> <a href=../../api/trajectories/ class=md-nav__link> <span class=md-ellipsis> Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../../api/classification/ class=md-nav__link> <span class=md-ellipsis> Classification </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../api/enrichment/ class=md-nav__link> <span class=md-ellipsis> Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../../api/plots/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> <li class=md-nav__item> <a href=../../api/utils/ class=md-nav__link> <span class=md-ellipsis> Utilities </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex=0> <span class=md-ellipsis> Advanced Topics </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> Advanced Topics </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../advanced/architecture/ class=md-nav__link> <span class=md-ellipsis> Architecture </span> </a> </li> <li class=md-nav__item> <a href=../../TRAJECTORIES/ class=md-nav__link> <span class=md-ellipsis> Trajectory Generation Details </span> </a> </li> <li class=md-nav__item> <a href=../../CLASSIFICATION/ class=md-nav__link> <span class=md-ellipsis> Classification Details </span> </a> </li> <li class=md-nav__item> <a href=../../ENRICHMENT_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> Enrichment Analysis Details </span> </a> </li> <li class=md-nav__item> <a href=../../ENRICHMENT_QUICK_REF/ class=md-nav__link> <span class=md-ellipsis> Enrichment Quick Reference </span> </a> </li> <li class=md-nav__item> <a href=../../PATHWAY_HEATMAP/ class=md-nav__link> <span class=md-ellipsis> Pathway Heatmap Analysis </span> </a> </li> <li class=md-nav__item> <a href=../../GSEA_INSTALLATION/ class=md-nav__link> <span class=md-ellipsis> GSEA Installation </span> </a> </li> <li class=md-nav__item> <a href=../../R_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> R Analysis Scripts </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/r-integration/ class=md-nav__link> <span class=md-ellipsis> R Integration </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/custom-models/ class=md-nav__link> <span class=md-ellipsis> Custom Models </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/performance/ class=md-nav__link> <span class=md-ellipsis> Performance Tuning </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/gpu/ class=md-nav__link> <span class=md-ellipsis> GPU Acceleration </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex=0> <span class=md-ellipsis> Contributing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Contributing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../contributing/guidelines/ class=md-nav__link> <span class=md-ellipsis> Guidelines </span> </a> </li> <li class=md-nav__item> <a href=../../contributing/development/ class=md-nav__link> <span class=md-ellipsis> Development Setup </span> </a> </li> <li class=md-nav__item> <a href=../../contributing/code-style/ class=md-nav__link> <span class=md-ellipsis> Code Style </span> </a> </li> <li class=md-nav__item> <a href=../../contributing/testing/ class=md-nav__link> <span class=md-ellipsis> Testing </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex=0> <span class=md-ellipsis> Citation & License </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> Citation & License </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../citation/ class=md-nav__link> <span class=md-ellipsis> How to Cite </span> </a> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> <li class=md-nav__item> <a href=../../acknowledgments/ class=md-nav__link> <span class=md-ellipsis> Acknowledgments </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#input-data-formats class=md-nav__link> <span class=md-ellipsis> Input Data Formats </span> </a> <nav class=md-nav aria-label="Input Data Formats"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-gene-expression-data class=md-nav__link> <span class=md-ellipsis> 1. Gene Expression Data </span> </a> </li> <li class=md-nav__item> <a href=#2-clinical-data class=md-nav__link> <span class=md-ellipsis> 2. Clinical Data </span> </a> </li> <li class=md-nav__item> <a href=#3-sample-metadata-optional class=md-nav__link> <span class=md-ellipsis> 3. Sample Metadata (Optional) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#tcga-data class=md-nav__link> <span class=md-ellipsis> TCGA Data </span> </a> <nav class=md-nav aria-label="TCGA Data"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#downloading-tcga-data class=md-nav__link> <span class=md-ellipsis> Downloading TCGA Data </span> </a> </li> <li class=md-nav__item> <a href=#tcga-data-structure class=md-nav__link> <span class=md-ellipsis> TCGA Data Structure </span> </a> </li> <li class=md-nav__item> <a href=#tcga-sample-ids class=md-nav__link> <span class=md-ellipsis> TCGA Sample IDs </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#custom-non-tcga-data class=md-nav__link> <span class=md-ellipsis> Custom (Non-TCGA) Data </span> </a> <nav class=md-nav aria-label="Custom (Non-TCGA) Data"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#preparing-your-own-data class=md-nav__link> <span class=md-ellipsis> Preparing Your Own Data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#data-preprocessing class=md-nav__link> <span class=md-ellipsis> Data Preprocessing </span> </a> <nav class=md-nav aria-label="Data Preprocessing"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#automatic-preprocessing class=md-nav__link> <span class=md-ellipsis> Automatic Preprocessing </span> </a> </li> <li class=md-nav__item> <a href=#preprocessing-parameters class=md-nav__link> <span class=md-ellipsis> Preprocessing Parameters </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#quality-control class=md-nav__link> <span class=md-ellipsis> Quality Control </span> </a> <nav class=md-nav aria-label="Quality Control"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#check-data-quality class=md-nav__link> <span class=md-ellipsis> Check Data Quality </span> </a> </li> <li class=md-nav__item> <a href=#sample-size-guidelines class=md-nav__link> <span class=md-ellipsis> Sample Size Guidelines </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#example-datasets class=md-nav__link> <span class=md-ellipsis> Example Datasets </span> </a> <nav class=md-nav aria-label="Example Datasets"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#kirc-kidney-renal-clear-cell-carcinoma class=md-nav__link> <span class=md-ellipsis> KIRC (Kidney Renal Clear Cell Carcinoma) </span> </a> </li> <li class=md-nav__item> <a href=#other-tcga-cancers class=md-nav__link> <span class=md-ellipsis> Other TCGA Cancers </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#sample-ids-dont-match class=md-nav__link> <span class=md-ellipsis> "Sample IDs don't match" </span> </a> </li> <li class=md-nav__item> <a href=#too-few-genes-after-filtering class=md-nav__link> <span class=md-ellipsis> "Too few genes after filtering" </span> </a> </li> <li class=md-nav__item> <a href=#data-looks-weird class=md-nav__link> <span class=md-ellipsis> "Data looks weird" </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=data-requirements>Data Requirements<a class=headerlink href=#data-requirements title="Permanent link">&para;</a></h1> <p>Understanding the input data formats and requirements for <code>renalprog</code>.</p> <h2 id=overview>Overview<a class=headerlink href=#overview title="Permanent link">&para;</a></h2> <p><code>renalprog</code> is designed to work with <strong>gene expression data</strong> and <strong>clinical annotations</strong>. The primary use case is TCGA (The Cancer Genome Atlas) data, but the package can work with any properly formatted RNA-seq or microarray data.</p> <h2 id=input-data-formats>Input Data Formats<a class=headerlink href=#input-data-formats title="Permanent link">&para;</a></h2> <h3 id=1-gene-expression-data>1. Gene Expression Data<a class=headerlink href=#1-gene-expression-data title="Permanent link">&para;</a></h3> <p><strong>Format</strong>: CSV or TSV file with genes as columns and samples as rows (or vice versa).</p> <p><strong>Example</strong> (<code>rnaseq.csv</code>): <div class="language-text highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>,Gene1,Gene2,Gene3,...,GeneN
</span><span id=__span-0-2><a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>Sample1,12.5,8.3,0.1,...,5.7
</span><span id=__span-0-3><a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a>Sample2,13.1,7.9,0.0,...,6.2
</span><span id=__span-0-4><a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a>Sample3,11.8,9.2,0.3,...,5.1
</span><span id=__span-0-5><a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a>...
</span></code></pre></div></p> <p><strong>Requirements</strong>:</p> <ul> <li><strong>Samples</strong>: Minimum 50, recommended 200+</li> <li><strong>Genes</strong>: Minimum 1000, recommended 5000+</li> <li><strong>Values</strong>: </li> <li>Can be raw counts, TPM, FPKM, or log-transformed</li> <li>Will be preprocessed by the pipeline</li> <li>Must be numeric (no missing values in final data)</li> <li><strong>Index</strong>: Sample IDs as row index OR column names</li> <li><strong>Columns</strong>: Gene IDs or symbols</li> </ul> <p><strong>Supported Formats</strong>:</p> <ul> <li>✅ Samples × Genes (preferred)</li> <li>✅ Genes × Samples (will be transposed)</li> <li>✅ Log-transformed or linear scale</li> <li>✅ Normalized or unnormalized (pipeline handles normalization)</li> </ul> <h3 id=2-clinical-data>2. Clinical Data<a class=headerlink href=#2-clinical-data title="Permanent link">&para;</a></h3> <p><strong>Format</strong>: CSV or TSV file with samples as rows and clinical variables as columns.</p> <p><strong>Example</strong> (<code>clinical.csv</code>): <div class="language-text highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a>sample_id,ajcc_pathologic_tumor_stage,age,gender,survival_days
</span><span id=__span-1-2><a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>Sample1,Stage I,65,M,1825
</span><span id=__span-1-3><a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>Sample2,Stage IIIA,72,F,730
</span><span id=__span-1-4><a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a>Sample3,Stage II,58,M,2190
</span><span id=__span-1-5><a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a>...
</span></code></pre></div></p> <p><strong>Required Columns</strong>:</p> <ul> <li><code>sample_id</code>: Must match gene expression sample IDs</li> <li><code>ajcc_pathologic_tumor_stage</code>: Cancer stage (or similar staging variable)</li> <li>Examples: "Stage I", "Stage II", "Stage IIIA", "Stage IV"</li> <li>Or: "T1N0M0", "T2N1M0", etc.</li> </ul> <p><strong>Optional Columns</strong> (highly recommended):</p> <ul> <li><code>age</code>: Patient age at diagnosis</li> <li><code>gender</code> or <code>sex</code>: M/F or Male/Female</li> <li><code>survival_days</code>: Days of follow-up</li> <li><code>vital_status</code>: Alive/Dead or 0/1</li> <li><code>tumor_grade</code>: G1, G2, G3, G4</li> <li><code>histological_type</code>: Tumor histology</li> </ul> <p><strong>Stage Format</strong>:</p> <p>The package can handle various stage formats:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1># AJCC staging</span>
</span><span id=__span-2-2><a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=s2>&quot;Stage I&quot;</span><span class=p>,</span> <span class=s2>&quot;Stage IA&quot;</span><span class=p>,</span> <span class=s2>&quot;Stage IB&quot;</span>
</span><span id=__span-2-3><a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a><span class=s2>&quot;Stage II&quot;</span><span class=p>,</span> <span class=s2>&quot;Stage IIA&quot;</span><span class=p>,</span> <span class=s2>&quot;Stage IIB&quot;</span>
</span><span id=__span-2-4><a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a><span class=s2>&quot;Stage III&quot;</span><span class=p>,</span> <span class=s2>&quot;Stage IIIA&quot;</span><span class=p>,</span> <span class=s2>&quot;Stage IIIB&quot;</span><span class=p>,</span> <span class=s2>&quot;Stage IIIC&quot;</span>
</span><span id=__span-2-5><a id=__codelineno-2-5 name=__codelineno-2-5 href=#__codelineno-2-5></a><span class=s2>&quot;Stage IV&quot;</span><span class=p>,</span> <span class=s2>&quot;Stage IVA&quot;</span><span class=p>,</span> <span class=s2>&quot;Stage IVB&quot;</span>
</span><span id=__span-2-6><a id=__codelineno-2-6 name=__codelineno-2-6 href=#__codelineno-2-6></a>
</span><span id=__span-2-7><a id=__codelineno-2-7 name=__codelineno-2-7 href=#__codelineno-2-7></a><span class=c1># TNM staging</span>
</span><span id=__span-2-8><a id=__codelineno-2-8 name=__codelineno-2-8 href=#__codelineno-2-8></a><span class=s2>&quot;T1N0M0&quot;</span><span class=p>,</span> <span class=s2>&quot;T2N1M0&quot;</span><span class=p>,</span> <span class=s2>&quot;T3N2M1&quot;</span>
</span><span id=__span-2-9><a id=__codelineno-2-9 name=__codelineno-2-9 href=#__codelineno-2-9></a>
</span><span id=__span-2-10><a id=__codelineno-2-10 name=__codelineno-2-10 href=#__codelineno-2-10></a><span class=c1># Simplified</span>
</span><span id=__span-2-11><a id=__codelineno-2-11 name=__codelineno-2-11 href=#__codelineno-2-11></a><span class=s2>&quot;Early&quot;</span><span class=p>,</span> <span class=s2>&quot;Late&quot;</span>
</span><span id=__span-2-12><a id=__codelineno-2-12 name=__codelineno-2-12 href=#__codelineno-2-12></a><span class=s2>&quot;Low&quot;</span><span class=p>,</span> <span class=s2>&quot;High&quot;</span>
</span></code></pre></div> <p>The pipeline automatically groups stages:</p> <ul> <li><strong>Early</strong>: Stage I, II</li> <li><strong>Late</strong>: Stage III, IV</li> </ul> <h3 id=3-sample-metadata-optional>3. Sample Metadata (Optional)<a class=headerlink href=#3-sample-metadata-optional title="Permanent link">&para;</a></h3> <p><strong>Format</strong>: CSV file with additional sample annotations.</p> <p><strong>Example</strong> (<code>phenotype.csv</code>): <div class="language-text highlight"><pre><span></span><code><span id=__span-3-1><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a>sample_id,sample_type,tissue_source_site,batch,platform
</span><span id=__span-3-2><a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a>Sample1,Primary Tumor,TCGA-XX,Batch1,IlluminaHiSeq
</span><span id=__span-3-3><a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a>Sample2,Primary Tumor,TCGA-YY,Batch1,IlluminaHiSeq
</span><span id=__span-3-4><a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a>Sample3,Primary Tumor,TCGA-ZZ,Batch2,IlluminaHiSeq
</span><span id=__span-3-5><a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a>...
</span></code></pre></div></p> <p>Useful for: - Batch effect analysis - Sample type filtering - Quality control</p> <h2 id=tcga-data>TCGA Data<a class=headerlink href=#tcga-data title="Permanent link">&para;</a></h2> <h3 id=downloading-tcga-data>Downloading TCGA Data<a class=headerlink href=#downloading-tcga-data title="Permanent link">&para;</a></h3> <p>The package includes utilities to download TCGA data from UCSC Xena Browser:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-4-1><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=kn>from</span><span class=w> </span><span class=nn>renalprog</span><span class=w> </span><span class=kn>import</span> <span class=n>dataset</span>
</span><span id=__span-4-2><a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>
</span><span id=__span-4-3><a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a><span class=c1># Download KIRC data</span>
</span><span id=__span-4-4><a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a><span class=n>rnaseq</span><span class=p>,</span> <span class=n>clinical</span><span class=p>,</span> <span class=n>pheno</span> <span class=o>=</span> <span class=n>dataset</span><span class=o>.</span><span class=n>download_data</span><span class=p>(</span>
</span><span id=__span-4-5><a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a>    <span class=n>destination</span><span class=o>=</span><span class=s1>&#39;data/raw&#39;</span><span class=p>,</span>
</span><span id=__span-4-6><a id=__codelineno-4-6 name=__codelineno-4-6 href=#__codelineno-4-6></a>    <span class=n>cancer_type</span><span class=o>=</span><span class=s1>&#39;KIRC&#39;</span><span class=p>,</span>
</span><span id=__span-4-7><a id=__codelineno-4-7 name=__codelineno-4-7 href=#__codelineno-4-7></a>    <span class=n>remove_gz</span><span class=o>=</span><span class=kc>True</span>
</span><span id=__span-4-8><a id=__codelineno-4-8 name=__codelineno-4-8 href=#__codelineno-4-8></a><span class=p>)</span>
</span></code></pre></div> <h3 id=tcga-data-structure>TCGA Data Structure<a class=headerlink href=#tcga-data-structure title="Permanent link">&para;</a></h3> <p>TCGA data includes:</p> <ol> <li><strong>Gene Expression</strong> (20,531 genes initially)</li> <li>Platform: Illumina HiSeq</li> <li>Normalization: RSEM + upper quartile normalized</li> <li> <p>File: <code>EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena</code></p> </li> <li> <p><strong>Clinical Data</strong> (survival and staging)</p> </li> <li>Curated clinical annotations</li> <li> <p>File: <code>Survival_SupplementalTable_S1_20171025_xena_sp</code></p> </li> <li> <p><strong>Phenotype</strong> (sample metadata)</p> </li> <li>Batch information</li> <li>Sample types (primary tumor, normal, metastatic)</li> <li>File: <code>TCGA_phenotype_denseDataOnlyDownload.tsv</code></li> </ol> <h3 id=tcga-sample-ids>TCGA Sample IDs<a class=headerlink href=#tcga-sample-ids title="Permanent link">&para;</a></h3> <p>TCGA uses barcodes like: <div class="language-text highlight"><pre><span></span><code><span id=__span-5-1><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a>TCGA-A3-3308-01A-01R-0864-07
</span><span id=__span-5-2><a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a>│    │  │    │ │ │  │    │
</span><span id=__span-5-3><a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a>│    │  │    │ │ │  │    └─ Sequencing center
</span><span id=__span-5-4><a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a>│    │  │    │ │ │  └────── Plate
</span><span id=__span-5-5><a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a>│    │  │    │ │ └─────────  RNA/DNA type
</span><span id=__span-5-6><a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a>│    │  │    │ └───────────  Aliquot
</span><span id=__span-5-7><a id=__codelineno-5-7 name=__codelineno-5-7 href=#__codelineno-5-7></a>│    │  │    └─────────────  Sample type (01=Primary, 11=Normal)
</span><span id=__span-5-8><a id=__codelineno-5-8 name=__codelineno-5-8 href=#__codelineno-5-8></a>│    │  └──────────────────  Patient ID
</span><span id=__span-5-9><a id=__codelineno-5-9 name=__codelineno-5-9 href=#__codelineno-5-9></a>│    └─────────────────────  Tissue source site
</span><span id=__span-5-10><a id=__codelineno-5-10 name=__codelineno-5-10 href=#__codelineno-5-10></a>└──────────────────────────  Project (TCGA)
</span></code></pre></div></p> <p>The package handles these automatically.</p> <h2 id=custom-non-tcga-data>Custom (Non-TCGA) Data<a class=headerlink href=#custom-non-tcga-data title="Permanent link">&para;</a></h2> <h3 id=preparing-your-own-data>Preparing Your Own Data<a class=headerlink href=#preparing-your-own-data title="Permanent link">&para;</a></h3> <p>If you have your own gene expression data:</p> <h4 id=step-1-format-gene-expression>Step 1: Format Gene Expression<a class=headerlink href=#step-1-format-gene-expression title="Permanent link">&para;</a></h4> <div class="language-python highlight"><pre><span></span><code><span id=__span-6-1><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=kn>import</span><span class=w> </span><span class=nn>pandas</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>pd</span>
</span><span id=__span-6-2><a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a>
</span><span id=__span-6-3><a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a><span class=c1># Load your data</span>
</span><span id=__span-6-4><a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a><span class=c1># Assume you have a samples × genes matrix</span>
</span><span id=__span-6-5><a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a><span class=n>your_data</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=s1>&#39;your_rnaseq_data.csv&#39;</span><span class=p>,</span> <span class=n>index_col</span><span class=o>=</span><span class=mi>0</span><span class=p>)</span>
</span><span id=__span-6-6><a id=__codelineno-6-6 name=__codelineno-6-6 href=#__codelineno-6-6></a>
</span><span id=__span-6-7><a id=__codelineno-6-7 name=__codelineno-6-7 href=#__codelineno-6-7></a><span class=c1># Ensure proper format</span>
</span><span id=__span-6-8><a id=__codelineno-6-8 name=__codelineno-6-8 href=#__codelineno-6-8></a><span class=k>assert</span> <span class=n>your_data</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>0</span><span class=p>]</span> <span class=o>&lt;</span> <span class=n>your_data</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>1</span><span class=p>],</span> <span class=s2>&quot;Transpose if needed&quot;</span>
</span><span id=__span-6-9><a id=__codelineno-6-9 name=__codelineno-6-9 href=#__codelineno-6-9></a><span class=c1># Should have more genes than samples</span>
</span><span id=__span-6-10><a id=__codelineno-6-10 name=__codelineno-6-10 href=#__codelineno-6-10></a>
</span><span id=__span-6-11><a id=__codelineno-6-11 name=__codelineno-6-11 href=#__codelineno-6-11></a><span class=c1># Check for missing values</span>
</span><span id=__span-6-12><a id=__codelineno-6-12 name=__codelineno-6-12 href=#__codelineno-6-12></a><span class=k>assert</span> <span class=ow>not</span> <span class=n>your_data</span><span class=o>.</span><span class=n>isnull</span><span class=p>()</span><span class=o>.</span><span class=n>any</span><span class=p>()</span><span class=o>.</span><span class=n>any</span><span class=p>(),</span> <span class=s2>&quot;Remove missing values&quot;</span>
</span><span id=__span-6-13><a id=__codelineno-6-13 name=__codelineno-6-13 href=#__codelineno-6-13></a>
</span><span id=__span-6-14><a id=__codelineno-6-14 name=__codelineno-6-14 href=#__codelineno-6-14></a><span class=c1># Save in standard format</span>
</span><span id=__span-6-15><a id=__codelineno-6-15 name=__codelineno-6-15 href=#__codelineno-6-15></a><span class=n>your_data</span><span class=o>.</span><span class=n>to_csv</span><span class=p>(</span><span class=s1>&#39;data/raw/your_data_formatted.csv&#39;</span><span class=p>)</span>
</span></code></pre></div> <h4 id=step-2-format-clinical-data>Step 2: Format Clinical Data<a class=headerlink href=#step-2-format-clinical-data title="Permanent link">&para;</a></h4> <div class="language-python highlight"><pre><span></span><code><span id=__span-7-1><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># Create clinical file</span>
</span><span id=__span-7-2><a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=n>clinical</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>DataFrame</span><span class=p>({</span>
</span><span id=__span-7-3><a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a>    <span class=s1>&#39;sample_id&#39;</span><span class=p>:</span> <span class=n>your_data</span><span class=o>.</span><span class=n>index</span><span class=p>,</span>
</span><span id=__span-7-4><a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a>    <span class=s1>&#39;ajcc_pathologic_tumor_stage&#39;</span><span class=p>:</span> <span class=n>your_stages</span><span class=p>,</span>  <span class=c1># e.g., [&quot;Stage I&quot;, &quot;Stage III&quot;, ...]</span>
</span><span id=__span-7-5><a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a>    <span class=s1>&#39;age&#39;</span><span class=p>:</span> <span class=n>your_ages</span><span class=p>,</span>
</span><span id=__span-7-6><a id=__codelineno-7-6 name=__codelineno-7-6 href=#__codelineno-7-6></a>    <span class=s1>&#39;gender&#39;</span><span class=p>:</span> <span class=n>your_gender</span>
</span><span id=__span-7-7><a id=__codelineno-7-7 name=__codelineno-7-7 href=#__codelineno-7-7></a><span class=p>})</span>
</span><span id=__span-7-8><a id=__codelineno-7-8 name=__codelineno-7-8 href=#__codelineno-7-8></a>
</span><span id=__span-7-9><a id=__codelineno-7-9 name=__codelineno-7-9 href=#__codelineno-7-9></a><span class=c1># Map to early/late if needed</span>
</span><span id=__span-7-10><a id=__codelineno-7-10 name=__codelineno-7-10 href=#__codelineno-7-10></a><span class=n>stage_map</span> <span class=o>=</span> <span class=p>{</span>
</span><span id=__span-7-11><a id=__codelineno-7-11 name=__codelineno-7-11 href=#__codelineno-7-11></a>    <span class=s1>&#39;Stage I&#39;</span><span class=p>:</span> <span class=s1>&#39;Early&#39;</span><span class=p>,</span>
</span><span id=__span-7-12><a id=__codelineno-7-12 name=__codelineno-7-12 href=#__codelineno-7-12></a>    <span class=s1>&#39;Stage II&#39;</span><span class=p>:</span> <span class=s1>&#39;Early&#39;</span><span class=p>,</span>
</span><span id=__span-7-13><a id=__codelineno-7-13 name=__codelineno-7-13 href=#__codelineno-7-13></a>    <span class=s1>&#39;Stage III&#39;</span><span class=p>:</span> <span class=s1>&#39;Late&#39;</span><span class=p>,</span>
</span><span id=__span-7-14><a id=__codelineno-7-14 name=__codelineno-7-14 href=#__codelineno-7-14></a>    <span class=s1>&#39;Stage IV&#39;</span><span class=p>:</span> <span class=s1>&#39;Late&#39;</span>
</span><span id=__span-7-15><a id=__codelineno-7-15 name=__codelineno-7-15 href=#__codelineno-7-15></a><span class=p>}</span>
</span><span id=__span-7-16><a id=__codelineno-7-16 name=__codelineno-7-16 href=#__codelineno-7-16></a><span class=n>clinical</span><span class=p>[</span><span class=s1>&#39;stage_binary&#39;</span><span class=p>]</span> <span class=o>=</span> <span class=n>clinical</span><span class=p>[</span><span class=s1>&#39;ajcc_pathologic_tumor_stage&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>map</span><span class=p>(</span><span class=n>stage_map</span><span class=p>)</span>
</span><span id=__span-7-17><a id=__codelineno-7-17 name=__codelineno-7-17 href=#__codelineno-7-17></a>
</span><span id=__span-7-18><a id=__codelineno-7-18 name=__codelineno-7-18 href=#__codelineno-7-18></a><span class=n>clinical</span><span class=o>.</span><span class=n>to_csv</span><span class=p>(</span><span class=s1>&#39;data/raw/your_clinical_formatted.csv&#39;</span><span class=p>,</span> <span class=n>index</span><span class=o>=</span><span class=kc>False</span><span class=p>)</span>
</span></code></pre></div> <h4 id=step-3-validate>Step 3: Validate<a class=headerlink href=#step-3-validate title="Permanent link">&para;</a></h4> <div class="language-python highlight"><pre><span></span><code><span id=__span-8-1><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=kn>from</span><span class=w> </span><span class=nn>renalprog</span><span class=w> </span><span class=kn>import</span> <span class=n>dataset</span>
</span><span id=__span-8-2><a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a>
</span><span id=__span-8-3><a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a><span class=c1># Try loading</span>
</span><span id=__span-8-4><a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=n>rnaseq</span> <span class=o>=</span> <span class=n>dataset</span><span class=o>.</span><span class=n>load_data</span><span class=p>(</span><span class=s1>&#39;data/raw/your_data_formatted.csv&#39;</span><span class=p>)</span>
</span><span id=__span-8-5><a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a><span class=n>clinical</span> <span class=o>=</span> <span class=n>dataset</span><span class=o>.</span><span class=n>load_data</span><span class=p>(</span><span class=s1>&#39;data/raw/your_clinical_formatted.csv&#39;</span><span class=p>)</span>
</span><span id=__span-8-6><a id=__codelineno-8-6 name=__codelineno-8-6 href=#__codelineno-8-6></a>
</span><span id=__span-8-7><a id=__codelineno-8-7 name=__codelineno-8-7 href=#__codelineno-8-7></a><span class=c1># Check alignment</span>
</span><span id=__span-8-8><a id=__codelineno-8-8 name=__codelineno-8-8 href=#__codelineno-8-8></a><span class=k>assert</span> <span class=nb>set</span><span class=p>(</span><span class=n>rnaseq</span><span class=o>.</span><span class=n>index</span><span class=p>)</span> <span class=o>==</span> <span class=nb>set</span><span class=p>(</span><span class=n>clinical</span><span class=p>[</span><span class=s1>&#39;sample_id&#39;</span><span class=p>]),</span> <span class=s2>&quot;Sample IDs must match!&quot;</span>
</span><span id=__span-8-9><a id=__codelineno-8-9 name=__codelineno-8-9 href=#__codelineno-8-9></a>
</span><span id=__span-8-10><a id=__codelineno-8-10 name=__codelineno-8-10 href=#__codelineno-8-10></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Samples: </span><span class=si>{</span><span class=n>rnaseq</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>0</span><span class=p>]</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-8-11><a id=__codelineno-8-11 name=__codelineno-8-11 href=#__codelineno-8-11></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Genes: </span><span class=si>{</span><span class=n>rnaseq</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>1</span><span class=p>]</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-8-12><a id=__codelineno-8-12 name=__codelineno-8-12 href=#__codelineno-8-12></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Stages: </span><span class=si>{</span><span class=n>clinical</span><span class=p>[</span><span class=s1>&#39;ajcc_pathologic_tumor_stage&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>value_counts</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span></code></pre></div> <h2 id=data-preprocessing>Data Preprocessing<a class=headerlink href=#data-preprocessing title="Permanent link">&para;</a></h2> <h3 id=automatic-preprocessing>Automatic Preprocessing<a class=headerlink href=#automatic-preprocessing title="Permanent link">&para;</a></h3> <p>The pipeline automatically:</p> <ol> <li><strong>Filters low-expression genes</strong></li> <li>Removes genes with low mean expression</li> <li>Removes genes with low variance</li> <li> <p>Keeps genes expressed in ≥20% of samples</p> </li> <li> <p><strong>Detects outliers</strong></p> </li> <li>Uses Mahalanobis distance</li> <li>Removes samples &gt;3 SD from mean</li> <li> <p>Configurable significance level</p> </li> <li> <p><strong>Normalizes</strong> (optional)</p> </li> <li>Log transformation</li> <li>Z-score normalization</li> <li>Quantile normalization</li> </ol> <h3 id=preprocessing-parameters>Preprocessing Parameters<a class=headerlink href=#preprocessing-parameters title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-9-1><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=kn>from</span><span class=w> </span><span class=nn>renalprog</span><span class=w> </span><span class=kn>import</span> <span class=n>features</span>
</span><span id=__span-9-2><a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>
</span><span id=__span-9-3><a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=n>processed</span><span class=p>,</span> <span class=n>info</span> <span class=o>=</span> <span class=n>features</span><span class=o>.</span><span class=n>preprocess_rnaseq</span><span class=p>(</span>
</span><span id=__span-9-4><a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a>    <span class=n>data</span><span class=o>=</span><span class=n>rnaseq</span><span class=p>,</span>
</span><span id=__span-9-5><a id=__codelineno-9-5 name=__codelineno-9-5 href=#__codelineno-9-5></a>    <span class=n>filter_expression</span><span class=o>=</span><span class=kc>True</span><span class=p>,</span>
</span><span id=__span-9-6><a id=__codelineno-9-6 name=__codelineno-9-6 href=#__codelineno-9-6></a>    <span class=n>mean_threshold</span><span class=o>=</span><span class=mf>0.5</span><span class=p>,</span>        <span class=c1># Minimum mean expression</span>
</span><span id=__span-9-7><a id=__codelineno-9-7 name=__codelineno-9-7 href=#__codelineno-9-7></a>    <span class=n>var_threshold</span><span class=o>=</span><span class=mf>0.5</span><span class=p>,</span>         <span class=c1># Minimum variance</span>
</span><span id=__span-9-8><a id=__codelineno-9-8 name=__codelineno-9-8 href=#__codelineno-9-8></a>    <span class=n>min_sample_fraction</span><span class=o>=</span><span class=mf>0.2</span><span class=p>,</span>   <span class=c1># Gene must be in ≥20% samples</span>
</span><span id=__span-9-9><a id=__codelineno-9-9 name=__codelineno-9-9 href=#__codelineno-9-9></a>    <span class=n>detect_outliers</span><span class=o>=</span><span class=kc>True</span><span class=p>,</span>
</span><span id=__span-9-10><a id=__codelineno-9-10 name=__codelineno-9-10 href=#__codelineno-9-10></a>    <span class=n>alpha</span><span class=o>=</span><span class=mf>0.05</span><span class=p>,</span>                <span class=c1># Outlier detection significance</span>
</span><span id=__span-9-11><a id=__codelineno-9-11 name=__codelineno-9-11 href=#__codelineno-9-11></a>    <span class=n>seed</span><span class=o>=</span><span class=mi>2023</span>
</span><span id=__span-9-12><a id=__codelineno-9-12 name=__codelineno-9-12 href=#__codelineno-9-12></a><span class=p>)</span>
</span><span id=__span-9-13><a id=__codelineno-9-13 name=__codelineno-9-13 href=#__codelineno-9-13></a>
</span><span id=__span-9-14><a id=__codelineno-9-14 name=__codelineno-9-14 href=#__codelineno-9-14></a><span class=c1># Check what was filtered</span>
</span><span id=__span-9-15><a id=__codelineno-9-15 name=__codelineno-9-15 href=#__codelineno-9-15></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Original: </span><span class=si>{</span><span class=n>rnaseq</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>1</span><span class=p>]</span><span class=si>}</span><span class=s2> genes&quot;</span><span class=p>)</span>
</span><span id=__span-9-16><a id=__codelineno-9-16 name=__codelineno-9-16 href=#__codelineno-9-16></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;After filtering: </span><span class=si>{</span><span class=n>processed</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>1</span><span class=p>]</span><span class=si>}</span><span class=s2> genes&quot;</span><span class=p>)</span>
</span><span id=__span-9-17><a id=__codelineno-9-17 name=__codelineno-9-17 href=#__codelineno-9-17></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Removed: </span><span class=si>{</span><span class=n>info</span><span class=p>[</span><span class=s1>&#39;n_outliers&#39;</span><span class=p>]</span><span class=si>}</span><span class=s2> outlier samples&quot;</span><span class=p>)</span>
</span></code></pre></div> <h2 id=quality-control>Quality Control<a class=headerlink href=#quality-control title="Permanent link">&para;</a></h2> <h3 id=check-data-quality>Check Data Quality<a class=headerlink href=#check-data-quality title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-10-1><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=kn>import</span><span class=w> </span><span class=nn>pandas</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>pd</span>
</span><span id=__span-10-2><a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=kn>import</span><span class=w> </span><span class=nn>numpy</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>np</span>
</span><span id=__span-10-3><a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a>
</span><span id=__span-10-4><a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=c1># Load data</span>
</span><span id=__span-10-5><a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a><span class=n>rnaseq</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=s1>&#39;data/processed/rnaseq.csv&#39;</span><span class=p>,</span> <span class=n>index_col</span><span class=o>=</span><span class=mi>0</span><span class=p>)</span>
</span><span id=__span-10-6><a id=__codelineno-10-6 name=__codelineno-10-6 href=#__codelineno-10-6></a>
</span><span id=__span-10-7><a id=__codelineno-10-7 name=__codelineno-10-7 href=#__codelineno-10-7></a><span class=c1># 1. Check for missing values</span>
</span><span id=__span-10-8><a id=__codelineno-10-8 name=__codelineno-10-8 href=#__codelineno-10-8></a><span class=n>missing</span> <span class=o>=</span> <span class=n>rnaseq</span><span class=o>.</span><span class=n>isnull</span><span class=p>()</span><span class=o>.</span><span class=n>sum</span><span class=p>()</span><span class=o>.</span><span class=n>sum</span><span class=p>()</span>
</span><span id=__span-10-9><a id=__codelineno-10-9 name=__codelineno-10-9 href=#__codelineno-10-9></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Missing values: </span><span class=si>{</span><span class=n>missing</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-10-10><a id=__codelineno-10-10 name=__codelineno-10-10 href=#__codelineno-10-10></a><span class=k>assert</span> <span class=n>missing</span> <span class=o>==</span> <span class=mi>0</span><span class=p>,</span> <span class=s2>&quot;Data contains missing values!&quot;</span>
</span><span id=__span-10-11><a id=__codelineno-10-11 name=__codelineno-10-11 href=#__codelineno-10-11></a>
</span><span id=__span-10-12><a id=__codelineno-10-12 name=__codelineno-10-12 href=#__codelineno-10-12></a><span class=c1># 2. Check for duplicates</span>
</span><span id=__span-10-13><a id=__codelineno-10-13 name=__codelineno-10-13 href=#__codelineno-10-13></a><span class=n>duplicates</span> <span class=o>=</span> <span class=n>rnaseq</span><span class=o>.</span><span class=n>index</span><span class=o>.</span><span class=n>duplicated</span><span class=p>()</span><span class=o>.</span><span class=n>sum</span><span class=p>()</span>
</span><span id=__span-10-14><a id=__codelineno-10-14 name=__codelineno-10-14 href=#__codelineno-10-14></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Duplicate samples: </span><span class=si>{</span><span class=n>duplicates</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-10-15><a id=__codelineno-10-15 name=__codelineno-10-15 href=#__codelineno-10-15></a><span class=k>assert</span> <span class=n>duplicates</span> <span class=o>==</span> <span class=mi>0</span><span class=p>,</span> <span class=s2>&quot;Duplicate samples found!&quot;</span>
</span><span id=__span-10-16><a id=__codelineno-10-16 name=__codelineno-10-16 href=#__codelineno-10-16></a>
</span><span id=__span-10-17><a id=__codelineno-10-17 name=__codelineno-10-17 href=#__codelineno-10-17></a><span class=c1># 3. Check value range</span>
</span><span id=__span-10-18><a id=__codelineno-10-18 name=__codelineno-10-18 href=#__codelineno-10-18></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Min value: </span><span class=si>{</span><span class=n>rnaseq</span><span class=o>.</span><span class=n>min</span><span class=p>()</span><span class=o>.</span><span class=n>min</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-10-19><a id=__codelineno-10-19 name=__codelineno-10-19 href=#__codelineno-10-19></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Max value: </span><span class=si>{</span><span class=n>rnaseq</span><span class=o>.</span><span class=n>max</span><span class=p>()</span><span class=o>.</span><span class=n>max</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-10-20><a id=__codelineno-10-20 name=__codelineno-10-20 href=#__codelineno-10-20></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Mean expression: </span><span class=si>{</span><span class=n>rnaseq</span><span class=o>.</span><span class=n>mean</span><span class=p>()</span><span class=o>.</span><span class=n>mean</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-10-21><a id=__codelineno-10-21 name=__codelineno-10-21 href=#__codelineno-10-21></a>
</span><span id=__span-10-22><a id=__codelineno-10-22 name=__codelineno-10-22 href=#__codelineno-10-22></a><span class=c1># 4. Check distribution</span>
</span><span id=__span-10-23><a id=__codelineno-10-23 name=__codelineno-10-23 href=#__codelineno-10-23></a><span class=kn>import</span><span class=w> </span><span class=nn>matplotlib.pyplot</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>plt</span>
</span><span id=__span-10-24><a id=__codelineno-10-24 name=__codelineno-10-24 href=#__codelineno-10-24></a>
</span><span id=__span-10-25><a id=__codelineno-10-25 name=__codelineno-10-25 href=#__codelineno-10-25></a><span class=n>fig</span><span class=p>,</span> <span class=n>axes</span> <span class=o>=</span> <span class=n>plt</span><span class=o>.</span><span class=n>subplots</span><span class=p>(</span><span class=mi>1</span><span class=p>,</span> <span class=mi>2</span><span class=p>,</span> <span class=n>figsize</span><span class=o>=</span><span class=p>(</span><span class=mi>12</span><span class=p>,</span> <span class=mi>4</span><span class=p>))</span>
</span><span id=__span-10-26><a id=__codelineno-10-26 name=__codelineno-10-26 href=#__codelineno-10-26></a>
</span><span id=__span-10-27><a id=__codelineno-10-27 name=__codelineno-10-27 href=#__codelineno-10-27></a><span class=c1># Overall distribution</span>
</span><span id=__span-10-28><a id=__codelineno-10-28 name=__codelineno-10-28 href=#__codelineno-10-28></a><span class=n>rnaseq</span><span class=o>.</span><span class=n>values</span><span class=o>.</span><span class=n>flatten</span><span class=p>()</span><span class=o>.</span><span class=n>hist</span><span class=p>(</span><span class=n>bins</span><span class=o>=</span><span class=mi>50</span><span class=p>,</span> <span class=n>ax</span><span class=o>=</span><span class=n>axes</span><span class=p>[</span><span class=mi>0</span><span class=p>])</span>
</span><span id=__span-10-29><a id=__codelineno-10-29 name=__codelineno-10-29 href=#__codelineno-10-29></a><span class=n>axes</span><span class=p>[</span><span class=mi>0</span><span class=p>]</span><span class=o>.</span><span class=n>set_title</span><span class=p>(</span><span class=s1>&#39;Expression Value Distribution&#39;</span><span class=p>)</span>
</span><span id=__span-10-30><a id=__codelineno-10-30 name=__codelineno-10-30 href=#__codelineno-10-30></a><span class=n>axes</span><span class=p>[</span><span class=mi>0</span><span class=p>]</span><span class=o>.</span><span class=n>set_xlabel</span><span class=p>(</span><span class=s1>&#39;Expression&#39;</span><span class=p>)</span>
</span><span id=__span-10-31><a id=__codelineno-10-31 name=__codelineno-10-31 href=#__codelineno-10-31></a>
</span><span id=__span-10-32><a id=__codelineno-10-32 name=__codelineno-10-32 href=#__codelineno-10-32></a><span class=c1># Per-sample means</span>
</span><span id=__span-10-33><a id=__codelineno-10-33 name=__codelineno-10-33 href=#__codelineno-10-33></a><span class=n>rnaseq</span><span class=o>.</span><span class=n>mean</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>hist</span><span class=p>(</span><span class=n>bins</span><span class=o>=</span><span class=mi>30</span><span class=p>,</span> <span class=n>ax</span><span class=o>=</span><span class=n>axes</span><span class=p>[</span><span class=mi>1</span><span class=p>])</span>
</span><span id=__span-10-34><a id=__codelineno-10-34 name=__codelineno-10-34 href=#__codelineno-10-34></a><span class=n>axes</span><span class=p>[</span><span class=mi>1</span><span class=p>]</span><span class=o>.</span><span class=n>set_title</span><span class=p>(</span><span class=s1>&#39;Per-Sample Mean Expression&#39;</span><span class=p>)</span>
</span><span id=__span-10-35><a id=__codelineno-10-35 name=__codelineno-10-35 href=#__codelineno-10-35></a><span class=n>axes</span><span class=p>[</span><span class=mi>1</span><span class=p>]</span><span class=o>.</span><span class=n>set_xlabel</span><span class=p>(</span><span class=s1>&#39;Mean Expression&#39;</span><span class=p>)</span>
</span><span id=__span-10-36><a id=__codelineno-10-36 name=__codelineno-10-36 href=#__codelineno-10-36></a>
</span><span id=__span-10-37><a id=__codelineno-10-37 name=__codelineno-10-37 href=#__codelineno-10-37></a><span class=n>plt</span><span class=o>.</span><span class=n>savefig</span><span class=p>(</span><span class=s1>&#39;reports/figures/data_quality.png&#39;</span><span class=p>)</span>
</span></code></pre></div> <h3 id=sample-size-guidelines>Sample Size Guidelines<a class=headerlink href=#sample-size-guidelines title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>Analysis Type</th> <th>Minimum Samples</th> <th>Recommended</th> <th>Optimal</th> </tr> </thead> <tbody> <tr> <td>VAE Training</td> <td>50</td> <td>200</td> <td>500+</td> </tr> <tr> <td>Classification</td> <td>100</td> <td>300</td> <td>1000+</td> </tr> <tr> <td>Trajectory Generation</td> <td>20 per stage</td> <td>50 per stage</td> <td>100+ per stage</td> </tr> <tr> <td>Enrichment Analysis</td> <td>N/A</td> <td>N/A</td> <td>N/A (uses trajectories)</td> </tr> </tbody> </table> <p><strong>Note</strong>: More samples generally improve results, but you need balanced stages:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-11-1><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=c1># Check stage balance</span>
</span><span id=__span-11-2><a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=n>clinical</span><span class=p>[</span><span class=s1>&#39;ajcc_pathologic_tumor_stage&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>value_counts</span><span class=p>()</span>
</span><span id=__span-11-3><a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a>
</span><span id=__span-11-4><a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a><span class=c1># Should have:</span>
</span><span id=__span-11-5><a id=__codelineno-11-5 name=__codelineno-11-5 href=#__codelineno-11-5></a><span class=c1># Early stages: ≥25% of samples</span>
</span><span id=__span-11-6><a id=__codelineno-11-6 name=__codelineno-11-6 href=#__codelineno-11-6></a><span class=c1># Late stages: ≥25% of samples</span>
</span></code></pre></div> <h2 id=example-datasets>Example Datasets<a class=headerlink href=#example-datasets title="Permanent link">&para;</a></h2> <h3 id=kirc-kidney-renal-clear-cell-carcinoma>KIRC (Kidney Renal Clear Cell Carcinoma)<a class=headerlink href=#kirc-kidney-renal-clear-cell-carcinoma title="Permanent link">&para;</a></h3> <p>The main dataset used in publication:</p> <ul> <li><strong>Samples</strong>: 534 (498 after filtering)</li> <li><strong>Genes</strong>: 20,531 (5,127 after filtering)</li> <li><strong>Stages</strong>: I (216), II (50), III (120), IV (80)</li> <li><strong>Source</strong>: TCGA via UCSC Xena</li> </ul> <h3 id=other-tcga-cancers>Other TCGA Cancers<a class=headerlink href=#other-tcga-cancers title="Permanent link">&para;</a></h3> <p>The pipeline works with other TCGA cancers:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-12-1><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=c1># Download different cancer type</span>
</span><span id=__span-12-2><a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=n>rnaseq</span><span class=p>,</span> <span class=n>clinical</span><span class=p>,</span> <span class=n>pheno</span> <span class=o>=</span> <span class=n>dataset</span><span class=o>.</span><span class=n>download_data</span><span class=p>(</span>
</span><span id=__span-12-3><a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a>    <span class=n>destination</span><span class=o>=</span><span class=s1>&#39;data/raw&#39;</span><span class=p>,</span>
</span><span id=__span-12-4><a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a>    <span class=n>cancer_type</span><span class=o>=</span><span class=s1>&#39;BRCA&#39;</span>  <span class=c1># or LUAD, COAD, etc.</span>
</span><span id=__span-12-5><a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a><span class=p>)</span>
</span></code></pre></div> <p>Supported: BRCA, LUAD, LUSC, COAD, STAD, LIHC, PRAD, etc.</p> <h2 id=troubleshooting>Troubleshooting<a class=headerlink href=#troubleshooting title="Permanent link">&para;</a></h2> <h3 id=sample-ids-dont-match>"Sample IDs don't match"<a class=headerlink href=#sample-ids-dont-match title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-13-1><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=c1># Find mismatches</span>
</span><span id=__span-13-2><a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=n>rnaseq_ids</span> <span class=o>=</span> <span class=nb>set</span><span class=p>(</span><span class=n>rnaseq</span><span class=o>.</span><span class=n>index</span><span class=p>)</span>
</span><span id=__span-13-3><a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a><span class=n>clinical_ids</span> <span class=o>=</span> <span class=nb>set</span><span class=p>(</span><span class=n>clinical</span><span class=p>[</span><span class=s1>&#39;sample_id&#39;</span><span class=p>])</span>
</span><span id=__span-13-4><a id=__codelineno-13-4 name=__codelineno-13-4 href=#__codelineno-13-4></a>
</span><span id=__span-13-5><a id=__codelineno-13-5 name=__codelineno-13-5 href=#__codelineno-13-5></a><span class=n>only_in_rnaseq</span> <span class=o>=</span> <span class=n>rnaseq_ids</span> <span class=o>-</span> <span class=n>clinical_ids</span>
</span><span id=__span-13-6><a id=__codelineno-13-6 name=__codelineno-13-6 href=#__codelineno-13-6></a><span class=n>only_in_clinical</span> <span class=o>=</span> <span class=n>clinical_ids</span> <span class=o>-</span> <span class=n>rnaseq_ids</span>
</span><span id=__span-13-7><a id=__codelineno-13-7 name=__codelineno-13-7 href=#__codelineno-13-7></a>
</span><span id=__span-13-8><a id=__codelineno-13-8 name=__codelineno-13-8 href=#__codelineno-13-8></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Only in RNA-seq: </span><span class=si>{</span><span class=nb>len</span><span class=p>(</span><span class=n>only_in_rnaseq</span><span class=p>)</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-13-9><a id=__codelineno-13-9 name=__codelineno-13-9 href=#__codelineno-13-9></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Only in clinical: </span><span class=si>{</span><span class=nb>len</span><span class=p>(</span><span class=n>only_in_clinical</span><span class=p>)</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-13-10><a id=__codelineno-13-10 name=__codelineno-13-10 href=#__codelineno-13-10></a>
</span><span id=__span-13-11><a id=__codelineno-13-11 name=__codelineno-13-11 href=#__codelineno-13-11></a><span class=c1># Keep only matching samples</span>
</span><span id=__span-13-12><a id=__codelineno-13-12 name=__codelineno-13-12 href=#__codelineno-13-12></a><span class=n>common_ids</span> <span class=o>=</span> <span class=n>rnaseq_ids</span> <span class=o>&amp;</span> <span class=n>clinical_ids</span>
</span><span id=__span-13-13><a id=__codelineno-13-13 name=__codelineno-13-13 href=#__codelineno-13-13></a><span class=n>rnaseq</span> <span class=o>=</span> <span class=n>rnaseq</span><span class=o>.</span><span class=n>loc</span><span class=p>[</span><span class=n>common_ids</span><span class=p>]</span>
</span><span id=__span-13-14><a id=__codelineno-13-14 name=__codelineno-13-14 href=#__codelineno-13-14></a><span class=n>clinical</span> <span class=o>=</span> <span class=n>clinical</span><span class=p>[</span><span class=n>clinical</span><span class=p>[</span><span class=s1>&#39;sample_id&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>isin</span><span class=p>(</span><span class=n>common_ids</span><span class=p>)]</span>
</span></code></pre></div> <h3 id=too-few-genes-after-filtering>"Too few genes after filtering"<a class=headerlink href=#too-few-genes-after-filtering title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-14-1><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=c1># Relax filtering thresholds</span>
</span><span id=__span-14-2><a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a><span class=n>processed</span><span class=p>,</span> <span class=n>info</span> <span class=o>=</span> <span class=n>features</span><span class=o>.</span><span class=n>preprocess_rnaseq</span><span class=p>(</span>
</span><span id=__span-14-3><a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a>    <span class=n>data</span><span class=o>=</span><span class=n>rnaseq</span><span class=p>,</span>
</span><span id=__span-14-4><a id=__codelineno-14-4 name=__codelineno-14-4 href=#__codelineno-14-4></a>    <span class=n>mean_threshold</span><span class=o>=</span><span class=mf>0.1</span><span class=p>,</span>     <span class=c1># Lower threshold</span>
</span><span id=__span-14-5><a id=__codelineno-14-5 name=__codelineno-14-5 href=#__codelineno-14-5></a>    <span class=n>var_threshold</span><span class=o>=</span><span class=mf>0.1</span><span class=p>,</span>      <span class=c1># Lower threshold</span>
</span><span id=__span-14-6><a id=__codelineno-14-6 name=__codelineno-14-6 href=#__codelineno-14-6></a>    <span class=n>min_sample_fraction</span><span class=o>=</span><span class=mf>0.1</span> <span class=c1># Require expression in only 10% samples</span>
</span><span id=__span-14-7><a id=__codelineno-14-7 name=__codelineno-14-7 href=#__codelineno-14-7></a><span class=p>)</span>
</span></code></pre></div> <h3 id=data-looks-weird>"Data looks weird"<a class=headerlink href=#data-looks-weird title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-15-1><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=c1># Check if data needs transformation</span>
</span><span id=__span-15-2><a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Min:&quot;</span><span class=p>,</span> <span class=n>rnaseq</span><span class=o>.</span><span class=n>min</span><span class=p>()</span><span class=o>.</span><span class=n>min</span><span class=p>())</span>
</span><span id=__span-15-3><a id=__codelineno-15-3 name=__codelineno-15-3 href=#__codelineno-15-3></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Max:&quot;</span><span class=p>,</span> <span class=n>rnaseq</span><span class=o>.</span><span class=n>max</span><span class=p>()</span><span class=o>.</span><span class=n>max</span><span class=p>())</span>
</span><span id=__span-15-4><a id=__codelineno-15-4 name=__codelineno-15-4 href=#__codelineno-15-4></a>
</span><span id=__span-15-5><a id=__codelineno-15-5 name=__codelineno-15-5 href=#__codelineno-15-5></a><span class=c1># If values are large (&gt;100), likely not log-transformed</span>
</span><span id=__span-15-6><a id=__codelineno-15-6 name=__codelineno-15-6 href=#__codelineno-15-6></a><span class=k>if</span> <span class=n>rnaseq</span><span class=o>.</span><span class=n>max</span><span class=p>()</span><span class=o>.</span><span class=n>max</span><span class=p>()</span> <span class=o>&gt;</span> <span class=mi>100</span><span class=p>:</span>
</span><span id=__span-15-7><a id=__codelineno-15-7 name=__codelineno-15-7 href=#__codelineno-15-7></a>    <span class=n>rnaseq_log</span> <span class=o>=</span> <span class=n>np</span><span class=o>.</span><span class=n>log2</span><span class=p>(</span><span class=n>rnaseq</span> <span class=o>+</span> <span class=mi>1</span><span class=p>)</span>
</span><span id=__span-15-8><a id=__codelineno-15-8 name=__codelineno-15-8 href=#__codelineno-15-8></a>    <span class=n>rnaseq</span> <span class=o>=</span> <span class=n>rnaseq_log</span>
</span></code></pre></div> <h2 id=next-steps>Next Steps<a class=headerlink href=#next-steps title="Permanent link">&para;</a></h2> <ul> <li><a href=../quickstart/ >Quick Start Tutorial</a></li> <li><a href=../step1-data-processing/ >Data Processing Pipeline</a></li> <li><a href=../../api/dataset/ >API Reference</a></li> </ul> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="December 19, 2025 16:49:24 UTC"><span class=timeago datetime=2025-12-19T16:49:24+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="December 19, 2025 16:49:24 UTC">2025-12-19</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="December 19, 2025 16:49:24 UTC"><span class=timeago datetime=2025-12-19T16:49:24+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="December 19, 2025 16:49:24 UTC">2025-12-19</span> </span> </aside> <form class=md-feedback name=feedback hidden> <fieldset> <legend class=md-feedback__title> Was this page helpful? </legend> <div class=md-feedback__inner> <div class=md-feedback__list> <button class="md-feedback__icon md-icon" type=submit title="This page was helpful" data-md-value=1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m7 0c0 .8-.7 1.5-1.5 1.5S14 10.3 14 9.5 14.7 8 15.5 8s1.5.7 1.5 1.5m-5 7.73c-1.75 0-3.29-.73-4.19-1.81L9.23 14c.45.72 1.52 1.23 2.77 1.23s2.32-.51 2.77-1.23l1.42 1.42c-.9 1.08-2.44 1.81-4.19 1.81"/></svg> </button> <button class="md-feedback__icon md-icon" type=submit title="This page could be improved" data-md-value=0> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10m-6.5-4c.8 0 1.5.7 1.5 1.5s-.7 1.5-1.5 1.5-1.5-.7-1.5-1.5.7-1.5 1.5-1.5M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m2 4.5c1.75 0 3.29.72 4.19 1.81l-1.42 1.42C14.32 16.5 13.25 16 12 16s-2.32.5-2.77 1.23l-1.42-1.42C8.71 14.72 10.25 14 12 14"/></svg> </button> </div> <div class=md-feedback__note> <div data-md-value=1 hidden> Thanks for your feedback! </div> <div data-md-value=0 hidden> Thanks for your feedback! Help us improve by opening an issue. </div> </div> </div> </fieldset> </form> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../quickstart/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Quick Start"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Quick Start </div> </div> </a> <a href=../ class="md-footer__link md-footer__link--next" aria-label="Next: Overview"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Overview </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2022-2025 EVENFLOW Project </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/gprolcastelo/renalprog target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://pypi.org/project/renalprog/ target=_blank rel=noopener title=pypi.org class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.8 200.5c-7.7-30.9-22.3-54.2-53.4-54.2h-40.1v47.4c0 36.8-31.2 67.8-66.8 67.8H172.7c-29.2 0-53.4 25-53.4 54.3v101.8c0 29 25.2 46 53.4 54.3 33.8 9.9 66.3 11.7 106.8 0 26.9-7.8 53.4-23.5 53.4-54.3v-40.7H226.2v-13.6h160.2c31.1 0 42.6-21.7 53.4-54.2 11.2-33.5 10.7-65.7 0-108.6M286.2 444.7a20.4 20.4 0 1 1 0-40.7 20.4 20.4 0 1 1 0 40.7M167.8 248.1h106.8c29.7 0 53.4-24.5 53.4-54.3V91.9c0-29-24.4-50.7-53.4-55.6-35.8-5.9-74.7-5.6-106.8.1-45.2 8-53.4 24.7-53.4 55.6v40.7h106.9v13.6h-147c-31.1 0-58.3 18.7-66.8 54.2-9.8 40.7-10.2 66.1 0 108.6 7.6 31.6 25.7 54.2 56.8 54.2H101v-48.8c0-35.3 30.5-66.4 66.8-66.4m-6.6-183.4a20.4 20.4 0 1 1 0 40.8 20.4 20.4 0 1 1 0-40.8"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "search.suggest", "search.highlight", "search.share", "content.code.copy", "content.code.annotate", "toc.follow"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../../js/timeago.min.js></script> <script src=../../js/timeago_mkdocs_material.js></script> <script src=../../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>