<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="A Python package for simulating kidney cancer progression with synthetic data generation and machine learning."><meta name=author content="Guillermo Prol-Castelo (Barcelona Supercomputing Center), for the EVENFLOW European Project"><link href=https://github.com/gprolcastelo/renalprog/tutorials/complete-pipeline/ rel=canonical><link href=../pretrained-models/ rel=prev><link href=../step1-data-processing/ rel=next><link rel=icon href=../../assets/images/kidneys.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>Complete Pipeline - renalprog</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../css/timeago.css><link rel=stylesheet href=../../assets/_mkdocstrings.css><link rel=stylesheet href=../../stylesheets/extra.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#complete-pipeline-tutorial class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=renalprog class="md-header__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../../assets/images/kidneys.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> renalprog </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Complete Pipeline </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../../getting-started/ class=md-tabs__link> Getting Started </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=../ class=md-tabs__link> Tutorials </a> </li> <li class=md-tabs__item> <a href=../../reproducibility/ class=md-tabs__link> Reproducibility </a> </li> <li class=md-tabs__item> <a href=../../api/ class=md-tabs__link> API Reference </a> </li> <li class=md-tabs__item> <a href=../../advanced/architecture/ class=md-tabs__link> Advanced Topics </a> </li> <li class=md-tabs__item> <a href=../../contributing/guidelines/ class=md-tabs__link> Contributing </a> </li> <li class=md-tabs__item> <a href=../../citation/ class=md-tabs__link> Citation & License </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=renalprog class="md-nav__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../../assets/images/kidneys.png alt=logo> </a> renalprog </label> <div class=md-nav__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> <span class=md-ellipsis> Getting Started </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Getting Started </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../getting-started/ class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=../quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class=md-nav__item> <a href=../data-requirements/ class=md-nav__link> <span class=md-ellipsis> Data Requirements </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3 checked> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex> <span class=md-ellipsis> Tutorials </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=true> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../pretrained-models/ class=md-nav__link> <span class=md-ellipsis> Using Pretrained Models </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Complete Pipeline </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Complete Pipeline </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> <span class=md-ellipsis> Prerequisites </span> </a> </li> <li class=md-nav__item> <a href=#pipeline-execution class=md-nav__link> <span class=md-ellipsis> Pipeline Execution </span> </a> <nav class=md-nav aria-label="Pipeline Execution"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#option-1-run-all-steps-sequentially class=md-nav__link> <span class=md-ellipsis> Option 1: Run All Steps Sequentially </span> </a> </li> <li class=md-nav__item> <a href=#option-2-run-with-makefile class=md-nav__link> <span class=md-ellipsis> Option 2: Run with Makefile </span> </a> </li> <li class=md-nav__item> <a href=#option-3-use-job-scripts-hpc class=md-nav__link> <span class=md-ellipsis> Option 3: Use Job Scripts (HPC) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#detailed-step-by-step-guide class=md-nav__link> <span class=md-ellipsis> Detailed Step-by-Step Guide </span> </a> <nav class=md-nav aria-label="Detailed Step-by-Step Guide"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-data-processing class=md-nav__link> <span class=md-ellipsis> Step 1: Data Processing </span> </a> </li> <li class=md-nav__item> <a href=#step-2-vae-training class=md-nav__link> <span class=md-ellipsis> Step 2: VAE Training </span> </a> </li> <li class=md-nav__item> <a href=#step-3-reconstruction-validation class=md-nav__link> <span class=md-ellipsis> Step 3: Reconstruction Validation </span> </a> </li> <li class=md-nav__item> <a href=#step-4-trajectory-generation class=md-nav__link> <span class=md-ellipsis> Step 4: Trajectory Generation </span> </a> </li> <li class=md-nav__item> <a href=#step-5-classification class=md-nav__link> <span class=md-ellipsis> Step 5: Classification </span> </a> </li> <li class=md-nav__item> <a href=#step-6-enrichment-analysis class=md-nav__link> <span class=md-ellipsis> Step 6: Enrichment Analysis </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#checking-pipeline-outputs class=md-nav__link> <span class=md-ellipsis> Checking Pipeline Outputs </span> </a> </li> <li class=md-nav__item> <a href=#expected-disk-usage class=md-nav__link> <span class=md-ellipsis> Expected Disk Usage </span> </a> </li> <li class=md-nav__item> <a href=#performance-benchmarks class=md-nav__link> <span class=md-ellipsis> Performance Benchmarks </span> </a> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#pipeline-fails-at-step-x class=md-nav__link> <span class=md-ellipsis> Pipeline Fails at Step X </span> </a> </li> <li class=md-nav__item> <a href=#out-of-disk-space class=md-nav__link> <span class=md-ellipsis> Out of Disk Space </span> </a> </li> <li class=md-nav__item> <a href=#out-of-memory class=md-nav__link> <span class=md-ellipsis> Out of Memory </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> <li class=md-nav__item> <a href=#citation class=md-nav__link> <span class=md-ellipsis> Citation </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../step1-data-processing/ class=md-nav__link> <span class=md-ellipsis> Step 1 - Data Processing </span> </a> </li> <li class=md-nav__item> <a href=../step2-vae-training/ class=md-nav__link> <span class=md-ellipsis> Step 2 - VAE Training </span> </a> </li> <li class=md-nav__item> <a href=../step3-reconstruction/ class=md-nav__link> <span class=md-ellipsis> Step 3 - Reconstruction </span> </a> </li> <li class=md-nav__item> <a href=../step4-trajectories/ class=md-nav__link> <span class=md-ellipsis> Step 4 - Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../step5-classification/ class=md-nav__link> <span class=md-ellipsis> Step 5 - Classification </span> </a> </li> <li class=md-nav__item> <a href=../step6-enrichment/ class=md-nav__link> <span class=md-ellipsis> Step 6 - Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../visualization/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> Reproducibility </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Reproducibility </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../reproducibility/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../reproducibility/requirements/ class=md-nav__link> <span class=md-ellipsis> System Requirements </span> </a> </li> <li class=md-nav__item> <a href=../../reproducibility/data-preparation/ class=md-nav__link> <span class=md-ellipsis> Data Preparation </span> </a> </li> <li class=md-nav__item> <a href=../../reproducibility/pipeline/ class=md-nav__link> <span class=md-ellipsis> Running the Pipeline </span> </a> </li> <li class=md-nav__item> <a href=../../reproducibility/results/ class=md-nav__link> <span class=md-ellipsis> Expected Results </span> </a> </li> <li class=md-nav__item> <a href=../../reproducibility/troubleshooting/ class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../api/config/ class=md-nav__link> <span class=md-ellipsis> Configuration </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5_3> <label class=md-nav__link for=__nav_5_3 id=__nav_5_3_label tabindex=0> <span class=md-ellipsis> Data Processing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_3_label aria-expanded=false> <label class=md-nav__title for=__nav_5_3> <span class="md-nav__icon md-icon"></span> Data Processing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/dataset/ class=md-nav__link> <span class=md-ellipsis> Dataset </span> </a> </li> <li class=md-nav__item> <a href=../../api/features/ class=md-nav__link> <span class=md-ellipsis> Features </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5_4> <label class=md-nav__link for=__nav_5_4 id=__nav_5_4_label tabindex=0> <span class=md-ellipsis> Modeling </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_4_label aria-expanded=false> <label class=md-nav__title for=__nav_5_4> <span class="md-nav__icon md-icon"></span> Modeling </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/models/ class=md-nav__link> <span class=md-ellipsis> VAE Models </span> </a> </li> <li class=md-nav__item> <a href=../../api/training/ class=md-nav__link> <span class=md-ellipsis> Training </span> </a> </li> <li class=md-nav__item> <a href=../../api/prediction/ class=md-nav__link> <span class=md-ellipsis> Prediction </span> </a> </li> <li class=md-nav__item> <a href=../../api/trajectories/ class=md-nav__link> <span class=md-ellipsis> Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../../api/classification/ class=md-nav__link> <span class=md-ellipsis> Classification </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../api/enrichment/ class=md-nav__link> <span class=md-ellipsis> Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../../api/plots/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> <li class=md-nav__item> <a href=../../api/utils/ class=md-nav__link> <span class=md-ellipsis> Utilities </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex=0> <span class=md-ellipsis> Advanced Topics </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> Advanced Topics </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../advanced/architecture/ class=md-nav__link> <span class=md-ellipsis> Architecture </span> </a> </li> <li class=md-nav__item> <a href=../../TRAJECTORIES/ class=md-nav__link> <span class=md-ellipsis> Trajectory Generation Details </span> </a> </li> <li class=md-nav__item> <a href=../../CLASSIFICATION/ class=md-nav__link> <span class=md-ellipsis> Classification Details </span> </a> </li> <li class=md-nav__item> <a href=../../ENRICHMENT_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> Enrichment Analysis Details </span> </a> </li> <li class=md-nav__item> <a href=../../ENRICHMENT_QUICK_REF/ class=md-nav__link> <span class=md-ellipsis> Enrichment Quick Reference </span> </a> </li> <li class=md-nav__item> <a href=../../PATHWAY_HEATMAP/ class=md-nav__link> <span class=md-ellipsis> Pathway Heatmap Analysis </span> </a> </li> <li class=md-nav__item> <a href=../../GSEA_INSTALLATION/ class=md-nav__link> <span class=md-ellipsis> GSEA Installation </span> </a> </li> <li class=md-nav__item> <a href=../../R_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> R Analysis Scripts </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/r-integration/ class=md-nav__link> <span class=md-ellipsis> R Integration </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/custom-models/ class=md-nav__link> <span class=md-ellipsis> Custom Models </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/performance/ class=md-nav__link> <span class=md-ellipsis> Performance Tuning </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/gpu/ class=md-nav__link> <span class=md-ellipsis> GPU Acceleration </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex=0> <span class=md-ellipsis> Contributing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Contributing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../contributing/guidelines/ class=md-nav__link> <span class=md-ellipsis> Guidelines </span> </a> </li> <li class=md-nav__item> <a href=../../contributing/development/ class=md-nav__link> <span class=md-ellipsis> Development Setup </span> </a> </li> <li class=md-nav__item> <a href=../../contributing/code-style/ class=md-nav__link> <span class=md-ellipsis> Code Style </span> </a> </li> <li class=md-nav__item> <a href=../../contributing/testing/ class=md-nav__link> <span class=md-ellipsis> Testing </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex=0> <span class=md-ellipsis> Citation & License </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> Citation & License </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../citation/ class=md-nav__link> <span class=md-ellipsis> How to Cite </span> </a> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> <li class=md-nav__item> <a href=../../acknowledgments/ class=md-nav__link> <span class=md-ellipsis> Acknowledgments </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> <span class=md-ellipsis> Prerequisites </span> </a> </li> <li class=md-nav__item> <a href=#pipeline-execution class=md-nav__link> <span class=md-ellipsis> Pipeline Execution </span> </a> <nav class=md-nav aria-label="Pipeline Execution"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#option-1-run-all-steps-sequentially class=md-nav__link> <span class=md-ellipsis> Option 1: Run All Steps Sequentially </span> </a> </li> <li class=md-nav__item> <a href=#option-2-run-with-makefile class=md-nav__link> <span class=md-ellipsis> Option 2: Run with Makefile </span> </a> </li> <li class=md-nav__item> <a href=#option-3-use-job-scripts-hpc class=md-nav__link> <span class=md-ellipsis> Option 3: Use Job Scripts (HPC) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#detailed-step-by-step-guide class=md-nav__link> <span class=md-ellipsis> Detailed Step-by-Step Guide </span> </a> <nav class=md-nav aria-label="Detailed Step-by-Step Guide"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-data-processing class=md-nav__link> <span class=md-ellipsis> Step 1: Data Processing </span> </a> </li> <li class=md-nav__item> <a href=#step-2-vae-training class=md-nav__link> <span class=md-ellipsis> Step 2: VAE Training </span> </a> </li> <li class=md-nav__item> <a href=#step-3-reconstruction-validation class=md-nav__link> <span class=md-ellipsis> Step 3: Reconstruction Validation </span> </a> </li> <li class=md-nav__item> <a href=#step-4-trajectory-generation class=md-nav__link> <span class=md-ellipsis> Step 4: Trajectory Generation </span> </a> </li> <li class=md-nav__item> <a href=#step-5-classification class=md-nav__link> <span class=md-ellipsis> Step 5: Classification </span> </a> </li> <li class=md-nav__item> <a href=#step-6-enrichment-analysis class=md-nav__link> <span class=md-ellipsis> Step 6: Enrichment Analysis </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#checking-pipeline-outputs class=md-nav__link> <span class=md-ellipsis> Checking Pipeline Outputs </span> </a> </li> <li class=md-nav__item> <a href=#expected-disk-usage class=md-nav__link> <span class=md-ellipsis> Expected Disk Usage </span> </a> </li> <li class=md-nav__item> <a href=#performance-benchmarks class=md-nav__link> <span class=md-ellipsis> Performance Benchmarks </span> </a> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#pipeline-fails-at-step-x class=md-nav__link> <span class=md-ellipsis> Pipeline Fails at Step X </span> </a> </li> <li class=md-nav__item> <a href=#out-of-disk-space class=md-nav__link> <span class=md-ellipsis> Out of Disk Space </span> </a> </li> <li class=md-nav__item> <a href=#out-of-memory class=md-nav__link> <span class=md-ellipsis> Out of Memory </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> <li class=md-nav__item> <a href=#citation class=md-nav__link> <span class=md-ellipsis> Citation </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=complete-pipeline-tutorial>Complete Pipeline Tutorial<a class=headerlink href=#complete-pipeline-tutorial title="Permanent link">&para;</a></h1> <p>This tutorial walks through the complete <code>renalprog</code> analysis pipeline from raw TCGA data to pathway enrichment results. This represents the full workflow used in the publication.</p> <h2 id=overview>Overview<a class=headerlink href=#overview title="Permanent link">&para;</a></h2> <p>The pipeline consists of six main steps:</p> <ol> <li><strong>Data Processing</strong>: Download and preprocess TCGA data</li> <li><strong>VAE Training</strong>: Train deep generative models </li> <li><strong>Reconstruction Validation</strong>: Assess model quality</li> <li><strong>Trajectory Generation</strong>: Create synthetic progression paths</li> <li><strong>Classification</strong>: Stage prediction and biomarker discovery</li> <li><strong>Enrichment Analysis</strong>: Pathway-level interpretation</li> </ol> <p><strong>Total Time</strong>: ~8-12 hours (varies with hardware)</p> <p><strong>Requirements</strong>: - 16+ GB RAM - 50+ GB free disk space - GPU recommended for steps 2-4 - R 4.0+ for step 6</p> <h2 id=prerequisites>Prerequisites<a class=headerlink href=#prerequisites title="Permanent link">&para;</a></h2> <p>Ensure you have completed the installation:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>mamba<span class=w> </span>env<span class=w> </span>create<span class=w> </span>-f<span class=w> </span>environment.yml
</span><span id=__span-0-2><a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>mamba<span class=w> </span>activate<span class=w> </span>renalprog
</span><span id=__span-0-3><a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a>pip<span class=w> </span>install<span class=w> </span>-e<span class=w> </span>.
</span></code></pre></div> <h2 id=pipeline-execution>Pipeline Execution<a class=headerlink href=#pipeline-execution title="Permanent link">&para;</a></h2> <h3 id=option-1-run-all-steps-sequentially>Option 1: Run All Steps Sequentially<a class=headerlink href=#option-1-run-all-steps-sequentially title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1># Navigate to scripts directory</span>
</span><span id=__span-1-2><a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=nb>cd</span><span class=w> </span>scripts/pipeline_steps
</span><span id=__span-1-3><a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>
</span><span id=__span-1-4><a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=c1># Step 1: Data Processing (30 minutes)</span>
</span><span id=__span-1-5><a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a>python<span class=w> </span>1_data_processing.py
</span><span id=__span-1-6><a id=__codelineno-1-6 name=__codelineno-1-6 href=#__codelineno-1-6></a>
</span><span id=__span-1-7><a id=__codelineno-1-7 name=__codelineno-1-7 href=#__codelineno-1-7></a><span class=c1># Step 2: VAE Training (2-4 hours with GPU)</span>
</span><span id=__span-1-8><a id=__codelineno-1-8 name=__codelineno-1-8 href=#__codelineno-1-8></a>python<span class=w> </span>2_models.py
</span><span id=__span-1-9><a id=__codelineno-1-9 name=__codelineno-1-9 href=#__codelineno-1-9></a>
</span><span id=__span-1-10><a id=__codelineno-1-10 name=__codelineno-1-10 href=#__codelineno-1-10></a><span class=c1># Step 3: Reconstruction Check (10 minutes)</span>
</span><span id=__span-1-11><a id=__codelineno-1-11 name=__codelineno-1-11 href=#__codelineno-1-11></a>python<span class=w> </span>3_check_reconstruction.py
</span><span id=__span-1-12><a id=__codelineno-1-12 name=__codelineno-1-12 href=#__codelineno-1-12></a>
</span><span id=__span-1-13><a id=__codelineno-1-13 name=__codelineno-1-13 href=#__codelineno-1-13></a><span class=c1># Step 4: Trajectory Generation (30 minutes)</span>
</span><span id=__span-1-14><a id=__codelineno-1-14 name=__codelineno-1-14 href=#__codelineno-1-14></a>python<span class=w> </span>4_trajectories.py
</span><span id=__span-1-15><a id=__codelineno-1-15 name=__codelineno-1-15 href=#__codelineno-1-15></a>
</span><span id=__span-1-16><a id=__codelineno-1-16 name=__codelineno-1-16 href=#__codelineno-1-16></a><span class=c1># Step 5: Classification (1 hour)</span>
</span><span id=__span-1-17><a id=__codelineno-1-17 name=__codelineno-1-17 href=#__codelineno-1-17></a>python<span class=w> </span>5_classification.py
</span><span id=__span-1-18><a id=__codelineno-1-18 name=__codelineno-1-18 href=#__codelineno-1-18></a>
</span><span id=__span-1-19><a id=__codelineno-1-19 name=__codelineno-1-19 href=#__codelineno-1-19></a><span class=c1># Step 6: Enrichment Analysis (2-6 hours with 8 cores)</span>
</span><span id=__span-1-20><a id=__codelineno-1-20 name=__codelineno-1-20 href=#__codelineno-1-20></a>python<span class=w> </span>6_enrichment_analysis.py
</span></code></pre></div> <h3 id=option-2-run-with-makefile>Option 2: Run with Makefile<a class=headerlink href=#option-2-run-with-makefile title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1># Run entire pipeline</span>
</span><span id=__span-2-2><a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a>make<span class=w> </span>pipeline
</span><span id=__span-2-3><a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a>
</span><span id=__span-2-4><a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a><span class=c1># Or run individual steps</span>
</span><span id=__span-2-5><a id=__codelineno-2-5 name=__codelineno-2-5 href=#__codelineno-2-5></a>make<span class=w> </span>data
</span><span id=__span-2-6><a id=__codelineno-2-6 name=__codelineno-2-6 href=#__codelineno-2-6></a>make<span class=w> </span>models
</span><span id=__span-2-7><a id=__codelineno-2-7 name=__codelineno-2-7 href=#__codelineno-2-7></a>make<span class=w> </span>trajectories
</span><span id=__span-2-8><a id=__codelineno-2-8 name=__codelineno-2-8 href=#__codelineno-2-8></a>make<span class=w> </span>classification
</span><span id=__span-2-9><a id=__codelineno-2-9 name=__codelineno-2-9 href=#__codelineno-2-9></a>make<span class=w> </span>enrichment
</span></code></pre></div> <h3 id=option-3-use-job-scripts-hpc>Option 3: Use Job Scripts (HPC)<a class=headerlink href=#option-3-use-job-scripts-hpc title="Permanent link">&para;</a></h3> <p>If you're on an HPC cluster with SLURM:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-3-1><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1># Submit all jobs</span>
</span><span id=__span-3-2><a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a>sbatch<span class=w> </span>jobs/job_data
</span><span id=__span-3-3><a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a>sbatch<span class=w> </span>jobs/job_models
</span><span id=__span-3-4><a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a>sbatch<span class=w> </span>jobs/job_trajectories
</span><span id=__span-3-5><a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a>sbatch<span class=w> </span>jobs/job_classification
</span><span id=__span-3-6><a id=__codelineno-3-6 name=__codelineno-3-6 href=#__codelineno-3-6></a>sbatch<span class=w> </span>jobs/job_enrichment
</span></code></pre></div> <h2 id=detailed-step-by-step-guide>Detailed Step-by-Step Guide<a class=headerlink href=#detailed-step-by-step-guide title="Permanent link">&para;</a></h2> <h3 id=step-1-data-processing>Step 1: Data Processing<a class=headerlink href=#step-1-data-processing title="Permanent link">&para;</a></h3> <p><strong>Script</strong>: <code>scripts/pipeline_steps/1_data_processing.py</code></p> <p><strong>What it does</strong>: - Downloads TCGA KIRC RNA-seq and clinical data - Filters low-expression genes - Removes outliers using Mahalanobis distance - Creates train/test splits</p> <p><strong>Expected outputs</strong>: <div class="language-text highlight"><pre><span></span><code><span id=__span-4-1><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a>data/interim/preprocessed_KIRC_data/
</span><span id=__span-4-2><a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>├── preprocessed_rnaseq.csv         # Filtered gene expression (498 samples × ~5000 genes)
</span><span id=__span-4-3><a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>├── preprocessing_info.json         # Filter statistics
</span><span id=__span-4-4><a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a>└── stages.csv                      # Cancer stage labels
</span></code></pre></div></p> <p><strong>Validation</strong>: <div class="language-python highlight"><pre><span></span><code><span id=__span-5-1><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=kn>import</span><span class=w> </span><span class=nn>pandas</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>pd</span>
</span><span id=__span-5-2><a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a>
</span><span id=__span-5-3><a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a><span class=n>rnaseq</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=s1>&#39;data/interim/preprocessed_KIRC_data/preprocessed_rnaseq.csv&#39;</span><span class=p>,</span> <span class=n>index_col</span><span class=o>=</span><span class=mi>0</span><span class=p>)</span>
</span><span id=__span-5-4><a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Samples: </span><span class=si>{</span><span class=n>rnaseq</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>0</span><span class=p>]</span><span class=si>}</span><span class=s2>, Genes: </span><span class=si>{</span><span class=n>rnaseq</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>1</span><span class=p>]</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-5-5><a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a><span class=c1># Expected: Samples: 498, Genes: ~5000</span>
</span><span id=__span-5-6><a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a>
</span><span id=__span-5-7><a id=__codelineno-5-7 name=__codelineno-5-7 href=#__codelineno-5-7></a><span class=c1># Check for missing values</span>
</span><span id=__span-5-8><a id=__codelineno-5-8 name=__codelineno-5-8 href=#__codelineno-5-8></a><span class=k>assert</span> <span class=n>rnaseq</span><span class=o>.</span><span class=n>isnull</span><span class=p>()</span><span class=o>.</span><span class=n>sum</span><span class=p>()</span><span class=o>.</span><span class=n>sum</span><span class=p>()</span> <span class=o>==</span> <span class=mi>0</span><span class=p>,</span> <span class=s2>&quot;Missing values detected!&quot;</span>
</span><span id=__span-5-9><a id=__codelineno-5-9 name=__codelineno-5-9 href=#__codelineno-5-9></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;✓ Data quality check passed&quot;</span><span class=p>)</span>
</span></code></pre></div></p> <p><strong>Time</strong>: ~30 minutes<br> <strong>See</strong>: <a href=../step1-data-processing/ >Data Processing Tutorial</a></p> <hr> <h3 id=step-2-vae-training>Step 2: VAE Training<a class=headerlink href=#step-2-vae-training title="Permanent link">&para;</a></h3> <p><strong>Script</strong>: <code>scripts/pipeline_steps/2_models.py</code></p> <p><strong>What it does</strong>: - Trains Variational Autoencoder on gene expression - Learns low-dimensional latent representation - Trains reconstruction network for improved decoding - Saves models and training histories</p> <p><strong>Configuration</strong> (in script): <div class="language-python highlight"><pre><span></span><code><span id=__span-6-1><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=n>vae_config</span><span class=o>.</span><span class=n>INPUT_DIM</span> <span class=o>=</span> <span class=n>X_train</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>1</span><span class=p>]</span>  <span class=c1># Number of genes</span>
</span><span id=__span-6-2><a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=n>vae_config</span><span class=o>.</span><span class=n>MID_DIM</span> <span class=o>=</span> <span class=mi>512</span>                  <span class=c1># Hidden layer size</span>
</span><span id=__span-6-3><a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a><span class=n>vae_config</span><span class=o>.</span><span class=n>LATENT_DIM</span> <span class=o>=</span> <span class=mi>256</span>               <span class=c1># Latent space dimensionality</span>
</span><span id=__span-6-4><a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a><span class=n>vae_config</span><span class=o>.</span><span class=n>EPOCHS</span> <span class=o>=</span> <span class=mi>600</span>                   <span class=c1># Total epochs (3 cycles × 200)</span>
</span><span id=__span-6-5><a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a><span class=n>vae_config</span><span class=o>.</span><span class=n>BATCH_SIZE</span> <span class=o>=</span> <span class=mi>8</span>
</span></code></pre></div></p> <p><strong>Expected outputs</strong>: <div class="language-text highlight"><pre><span></span><code><span id=__span-7-1><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a>models/20251217_models_KIRC/
</span><span id=__span-7-2><a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a>├── vae/
</span><span id=__span-7-3><a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a>│   ├── vae_model.pt                 # Trained VAE
</span><span id=__span-7-4><a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a>│   ├── vae_config.json              # Model configuration
</span><span id=__span-7-5><a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a>│   └── vae_training_history.png     # Loss curves
</span><span id=__span-7-6><a id=__codelineno-7-6 name=__codelineno-7-6 href=#__codelineno-7-6></a>├── reconstruction/
</span><span id=__span-7-7><a id=__codelineno-7-7 name=__codelineno-7-7 href=#__codelineno-7-7></a>│   ├── reconstruction_network.pt    # Post-processing network
</span><span id=__span-7-8><a id=__codelineno-7-8 name=__codelineno-7-8 href=#__codelineno-7-8></a>│   └── reconstruction_network_history.png
</span><span id=__span-7-9><a id=__codelineno-7-9 name=__codelineno-7-9 href=#__codelineno-7-9></a>└── training_data/
</span><span id=__span-7-10><a id=__codelineno-7-10 name=__codelineno-7-10 href=#__codelineno-7-10></a>    ├── X_train.csv
</span><span id=__span-7-11><a id=__codelineno-7-11 name=__codelineno-7-11 href=#__codelineno-7-11></a>    └── X_test.csv
</span></code></pre></div></p> <p><strong>Validation</strong>: <div class="language-python highlight"><pre><span></span><code><span id=__span-8-1><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=kn>import</span><span class=w> </span><span class=nn>torch</span>
</span><span id=__span-8-2><a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=kn>from</span><span class=w> </span><span class=nn>renalprog.modeling.models</span><span class=w> </span><span class=kn>import</span> <span class=n>VAE</span>
</span><span id=__span-8-3><a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a>
</span><span id=__span-8-4><a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=c1># Load model</span>
</span><span id=__span-8-5><a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a><span class=n>vae</span> <span class=o>=</span> <span class=n>VAE</span><span class=o>.</span><span class=n>load</span><span class=p>(</span><span class=s1>&#39;models/20251217_models_KIRC/vae/vae_model.pt&#39;</span><span class=p>)</span>
</span><span id=__span-8-6><a id=__codelineno-8-6 name=__codelineno-8-6 href=#__codelineno-8-6></a>
</span><span id=__span-8-7><a id=__codelineno-8-7 name=__codelineno-8-7 href=#__codelineno-8-7></a><span class=c1># Check reconstruction</span>
</span><span id=__span-8-8><a id=__codelineno-8-8 name=__codelineno-8-8 href=#__codelineno-8-8></a><span class=n>X_test</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=s1>&#39;models/20251217_models_KIRC/training_data/X_test.csv&#39;</span><span class=p>,</span> <span class=n>index_col</span><span class=o>=</span><span class=mi>0</span><span class=p>)</span>
</span><span id=__span-8-9><a id=__codelineno-8-9 name=__codelineno-8-9 href=#__codelineno-8-9></a><span class=n>X_tensor</span> <span class=o>=</span> <span class=n>torch</span><span class=o>.</span><span class=n>FloatTensor</span><span class=p>(</span><span class=n>X_test</span><span class=o>.</span><span class=n>values</span><span class=p>)</span>
</span><span id=__span-8-10><a id=__codelineno-8-10 name=__codelineno-8-10 href=#__codelineno-8-10></a>
</span><span id=__span-8-11><a id=__codelineno-8-11 name=__codelineno-8-11 href=#__codelineno-8-11></a><span class=k>with</span> <span class=n>torch</span><span class=o>.</span><span class=n>no_grad</span><span class=p>():</span>
</span><span id=__span-8-12><a id=__codelineno-8-12 name=__codelineno-8-12 href=#__codelineno-8-12></a>    <span class=n>X_recon</span> <span class=o>=</span> <span class=n>vae</span><span class=p>(</span><span class=n>X_tensor</span><span class=p>)[</span><span class=mi>0</span><span class=p>]</span>
</span><span id=__span-8-13><a id=__codelineno-8-13 name=__codelineno-8-13 href=#__codelineno-8-13></a>    <span class=n>mse</span> <span class=o>=</span> <span class=p>((</span><span class=n>X_tensor</span> <span class=o>-</span> <span class=n>X_recon</span><span class=p>)</span> <span class=o>**</span> <span class=mi>2</span><span class=p>)</span><span class=o>.</span><span class=n>mean</span><span class=p>()</span>
</span><span id=__span-8-14><a id=__codelineno-8-14 name=__codelineno-8-14 href=#__codelineno-8-14></a>    <span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Reconstruction MSE: </span><span class=si>{</span><span class=n>mse</span><span class=si>:</span><span class=s2>.4f</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-8-15><a id=__codelineno-8-15 name=__codelineno-8-15 href=#__codelineno-8-15></a>    <span class=c1># Expected: MSE &lt; 1.0 for well-trained model</span>
</span></code></pre></div></p> <p><strong>Time</strong>: 2-4 hours (GPU), 8-12 hours (CPU)<br> <strong>See</strong>: <a href=../step2-vae-training/ >VAE Training Tutorial</a></p> <hr> <h3 id=step-3-reconstruction-validation>Step 3: Reconstruction Validation<a class=headerlink href=#step-3-reconstruction-validation title="Permanent link">&para;</a></h3> <p><strong>Script</strong>: <code>scripts/pipeline_steps/3_check_reconstruction.py</code></p> <p><strong>What it does</strong>: - Visualizes latent space with UMAP/t-SNE - Compares original vs. reconstructed gene expression - Generates quality control plots</p> <p><strong>Expected outputs</strong>: <div class="language-text highlight"><pre><span></span><code><span id=__span-9-1><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a>reports/figures/reconstruction/
</span><span id=__span-9-2><a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>├── latent_space_umap.png           # UMAP colored by stage
</span><span id=__span-9-3><a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a>├── latent_space_tsne.png           # t-SNE colored by stage
</span><span id=__span-9-4><a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a>├── reconstruction_scatter.png      # Original vs. reconstructed
</span><span id=__span-9-5><a id=__codelineno-9-5 name=__codelineno-9-5 href=#__codelineno-9-5></a>└── gene_correlation.png            # Per-gene reconstruction quality
</span></code></pre></div></p> <p><strong>Validation</strong>: <div class="language-python highlight"><pre><span></span><code><span id=__span-10-1><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=c1># Check that plots exist</span>
</span><span id=__span-10-2><a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=kn>import</span><span class=w> </span><span class=nn>os</span>
</span><span id=__span-10-3><a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a>
</span><span id=__span-10-4><a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=n>plot_dir</span> <span class=o>=</span> <span class=s1>&#39;reports/figures/reconstruction/&#39;</span>
</span><span id=__span-10-5><a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a><span class=n>required_plots</span> <span class=o>=</span> <span class=p>[</span>
</span><span id=__span-10-6><a id=__codelineno-10-6 name=__codelineno-10-6 href=#__codelineno-10-6></a>    <span class=s1>&#39;latent_space_umap.png&#39;</span><span class=p>,</span>
</span><span id=__span-10-7><a id=__codelineno-10-7 name=__codelineno-10-7 href=#__codelineno-10-7></a>    <span class=s1>&#39;reconstruction_scatter.png&#39;</span>
</span><span id=__span-10-8><a id=__codelineno-10-8 name=__codelineno-10-8 href=#__codelineno-10-8></a><span class=p>]</span>
</span><span id=__span-10-9><a id=__codelineno-10-9 name=__codelineno-10-9 href=#__codelineno-10-9></a>
</span><span id=__span-10-10><a id=__codelineno-10-10 name=__codelineno-10-10 href=#__codelineno-10-10></a><span class=k>for</span> <span class=n>plot</span> <span class=ow>in</span> <span class=n>required_plots</span><span class=p>:</span>
</span><span id=__span-10-11><a id=__codelineno-10-11 name=__codelineno-10-11 href=#__codelineno-10-11></a>    <span class=k>assert</span> <span class=n>os</span><span class=o>.</span><span class=n>path</span><span class=o>.</span><span class=n>exists</span><span class=p>(</span><span class=n>os</span><span class=o>.</span><span class=n>path</span><span class=o>.</span><span class=n>join</span><span class=p>(</span><span class=n>plot_dir</span><span class=p>,</span> <span class=n>plot</span><span class=p>)),</span> <span class=sa>f</span><span class=s2>&quot;Missing: </span><span class=si>{</span><span class=n>plot</span><span class=si>}</span><span class=s2>&quot;</span>
</span><span id=__span-10-12><a id=__codelineno-10-12 name=__codelineno-10-12 href=#__codelineno-10-12></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;✓ All validation plots generated&quot;</span><span class=p>)</span>
</span></code></pre></div></p> <p><strong>Time</strong>: ~10 minutes<br> <strong>See</strong>: <a href=../step3-reconstruction/ >Reconstruction Tutorial</a></p> <hr> <h3 id=step-4-trajectory-generation>Step 4: Trajectory Generation<a class=headerlink href=#step-4-trajectory-generation title="Permanent link">&para;</a></h3> <p><strong>Script</strong>: <code>scripts/pipeline_steps/4_trajectories.py</code></p> <p><strong>What it does</strong>: - Generates synthetic patient trajectories - Interpolates between early and late stage samples - Decodes trajectories to gene expression space - Creates multiple trajectory types</p> <p><strong>Configuration</strong>: <div class="language-python highlight"><pre><span></span><code><span id=__span-11-1><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=n>n_trajectories</span> <span class=o>=</span> <span class=mi>500</span>                <span class=c1># Number of trajectories to generate</span>
</span><span id=__span-11-2><a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=n>n_timepoints</span> <span class=o>=</span> <span class=mi>20</span>                   <span class=c1># Timepoints per trajectory</span>
</span><span id=__span-11-3><a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a><span class=n>trajectory_types</span> <span class=o>=</span> <span class=p>[</span>
</span><span id=__span-11-4><a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a>    <span class=s1>&#39;early_to_late&#39;</span><span class=p>,</span>                <span class=c1># Early stage → Late stage</span>
</span><span id=__span-11-5><a id=__codelineno-11-5 name=__codelineno-11-5 href=#__codelineno-11-5></a>    <span class=s1>&#39;stage1_to_stage4&#39;</span><span class=p>,</span>             <span class=c1># Stage I → Stage IV</span>
</span><span id=__span-11-6><a id=__codelineno-11-6 name=__codelineno-11-6 href=#__codelineno-11-6></a>    <span class=s1>&#39;matched_pairs&#39;</span>                 <span class=c1># Patient-specific progressions</span>
</span><span id=__span-11-7><a id=__codelineno-11-7 name=__codelineno-11-7 href=#__codelineno-11-7></a><span class=p>]</span>
</span></code></pre></div></p> <p><strong>Expected outputs</strong>: <div class="language-text highlight"><pre><span></span><code><span id=__span-12-1><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a>data/interim/20251217_synthetic_data/kirc/
</span><span id=__span-12-2><a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a>├── early_to_late/
</span><span id=__span-12-3><a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a>│   ├── TCGA-XXX-YYY_to_TCGA-AAA-BBB.csv
</span><span id=__span-12-4><a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a>│   ├── TCGA-XXX-ZZZ_to_TCGA-CCC-DDD.csv
</span><span id=__span-12-5><a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a>│   └── ...                         # 500 trajectory files
</span><span id=__span-12-6><a id=__codelineno-12-6 name=__codelineno-12-6 href=#__codelineno-12-6></a>├── trajectory_metadata.csv         # Start/end points, stages
</span><span id=__span-12-7><a id=__codelineno-12-7 name=__codelineno-12-7 href=#__codelineno-12-7></a>└── generation_params.json          # Generation parameters
</span></code></pre></div></p> <p><strong>Validation</strong>: <div class="language-python highlight"><pre><span></span><code><span id=__span-13-1><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=kn>import</span><span class=w> </span><span class=nn>pandas</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>pd</span>
</span><span id=__span-13-2><a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=kn>import</span><span class=w> </span><span class=nn>glob</span>
</span><span id=__span-13-3><a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a>
</span><span id=__span-13-4><a id=__codelineno-13-4 name=__codelineno-13-4 href=#__codelineno-13-4></a><span class=c1># Count trajectories</span>
</span><span id=__span-13-5><a id=__codelineno-13-5 name=__codelineno-13-5 href=#__codelineno-13-5></a><span class=n>trajectory_files</span> <span class=o>=</span> <span class=n>glob</span><span class=o>.</span><span class=n>glob</span><span class=p>(</span><span class=s1>&#39;data/interim/*/kirc/early_to_late/*.csv&#39;</span><span class=p>)</span>
</span><span id=__span-13-6><a id=__codelineno-13-6 name=__codelineno-13-6 href=#__codelineno-13-6></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Generated </span><span class=si>{</span><span class=nb>len</span><span class=p>(</span><span class=n>trajectory_files</span><span class=p>)</span><span class=si>}</span><span class=s2> trajectories&quot;</span><span class=p>)</span>
</span><span id=__span-13-7><a id=__codelineno-13-7 name=__codelineno-13-7 href=#__codelineno-13-7></a><span class=c1># Expected: 500 trajectories</span>
</span><span id=__span-13-8><a id=__codelineno-13-8 name=__codelineno-13-8 href=#__codelineno-13-8></a>
</span><span id=__span-13-9><a id=__codelineno-13-9 name=__codelineno-13-9 href=#__codelineno-13-9></a><span class=c1># Check trajectory format</span>
</span><span id=__span-13-10><a id=__codelineno-13-10 name=__codelineno-13-10 href=#__codelineno-13-10></a><span class=n>traj</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=n>trajectory_files</span><span class=p>[</span><span class=mi>0</span><span class=p>],</span> <span class=n>index_col</span><span class=o>=</span><span class=mi>0</span><span class=p>)</span>
</span><span id=__span-13-11><a id=__codelineno-13-11 name=__codelineno-13-11 href=#__codelineno-13-11></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Trajectory shape: </span><span class=si>{</span><span class=n>traj</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-13-12><a id=__codelineno-13-12 name=__codelineno-13-12 href=#__codelineno-13-12></a><span class=c1># Expected: (20 timepoints, ~5000 genes)</span>
</span><span id=__span-13-13><a id=__codelineno-13-13 name=__codelineno-13-13 href=#__codelineno-13-13></a>
</span><span id=__span-13-14><a id=__codelineno-13-14 name=__codelineno-13-14 href=#__codelineno-13-14></a><span class=k>assert</span> <span class=n>traj</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>0</span><span class=p>]</span> <span class=o>==</span> <span class=mi>20</span><span class=p>,</span> <span class=s2>&quot;Wrong number of timepoints&quot;</span>
</span><span id=__span-13-15><a id=__codelineno-13-15 name=__codelineno-13-15 href=#__codelineno-13-15></a><span class=k>assert</span> <span class=n>traj</span><span class=o>.</span><span class=n>isnull</span><span class=p>()</span><span class=o>.</span><span class=n>sum</span><span class=p>()</span><span class=o>.</span><span class=n>sum</span><span class=p>()</span> <span class=o>==</span> <span class=mi>0</span><span class=p>,</span> <span class=s2>&quot;Missing values in trajectory&quot;</span>
</span><span id=__span-13-16><a id=__codelineno-13-16 name=__codelineno-13-16 href=#__codelineno-13-16></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;✓ Trajectory validation passed&quot;</span><span class=p>)</span>
</span></code></pre></div></p> <p><strong>Time</strong>: ~30 minutes<br> <strong>See</strong>: <a href=../step4-trajectories/ >Trajectory Tutorial</a></p> <hr> <h3 id=step-5-classification>Step 5: Classification<a class=headerlink href=#step-5-classification title="Permanent link">&para;</a></h3> <p><strong>Script</strong>: <code>scripts/pipeline_steps/5_classification.py</code></p> <p><strong>What it does</strong>: - Trains XGBoost classifier for stage prediction - Calculates SHAP values for interpretability - Identifies important gene signatures - Evaluates performance with cross-validation</p> <p><strong>Expected outputs</strong>: <div class="language-text highlight"><pre><span></span><code><span id=__span-14-1><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a>models/20251217_classification_kirc/
</span><span id=__span-14-2><a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a>├── classifier.pkl                  # Trained XGBoost model
</span><span id=__span-14-3><a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a>├── classification_metrics.json     # Accuracy, ROC AUC, etc.
</span><span id=__span-14-4><a id=__codelineno-14-4 name=__codelineno-14-4 href=#__codelineno-14-4></a>├── shap_values.npy                 # SHAP importance values
</span><span id=__span-14-5><a id=__codelineno-14-5 name=__codelineno-14-5 href=#__codelineno-14-5></a>├── important_genes.csv             # Top biomarker genes
</span><span id=__span-14-6><a id=__codelineno-14-6 name=__codelineno-14-6 href=#__codelineno-14-6></a>└── figures/
</span><span id=__span-14-7><a id=__codelineno-14-7 name=__codelineno-14-7 href=#__codelineno-14-7></a>    ├── confusion_matrix.png
</span><span id=__span-14-8><a id=__codelineno-14-8 name=__codelineno-14-8 href=#__codelineno-14-8></a>    ├── roc_curve.png
</span><span id=__span-14-9><a id=__codelineno-14-9 name=__codelineno-14-9 href=#__codelineno-14-9></a>    ├── shap_summary.png
</span><span id=__span-14-10><a id=__codelineno-14-10 name=__codelineno-14-10 href=#__codelineno-14-10></a>    └── shap_waterfall.png
</span></code></pre></div></p> <p><strong>Validation</strong>: <div class="language-python highlight"><pre><span></span><code><span id=__span-15-1><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=kn>import</span><span class=w> </span><span class=nn>json</span>
</span><span id=__span-15-2><a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a>
</span><span id=__span-15-3><a id=__codelineno-15-3 name=__codelineno-15-3 href=#__codelineno-15-3></a><span class=c1># Check performance metrics</span>
</span><span id=__span-15-4><a id=__codelineno-15-4 name=__codelineno-15-4 href=#__codelineno-15-4></a><span class=k>with</span> <span class=nb>open</span><span class=p>(</span><span class=s1>&#39;models/20251217_classification_kirc/classification_metrics.json&#39;</span><span class=p>)</span> <span class=k>as</span> <span class=n>f</span><span class=p>:</span>
</span><span id=__span-15-5><a id=__codelineno-15-5 name=__codelineno-15-5 href=#__codelineno-15-5></a>    <span class=n>metrics</span> <span class=o>=</span> <span class=n>json</span><span class=o>.</span><span class=n>load</span><span class=p>(</span><span class=n>f</span><span class=p>)</span>
</span><span id=__span-15-6><a id=__codelineno-15-6 name=__codelineno-15-6 href=#__codelineno-15-6></a>
</span><span id=__span-15-7><a id=__codelineno-15-7 name=__codelineno-15-7 href=#__codelineno-15-7></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Test Accuracy: </span><span class=si>{</span><span class=n>metrics</span><span class=p>[</span><span class=s1>&#39;test_accuracy&#39;</span><span class=p>]</span><span class=si>:</span><span class=s2>.3f</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-15-8><a id=__codelineno-15-8 name=__codelineno-15-8 href=#__codelineno-15-8></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;ROC AUC: </span><span class=si>{</span><span class=n>metrics</span><span class=p>[</span><span class=s1>&#39;test_roc_auc&#39;</span><span class=p>]</span><span class=si>:</span><span class=s2>.3f</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-15-9><a id=__codelineno-15-9 name=__codelineno-15-9 href=#__codelineno-15-9></a>
</span><span id=__span-15-10><a id=__codelineno-15-10 name=__codelineno-15-10 href=#__codelineno-15-10></a><span class=c1># Expected performance for KIRC dataset</span>
</span><span id=__span-15-11><a id=__codelineno-15-11 name=__codelineno-15-11 href=#__codelineno-15-11></a><span class=k>assert</span> <span class=n>metrics</span><span class=p>[</span><span class=s1>&#39;test_accuracy&#39;</span><span class=p>]</span> <span class=o>&gt;</span> <span class=mf>0.85</span><span class=p>,</span> <span class=s2>&quot;Low accuracy!&quot;</span>
</span><span id=__span-15-12><a id=__codelineno-15-12 name=__codelineno-15-12 href=#__codelineno-15-12></a><span class=k>assert</span> <span class=n>metrics</span><span class=p>[</span><span class=s1>&#39;test_roc_auc&#39;</span><span class=p>]</span> <span class=o>&gt;</span> <span class=mf>0.90</span><span class=p>,</span> <span class=s2>&quot;Low ROC AUC!&quot;</span>
</span><span id=__span-15-13><a id=__codelineno-15-13 name=__codelineno-15-13 href=#__codelineno-15-13></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;✓ Classification performance acceptable&quot;</span><span class=p>)</span>
</span></code></pre></div></p> <p><strong>Time</strong>: ~1 hour<br> <strong>See</strong>: <a href=../step5-classification/ >Classification Tutorial</a></p> <hr> <h3 id=step-6-enrichment-analysis>Step 6: Enrichment Analysis<a class=headerlink href=#step-6-enrichment-analysis title="Permanent link">&para;</a></h3> <p><strong>Script</strong>: <code>scripts/pipeline_steps/6_enrichment_analysis.py</code></p> <p><strong>What it does</strong>: - Calculates differential expression (DESeq-style) for each timepoint - Runs GSEA on ranked gene lists - Identifies enriched pathways along trajectories - Combines results into single dataset</p> <p><strong>Prerequisites</strong>: - GSEA CLI tool installed (<a href=../../GSEA_INSTALLATION/ >installation guide</a>) - ReactomePathways.gmt in <code>data/external/</code></p> <p><strong>Configuration</strong>: <div class="language-python highlight"><pre><span></span><code><span id=__span-16-1><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=n>n_threads</span> <span class=o>=</span> <span class=mi>8</span>                       <span class=c1># Parallel processing threads</span>
</span><span id=__span-16-2><a id=__codelineno-16-2 name=__codelineno-16-2 href=#__codelineno-16-2></a><span class=n>gsea_path</span> <span class=o>=</span> <span class=s1>&#39;./GSEA_4.3.2/gsea-cli.sh&#39;</span>
</span><span id=__span-16-3><a id=__codelineno-16-3 name=__codelineno-16-3 href=#__codelineno-16-3></a><span class=n>pathways_file</span> <span class=o>=</span> <span class=s1>&#39;data/external/ReactomePathways.gmt&#39;</span>
</span></code></pre></div></p> <p><strong>Expected outputs</strong>: <div class="language-text highlight"><pre><span></span><code><span id=__span-17-1><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a>data/processed/20251217_enrichment/
</span><span id=__span-17-2><a id=__codelineno-17-2 name=__codelineno-17-2 href=#__codelineno-17-2></a>├── deseq/
</span><span id=__span-17-3><a id=__codelineno-17-3 name=__codelineno-17-3 href=#__codelineno-17-3></a>│   └── early_to_late/
</span><span id=__span-17-4><a id=__codelineno-17-4 name=__codelineno-17-4 href=#__codelineno-17-4></a>│       ├── TCGA-XXX_to_YYY/
</span><span id=__span-17-5><a id=__codelineno-17-5 name=__codelineno-17-5 href=#__codelineno-17-5></a>│       │   ├── patient_tp0_foldchange.rnk
</span><span id=__span-17-6><a id=__codelineno-17-6 name=__codelineno-17-6 href=#__codelineno-17-6></a>│       │   ├── patient_tp1_foldchange.rnk
</span><span id=__span-17-7><a id=__codelineno-17-7 name=__codelineno-17-7 href=#__codelineno-17-7></a>│       │   └── gsea_tp0/
</span><span id=__span-17-8><a id=__codelineno-17-8 name=__codelineno-17-8 href=#__codelineno-17-8></a>│       │       ├── gsea_report_for_na_pos_*.tsv
</span><span id=__span-17-9><a id=__codelineno-17-9 name=__codelineno-17-9 href=#__codelineno-17-9></a>│       │       └── gsea_report_for_na_neg_*.tsv
</span><span id=__span-17-10><a id=__codelineno-17-10 name=__codelineno-17-10 href=#__codelineno-17-10></a>│       └── ...
</span><span id=__span-17-11><a id=__codelineno-17-11 name=__codelineno-17-11 href=#__codelineno-17-11></a>└── trajectory_enrichment.csv       # Final combined results
</span></code></pre></div></p> <p><strong>Final result format</strong>: <div class="language-text highlight"><pre><span></span><code><span id=__span-18-1><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a>Patient,Idx,Transition,NAME,ES,NES,FDR q-val
</span><span id=__span-18-2><a id=__codelineno-18-2 name=__codelineno-18-2 href=#__codelineno-18-2></a>TCGA-3Z-A93Z-01_to_TCGA-A3-A8OW-01,0,early_to_late,Cell Cycle,0.65,2.13,0.001
</span><span id=__span-18-3><a id=__codelineno-18-3 name=__codelineno-18-3 href=#__codelineno-18-3></a>TCGA-3Z-A93Z-01_to_TCGA-A3-A8OW-01,0,early_to_late,DNA Repair,0.52,1.87,0.012
</span><span id=__span-18-4><a id=__codelineno-18-4 name=__codelineno-18-4 href=#__codelineno-18-4></a>...
</span></code></pre></div></p> <p><strong>Validation</strong>: <div class="language-python highlight"><pre><span></span><code><span id=__span-19-1><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=kn>import</span><span class=w> </span><span class=nn>pandas</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>pd</span>
</span><span id=__span-19-2><a id=__codelineno-19-2 name=__codelineno-19-2 href=#__codelineno-19-2></a>
</span><span id=__span-19-3><a id=__codelineno-19-3 name=__codelineno-19-3 href=#__codelineno-19-3></a><span class=c1># Load enrichment results</span>
</span><span id=__span-19-4><a id=__codelineno-19-4 name=__codelineno-19-4 href=#__codelineno-19-4></a><span class=n>enrichment</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=s1>&#39;data/processed/20251217_enrichment/trajectory_enrichment.csv&#39;</span><span class=p>)</span>
</span><span id=__span-19-5><a id=__codelineno-19-5 name=__codelineno-19-5 href=#__codelineno-19-5></a>
</span><span id=__span-19-6><a id=__codelineno-19-6 name=__codelineno-19-6 href=#__codelineno-19-6></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Total enrichment results: </span><span class=si>{</span><span class=nb>len</span><span class=p>(</span><span class=n>enrichment</span><span class=p>)</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-19-7><a id=__codelineno-19-7 name=__codelineno-19-7 href=#__codelineno-19-7></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Unique patients: </span><span class=si>{</span><span class=n>enrichment</span><span class=p>[</span><span class=s1>&#39;Patient&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>nunique</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-19-8><a id=__codelineno-19-8 name=__codelineno-19-8 href=#__codelineno-19-8></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Unique pathways: </span><span class=si>{</span><span class=n>enrichment</span><span class=p>[</span><span class=s1>&#39;NAME&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>nunique</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-19-9><a id=__codelineno-19-9 name=__codelineno-19-9 href=#__codelineno-19-9></a>
</span><span id=__span-19-10><a id=__codelineno-19-10 name=__codelineno-19-10 href=#__codelineno-19-10></a><span class=c1># Check for significant pathways</span>
</span><span id=__span-19-11><a id=__codelineno-19-11 name=__codelineno-19-11 href=#__codelineno-19-11></a><span class=n>sig_pathways</span> <span class=o>=</span> <span class=n>enrichment</span><span class=p>[</span><span class=n>enrichment</span><span class=p>[</span><span class=s1>&#39;FDR q-val&#39;</span><span class=p>]</span> <span class=o>&lt;</span> <span class=mf>0.05</span><span class=p>]</span>
</span><span id=__span-19-12><a id=__codelineno-19-12 name=__codelineno-19-12 href=#__codelineno-19-12></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Significant pathways (FDR &lt; 0.05): </span><span class=si>{</span><span class=nb>len</span><span class=p>(</span><span class=n>sig_pathways</span><span class=p>)</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-19-13><a id=__codelineno-19-13 name=__codelineno-19-13 href=#__codelineno-19-13></a>
</span><span id=__span-19-14><a id=__codelineno-19-14 name=__codelineno-19-14 href=#__codelineno-19-14></a><span class=k>assert</span> <span class=nb>len</span><span class=p>(</span><span class=n>enrichment</span><span class=p>)</span> <span class=o>&gt;</span> <span class=mi>0</span><span class=p>,</span> <span class=s2>&quot;No enrichment results!&quot;</span>
</span><span id=__span-19-15><a id=__codelineno-19-15 name=__codelineno-19-15 href=#__codelineno-19-15></a><span class=k>assert</span> <span class=n>enrichment</span><span class=p>[</span><span class=s1>&#39;Patient&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>nunique</span><span class=p>()</span> <span class=o>==</span> <span class=mi>500</span><span class=p>,</span> <span class=s2>&quot;Missing trajectories!&quot;</span>
</span><span id=__span-19-16><a id=__codelineno-19-16 name=__codelineno-19-16 href=#__codelineno-19-16></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;✓ Enrichment analysis complete&quot;</span><span class=p>)</span>
</span></code></pre></div></p> <p><strong>Time</strong>: 2-6 hours (depends on CPU cores)<br> <strong>See</strong>: <a href=../step6-enrichment/ >Enrichment Tutorial</a></p> <hr> <h2 id=checking-pipeline-outputs>Checking Pipeline Outputs<a class=headerlink href=#checking-pipeline-outputs title="Permanent link">&para;</a></h2> <p>After running the complete pipeline, verify all outputs:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-20-1><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a><span class=c1># Run validation script</span>
</span><span id=__span-20-2><a id=__codelineno-20-2 name=__codelineno-20-2 href=#__codelineno-20-2></a>python<span class=w> </span>scripts/validate_pipeline_outputs.py
</span></code></pre></div> <p>Or manually check key files:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-21-1><a id=__codelineno-21-1 name=__codelineno-21-1 href=#__codelineno-21-1></a><span class=kn>import</span><span class=w> </span><span class=nn>os</span>
</span><span id=__span-21-2><a id=__codelineno-21-2 name=__codelineno-21-2 href=#__codelineno-21-2></a><span class=kn>import</span><span class=w> </span><span class=nn>pandas</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>pd</span>
</span><span id=__span-21-3><a id=__codelineno-21-3 name=__codelineno-21-3 href=#__codelineno-21-3></a>
</span><span id=__span-21-4><a id=__codelineno-21-4 name=__codelineno-21-4 href=#__codelineno-21-4></a><span class=c1># Check each step&#39;s outputs</span>
</span><span id=__span-21-5><a id=__codelineno-21-5 name=__codelineno-21-5 href=#__codelineno-21-5></a><span class=n>checks</span> <span class=o>=</span> <span class=p>{</span>
</span><span id=__span-21-6><a id=__codelineno-21-6 name=__codelineno-21-6 href=#__codelineno-21-6></a>    <span class=s1>&#39;Step 1&#39;</span><span class=p>:</span> <span class=s1>&#39;data/interim/preprocessed_KIRC_data/preprocessed_rnaseq.csv&#39;</span><span class=p>,</span>
</span><span id=__span-21-7><a id=__codelineno-21-7 name=__codelineno-21-7 href=#__codelineno-21-7></a>    <span class=s1>&#39;Step 2&#39;</span><span class=p>:</span> <span class=s1>&#39;models/20251217_models_KIRC/vae/vae_model.pt&#39;</span><span class=p>,</span>
</span><span id=__span-21-8><a id=__codelineno-21-8 name=__codelineno-21-8 href=#__codelineno-21-8></a>    <span class=s1>&#39;Step 3&#39;</span><span class=p>:</span> <span class=s1>&#39;reports/figures/reconstruction/latent_space_umap.png&#39;</span><span class=p>,</span>
</span><span id=__span-21-9><a id=__codelineno-21-9 name=__codelineno-21-9 href=#__codelineno-21-9></a>    <span class=s1>&#39;Step 4&#39;</span><span class=p>:</span> <span class=s1>&#39;data/interim/20251217_synthetic_data/kirc/trajectory_metadata.csv&#39;</span><span class=p>,</span>
</span><span id=__span-21-10><a id=__codelineno-21-10 name=__codelineno-21-10 href=#__codelineno-21-10></a>    <span class=s1>&#39;Step 5&#39;</span><span class=p>:</span> <span class=s1>&#39;models/20251217_classification_kirc/classifier.pkl&#39;</span><span class=p>,</span>
</span><span id=__span-21-11><a id=__codelineno-21-11 name=__codelineno-21-11 href=#__codelineno-21-11></a>    <span class=s1>&#39;Step 6&#39;</span><span class=p>:</span> <span class=s1>&#39;data/processed/20251217_enrichment/trajectory_enrichment.csv&#39;</span><span class=p>,</span>
</span><span id=__span-21-12><a id=__codelineno-21-12 name=__codelineno-21-12 href=#__codelineno-21-12></a><span class=p>}</span>
</span><span id=__span-21-13><a id=__codelineno-21-13 name=__codelineno-21-13 href=#__codelineno-21-13></a>
</span><span id=__span-21-14><a id=__codelineno-21-14 name=__codelineno-21-14 href=#__codelineno-21-14></a><span class=k>for</span> <span class=n>step</span><span class=p>,</span> <span class=n>filepath</span> <span class=ow>in</span> <span class=n>checks</span><span class=o>.</span><span class=n>items</span><span class=p>():</span>
</span><span id=__span-21-15><a id=__codelineno-21-15 name=__codelineno-21-15 href=#__codelineno-21-15></a>    <span class=n>exists</span> <span class=o>=</span> <span class=n>os</span><span class=o>.</span><span class=n>path</span><span class=o>.</span><span class=n>exists</span><span class=p>(</span><span class=n>filepath</span><span class=p>)</span>
</span><span id=__span-21-16><a id=__codelineno-21-16 name=__codelineno-21-16 href=#__codelineno-21-16></a>    <span class=n>status</span> <span class=o>=</span> <span class=s2>&quot;✓&quot;</span> <span class=k>if</span> <span class=n>exists</span> <span class=k>else</span> <span class=s2>&quot;✗&quot;</span>
</span><span id=__span-21-17><a id=__codelineno-21-17 name=__codelineno-21-17 href=#__codelineno-21-17></a>    <span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;</span><span class=si>{</span><span class=n>status</span><span class=si>}</span><span class=s2> </span><span class=si>{</span><span class=n>step</span><span class=si>}</span><span class=s2>: </span><span class=si>{</span><span class=n>filepath</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span></code></pre></div> <h2 id=expected-disk-usage>Expected Disk Usage<a class=headerlink href=#expected-disk-usage title="Permanent link">&para;</a></h2> <p>After completing the pipeline:</p> <div class="language-text highlight"><pre><span></span><code><span id=__span-22-1><a id=__codelineno-22-1 name=__codelineno-22-1 href=#__codelineno-22-1></a>data/raw/                    ~5 GB    (TCGA downloads)
</span><span id=__span-22-2><a id=__codelineno-22-2 name=__codelineno-22-2 href=#__codelineno-22-2></a>data/interim/                ~15 GB   (trajectories + intermediate files)
</span><span id=__span-22-3><a id=__codelineno-22-3 name=__codelineno-22-3 href=#__codelineno-22-3></a>data/processed/              ~10 GB   (enrichment results)
</span><span id=__span-22-4><a id=__codelineno-22-4 name=__codelineno-22-4 href=#__codelineno-22-4></a>models/                      ~1 GB    (trained models)
</span><span id=__span-22-5><a id=__codelineno-22-5 name=__codelineno-22-5 href=#__codelineno-22-5></a>reports/                     ~100 MB  (figures)
</span><span id=__span-22-6><a id=__codelineno-22-6 name=__codelineno-22-6 href=#__codelineno-22-6></a>-------------------------------------------
</span><span id=__span-22-7><a id=__codelineno-22-7 name=__codelineno-22-7 href=#__codelineno-22-7></a>Total:                       ~31 GB
</span></code></pre></div> <h2 id=performance-benchmarks>Performance Benchmarks<a class=headerlink href=#performance-benchmarks title="Permanent link">&para;</a></h2> <p>Typical runtimes on different systems:</p> <table> <thead> <tr> <th>System</th> <th>Step 2</th> <th>Step 6</th> <th>Total</th> <th>Config</th> </tr> </thead> <tbody> <tr> <td>HPC Cluster</td> <td>1 hr</td> <td>1 hr</td> <td>3 hrs</td> <td>48 cores, A100 GPU</td> </tr> <tr> <td>Desktop (2023)</td> <td>2 hrs</td> <td>2 hrs</td> <td>6 hrs</td> <td>16 cores, RTX 4080</td> </tr> <tr> <td>Desktop (2020)</td> <td>3 hrs</td> <td>4 hrs</td> <td>9 hrs</td> <td>8 cores, GTX 1080</td> </tr> <tr> <td>Laptop (2018)</td> <td>8 hrs</td> <td>6 hrs</td> <td>16 hrs</td> <td>4 cores, CPU only</td> </tr> </tbody> </table> <h2 id=troubleshooting>Troubleshooting<a class=headerlink href=#troubleshooting title="Permanent link">&para;</a></h2> <h3 id=pipeline-fails-at-step-x>Pipeline Fails at Step X<a class=headerlink href=#pipeline-fails-at-step-x title="Permanent link">&para;</a></h3> <p>Resume from the failing step:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-23-1><a id=__codelineno-23-1 name=__codelineno-23-1 href=#__codelineno-23-1></a><span class=c1># Fix the issue, then run from that step onward</span>
</span><span id=__span-23-2><a id=__codelineno-23-2 name=__codelineno-23-2 href=#__codelineno-23-2></a>python<span class=w> </span>scripts/pipeline_steps/X_step_name.py
</span></code></pre></div> <h3 id=out-of-disk-space>Out of Disk Space<a class=headerlink href=#out-of-disk-space title="Permanent link">&para;</a></h3> <p>Clean up intermediate files:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-24-1><a id=__codelineno-24-1 name=__codelineno-24-1 href=#__codelineno-24-1></a><span class=c1># Remove GSEA temporary files</span>
</span><span id=__span-24-2><a id=__codelineno-24-2 name=__codelineno-24-2 href=#__codelineno-24-2></a>rm<span class=w> </span>-rf<span class=w> </span>data/processed/*/deseq/*/gsea_*
</span><span id=__span-24-3><a id=__codelineno-24-3 name=__codelineno-24-3 href=#__codelineno-24-3></a>
</span><span id=__span-24-4><a id=__codelineno-24-4 name=__codelineno-24-4 href=#__codelineno-24-4></a><span class=c1># Remove old dated directories</span>
</span><span id=__span-24-5><a id=__codelineno-24-5 name=__codelineno-24-5 href=#__codelineno-24-5></a>rm<span class=w> </span>-rf<span class=w> </span>data/interim/202412*<span class=w>  </span><span class=c1># Keep only latest</span>
</span></code></pre></div> <h3 id=out-of-memory>Out of Memory<a class=headerlink href=#out-of-memory title="Permanent link">&para;</a></h3> <p>Reduce batch sizes and threads:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-25-1><a id=__codelineno-25-1 name=__codelineno-25-1 href=#__codelineno-25-1></a><span class=c1># In 2_models.py</span>
</span><span id=__span-25-2><a id=__codelineno-25-2 name=__codelineno-25-2 href=#__codelineno-25-2></a><span class=n>vae_config</span><span class=o>.</span><span class=n>BATCH_SIZE</span> <span class=o>=</span> <span class=mi>4</span>  <span class=c1># Reduce from 8</span>
</span><span id=__span-25-3><a id=__codelineno-25-3 name=__codelineno-25-3 href=#__codelineno-25-3></a>
</span><span id=__span-25-4><a id=__codelineno-25-4 name=__codelineno-25-4 href=#__codelineno-25-4></a><span class=c1># In 6_enrichment_analysis.py</span>
</span><span id=__span-25-5><a id=__codelineno-25-5 name=__codelineno-25-5 href=#__codelineno-25-5></a><span class=n>n_threads</span> <span class=o>=</span> <span class=mi>2</span>  <span class=c1># Reduce from 8</span>
</span></code></pre></div> <h2 id=next-steps>Next Steps<a class=headerlink href=#next-steps title="Permanent link">&para;</a></h2> <p>After completing the pipeline:</p> <ol> <li><strong>Analyze Results</strong>: Explore the enrichment results</li> <li><strong>Create Visualizations</strong>: Generate publication figures</li> <li><strong>Interpret Pathways</strong>: Map significant pathways to biology</li> <li><strong>Validate Findings</strong>: Test predictions experimentally</li> </ol> <p>See: - <a href=../visualization/ >Visualization Tutorial</a> - <a href=../../reproducibility/results/ >Result Interpretation Guide</a> - <a href=../../api/ >API Reference</a> for custom analyses</p> <h2 id=citation>Citation<a class=headerlink href=#citation title="Permanent link">&para;</a></h2> <p>If you use this pipeline in your research, please cite:</p> <div class="language-bibtex highlight"><pre><span></span><code><span id=__span-26-1><a id=__codelineno-26-1 name=__codelineno-26-1 href=#__codelineno-26-1></a><span class=nc>@software</span><span class=p>{</span><span class=nl>renalprog2024</span><span class=p>,</span>
</span><span id=__span-26-2><a id=__codelineno-26-2 name=__codelineno-26-2 href=#__codelineno-26-2></a><span class=w>  </span><span class=na>author</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{Prol-Castelo, Guillermo}</span><span class=p>,</span>
</span><span id=__span-26-3><a id=__codelineno-26-3 name=__codelineno-26-3 href=#__codelineno-26-3></a><span class=w>  </span><span class=na>title</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{renalprog: Cancer Progression Forecasting with Generative AI}</span><span class=p>,</span>
</span><span id=__span-26-4><a id=__codelineno-26-4 name=__codelineno-26-4 href=#__codelineno-26-4></a><span class=w>  </span><span class=na>year</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{2024}</span><span class=p>,</span>
</span><span id=__span-26-5><a id=__codelineno-26-5 name=__codelineno-26-5 href=#__codelineno-26-5></a><span class=w>  </span><span class=na>url</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{https://github.com/gprolcastelo/renalprog}</span>
</span><span id=__span-26-6><a id=__codelineno-26-6 name=__codelineno-26-6 href=#__codelineno-26-6></a><span class=p>}</span>
</span></code></pre></div> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="December 19, 2025 16:49:24 UTC"><span class=timeago datetime=2025-12-19T16:49:24+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="December 19, 2025 16:49:24 UTC">2025-12-19</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="December 19, 2025 16:49:24 UTC"><span class=timeago datetime=2025-12-19T16:49:24+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="December 19, 2025 16:49:24 UTC">2025-12-19</span> </span> </aside> <form class=md-feedback name=feedback hidden> <fieldset> <legend class=md-feedback__title> Was this page helpful? </legend> <div class=md-feedback__inner> <div class=md-feedback__list> <button class="md-feedback__icon md-icon" type=submit title="This page was helpful" data-md-value=1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m7 0c0 .8-.7 1.5-1.5 1.5S14 10.3 14 9.5 14.7 8 15.5 8s1.5.7 1.5 1.5m-5 7.73c-1.75 0-3.29-.73-4.19-1.81L9.23 14c.45.72 1.52 1.23 2.77 1.23s2.32-.51 2.77-1.23l1.42 1.42c-.9 1.08-2.44 1.81-4.19 1.81"/></svg> </button> <button class="md-feedback__icon md-icon" type=submit title="This page could be improved" data-md-value=0> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10m-6.5-4c.8 0 1.5.7 1.5 1.5s-.7 1.5-1.5 1.5-1.5-.7-1.5-1.5.7-1.5 1.5-1.5M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m2 4.5c1.75 0 3.29.72 4.19 1.81l-1.42 1.42C14.32 16.5 13.25 16 12 16s-2.32.5-2.77 1.23l-1.42-1.42C8.71 14.72 10.25 14 12 14"/></svg> </button> </div> <div class=md-feedback__note> <div data-md-value=1 hidden> Thanks for your feedback! </div> <div data-md-value=0 hidden> Thanks for your feedback! Help us improve by opening an issue. </div> </div> </div> </fieldset> </form> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../pretrained-models/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Using Pretrained Models"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Using Pretrained Models </div> </div> </a> <a href=../step1-data-processing/ class="md-footer__link md-footer__link--next" aria-label="Next: Step 1 - Data Processing"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Step 1 - Data Processing </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2022-2025 EVENFLOW Project </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/gprolcastelo/renalprog target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://pypi.org/project/renalprog/ target=_blank rel=noopener title=pypi.org class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.8 200.5c-7.7-30.9-22.3-54.2-53.4-54.2h-40.1v47.4c0 36.8-31.2 67.8-66.8 67.8H172.7c-29.2 0-53.4 25-53.4 54.3v101.8c0 29 25.2 46 53.4 54.3 33.8 9.9 66.3 11.7 106.8 0 26.9-7.8 53.4-23.5 53.4-54.3v-40.7H226.2v-13.6h160.2c31.1 0 42.6-21.7 53.4-54.2 11.2-33.5 10.7-65.7 0-108.6M286.2 444.7a20.4 20.4 0 1 1 0-40.7 20.4 20.4 0 1 1 0 40.7M167.8 248.1h106.8c29.7 0 53.4-24.5 53.4-54.3V91.9c0-29-24.4-50.7-53.4-55.6-35.8-5.9-74.7-5.6-106.8.1-45.2 8-53.4 24.7-53.4 55.6v40.7h106.9v13.6h-147c-31.1 0-58.3 18.7-66.8 54.2-9.8 40.7-10.2 66.1 0 108.6 7.6 31.6 25.7 54.2 56.8 54.2H101v-48.8c0-35.3 30.5-66.4 66.8-66.4m-6.6-183.4a20.4 20.4 0 1 1 0 40.8 20.4 20.4 0 1 1 0-40.8"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "search.suggest", "search.highlight", "search.share", "content.code.copy", "content.code.annotate", "toc.follow"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../../js/timeago.min.js></script> <script src=../../js/timeago_mkdocs_material.js></script> <script src=../../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>