<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="A Python package for simulating kidney cancer progression with synthetic data generation and machine learning."><meta name=author content="Guillermo Prol-Castelo (Barcelona Supercomputing Center), for the EVENFLOW European Project"><link href=https://github.com/gprolcastelo/renalprog/tutorials/step1-data-processing/ rel=canonical><link href=../complete-pipeline/ rel=prev><link href=../step2-vae-training/ rel=next><link rel=icon href=../../assets/images/kidneys.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>Step 1 - Data Processing - renalprog</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../css/timeago.css><link rel=stylesheet href=../../assets/_mkdocstrings.css><link rel=stylesheet href=../../stylesheets/extra.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#step-1-data-processing-pipeline class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=renalprog class="md-header__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../../assets/images/kidneys.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> renalprog </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Step 1 - Data Processing </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../installation/ class=md-tabs__link> Getting Started </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=../ class=md-tabs__link> Tutorials </a> </li> <li class=md-tabs__item> <a href=../../api/ class=md-tabs__link> API Reference </a> </li> <li class=md-tabs__item> <a href=../../advanced/ENRICHMENT_ANALYSIS/ class=md-tabs__link> Advanced Topics </a> </li> <li class=md-tabs__item> <a href=../../contributing/guidelines/ class=md-tabs__link> Contributing </a> </li> <li class=md-tabs__item> <a href=../../citation/ class=md-tabs__link> Citation & License </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=renalprog class="md-nav__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../../assets/images/kidneys.png alt=logo> </a> renalprog </label> <div class=md-nav__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> <span class=md-ellipsis> Getting Started </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Getting Started </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../installation/ class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=../quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class=md-nav__item> <a href=../data-requirements/ class=md-nav__link> <span class=md-ellipsis> Data Requirements </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3 checked> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex> <span class=md-ellipsis> Tutorials </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=true> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../pretrained-models/ class=md-nav__link> <span class=md-ellipsis> Using Pretrained Models </span> </a> </li> <li class=md-nav__item> <a href=../complete-pipeline/ class=md-nav__link> <span class=md-ellipsis> Complete Pipeline </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Step 1 - Data Processing </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Step 1 - Data Processing </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> <span class=md-ellipsis> Prerequisites </span> </a> </li> <li class=md-nav__item> <a href=#usage class=md-nav__link> <span class=md-ellipsis> Usage </span> </a> <nav class=md-nav aria-label=Usage> <ul class=md-nav__list> <li class=md-nav__item> <a href=#option-1-download-preprocessed-data-from-zenodo-recommended class=md-nav__link> <span class=md-ellipsis> Option 1: Download Preprocessed Data from Zenodo (Recommended) </span> </a> </li> <li class=md-nav__item> <a href=#option-2-process-raw-tcga-data-from-scratch class=md-nav__link> <span class=md-ellipsis> Option 2: Process Raw TCGA Data from Scratch </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#command-line-arguments class=md-nav__link> <span class=md-ellipsis> Command-Line Arguments </span> </a> </li> <li class=md-nav__item> <a href=#data-sources class=md-nav__link> <span class=md-ellipsis> Data Sources </span> </a> <nav class=md-nav aria-label="Data Sources"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#zenodo-preprocessed-data class=md-nav__link> <span class=md-ellipsis> Zenodo Preprocessed Data </span> </a> </li> <li class=md-nav__item> <a href=#raw-tcga-data class=md-nav__link> <span class=md-ellipsis> Raw TCGA Data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#processing-steps class=md-nav__link> <span class=md-ellipsis> Processing Steps </span> </a> <nav class=md-nav aria-label="Processing Steps"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#when-using-zenodo_preprocessed class=md-nav__link> <span class=md-ellipsis> When Using --zenodo_preprocessed </span> </a> </li> <li class=md-nav__item> <a href=#when-processing-raw-data class=md-nav__link> <span class=md-ellipsis> When Processing Raw Data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#output-structure class=md-nav__link> <span class=md-ellipsis> Output Structure </span> </a> <nav class=md-nav aria-label="Output Structure"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#using-zenodo_preprocessed class=md-nav__link> <span class=md-ellipsis> Using --zenodo_preprocessed </span> </a> </li> <li class=md-nav__item> <a href=#processing-raw-data class=md-nav__link> <span class=md-ellipsis> Processing Raw Data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#preprocessing-details class=md-nav__link> <span class=md-ellipsis> Preprocessing Details </span> </a> <nav class=md-nav aria-label="Preprocessing Details"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#gene-expression-filtering class=md-nav__link> <span class=md-ellipsis> Gene Expression Filtering </span> </a> </li> <li class=md-nav__item> <a href=#outlier-detection class=md-nav__link> <span class=md-ellipsis> Outlier Detection </span> </a> </li> <li class=md-nav__item> <a href=#duplicate-handling class=md-nav__link> <span class=md-ellipsis> Duplicate Handling </span> </a> </li> <li class=md-nav__item> <a href=#stage-classification class=md-nav__link> <span class=md-ellipsis> Stage Classification </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#complete-example-workflow class=md-nav__link> <span class=md-ellipsis> Complete Example Workflow </span> </a> <nav class=md-nav aria-label="Complete Example Workflow"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#reproducible-workflow-zenodo class=md-nav__link> <span class=md-ellipsis> Reproducible Workflow (Zenodo) </span> </a> </li> <li class=md-nav__item> <a href=#custom-workflow-raw-processing class=md-nav__link> <span class=md-ellipsis> Custom Workflow (Raw Processing) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#verification class=md-nav__link> <span class=md-ellipsis> Verification </span> </a> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#download-failures class=md-nav__link> <span class=md-ellipsis> Download Failures </span> </a> </li> <li class=md-nav__item> <a href=#clinical-data-mismatch class=md-nav__link> <span class=md-ellipsis> Clinical Data Mismatch </span> </a> </li> <li class=md-nav__item> <a href=#memory-issues class=md-nav__link> <span class=md-ellipsis> Memory Issues </span> </a> </li> <li class=md-nav__item> <a href=#missing-raw-data-files class=md-nav__link> <span class=md-ellipsis> Missing Raw Data Files </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> <li class=md-nav__item> <a href=#additional-resources class=md-nav__link> <span class=md-ellipsis> Additional Resources </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../step2-vae-training/ class=md-nav__link> <span class=md-ellipsis> Step 2 - VAE Training </span> </a> </li> <li class=md-nav__item> <a href=../step3-reconstruction/ class=md-nav__link> <span class=md-ellipsis> Step 3 - Reconstruction </span> </a> </li> <li class=md-nav__item> <a href=../step4-trajectories/ class=md-nav__link> <span class=md-ellipsis> Step 4 - Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../step5-classification/ class=md-nav__link> <span class=md-ellipsis> Step 5 - Classification </span> </a> </li> <li class=md-nav__item> <a href=../step6-enrichment/ class=md-nav__link> <span class=md-ellipsis> Step 6 - Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../visualization/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../api/config/ class=md-nav__link> <span class=md-ellipsis> Configuration </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4_3> <label class=md-nav__link for=__nav_4_3 id=__nav_4_3_label tabindex=0> <span class=md-ellipsis> Data Processing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_3_label aria-expanded=false> <label class=md-nav__title for=__nav_4_3> <span class="md-nav__icon md-icon"></span> Data Processing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/dataset/ class=md-nav__link> <span class=md-ellipsis> Dataset </span> </a> </li> <li class=md-nav__item> <a href=../../api/features/ class=md-nav__link> <span class=md-ellipsis> Features </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4_4> <label class=md-nav__link for=__nav_4_4 id=__nav_4_4_label tabindex=0> <span class=md-ellipsis> Modeling </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4_4> <span class="md-nav__icon md-icon"></span> Modeling </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/models/ class=md-nav__link> <span class=md-ellipsis> VAE Models </span> </a> </li> <li class=md-nav__item> <a href=../../api/training/ class=md-nav__link> <span class=md-ellipsis> Training </span> </a> </li> <li class=md-nav__item> <a href=../../api/prediction/ class=md-nav__link> <span class=md-ellipsis> Prediction </span> </a> </li> <li class=md-nav__item> <a href=../../api/trajectories/ class=md-nav__link> <span class=md-ellipsis> Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../../api/classification/ class=md-nav__link> <span class=md-ellipsis> Classification </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../api/enrichment/ class=md-nav__link> <span class=md-ellipsis> Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../../api/plots/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> <li class=md-nav__item> <a href=../../api/utils/ class=md-nav__link> <span class=md-ellipsis> Utilities </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> <span class=md-ellipsis> Advanced Topics </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Advanced Topics </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../advanced/ENRICHMENT_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> Enrichment Analysis Details </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/PATHWAY_HEATMAP/ class=md-nav__link> <span class=md-ellipsis> Pathway Heatmap Analysis </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/GSEA_INSTALLATION/ class=md-nav__link> <span class=md-ellipsis> GSEA Installation </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/R_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> R Analysis Scripts </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/performance/ class=md-nav__link> <span class=md-ellipsis> Performance Tuning </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex=0> <span class=md-ellipsis> Contributing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> Contributing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../contributing/guidelines/ class=md-nav__link> <span class=md-ellipsis> Guidelines </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex=0> <span class=md-ellipsis> Citation & License </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Citation & License </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../citation/ class=md-nav__link> <span class=md-ellipsis> How to Cite </span> </a> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> <li class=md-nav__item> <a href=../../acknowledgments/ class=md-nav__link> <span class=md-ellipsis> Acknowledgments </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> <span class=md-ellipsis> Prerequisites </span> </a> </li> <li class=md-nav__item> <a href=#usage class=md-nav__link> <span class=md-ellipsis> Usage </span> </a> <nav class=md-nav aria-label=Usage> <ul class=md-nav__list> <li class=md-nav__item> <a href=#option-1-download-preprocessed-data-from-zenodo-recommended class=md-nav__link> <span class=md-ellipsis> Option 1: Download Preprocessed Data from Zenodo (Recommended) </span> </a> </li> <li class=md-nav__item> <a href=#option-2-process-raw-tcga-data-from-scratch class=md-nav__link> <span class=md-ellipsis> Option 2: Process Raw TCGA Data from Scratch </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#command-line-arguments class=md-nav__link> <span class=md-ellipsis> Command-Line Arguments </span> </a> </li> <li class=md-nav__item> <a href=#data-sources class=md-nav__link> <span class=md-ellipsis> Data Sources </span> </a> <nav class=md-nav aria-label="Data Sources"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#zenodo-preprocessed-data class=md-nav__link> <span class=md-ellipsis> Zenodo Preprocessed Data </span> </a> </li> <li class=md-nav__item> <a href=#raw-tcga-data class=md-nav__link> <span class=md-ellipsis> Raw TCGA Data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#processing-steps class=md-nav__link> <span class=md-ellipsis> Processing Steps </span> </a> <nav class=md-nav aria-label="Processing Steps"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#when-using-zenodo_preprocessed class=md-nav__link> <span class=md-ellipsis> When Using --zenodo_preprocessed </span> </a> </li> <li class=md-nav__item> <a href=#when-processing-raw-data class=md-nav__link> <span class=md-ellipsis> When Processing Raw Data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#output-structure class=md-nav__link> <span class=md-ellipsis> Output Structure </span> </a> <nav class=md-nav aria-label="Output Structure"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#using-zenodo_preprocessed class=md-nav__link> <span class=md-ellipsis> Using --zenodo_preprocessed </span> </a> </li> <li class=md-nav__item> <a href=#processing-raw-data class=md-nav__link> <span class=md-ellipsis> Processing Raw Data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#preprocessing-details class=md-nav__link> <span class=md-ellipsis> Preprocessing Details </span> </a> <nav class=md-nav aria-label="Preprocessing Details"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#gene-expression-filtering class=md-nav__link> <span class=md-ellipsis> Gene Expression Filtering </span> </a> </li> <li class=md-nav__item> <a href=#outlier-detection class=md-nav__link> <span class=md-ellipsis> Outlier Detection </span> </a> </li> <li class=md-nav__item> <a href=#duplicate-handling class=md-nav__link> <span class=md-ellipsis> Duplicate Handling </span> </a> </li> <li class=md-nav__item> <a href=#stage-classification class=md-nav__link> <span class=md-ellipsis> Stage Classification </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#complete-example-workflow class=md-nav__link> <span class=md-ellipsis> Complete Example Workflow </span> </a> <nav class=md-nav aria-label="Complete Example Workflow"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#reproducible-workflow-zenodo class=md-nav__link> <span class=md-ellipsis> Reproducible Workflow (Zenodo) </span> </a> </li> <li class=md-nav__item> <a href=#custom-workflow-raw-processing class=md-nav__link> <span class=md-ellipsis> Custom Workflow (Raw Processing) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#verification class=md-nav__link> <span class=md-ellipsis> Verification </span> </a> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#download-failures class=md-nav__link> <span class=md-ellipsis> Download Failures </span> </a> </li> <li class=md-nav__item> <a href=#clinical-data-mismatch class=md-nav__link> <span class=md-ellipsis> Clinical Data Mismatch </span> </a> </li> <li class=md-nav__item> <a href=#memory-issues class=md-nav__link> <span class=md-ellipsis> Memory Issues </span> </a> </li> <li class=md-nav__item> <a href=#missing-raw-data-files class=md-nav__link> <span class=md-ellipsis> Missing Raw Data Files </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> <li class=md-nav__item> <a href=#additional-resources class=md-nav__link> <span class=md-ellipsis> Additional Resources </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=step-1-data-processing-pipeline>Step 1: Data Processing Pipeline<a class=headerlink href=#step-1-data-processing-pipeline title="Permanent link">&para;</a></h1> <p>This guide explains how to download and preprocess TCGA RNA-seq data for the RenalProg pipeline.</p> <h2 id=overview>Overview<a class=headerlink href=#overview title="Permanent link">&para;</a></h2> <p>The data processing pipeline performs the following steps:</p> <ol> <li><strong>Option 1: Download Preprocessed Data</strong> - Download preprocessed data from Zenodo (recommended for reproducibility)</li> <li><strong>Option 2: Process Raw Data</strong> - Download and process raw TCGA data from scratch</li> </ol> <p>The script provides flexible options for: - Downloading preprocessed RNAseq data from Zenodo repositories - Downloading raw TCGA data and processing it locally - Processing multiple cancer types (KIRC, BRCA) - Automatic clinical data filtering to match preprocessed samples</p> <h2 id=prerequisites>Prerequisites<a class=headerlink href=#prerequisites title="Permanent link">&para;</a></h2> <p>Before running the data processing pipeline, ensure you have:</p> <ul> <li><strong>Python environment</strong>: With required packages installed (pandas, numpy, scipy, etc.)</li> <li><strong>Internet connection</strong>: For downloading data from Zenodo or TCGA</li> <li><strong>Sufficient disk space</strong>: ~2-5 GB for raw data, ~500 MB for preprocessed data</li> </ul> <h2 id=usage>Usage<a class=headerlink href=#usage title="Permanent link">&para;</a></h2> <h3 id=option-1-download-preprocessed-data-from-zenodo-recommended>Option 1: Download Preprocessed Data from Zenodo (Recommended)<a class=headerlink href=#option-1-download-preprocessed-data-from-zenodo-recommended title="Permanent link">&para;</a></h3> <p>This is the fastest and most reproducible option. It downloads preprocessed RNAseq data from publicly available Zenodo repositories.</p> <h4 id=basic-usage-kirc>Basic Usage - KIRC<a class=headerlink href=#basic-usage-kirc title="Permanent link">&para;</a></h4> <div class="language-bash highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>python<span class=w> </span>scripts/pipeline_steps/1_data_processing.py<span class=w> </span>--zenodo_preprocessed
</span></code></pre></div> <p>This will: - Download preprocessed KIRC RNAseq data from Zenodo - Use existing clinical/phenotype files (assumes they're already downloaded) - Process and filter clinical data to match the RNAseq patients - Save everything to the appropriate directories</p> <h4 id=with-raw-data-download-kirc>With Raw Data Download - KIRC<a class=headerlink href=#with-raw-data-download-kirc title="Permanent link">&para;</a></h4> <div class="language-bash highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a>python<span class=w> </span>scripts/pipeline_steps/1_data_processing.py<span class=w> </span>--zenodo_preprocessed<span class=w> </span>--download_raw
</span></code></pre></div> <p>This will: - Download preprocessed KIRC RNAseq data from Zenodo - Download raw clinical and phenotype data from TCGA - Process and filter clinical data to match the RNAseq patients</p> <h4 id=for-brca-cancer-type>For BRCA Cancer Type<a class=headerlink href=#for-brca-cancer-type title="Permanent link">&para;</a></h4> <div class="language-bash highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a>python<span class=w> </span>scripts/pipeline_steps/1_data_processing.py<span class=w> </span>--zenodo_preprocessed<span class=w> </span>--cancer_type<span class=w> </span>BRCA<span class=w> </span>--download_raw
</span></code></pre></div> <h3 id=option-2-process-raw-tcga-data-from-scratch>Option 2: Process Raw TCGA Data from Scratch<a class=headerlink href=#option-2-process-raw-tcga-data-from-scratch title="Permanent link">&para;</a></h3> <p>This option downloads raw TCGA data and performs all preprocessing steps locally.</p> <h4 id=process-kirc-assumes-raw-data-already-downloaded>Process KIRC (assumes raw data already downloaded)<a class=headerlink href=#process-kirc-assumes-raw-data-already-downloaded title="Permanent link">&para;</a></h4> <div class="language-bash highlight"><pre><span></span><code><span id=__span-3-1><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a>python<span class=w> </span>scripts/pipeline_steps/1_data_processing.py
</span></code></pre></div> <h4 id=download-and-process-kirc>Download and Process KIRC<a class=headerlink href=#download-and-process-kirc title="Permanent link">&para;</a></h4> <div class="language-bash highlight"><pre><span></span><code><span id=__span-4-1><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a>python<span class=w> </span>scripts/pipeline_steps/1_data_processing.py<span class=w> </span>--download_raw
</span></code></pre></div> <h4 id=process-brca>Process BRCA<a class=headerlink href=#process-brca title="Permanent link">&para;</a></h4> <div class="language-bash highlight"><pre><span></span><code><span id=__span-5-1><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a>python<span class=w> </span>scripts/pipeline_steps/1_data_processing.py<span class=w> </span>--cancer_type<span class=w> </span>BRCA<span class=w> </span>--download_raw
</span></code></pre></div> <h2 id=command-line-arguments>Command-Line Arguments<a class=headerlink href=#command-line-arguments title="Permanent link">&para;</a></h2> <table> <thead> <tr> <th>Argument</th> <th>Type</th> <th>Default</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>--zenodo_preprocessed</code></td> <td>Flag</td> <td>False</td> <td>Download preprocessed data from Zenodo instead of processing raw data</td> </tr> <tr> <td><code>--cancer_type</code></td> <td>String</td> <td>'KIRC'</td> <td>Cancer type to process. Choices: 'KIRC', 'BRCA'</td> </tr> <tr> <td><code>--download_raw</code></td> <td>Flag</td> <td>False</td> <td>Download raw TCGA data (otherwise assumes data is already downloaded)</td> </tr> </tbody> </table> <h2 id=data-sources>Data Sources<a class=headerlink href=#data-sources title="Permanent link">&para;</a></h2> <h3 id=zenodo-preprocessed-data>Zenodo Preprocessed Data<a class=headerlink href=#zenodo-preprocessed-data title="Permanent link">&para;</a></h3> <h4 id=kirc>KIRC<a class=headerlink href=#kirc title="Permanent link">&para;</a></h4> <ul> <li><strong>Repository</strong>: https://zenodo.org/records/17987300</li> <li><strong>File</strong>: Static_KIRC.csv</li> <li><strong>Direct Download</strong>: https://zenodo.org/records/17987300/files/Static_KIRC.csv?download=1</li> </ul> <h4 id=brca>BRCA<a class=headerlink href=#brca title="Permanent link">&para;</a></h4> <ul> <li><strong>Repository</strong>: https://zenodo.org/records/17986123</li> <li><strong>File</strong>: Static_BRCA.csv</li> <li><strong>Direct Download</strong>: https://zenodo.org/records/17986123/files/Static_BRCA.csv?download=1</li> </ul> <h3 id=raw-tcga-data>Raw TCGA Data<a class=headerlink href=#raw-tcga-data title="Permanent link">&para;</a></h3> <p>When using <code>--download_raw</code>, the script downloads: - RNA-seq expression data from UCSC Xena - Clinical survival data - Phenotype data</p> <h2 id=processing-steps>Processing Steps<a class=headerlink href=#processing-steps title="Permanent link">&para;</a></h2> <h3 id=when-using-zenodo_preprocessed>When Using <code>--zenodo_preprocessed</code><a class=headerlink href=#when-using-zenodo_preprocessed title="Permanent link">&para;</a></h3> <ol> <li><strong>Download preprocessed RNAseq data</strong> from Zenodo</li> <li><strong>Download clinical/phenotype data</strong> (if <code>--download_raw</code> is specified)</li> <li><strong>Process clinical data</strong>:</li> <li>Load raw clinical and phenotype files</li> <li>Filter for specified cancer type</li> <li>Convert stage information (early/late classification)</li> <li>Filter to match only patients in the preprocessed RNAseq data</li> <li><strong>Copy control data from repository</strong>:</li> <li>Control samples are included in the repository at <code>controls/{CANCER_TYPE}/</code></li> <li>Copies <code>rnaseq_controls.csv</code> and <code>clinical_controls.csv</code> to preprocessed data directory</li> <li>No need to download or process controls separately</li> <li><strong>Save outputs</strong>:</li> <li><code>preprocessed_rnaseq.csv</code>: Preprocessed gene expression data</li> <li><code>clinical_data.csv</code>: Filtered clinical metadata</li> <li><code>stages.csv</code>: Stage information for matched patients</li> <li><code>controls/{CANCER_TYPE}_control_rnaseq.csv</code>: Control RNAseq data (copied from repo)</li> <li><code>controls/{CANCER_TYPE}_control_clinical.csv</code>: Control clinical data (copied from repo)</li> </ol> <h3 id=when-processing-raw-data>When Processing Raw Data<a class=headerlink href=#when-processing-raw-data title="Permanent link">&para;</a></h3> <ol> <li><strong>Download raw data</strong> (if <code>--download_raw</code> is specified)</li> <li><strong>Process downloaded data</strong>:</li> <li>Filter for specified cancer type</li> <li>Merge clinical and phenotype information</li> <li>Extract control samples (healthy tissue)</li> <li>Convert stages to early/late classification</li> <li><strong>Preprocess RNAseq data</strong>:</li> <li>Filter low-expression genes</li> <li>Detect and remove outlier samples using Mahalanobis distance</li> <li>Remove duplicate genes</li> <li><strong>Process control samples</strong>:</li> <li>Filter control data to match preprocessed genes</li> <li><strong>Save outputs</strong>:</li> <li><code>preprocessed_rnaseq.csv</code>: Preprocessed gene expression data</li> <li><code>preprocessing_info.csv/json</code>: Metadata about preprocessing steps</li> <li><code>stages.csv</code>: Stage information</li> <li><code>controls/preprocessed_control_rnaseq.csv</code>: Preprocessed control samples</li> </ol> <h2 id=output-structure>Output Structure<a class=headerlink href=#output-structure title="Permanent link">&para;</a></h2> <h3 id=using-zenodo_preprocessed>Using <code>--zenodo_preprocessed</code><a class=headerlink href=#using-zenodo_preprocessed title="Permanent link">&para;</a></h3> <div class="language-text highlight"><pre><span></span><code><span id=__span-6-1><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a>data/interim/preprocessed_{CANCER_TYPE}_data/
</span><span id=__span-6-2><a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a>├── Static_{CANCER_TYPE}.csv                           # Original downloaded file from Zenodo
</span><span id=__span-6-3><a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a>├── preprocessed_rnaseq.csv                            # Standard format (genes × patients)
</span><span id=__span-6-4><a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a>├── clinical_data.csv                                  # Filtered clinical data
</span><span id=__span-6-5><a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a>├── stages.csv                                         # Stage information
</span><span id=__span-6-6><a id=__codelineno-6-6 name=__codelineno-6-6 href=#__codelineno-6-6></a>└── controls/
</span><span id=__span-6-7><a id=__codelineno-6-7 name=__codelineno-6-7 href=#__codelineno-6-7></a>    ├── {CANCER_TYPE}_control_rnaseq.csv              # Control RNAseq (copied from repo)
</span><span id=__span-6-8><a id=__codelineno-6-8 name=__codelineno-6-8 href=#__codelineno-6-8></a>    └── {CANCER_TYPE}_control_clinical.csv            # Control clinical (copied from repo)
</span></code></pre></div> <p>Also creates (intermediate): <div class="language-text highlight"><pre><span></span><code><span id=__span-7-1><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a>data/interim/{CANCER_TYPE}_data/
</span><span id=__span-7-2><a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a>└── [Processed clinical/phenotype files before filtering]
</span></code></pre></div></p> <p><strong>Note</strong>: Control data is included in the repository at: <div class="language-text highlight"><pre><span></span><code><span id=__span-8-1><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a>controls/{CANCER_TYPE}/
</span><span id=__span-8-2><a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a>├── rnaseq_controls.csv
</span><span id=__span-8-3><a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a>└── clinical_controls.csv
</span></code></pre></div></p> <h3 id=processing-raw-data>Processing Raw Data<a class=headerlink href=#processing-raw-data title="Permanent link">&para;</a></h3> <div class="language-text highlight"><pre><span></span><code><span id=__span-9-1><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a>data/raw/
</span><span id=__span-9-2><a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>├── EB%2B%2BAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena
</span><span id=__span-9-3><a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a>├── Survival_SupplementalTable_S1_20171025_xena_sp
</span><span id=__span-9-4><a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a>└── TCGA_phenotype_denseDataOnlyDownload.tsv
</span><span id=__span-9-5><a id=__codelineno-9-5 name=__codelineno-9-5 href=#__codelineno-9-5></a>
</span><span id=__span-9-6><a id=__codelineno-9-6 name=__codelineno-9-6 href=#__codelineno-9-6></a>data/interim/{CANCER_TYPE}_data/
</span><span id=__span-9-7><a id=__codelineno-9-7 name=__codelineno-9-7 href=#__codelineno-9-7></a>├── {CANCER_TYPE}_rnaseq.csv
</span><span id=__span-9-8><a id=__codelineno-9-8 name=__codelineno-9-8 href=#__codelineno-9-8></a>├── {CANCER_TYPE}_clinical.csv
</span><span id=__span-9-9><a id=__codelineno-9-9 name=__codelineno-9-9 href=#__codelineno-9-9></a>└── controls/
</span><span id=__span-9-10><a id=__codelineno-9-10 name=__codelineno-9-10 href=#__codelineno-9-10></a>    ├── {CANCER_TYPE}_control_rnaseq.csv
</span><span id=__span-9-11><a id=__codelineno-9-11 name=__codelineno-9-11 href=#__codelineno-9-11></a>    └── {CANCER_TYPE}_control_clinical.csv
</span><span id=__span-9-12><a id=__codelineno-9-12 name=__codelineno-9-12 href=#__codelineno-9-12></a>
</span><span id=__span-9-13><a id=__codelineno-9-13 name=__codelineno-9-13 href=#__codelineno-9-13></a>data/interim/preprocessed_{CANCER_TYPE}_data/
</span><span id=__span-9-14><a id=__codelineno-9-14 name=__codelineno-9-14 href=#__codelineno-9-14></a>├── preprocessed_rnaseq.csv
</span><span id=__span-9-15><a id=__codelineno-9-15 name=__codelineno-9-15 href=#__codelineno-9-15></a>├── preprocessing_info.csv
</span><span id=__span-9-16><a id=__codelineno-9-16 name=__codelineno-9-16 href=#__codelineno-9-16></a>├── preprocessing_info.json
</span><span id=__span-9-17><a id=__codelineno-9-17 name=__codelineno-9-17 href=#__codelineno-9-17></a>├── stages.csv
</span><span id=__span-9-18><a id=__codelineno-9-18 name=__codelineno-9-18 href=#__codelineno-9-18></a>└── controls/
</span><span id=__span-9-19><a id=__codelineno-9-19 name=__codelineno-9-19 href=#__codelineno-9-19></a>    └── preprocessed_control_rnaseq.csv
</span></code></pre></div> <h2 id=preprocessing-details>Preprocessing Details<a class=headerlink href=#preprocessing-details title="Permanent link">&para;</a></h2> <p>When processing raw data, the following steps are applied:</p> <h3 id=gene-expression-filtering>Gene Expression Filtering<a class=headerlink href=#gene-expression-filtering title="Permanent link">&para;</a></h3> <ul> <li><strong>Low expression filtering</strong>: Removes genes with mean expression &lt; 0.5 and variance &lt; 0.5</li> <li><strong>Sample fraction</strong>: Genes must be expressed in at least 20% of samples</li> <li><strong>Statistical threshold</strong>: Alpha = 0.05 for statistical tests</li> </ul> <h3 id=outlier-detection>Outlier Detection<a class=headerlink href=#outlier-detection title="Permanent link">&para;</a></h3> <ul> <li><strong>Method</strong>: Mahalanobis distance with Minimum Covariance Determinant (MCD)</li> <li><strong>Purpose</strong>: Identifies and removes samples with atypical expression patterns</li> <li><strong>Seed</strong>: 2023 (for reproducibility)</li> </ul> <h3 id=duplicate-handling>Duplicate Handling<a class=headerlink href=#duplicate-handling title="Permanent link">&para;</a></h3> <ul> <li><strong>Strategy</strong>: Keep first occurrence, remove subsequent duplicates</li> <li><strong>Applied to</strong>: Both gene names and sample IDs</li> </ul> <h3 id=stage-classification>Stage Classification<a class=headerlink href=#stage-classification title="Permanent link">&para;</a></h3> <ul> <li><strong>Early stage</strong>: Stage I and Stage II</li> <li><strong>Late stage</strong>: Stage III and Stage IV</li> </ul> <h2 id=complete-example-workflow>Complete Example Workflow<a class=headerlink href=#complete-example-workflow title="Permanent link">&para;</a></h2> <h3 id=reproducible-workflow-zenodo>Reproducible Workflow (Zenodo)<a class=headerlink href=#reproducible-workflow-zenodo title="Permanent link">&para;</a></h3> <p>For maximum reproducibility using published preprocessed data:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-10-1><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=c1># Download preprocessed RNAseq and raw clinical data</span>
</span><span id=__span-10-2><a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a>python<span class=w> </span>scripts/pipeline_steps/1_data_processing.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-10-3><a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a><span class=w>  </span>--zenodo_preprocessed<span class=w> </span><span class=se>\</span>
</span><span id=__span-10-4><a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=w>  </span>--download_raw<span class=w> </span><span class=se>\</span>
</span><span id=__span-10-5><a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a><span class=w>  </span>--cancer_type<span class=w> </span>KIRC
</span></code></pre></div> <h3 id=custom-workflow-raw-processing>Custom Workflow (Raw Processing)<a class=headerlink href=#custom-workflow-raw-processing title="Permanent link">&para;</a></h3> <p>For full control over preprocessing parameters:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-11-1><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=c1># Download and process everything from scratch</span>
</span><span id=__span-11-2><a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a>python<span class=w> </span>scripts/pipeline_steps/1_data_processing.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-11-3><a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a><span class=w>  </span>--download_raw<span class=w> </span><span class=se>\</span>
</span><span id=__span-11-4><a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a><span class=w>  </span>--cancer_type<span class=w> </span>KIRC
</span></code></pre></div> <h2 id=verification>Verification<a class=headerlink href=#verification title="Permanent link">&para;</a></h2> <p>After running the script, verify the outputs:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-12-1><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=c1># Check preprocessed data dimensions</span>
</span><span id=__span-12-2><a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a>python<span class=w> </span>-c<span class=w> </span><span class=s2>&quot;</span>
</span><span id=__span-12-3><a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a><span class=s2>import pandas as pd</span>
</span><span id=__span-12-4><a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a><span class=s2>data = pd.read_csv(&#39;data/interim/preprocessed_KIRC_data/preprocessed_rnaseq.csv&#39;, index_col=0)</span>
</span><span id=__span-12-5><a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a><span class=s2>clinical = pd.read_csv(&#39;data/interim/preprocessed_KIRC_data/clinical_data.csv&#39;, index_col=0)</span>
</span><span id=__span-12-6><a id=__codelineno-12-6 name=__codelineno-12-6 href=#__codelineno-12-6></a><span class=s2>print(f&#39;RNAseq: {data.shape[0]} genes × {data.shape[1]} patients&#39;)</span>
</span><span id=__span-12-7><a id=__codelineno-12-7 name=__codelineno-12-7 href=#__codelineno-12-7></a><span class=s2>print(f&#39;Clinical: {clinical.shape[0]} patients × {clinical.shape[1]} features&#39;)</span>
</span><span id=__span-12-8><a id=__codelineno-12-8 name=__codelineno-12-8 href=#__codelineno-12-8></a><span class=s2>&quot;</span>
</span></code></pre></div> <h2 id=troubleshooting>Troubleshooting<a class=headerlink href=#troubleshooting title="Permanent link">&para;</a></h2> <h3 id=download-failures>Download Failures<a class=headerlink href=#download-failures title="Permanent link">&para;</a></h3> <p>If downloads from Zenodo fail: - Check your internet connection - Verify the Zenodo URLs are accessible - Try using a VPN if regional restrictions apply</p> <h3 id=clinical-data-mismatch>Clinical Data Mismatch<a class=headerlink href=#clinical-data-mismatch title="Permanent link">&para;</a></h3> <p>If patient IDs don't match between RNAseq and clinical data: - Ensure you're using the same cancer type for both datasets - Verify that raw data files are not corrupted - Check that the preprocessing completed successfully</p> <h3 id=memory-issues>Memory Issues<a class=headerlink href=#memory-issues title="Permanent link">&para;</a></h3> <p>For large datasets: - Process on a machine with sufficient RAM (recommend 16+ GB) - Consider processing cancer types separately - Monitor memory usage during preprocessing</p> <h3 id=missing-raw-data-files>Missing Raw Data Files<a class=headerlink href=#missing-raw-data-files title="Permanent link">&para;</a></h3> <p>If raw data files are not found: - Use the <code>--download_raw</code> flag to download them automatically - Or manually download from UCSC Xena and place in <code>data/raw/</code></p> <h2 id=next-steps>Next Steps<a class=headerlink href=#next-steps title="Permanent link">&para;</a></h2> <p>After completing the data processing:</p> <ol> <li><strong>Verify outputs</strong>: Check that all expected files were created</li> <li><strong>Inspect data quality</strong>: Review preprocessing statistics</li> <li><strong>Proceed to Step 2</strong>: Train VAE models on the preprocessed data</li> </ol> <div class="language-bash highlight"><pre><span></span><code><span id=__span-13-1><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a>python<span class=w> </span>scripts/pipeline_steps/2_models.py
</span></code></pre></div> <h2 id=additional-resources>Additional Resources<a class=headerlink href=#additional-resources title="Permanent link">&para;</a></h2> <ul> <li><a href=https://portal.gdc.cancer.gov/ >TCGA Data Portal</a></li> <li><a href=https://xenabrowser.net/ >UCSC Xena Browser</a></li> <li><a href=https://zenodo.org/ >Zenodo Repository Browser</a></li> <li><a href=../../advanced/ENRICHMENT_ANALYSIS/ >Feature Selection Guide</a></li> </ul> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="January 28, 2026 15:39:22 UTC"><span class=timeago datetime=2026-01-28T15:39:22+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="January 28, 2026 15:39:22 UTC">2026-01-28</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="December 19, 2025 16:49:24 UTC"><span class=timeago datetime=2025-12-19T16:49:24+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="December 19, 2025 16:49:24 UTC">2025-12-19</span> </span> </aside> <form class=md-feedback name=feedback hidden> <fieldset> <legend class=md-feedback__title> Was this page helpful? </legend> <div class=md-feedback__inner> <div class=md-feedback__list> <button class="md-feedback__icon md-icon" type=submit title="This page was helpful" data-md-value=1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m7 0c0 .8-.7 1.5-1.5 1.5S14 10.3 14 9.5 14.7 8 15.5 8s1.5.7 1.5 1.5m-5 7.73c-1.75 0-3.29-.73-4.19-1.81L9.23 14c.45.72 1.52 1.23 2.77 1.23s2.32-.51 2.77-1.23l1.42 1.42c-.9 1.08-2.44 1.81-4.19 1.81"/></svg> </button> <button class="md-feedback__icon md-icon" type=submit title="This page could be improved" data-md-value=0> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10m-6.5-4c.8 0 1.5.7 1.5 1.5s-.7 1.5-1.5 1.5-1.5-.7-1.5-1.5.7-1.5 1.5-1.5M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m2 4.5c1.75 0 3.29.72 4.19 1.81l-1.42 1.42C14.32 16.5 13.25 16 12 16s-2.32.5-2.77 1.23l-1.42-1.42C8.71 14.72 10.25 14 12 14"/></svg> </button> </div> <div class=md-feedback__note> <div data-md-value=1 hidden> Thanks for your feedback! </div> <div data-md-value=0 hidden> Thanks for your feedback! Help us improve by opening an issue. </div> </div> </div> </fieldset> </form> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../complete-pipeline/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Complete Pipeline"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Complete Pipeline </div> </div> </a> <a href=../step2-vae-training/ class="md-footer__link md-footer__link--next" aria-label="Next: Step 2 - VAE Training"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Step 2 - VAE Training </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2022-2025 EVENFLOW Project </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/gprolcastelo/renalprog target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://pypi.org/project/renalprog/ target=_blank rel=noopener title=pypi.org class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.8 200.5c-7.7-30.9-22.3-54.2-53.4-54.2h-40.1v47.4c0 36.8-31.2 67.8-66.8 67.8H172.7c-29.2 0-53.4 25-53.4 54.3v101.8c0 29 25.2 46 53.4 54.3 33.8 9.9 66.3 11.7 106.8 0 26.9-7.8 53.4-23.5 53.4-54.3v-40.7H226.2v-13.6h160.2c31.1 0 42.6-21.7 53.4-54.2 11.2-33.5 10.7-65.7 0-108.6M286.2 444.7a20.4 20.4 0 1 1 0-40.7 20.4 20.4 0 1 1 0 40.7M167.8 248.1h106.8c29.7 0 53.4-24.5 53.4-54.3V91.9c0-29-24.4-50.7-53.4-55.6-35.8-5.9-74.7-5.6-106.8.1-45.2 8-53.4 24.7-53.4 55.6v40.7h106.9v13.6h-147c-31.1 0-58.3 18.7-66.8 54.2-9.8 40.7-10.2 66.1 0 108.6 7.6 31.6 25.7 54.2 56.8 54.2H101v-48.8c0-35.3 30.5-66.4 66.8-66.4m-6.6-183.4a20.4 20.4 0 1 1 0 40.8 20.4 20.4 0 1 1 0-40.8"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "search.suggest", "search.highlight", "search.share", "content.code.copy", "content.code.annotate", "toc.follow"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../../js/timeago.min.js></script> <script src=../../js/timeago_mkdocs_material.js></script> <script src=../../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>