<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="A Python package for simulating kidney cancer progression with synthetic data generation and machine learning."><meta name=author content="Guillermo Prol-Castelo (Barcelona Supercomputing Center), for the EVENFLOW European Project"><link href=https://github.com/gprolcastelo/renalprog/tutorials/step6-enrichment/ rel=canonical><link href=../step5-classification/ rel=prev><link href=../visualization/ rel=next><link rel=icon href=../../assets/images/kidneys.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>Step 6 - Enrichment - renalprog</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../css/timeago.css><link rel=stylesheet href=../../assets/_mkdocstrings.css><link rel=stylesheet href=../../stylesheets/extra.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#step-6-enrichment-analysis-pipeline class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=renalprog class="md-header__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../../assets/images/kidneys.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> renalprog </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Step 6 - Enrichment </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../installation/ class=md-tabs__link> Getting Started </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=../ class=md-tabs__link> Tutorials </a> </li> <li class=md-tabs__item> <a href=../../api/ class=md-tabs__link> API Reference </a> </li> <li class=md-tabs__item> <a href=../../advanced/ENRICHMENT_ANALYSIS/ class=md-tabs__link> Advanced Topics </a> </li> <li class=md-tabs__item> <a href=../../contributing/guidelines/ class=md-tabs__link> Contributing </a> </li> <li class=md-tabs__item> <a href=../../citation/ class=md-tabs__link> Citation & License </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=renalprog class="md-nav__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../../assets/images/kidneys.png alt=logo> </a> renalprog </label> <div class=md-nav__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> <span class=md-ellipsis> Getting Started </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Getting Started </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../installation/ class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=../quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class=md-nav__item> <a href=../data-requirements/ class=md-nav__link> <span class=md-ellipsis> Data Requirements </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3 checked> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex> <span class=md-ellipsis> Tutorials </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=true> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../pretrained-models/ class=md-nav__link> <span class=md-ellipsis> Using Pretrained Models </span> </a> </li> <li class=md-nav__item> <a href=../complete-pipeline/ class=md-nav__link> <span class=md-ellipsis> Complete Pipeline </span> </a> </li> <li class=md-nav__item> <a href=../step1-data-processing/ class=md-nav__link> <span class=md-ellipsis> Step 1 - Data Processing </span> </a> </li> <li class=md-nav__item> <a href=../step2-vae-training/ class=md-nav__link> <span class=md-ellipsis> Step 2 - VAE Training </span> </a> </li> <li class=md-nav__item> <a href=../step3-reconstruction/ class=md-nav__link> <span class=md-ellipsis> Step 3 - Reconstruction </span> </a> </li> <li class=md-nav__item> <a href=../step4-trajectories/ class=md-nav__link> <span class=md-ellipsis> Step 4 - Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../step5-classification/ class=md-nav__link> <span class=md-ellipsis> Step 5 - Classification </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Step 6 - Enrichment </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Step 6 - Enrichment </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> <span class=md-ellipsis> Prerequisites </span> </a> </li> <li class=md-nav__item> <a href=#usage class=md-nav__link> <span class=md-ellipsis> Usage </span> </a> <nav class=md-nav aria-label=Usage> <ul class=md-nav__list> <li class=md-nav__item> <a href=#basic-usage-with-defaults class=md-nav__link> <span class=md-ellipsis> Basic Usage (with defaults) </span> </a> </li> <li class=md-nav__item> <a href=#custom-configuration class=md-nav__link> <span class=md-ellipsis> Custom Configuration </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#configuration-parameters class=md-nav__link> <span class=md-ellipsis> Configuration Parameters </span> </a> <nav class=md-nav aria-label="Configuration Parameters"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#required-parameter class=md-nav__link> <span class=md-ellipsis> Required Parameter </span> </a> </li> <li class=md-nav__item> <a href=#optional-parameters-environment-variables class=md-nav__link> <span class=md-ellipsis> Optional Parameters (Environment Variables) </span> </a> </li> <li class=md-nav__item> <a href=#default-paths-for-other-cancer-types class=md-nav__link> <span class=md-ellipsis> Default Paths for Other Cancer Types </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#available-gsea-parameters class=md-nav__link> <span class=md-ellipsis> Available GSEA Parameters </span> </a> </li> <li class=md-nav__item> <a href=#functions-supporting-gsea-parameters class=md-nav__link> <span class=md-ellipsis> Functions Supporting GSEA Parameters </span> </a> </li> <li class=md-nav__item> <a href=#pipeline-execution-steps class=md-nav__link> <span class=md-ellipsis> Pipeline Execution Steps </span> </a> <nav class=md-nav aria-label="Pipeline Execution Steps"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-generate-deseq2-greasy-file class=md-nav__link> <span class=md-ellipsis> Step 1: Generate DESeq2 Greasy File </span> </a> </li> <li class=md-nav__item> <a href=#step-2-generate-gsea-greasy-file class=md-nav__link> <span class=md-ellipsis> Step 2: Generate GSEA Greasy File </span> </a> </li> <li class=md-nav__item> <a href=#step-3-final-trajectory-formatting class=md-nav__link> <span class=md-ellipsis> Step 3: Final Trajectory Formatting </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#complete-example-workflow class=md-nav__link> <span class=md-ellipsis> Complete Example Workflow </span> </a> </li> <li class=md-nav__item> <a href=#background-execution class=md-nav__link> <span class=md-ellipsis> Background Execution </span> </a> </li> <li class=md-nav__item> <a href=#output-files class=md-nav__link> <span class=md-ellipsis> Output Files </span> </a> <nav class=md-nav aria-label="Output Files"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#generated-greasy-files class=md-nav__link> <span class=md-ellipsis> Generated Greasy Files: </span> </a> </li> <li class=md-nav__item> <a href=#final-output class=md-nav__link> <span class=md-ellipsis> Final Output: </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#gsea-not-found class=md-nav__link> <span class=md-ellipsis> GSEA Not Found </span> </a> </li> <li class=md-nav__item> <a href=#no-csv-files-found class=md-nav__link> <span class=md-ellipsis> No CSV Files Found </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#additional-resources class=md-nav__link> <span class=md-ellipsis> Additional Resources </span> </a> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../visualization/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../api/config/ class=md-nav__link> <span class=md-ellipsis> Configuration </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4_3> <label class=md-nav__link for=__nav_4_3 id=__nav_4_3_label tabindex=0> <span class=md-ellipsis> Data Processing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_3_label aria-expanded=false> <label class=md-nav__title for=__nav_4_3> <span class="md-nav__icon md-icon"></span> Data Processing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/dataset/ class=md-nav__link> <span class=md-ellipsis> Dataset </span> </a> </li> <li class=md-nav__item> <a href=../../api/features/ class=md-nav__link> <span class=md-ellipsis> Features </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4_4> <label class=md-nav__link for=__nav_4_4 id=__nav_4_4_label tabindex=0> <span class=md-ellipsis> Modeling </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4_4> <span class="md-nav__icon md-icon"></span> Modeling </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/models/ class=md-nav__link> <span class=md-ellipsis> VAE Models </span> </a> </li> <li class=md-nav__item> <a href=../../api/training/ class=md-nav__link> <span class=md-ellipsis> Training </span> </a> </li> <li class=md-nav__item> <a href=../../api/prediction/ class=md-nav__link> <span class=md-ellipsis> Prediction </span> </a> </li> <li class=md-nav__item> <a href=../../api/trajectories/ class=md-nav__link> <span class=md-ellipsis> Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../../api/classification/ class=md-nav__link> <span class=md-ellipsis> Classification </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../api/enrichment/ class=md-nav__link> <span class=md-ellipsis> Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../../api/plots/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> <li class=md-nav__item> <a href=../../api/utils/ class=md-nav__link> <span class=md-ellipsis> Utilities </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> <span class=md-ellipsis> Advanced Topics </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Advanced Topics </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../advanced/ENRICHMENT_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> Enrichment Analysis Details </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/PATHWAY_HEATMAP/ class=md-nav__link> <span class=md-ellipsis> Pathway Heatmap Analysis </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/GSEA_INSTALLATION/ class=md-nav__link> <span class=md-ellipsis> GSEA Installation </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/R_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> R Analysis Scripts </span> </a> </li> <li class=md-nav__item> <a href=../../advanced/performance/ class=md-nav__link> <span class=md-ellipsis> Performance Tuning </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex=0> <span class=md-ellipsis> Contributing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> Contributing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../contributing/guidelines/ class=md-nav__link> <span class=md-ellipsis> Guidelines </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex=0> <span class=md-ellipsis> Citation & License </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Citation & License </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../citation/ class=md-nav__link> <span class=md-ellipsis> How to Cite </span> </a> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> <li class=md-nav__item> <a href=../../acknowledgments/ class=md-nav__link> <span class=md-ellipsis> Acknowledgments </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> <span class=md-ellipsis> Prerequisites </span> </a> </li> <li class=md-nav__item> <a href=#usage class=md-nav__link> <span class=md-ellipsis> Usage </span> </a> <nav class=md-nav aria-label=Usage> <ul class=md-nav__list> <li class=md-nav__item> <a href=#basic-usage-with-defaults class=md-nav__link> <span class=md-ellipsis> Basic Usage (with defaults) </span> </a> </li> <li class=md-nav__item> <a href=#custom-configuration class=md-nav__link> <span class=md-ellipsis> Custom Configuration </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#configuration-parameters class=md-nav__link> <span class=md-ellipsis> Configuration Parameters </span> </a> <nav class=md-nav aria-label="Configuration Parameters"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#required-parameter class=md-nav__link> <span class=md-ellipsis> Required Parameter </span> </a> </li> <li class=md-nav__item> <a href=#optional-parameters-environment-variables class=md-nav__link> <span class=md-ellipsis> Optional Parameters (Environment Variables) </span> </a> </li> <li class=md-nav__item> <a href=#default-paths-for-other-cancer-types class=md-nav__link> <span class=md-ellipsis> Default Paths for Other Cancer Types </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#available-gsea-parameters class=md-nav__link> <span class=md-ellipsis> Available GSEA Parameters </span> </a> </li> <li class=md-nav__item> <a href=#functions-supporting-gsea-parameters class=md-nav__link> <span class=md-ellipsis> Functions Supporting GSEA Parameters </span> </a> </li> <li class=md-nav__item> <a href=#pipeline-execution-steps class=md-nav__link> <span class=md-ellipsis> Pipeline Execution Steps </span> </a> <nav class=md-nav aria-label="Pipeline Execution Steps"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-generate-deseq2-greasy-file class=md-nav__link> <span class=md-ellipsis> Step 1: Generate DESeq2 Greasy File </span> </a> </li> <li class=md-nav__item> <a href=#step-2-generate-gsea-greasy-file class=md-nav__link> <span class=md-ellipsis> Step 2: Generate GSEA Greasy File </span> </a> </li> <li class=md-nav__item> <a href=#step-3-final-trajectory-formatting class=md-nav__link> <span class=md-ellipsis> Step 3: Final Trajectory Formatting </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#complete-example-workflow class=md-nav__link> <span class=md-ellipsis> Complete Example Workflow </span> </a> </li> <li class=md-nav__item> <a href=#background-execution class=md-nav__link> <span class=md-ellipsis> Background Execution </span> </a> </li> <li class=md-nav__item> <a href=#output-files class=md-nav__link> <span class=md-ellipsis> Output Files </span> </a> <nav class=md-nav aria-label="Output Files"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#generated-greasy-files class=md-nav__link> <span class=md-ellipsis> Generated Greasy Files: </span> </a> </li> <li class=md-nav__item> <a href=#final-output class=md-nav__link> <span class=md-ellipsis> Final Output: </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#gsea-not-found class=md-nav__link> <span class=md-ellipsis> GSEA Not Found </span> </a> </li> <li class=md-nav__item> <a href=#no-csv-files-found class=md-nav__link> <span class=md-ellipsis> No CSV Files Found </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#additional-resources class=md-nav__link> <span class=md-ellipsis> Additional Resources </span> </a> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=step-6-enrichment-analysis-pipeline>Step 6: Enrichment Analysis Pipeline<a class=headerlink href=#step-6-enrichment-analysis-pipeline title="Permanent link">&para;</a></h1> <p>This guide explains how to run the enrichment analysis pipeline for trajectory data using DESeq2 and GSEA (Gene Set Enrichment Analysis).</p> <h2 id=overview>Overview<a class=headerlink href=#overview title="Permanent link">&para;</a></h2> <p>The enrichment pipeline performs three main steps:</p> <ol> <li><strong>DESeq2 Analysis</strong>: Performs differential expression analysis on trajectory data</li> <li><strong>GSEA Analysis</strong>: Runs Gene Set Enrichment Analysis on the DESeq2 results</li> <li><strong>Trajectory Formatting</strong>: Creates the final trajectory dataset with enrichment results</li> </ol> <p>The pipeline generates "greasy files" (batch job files) that need to be executed manually, allowing you to choose the best execution method for your computing environment.</p> <h2 id=prerequisites>Prerequisites<a class=headerlink href=#prerequisites title="Permanent link">&para;</a></h2> <p>Before running the enrichment pipeline, ensure you have:</p> <ul> <li><strong>GSEA installed</strong>: The pipeline expects GSEA CLI to be available at <code>GSEA_4.3.2/gsea-cli.sh</code></li> <li>See <a href=../../advanced/GSEA_INSTALLATION/ >GSEA Installation Guide</a> for setup instructions</li> <li><strong>Python environment</strong>: With required packages (DESeq2, pandas, etc.)</li> <li><strong>Trajectory data</strong>: CSV files containing trajectory information (see <a href=../step5-classification/ >previous step</a>).</li> <li><strong>Preprocessed data</strong>: RNASeq and clinical metadata files.</li> <li><strong>Control data</strong>: Control RNASeq and clinical metadata.</li> <li><strong>Note</strong>: Control data is included in the repository at <code>controls/{CANCER_TYPE}/</code> with files:<ul> <li><code>rnaseq_controls.csv</code></li> <li><code>clinical_controls.csv</code></li> </ul> </li> <li>You can use repo controls by setting <code>USE_REPO_CONTROLS=true</code></li> <li>Alternatively, controls from <code>data/interim/preprocessed_{CANCER_TYPE}_data/controls/</code> will be used by default</li> <li><strong>Pathways file</strong>: GMT format pathway file (default: <code>data/external/ReactomePathways.gmt</code>).</li> </ul> <h2 id=usage>Usage<a class=headerlink href=#usage title="Permanent link">&para;</a></h2> <h3 id=basic-usage-with-defaults>Basic Usage (with defaults)<a class=headerlink href=#basic-usage-with-defaults title="Permanent link">&para;</a></h3> <p>For KIRC cancer type with default paths:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>bash<span class=w> </span>scripts/enrichment/pipeline.sh<span class=w> </span>kirc
</span></code></pre></div> <h3 id=custom-configuration>Custom Configuration<a class=headerlink href=#custom-configuration title="Permanent link">&para;</a></h3> <p>You can override default paths by setting environment variables:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1># Example: Custom trajectory directory</span>
</span><span id=__span-1-2><a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=nv>CMD_DIR</span><span class=o>=</span>/path/to/custom/trajectories<span class=w> </span>bash<span class=w> </span>scripts/enrichment/pipeline.sh<span class=w> </span>kirc
</span><span id=__span-1-3><a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>
</span><span id=__span-1-4><a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=c1># Example: Use controls from repository root</span>
</span><span id=__span-1-5><a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a><span class=nv>USE_REPO_CONTROLS</span><span class=o>=</span><span class=nb>true</span><span class=w> </span>bash<span class=w> </span>scripts/enrichment/pipeline.sh<span class=w> </span>kirc
</span><span id=__span-1-6><a id=__codelineno-1-6 name=__codelineno-1-6 href=#__codelineno-1-6></a>
</span><span id=__span-1-7><a id=__codelineno-1-7 name=__codelineno-1-7 href=#__codelineno-1-7></a><span class=c1># Example: Multiple custom parameters</span>
</span><span id=__span-1-8><a id=__codelineno-1-8 name=__codelineno-1-8 href=#__codelineno-1-8></a><span class=nv>CMD_DIR</span><span class=o>=</span>/path/to/trajectories<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-9><a id=__codelineno-1-9 name=__codelineno-1-9 href=#__codelineno-1-9></a><span class=nv>DATA_DIR</span><span class=o>=</span>/path/to/rnaseq.csv<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-10><a id=__codelineno-1-10 name=__codelineno-1-10 href=#__codelineno-1-10></a><span class=nv>METADATA_DIR</span><span class=o>=</span>/path/to/clinical.csv<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-11><a id=__codelineno-1-11 name=__codelineno-1-11 href=#__codelineno-1-11></a><span class=nv>PATHWAYS_FILE</span><span class=o>=</span>/path/to/custom_pathways.gmt<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-12><a id=__codelineno-1-12 name=__codelineno-1-12 href=#__codelineno-1-12></a>bash<span class=w> </span>scripts/enrichment/pipeline.sh<span class=w> </span>kirc
</span></code></pre></div> <h2 id=configuration-parameters>Configuration Parameters<a class=headerlink href=#configuration-parameters title="Permanent link">&para;</a></h2> <h3 id=required-parameter>Required Parameter<a class=headerlink href=#required-parameter title="Permanent link">&para;</a></h3> <ul> <li><strong>cancer_type</strong>: Cancer type identifier (e.g., <code>kirc</code>, <code>brca</code>, <code>luad</code>)</li> <li>Passed as the first argument to the script</li> </ul> <h3 id=optional-parameters-environment-variables>Optional Parameters (Environment Variables)<a class=headerlink href=#optional-parameters-environment-variables title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Default (KIRC)</th> </tr> </thead> <tbody> <tr> <td><code>CMD_DIR</code></td> <td>Directory containing trajectory CSV files</td> <td><code>data/processed/synthetic_data/kirc/recnet/early_to_late/test_to_test/</code></td> </tr> <tr> <td><code>ERR_OUT_DIR</code></td> <td>Directory for error/output logs</td> <td><code>data/processed/synthetic_data/kirc/recnet/early_to_late/err_out_kirc_test</code></td> </tr> <tr> <td><code>ST_FILE</code></td> <td>Source-target file path</td> <td><code>data/processed/patient_trajectories_KIRC/random_connections_to_5_next_test.csv</code></td> </tr> <tr> <td><code>DATA_DIR</code></td> <td>RNASeq data file path</td> <td><code>data/interim/preprocessed_KIRC_data/KIRC_rnaseq.csv</code></td> </tr> <tr> <td><code>METADATA_DIR</code></td> <td>Clinical metadata file path</td> <td><code>data/interim/preprocessed_KIRC_data/KIRC_clinical.csv</code></td> </tr> <tr> <td><code>CONTROL_DATA</code></td> <td>Control RNASeq data file path</td> <td><code>data/interim/preprocessed_KIRC_data/controls/KIRC_control_rnaseq.csv</code></td> </tr> <tr> <td><code>CONTROL_METADATA</code></td> <td>Control clinical metadata file path</td> <td><code>data/interim/preprocessed_KIRC_data/controls/KIRC_control_clinical.csv</code></td> </tr> <tr> <td><code>META_NODES</code></td> <td>Nodes metadata file path</td> <td><code>data/processed/patient_trajectories_KIRC/nodes_metadata.csv</code></td> </tr> <tr> <td><code>STAGE_TRANSITION</code></td> <td>Stage transition identifier</td> <td><code>early_to_late</code></td> </tr> <tr> <td><code>PATHWAYS_FILE</code></td> <td>Pathways GMT file</td> <td><code>data/external/ReactomePathways.gmt</code></td> </tr> <tr> <td><code>USE_REPO_CONTROLS</code></td> <td>Use controls from repo root</td> <td><code>false</code> (uses preprocessed controls by default)</td> </tr> </tbody> </table> <h3 id=default-paths-for-other-cancer-types>Default Paths for Other Cancer Types<a class=headerlink href=#default-paths-for-other-cancer-types title="Permanent link">&para;</a></h3> <p>For cancer types other than KIRC, the script uses different default paths. You can override these by setting the environment variables explicitly.</p> <h2 id=available-gsea-parameters>Available GSEA Parameters<a class=headerlink href=#available-gsea-parameters title="Permanent link">&para;</a></h2> <p>GSEA parameters can be passed as arguments to the enrichment functions, making it easy to customize GSEA behavior without modifying configuration files.</p> <p>All functions that use <code>build_gsea_command</code> accept the following optional parameters:</p> <table> <thead> <tr> <th>Parameter</th> <th>Type</th> <th>Default</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>pathway_file</code></td> <td>str</td> <td><code>'data/external/ReactomePathways.gmt'</code></td> <td>Path to the GMT file containing gene sets</td> </tr> <tr> <td><code>mode</code></td> <td>str</td> <td><code>'Max_probe'</code></td> <td>GSEA collapse mode for handling multiple probes per gene</td> </tr> <tr> <td><code>norm</code></td> <td>str</td> <td><code>'meandiv'</code></td> <td>Normalization mode for gene set enrichment scores</td> </tr> <tr> <td><code>nperm</code></td> <td>int</td> <td><code>1000</code></td> <td>Number of permutations for significance testing</td> </tr> <tr> <td><code>rnd_seed</code></td> <td>str</td> <td><code>'timestamp'</code></td> <td>Random seed for reproducibility</td> </tr> <tr> <td><code>scoring_scheme</code></td> <td>str</td> <td><code>'weighted'</code></td> <td>Scoring scheme for enrichment calculation</td> </tr> <tr> <td><code>set_max</code></td> <td>int</td> <td><code>500</code></td> <td>Maximum size of gene sets to include</td> </tr> <tr> <td><code>set_min</code></td> <td>int</td> <td><code>15</code></td> <td>Minimum size of gene sets to include</td> </tr> </tbody> </table> <h2 id=functions-supporting-gsea-parameters>Functions Supporting GSEA Parameters<a class=headerlink href=#functions-supporting-gsea-parameters title="Permanent link">&para;</a></h2> <p>The following functions now accept GSEA parameters as kwargs:</p> <ol> <li><strong><code>build_gsea_command()</code></strong> - Core function that builds the GSEA command</li> <li><strong><code>get_rnk_single_patient()</code></strong> - Generate GSEA command for a single patient</li> <li><strong><code>fun_single_patient_and_gsea()</code></strong> - Perform analysis for a single patient</li> <li><strong><code>fun_synth_single_patient_and_gsea()</code></strong> - Perform analysis for synthetic patients</li> </ol> <h2 id=pipeline-execution-steps>Pipeline Execution Steps<a class=headerlink href=#pipeline-execution-steps title="Permanent link">&para;</a></h2> <h3 id=step-1-generate-deseq2-greasy-file>Step 1: Generate DESeq2 Greasy File<a class=headerlink href=#step-1-generate-deseq2-greasy-file title="Permanent link">&para;</a></h3> <p>The script will automatically generate a greasy file containing all DESeq2 commands:</p> <div class="language-text highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a>scripts/enrichment/greasy_deseq_file_&lt;cancer_type&gt;.txt
</span></code></pre></div> <p><strong>You must execute this file manually.</strong> The script will display instructions when this file is generated.</p> <h4 id=execution-options-for-deseq2-greasy-file>Execution Options for DESeq2 Greasy File:<a class=headerlink href=#execution-options-for-deseq2-greasy-file title="Permanent link">&para;</a></h4> <p>Even though the commands can be run sequentially, this would be very slow for large datasets. It is recommended to parallelize the execution, and highly recommended to use HPC.</p> <p>Here are several options to run the DESeq2 greasy file:</p> <p><strong>Option 1: Sequential Execution</strong> <div class="language-bash highlight"><pre><span></span><code><span id=__span-3-1><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a>bash<span class=w> </span>scripts/enrichment/greasy_deseq_file_kirc.txt
</span></code></pre></div></p> <p><strong>Option 2: GNU Parallel (Recommended)</strong> <div class="language-bash highlight"><pre><span></span><code><span id=__span-4-1><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1># Run with 8 parallel jobs</span>
</span><span id=__span-4-2><a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>parallel<span class=w> </span>-j<span class=w> </span><span class=m>8</span><span class=w> </span>&lt;<span class=w> </span>scripts/enrichment/greasy_deseq_file_kirc.txt
</span><span id=__span-4-3><a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>
</span><span id=__span-4-4><a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a><span class=c1># Run with all available cores</span>
</span><span id=__span-4-5><a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a>parallel<span class=w> </span>&lt;<span class=w> </span>scripts/enrichment/greasy_deseq_file_kirc.txt
</span></code></pre></div></p> <p><strong>Option 3: SLURM Job Scheduler for HPC</strong> <div class="language-bash highlight"><pre><span></span><code><span id=__span-5-1><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=ch>#!/bin/bash</span>
</span><span id=__span-5-2><a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=c1>#SBATCH --job-name=deseq_kirc</span>
</span><span id=__span-5-3><a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a><span class=c1>#SBATCH --output=logs/deseq_%j.out</span>
</span><span id=__span-5-4><a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=c1>#SBATCH --error=logs/deseq_%j.err</span>
</span><span id=__span-5-5><a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a><span class=c1>#SBATCH --time=02:00:00</span>
</span><span id=__span-5-6><a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a><span class=c1>#SBATCH --ntasks=10</span>
</span><span id=__span-5-7><a id=__codelineno-5-7 name=__codelineno-5-7 href=#__codelineno-5-7></a><span class=c1>#SBATCH --cpus-per-task=4</span>
</span><span id=__span-5-8><a id=__codelineno-5-8 name=__codelineno-5-8 href=#__codelineno-5-8></a>
</span><span id=__span-5-9><a id=__codelineno-5-9 name=__codelineno-5-9 href=#__codelineno-5-9></a><span class=c1># Load required modules</span>
</span><span id=__span-5-10><a id=__codelineno-5-10 name=__codelineno-5-10 href=#__codelineno-5-10></a>module<span class=w> </span>load<span class=w> </span>parallel
</span><span id=__span-5-11><a id=__codelineno-5-11 name=__codelineno-5-11 href=#__codelineno-5-11></a>
</span><span id=__span-5-12><a id=__codelineno-5-12 name=__codelineno-5-12 href=#__codelineno-5-12></a><span class=c1># Run greasy file with parallel</span>
</span><span id=__span-5-13><a id=__codelineno-5-13 name=__codelineno-5-13 href=#__codelineno-5-13></a>parallel<span class=w> </span>-j<span class=w> </span><span class=m>10</span><span class=w> </span>&lt;<span class=w> </span>scripts/enrichment/greasy_deseq_file_kirc.txt
</span></code></pre></div></p> <h3 id=step-2-generate-gsea-greasy-file>Step 2: Generate GSEA Greasy File<a class=headerlink href=#step-2-generate-gsea-greasy-file title="Permanent link">&para;</a></h3> <p>After DESeq2 completes, the script will generate a GSEA greasy file:</p> <div class="language-text highlight"><pre><span></span><code><span id=__span-6-1><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a>&lt;CMD_DIR&gt;/greasy_&lt;cancer_type&gt;.sh
</span></code></pre></div> <p><strong>You must execute this file manually.</strong> The script will display instructions when this file is generated.</p> <h4 id=execution-options-for-gsea-greasy-file>Execution Options for GSEA Greasy File:<a class=headerlink href=#execution-options-for-gsea-greasy-file title="Permanent link">&para;</a></h4> <p>As mentioned before, sequential execution is possible but slow. Parallel execution is recommended.</p> <p><strong>Option 1: Sequential Execution</strong> <div class="language-bash highlight"><pre><span></span><code><span id=__span-7-1><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a>bash<span class=w> </span>data/processed/.../greasy_kirc.sh
</span></code></pre></div></p> <p><strong>Option 2: GNU Parallel (Recommended)</strong> <div class="language-bash highlight"><pre><span></span><code><span id=__span-8-1><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># Run with 8 parallel jobs</span>
</span><span id=__span-8-2><a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a>parallel<span class=w> </span>-j<span class=w> </span><span class=m>8</span><span class=w> </span>&lt;<span class=w> </span>data/processed/.../greasy_kirc.sh
</span><span id=__span-8-3><a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a>
</span><span id=__span-8-4><a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=c1># Monitor progress with a progress bar</span>
</span><span id=__span-8-5><a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a>parallel<span class=w> </span>--progress<span class=w> </span>-j<span class=w> </span><span class=m>8</span><span class=w> </span>&lt;<span class=w> </span>data/processed/.../greasy_kirc.sh
</span></code></pre></div></p> <p><strong>Option 3: SLURM with GNU Parallel</strong> <div class="language-bash highlight"><pre><span></span><code><span id=__span-9-1><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=ch>#!/bin/bash</span>
</span><span id=__span-9-2><a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a><span class=c1>#SBATCH --job-name=gsea_kirc</span>
</span><span id=__span-9-3><a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=c1>#SBATCH --output=logs/gsea_%j.out</span>
</span><span id=__span-9-4><a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a><span class=c1>#SBATCH --error=logs/gsea_%j.err</span>
</span><span id=__span-9-5><a id=__codelineno-9-5 name=__codelineno-9-5 href=#__codelineno-9-5></a><span class=c1>#SBATCH --time=04:00:00</span>
</span><span id=__span-9-6><a id=__codelineno-9-6 name=__codelineno-9-6 href=#__codelineno-9-6></a><span class=c1>#SBATCH --ntasks=8</span>
</span><span id=__span-9-7><a id=__codelineno-9-7 name=__codelineno-9-7 href=#__codelineno-9-7></a><span class=c1>#SBATCH --cpus-per-task=1</span>
</span><span id=__span-9-8><a id=__codelineno-9-8 name=__codelineno-9-8 href=#__codelineno-9-8></a>
</span><span id=__span-9-9><a id=__codelineno-9-9 name=__codelineno-9-9 href=#__codelineno-9-9></a><span class=c1># Load required modules</span>
</span><span id=__span-9-10><a id=__codelineno-9-10 name=__codelineno-9-10 href=#__codelineno-9-10></a>module<span class=w> </span>load<span class=w> </span>parallel
</span><span id=__span-9-11><a id=__codelineno-9-11 name=__codelineno-9-11 href=#__codelineno-9-11></a>module<span class=w> </span>load<span class=w> </span>java<span class=w>  </span><span class=c1># Required for GSEA</span>
</span><span id=__span-9-12><a id=__codelineno-9-12 name=__codelineno-9-12 href=#__codelineno-9-12></a>
</span><span id=__span-9-13><a id=__codelineno-9-13 name=__codelineno-9-13 href=#__codelineno-9-13></a><span class=c1># Export GSEA function</span>
</span><span id=__span-9-14><a id=__codelineno-9-14 name=__codelineno-9-14 href=#__codelineno-9-14></a>gsea<span class=o>()</span><span class=w> </span><span class=o>{</span>
</span><span id=__span-9-15><a id=__codelineno-9-15 name=__codelineno-9-15 href=#__codelineno-9-15></a><span class=w>    </span><span class=s2>&quot;</span><span class=k>$(</span><span class=nb>pwd</span><span class=k>)</span><span class=s2>/GSEA_4.3.2/gsea-cli.sh&quot;</span><span class=w> </span><span class=s2>&quot;</span><span class=nv>$@</span><span class=s2>&quot;</span>
</span><span id=__span-9-16><a id=__codelineno-9-16 name=__codelineno-9-16 href=#__codelineno-9-16></a><span class=o>}</span>
</span><span id=__span-9-17><a id=__codelineno-9-17 name=__codelineno-9-17 href=#__codelineno-9-17></a><span class=nb>export</span><span class=w> </span>-f<span class=w> </span>gsea
</span><span id=__span-9-18><a id=__codelineno-9-18 name=__codelineno-9-18 href=#__codelineno-9-18></a>
</span><span id=__span-9-19><a id=__codelineno-9-19 name=__codelineno-9-19 href=#__codelineno-9-19></a><span class=c1># Run greasy file with parallel</span>
</span><span id=__span-9-20><a id=__codelineno-9-20 name=__codelineno-9-20 href=#__codelineno-9-20></a>parallel<span class=w> </span>-j<span class=w> </span><span class=m>8</span><span class=w> </span>&lt;<span class=w> </span>data/processed/.../greasy_kirc.sh
</span></code></pre></div></p> <h3 id=step-3-final-trajectory-formatting>Step 3: Final Trajectory Formatting<a class=headerlink href=#step-3-final-trajectory-formatting title="Permanent link">&para;</a></h3> <p>After GSEA completes, you need to run the trajectory formatting step manually. The script will display the exact command to execute with all the correct parameters.</p> <p><strong>You must execute this command manually.</strong> The script will display instructions.</p> <h4 id=execution>Execution:<a class=headerlink href=#execution title="Permanent link">&para;</a></h4> <p>Run the command displayed by the script:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-10-1><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a>python<span class=w> </span>scripts/enrichment/trajectory_formatting.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-10-2><a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=w>  </span>--path_synth<span class=w> </span>&lt;CMD_DIR&gt;<span class=w> </span><span class=se>\</span>
</span><span id=__span-10-3><a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a><span class=w>  </span>--pathways_file<span class=w> </span>&lt;PATHWAYS_FILE&gt;<span class=w> </span><span class=se>\</span>
</span><span id=__span-10-4><a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=w>  </span>--save_dir<span class=w> </span>&lt;parent_dir&gt;<span class=w> </span><span class=se>\</span>
</span><span id=__span-10-5><a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a><span class=w>  </span>--cancer_type<span class=w> </span>&lt;cancer_type&gt;
</span></code></pre></div> <p>Replace the placeholders with the actual values shown in the script output. This creates the final trajectory dataset with enrichment results.</p> <h2 id=complete-example-workflow>Complete Example Workflow<a class=headerlink href=#complete-example-workflow title="Permanent link">&para;</a></h2> <p>Here's a complete example for KIRC with custom paths:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-11-1><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=c1># Step 1: Set up environment variables</span>
</span><span id=__span-11-2><a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=nb>export</span><span class=w> </span><span class=nv>CMD_DIR</span><span class=o>=</span><span class=s2>&quot;/data/trajectories/kirc/test&quot;</span>
</span><span id=__span-11-3><a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a><span class=nb>export</span><span class=w> </span><span class=nv>DATA_DIR</span><span class=o>=</span><span class=s2>&quot;/data/preprocessed/kirc/rnaseq.csv&quot;</span>
</span><span id=__span-11-4><a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a><span class=nb>export</span><span class=w> </span><span class=nv>METADATA_DIR</span><span class=o>=</span><span class=s2>&quot;/data/preprocessed/kirc/clinical.csv&quot;</span>
</span><span id=__span-11-5><a id=__codelineno-11-5 name=__codelineno-11-5 href=#__codelineno-11-5></a><span class=nb>export</span><span class=w> </span><span class=nv>CONTROL_DATA</span><span class=o>=</span><span class=s2>&quot;/data/controls/kirc/rnaseq_control.csv&quot;</span>
</span><span id=__span-11-6><a id=__codelineno-11-6 name=__codelineno-11-6 href=#__codelineno-11-6></a><span class=nb>export</span><span class=w> </span><span class=nv>CONTROL_METADATA</span><span class=o>=</span><span class=s2>&quot;/data/controls/kirc/clinical_control.csv&quot;</span>
</span><span id=__span-11-7><a id=__codelineno-11-7 name=__codelineno-11-7 href=#__codelineno-11-7></a><span class=nb>export</span><span class=w> </span><span class=nv>PATHWAYS_FILE</span><span class=o>=</span><span class=s2>&quot;/data/pathways/ReactomePathways.gmt&quot;</span>
</span><span id=__span-11-8><a id=__codelineno-11-8 name=__codelineno-11-8 href=#__codelineno-11-8></a>
</span><span id=__span-11-9><a id=__codelineno-11-9 name=__codelineno-11-9 href=#__codelineno-11-9></a><span class=c1># Step 2: Run the enrichment pipeline script</span>
</span><span id=__span-11-10><a id=__codelineno-11-10 name=__codelineno-11-10 href=#__codelineno-11-10></a>bash<span class=w> </span>scripts/enrichment/pipeline.sh<span class=w> </span>kirc
</span><span id=__span-11-11><a id=__codelineno-11-11 name=__codelineno-11-11 href=#__codelineno-11-11></a>
</span><span id=__span-11-12><a id=__codelineno-11-12 name=__codelineno-11-12 href=#__codelineno-11-12></a><span class=c1># The script will pause after generating the DESeq2 greasy file</span>
</span><span id=__span-11-13><a id=__codelineno-11-13 name=__codelineno-11-13 href=#__codelineno-11-13></a><span class=c1># Step 3: Execute the DESeq2 greasy file</span>
</span><span id=__span-11-14><a id=__codelineno-11-14 name=__codelineno-11-14 href=#__codelineno-11-14></a>parallel<span class=w> </span>-j<span class=w> </span><span class=m>8</span><span class=w> </span>&lt;<span class=w> </span>scripts/enrichment/greasy_deseq_file_kirc.txt
</span><span id=__span-11-15><a id=__codelineno-11-15 name=__codelineno-11-15 href=#__codelineno-11-15></a>
</span><span id=__span-11-16><a id=__codelineno-11-16 name=__codelineno-11-16 href=#__codelineno-11-16></a><span class=c1># The script will then pause after generating the GSEA greasy file</span>
</span><span id=__span-11-17><a id=__codelineno-11-17 name=__codelineno-11-17 href=#__codelineno-11-17></a><span class=c1># Step 4: Execute the GSEA greasy file</span>
</span><span id=__span-11-18><a id=__codelineno-11-18 name=__codelineno-11-18 href=#__codelineno-11-18></a>parallel<span class=w> </span>-j<span class=w> </span><span class=m>8</span><span class=w> </span>&lt;<span class=w> </span>data/processed/.../greasy_kirc.sh
</span><span id=__span-11-19><a id=__codelineno-11-19 name=__codelineno-11-19 href=#__codelineno-11-19></a>
</span><span id=__span-11-20><a id=__codelineno-11-20 name=__codelineno-11-20 href=#__codelineno-11-20></a><span class=c1># Step 5: Execute the trajectory formatting step</span>
</span><span id=__span-11-21><a id=__codelineno-11-21 name=__codelineno-11-21 href=#__codelineno-11-21></a><span class=c1># Use the exact command shown by the script output</span>
</span><span id=__span-11-22><a id=__codelineno-11-22 name=__codelineno-11-22 href=#__codelineno-11-22></a>python<span class=w> </span>scripts/enrichment/trajectory_formatting.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-11-23><a id=__codelineno-11-23 name=__codelineno-11-23 href=#__codelineno-11-23></a><span class=w>  </span>--path_synth<span class=w> </span>/data/trajectories/kirc/test<span class=w> </span><span class=se>\</span>
</span><span id=__span-11-24><a id=__codelineno-11-24 name=__codelineno-11-24 href=#__codelineno-11-24></a><span class=w>  </span>--pathways_file<span class=w> </span>/data/pathways/ReactomePathways.gmt<span class=w> </span><span class=se>\</span>
</span><span id=__span-11-25><a id=__codelineno-11-25 name=__codelineno-11-25 href=#__codelineno-11-25></a><span class=w>  </span>--save_dir<span class=w> </span>/data/trajectories/kirc<span class=w> </span><span class=se>\</span>
</span><span id=__span-11-26><a id=__codelineno-11-26 name=__codelineno-11-26 href=#__codelineno-11-26></a><span class=w>  </span>--cancer_type<span class=w> </span>kirc
</span><span id=__span-11-27><a id=__codelineno-11-27 name=__codelineno-11-27 href=#__codelineno-11-27></a>
</span><span id=__span-11-28><a id=__codelineno-11-28 name=__codelineno-11-28 href=#__codelineno-11-28></a><span class=c1># Results will be saved to the parent directory of CMD_DIR</span>
</span></code></pre></div> <h2 id=background-execution>Background Execution<a class=headerlink href=#background-execution title="Permanent link">&para;</a></h2> <p>For long-running pipelines, you can run the script in the background:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-12-1><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a>nohup<span class=w> </span>bash<span class=w> </span>scripts/enrichment/pipeline.sh<span class=w> </span>kirc<span class=w> </span>&gt;<span class=w> </span>enrichment.log<span class=w> </span><span class=m>2</span>&gt;<span class=p>&amp;</span><span class=m>1</span><span class=w> </span><span class=p>&amp;</span>
</span></code></pre></div> <p>Monitor progress: <div class="language-bash highlight"><pre><span></span><code><span id=__span-13-1><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a>tail<span class=w> </span>-f<span class=w> </span>enrichment.log
</span></code></pre></div></p> <h2 id=output-files>Output Files<a class=headerlink href=#output-files title="Permanent link">&para;</a></h2> <h3 id=generated-greasy-files>Generated Greasy Files:<a class=headerlink href=#generated-greasy-files title="Permanent link">&para;</a></h3> <ul> <li><code>scripts/enrichment/greasy_deseq_file_&lt;cancer_type&gt;.txt</code>: DESeq2 commands</li> <li><code>&lt;CMD_DIR&gt;/greasy_&lt;cancer_type&gt;.sh</code>: GSEA commands</li> </ul> <h3 id=final-output>Final Output:<a class=headerlink href=#final-output title="Permanent link">&para;</a></h3> <ul> <li><code>&lt;parent_dir&gt;/&lt;cancer_type&gt;_trajectory_enrichment.csv</code>: Final trajectory dataset with enrichment results</li> <li>DESeq2 results in <code>&lt;CMD_DIR&gt;/</code> directory</li> <li>GSEA results in <code>&lt;CMD_DIR&gt;/</code> subdirectories</li> </ul> <h2 id=troubleshooting>Troubleshooting<a class=headerlink href=#troubleshooting title="Permanent link">&para;</a></h2> <h3 id=gsea-not-found>GSEA Not Found<a class=headerlink href=#gsea-not-found title="Permanent link">&para;</a></h3> <p>If you see errors about GSEA not being found: - Verify GSEA is installed at <code>GSEA_4.3.2/gsea-cli.sh</code> - Check execution permissions: <code>chmod +x GSEA_4.3.2/gsea-cli.sh</code> - See <a href=../../advanced/GSEA_INSTALLATION/ >GSEA Installation Guide</a></p> <h3 id=no-csv-files-found>No CSV Files Found<a class=headerlink href=#no-csv-files-found title="Permanent link">&para;</a></h3> <p>If the script reports no CSV files in CMD_DIR: - Verify CMD_DIR contains trajectory CSV files - Check file permissions - Ensure the trajectory generation step completed successfully</p> <h2 id=additional-resources>Additional Resources<a class=headerlink href=#additional-resources title="Permanent link">&para;</a></h2> <ul> <li><a href=https://www.gsea-msigdb.org/gsea/doc/ >GSEA Documentation</a></li> <li><a href=https://pydeseq2.readthedocs.io/en/latest/index.html>pyDESeq2 Documentation</a></li> <li><a href=https://www.gnu.org/software/parallel/parallel_tutorial.html>GNU Parallel Tutorial</a></li> <li><a href=../../advanced/ENRICHMENT_ANALYSIS/ >Enrichment Analysis Guide</a></li> </ul> <h2 id=next-steps>Next Steps<a class=headerlink href=#next-steps title="Permanent link">&para;</a></h2> <p>After completing the enrichment pipeline: - Analyze the final trajectory enrichment results - Visualize pathway changes along trajectories - Generate pathway heatmaps (see <a href=../../advanced/PATHWAY_HEATMAP/ >Pathway Heatmap Guide</a>) - Perform downstream analysis and interpretation</p> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="January 28, 2026 15:39:22 UTC"><span class=timeago datetime=2026-01-28T15:39:22+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="January 28, 2026 15:39:22 UTC">2026-01-28</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="December 19, 2025 16:49:24 UTC"><span class=timeago datetime=2025-12-19T16:49:24+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="December 19, 2025 16:49:24 UTC">2025-12-19</span> </span> </aside> <form class=md-feedback name=feedback hidden> <fieldset> <legend class=md-feedback__title> Was this page helpful? </legend> <div class=md-feedback__inner> <div class=md-feedback__list> <button class="md-feedback__icon md-icon" type=submit title="This page was helpful" data-md-value=1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m7 0c0 .8-.7 1.5-1.5 1.5S14 10.3 14 9.5 14.7 8 15.5 8s1.5.7 1.5 1.5m-5 7.73c-1.75 0-3.29-.73-4.19-1.81L9.23 14c.45.72 1.52 1.23 2.77 1.23s2.32-.51 2.77-1.23l1.42 1.42c-.9 1.08-2.44 1.81-4.19 1.81"/></svg> </button> <button class="md-feedback__icon md-icon" type=submit title="This page could be improved" data-md-value=0> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10m-6.5-4c.8 0 1.5.7 1.5 1.5s-.7 1.5-1.5 1.5-1.5-.7-1.5-1.5.7-1.5 1.5-1.5M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m2 4.5c1.75 0 3.29.72 4.19 1.81l-1.42 1.42C14.32 16.5 13.25 16 12 16s-2.32.5-2.77 1.23l-1.42-1.42C8.71 14.72 10.25 14 12 14"/></svg> </button> </div> <div class=md-feedback__note> <div data-md-value=1 hidden> Thanks for your feedback! </div> <div data-md-value=0 hidden> Thanks for your feedback! Help us improve by opening an issue. </div> </div> </div> </fieldset> </form> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../step5-classification/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Step 5 - Classification"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Step 5 - Classification </div> </div> </a> <a href=../visualization/ class="md-footer__link md-footer__link--next" aria-label="Next: Visualization"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Visualization </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2022-2025 EVENFLOW Project </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/gprolcastelo/renalprog target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://pypi.org/project/renalprog/ target=_blank rel=noopener title=pypi.org class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.8 200.5c-7.7-30.9-22.3-54.2-53.4-54.2h-40.1v47.4c0 36.8-31.2 67.8-66.8 67.8H172.7c-29.2 0-53.4 25-53.4 54.3v101.8c0 29 25.2 46 53.4 54.3 33.8 9.9 66.3 11.7 106.8 0 26.9-7.8 53.4-23.5 53.4-54.3v-40.7H226.2v-13.6h160.2c31.1 0 42.6-21.7 53.4-54.2 11.2-33.5 10.7-65.7 0-108.6M286.2 444.7a20.4 20.4 0 1 1 0-40.7 20.4 20.4 0 1 1 0 40.7M167.8 248.1h106.8c29.7 0 53.4-24.5 53.4-54.3V91.9c0-29-24.4-50.7-53.4-55.6-35.8-5.9-74.7-5.6-106.8.1-45.2 8-53.4 24.7-53.4 55.6v40.7h106.9v13.6h-147c-31.1 0-58.3 18.7-66.8 54.2-9.8 40.7-10.2 66.1 0 108.6 7.6 31.6 25.7 54.2 56.8 54.2H101v-48.8c0-35.3 30.5-66.4 66.8-66.4m-6.6-183.4a20.4 20.4 0 1 1 0 40.8 20.4 20.4 0 1 1 0-40.8"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "search.suggest", "search.highlight", "search.share", "content.code.copy", "content.code.annotate", "toc.follow"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../../js/timeago.min.js></script> <script src=../../js/timeago_mkdocs_material.js></script> <script src=../../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>