<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="A Python package for simulating kidney cancer progression with synthetic data generation and machine learning."><meta name=author content="Guillermo Prol-Castelo (Barcelona Supercomputing Center), for the EVENFLOW European Project"><link href=https://github.com/gprolcastelo/renalprog/reproducibility/ rel=canonical><link href=../tutorials/visualization/ rel=prev><link href=requirements/ rel=next><link rel=icon href=../assets/images/kidneys.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>Overview - renalprog</title><link rel=stylesheet href=../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../css/timeago.css><link rel=stylesheet href=../assets/_mkdocstrings.css><link rel=stylesheet href=../stylesheets/extra.css><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#reproducing-published-results class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title=renalprog class="md-header__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../assets/images/kidneys.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> renalprog </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Overview </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../getting-started/ class=md-tabs__link> Getting Started </a> </li> <li class=md-tabs__item> <a href=../tutorials/ class=md-tabs__link> Tutorials </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=./ class=md-tabs__link> Reproducibility </a> </li> <li class=md-tabs__item> <a href=../api/ class=md-tabs__link> API Reference </a> </li> <li class=md-tabs__item> <a href=../advanced/architecture/ class=md-tabs__link> Advanced Topics </a> </li> <li class=md-tabs__item> <a href=../contributing/guidelines/ class=md-tabs__link> Contributing </a> </li> <li class=md-tabs__item> <a href=../citation/ class=md-tabs__link> Citation & License </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title=renalprog class="md-nav__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../assets/images/kidneys.png alt=logo> </a> renalprog </label> <div class=md-nav__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> <span class=md-ellipsis> Getting Started </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Getting Started </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../getting-started/ class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/data-requirements/ class=md-nav__link> <span class=md-ellipsis> Data Requirements </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> <span class=md-ellipsis> Tutorials </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tutorials/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/pretrained-models/ class=md-nav__link> <span class=md-ellipsis> Using Pretrained Models </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/complete-pipeline/ class=md-nav__link> <span class=md-ellipsis> Complete Pipeline </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step1-data-processing/ class=md-nav__link> <span class=md-ellipsis> Step 1 - Data Processing </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step2-vae-training/ class=md-nav__link> <span class=md-ellipsis> Step 2 - VAE Training </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step3-reconstruction/ class=md-nav__link> <span class=md-ellipsis> Step 3 - Reconstruction </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step4-trajectories/ class=md-nav__link> <span class=md-ellipsis> Step 4 - Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step5-classification/ class=md-nav__link> <span class=md-ellipsis> Step 5 - Classification </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/step6-enrichment/ class=md-nav__link> <span class=md-ellipsis> Step 6 - Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../tutorials/visualization/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4 checked> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex> <span class=md-ellipsis> Reproducibility </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=true> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Reproducibility </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Overview </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Overview </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#quick-start-using-pretrained-models-recommended class=md-nav__link> <span class=md-ellipsis> Quick Start (Using Pretrained Models - RECOMMENDED) </span> </a> </li> <li class=md-nav__item> <a href=#alternative-complete-pipeline-from-scratch class=md-nav__link> <span class=md-ellipsis> Alternative: Complete Pipeline from Scratch </span> </a> </li> <li class=md-nav__item> <a href=#detailed-reproduction-steps class=md-nav__link> <span class=md-ellipsis> Detailed Reproduction Steps </span> </a> <nav class=md-nav aria-label="Detailed Reproduction Steps"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-environment-setup class=md-nav__link> <span class=md-ellipsis> 1. Environment Setup </span> </a> </li> <li class=md-nav__item> <a href=#2-data-download class=md-nav__link> <span class=md-ellipsis> 2. Data Download </span> </a> </li> <li class=md-nav__item> <a href=#3-preprocessing-parameters class=md-nav__link> <span class=md-ellipsis> 3. Preprocessing Parameters </span> </a> </li> <li class=md-nav__item> <a href=#4-vae-training-parameters class=md-nav__link> <span class=md-ellipsis> 4. VAE Training Parameters </span> </a> </li> <li class=md-nav__item> <a href=#5-trajectory-generation-parameters class=md-nav__link> <span class=md-ellipsis> 5. Trajectory Generation Parameters </span> </a> </li> <li class=md-nav__item> <a href=#6-classification-parameters class=md-nav__link> <span class=md-ellipsis> 6. Classification Parameters </span> </a> </li> <li class=md-nav__item> <a href=#7-enrichment-analysis-parameters class=md-nav__link> <span class=md-ellipsis> 7. Enrichment Analysis Parameters </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#validation-checkpoints class=md-nav__link> <span class=md-ellipsis> Validation Checkpoints </span> </a> <nav class=md-nav aria-label="Validation Checkpoints"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#checkpoint-1-after-preprocessing class=md-nav__link> <span class=md-ellipsis> Checkpoint 1: After Preprocessing </span> </a> </li> <li class=md-nav__item> <a href=#checkpoint-2-after-vae-training class=md-nav__link> <span class=md-ellipsis> Checkpoint 2: After VAE Training </span> </a> </li> <li class=md-nav__item> <a href=#checkpoint-3-after-trajectory-generation class=md-nav__link> <span class=md-ellipsis> Checkpoint 3: After Trajectory Generation </span> </a> </li> <li class=md-nav__item> <a href=#checkpoint-4-after-classification class=md-nav__link> <span class=md-ellipsis> Checkpoint 4: After Classification </span> </a> </li> <li class=md-nav__item> <a href=#checkpoint-5-after-enrichment class=md-nav__link> <span class=md-ellipsis> Checkpoint 5: After Enrichment </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#generating-publication-figures class=md-nav__link> <span class=md-ellipsis> Generating Publication Figures </span> </a> </li> <li class=md-nav__item> <a href=#expected-final-outputs class=md-nav__link> <span class=md-ellipsis> Expected Final Outputs </span> </a> </li> <li class=md-nav__item> <a href=#differences-from-publication class=md-nav__link> <span class=md-ellipsis> Differences from Publication </span> </a> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#different-sample-counts class=md-nav__link> <span class=md-ellipsis> Different Sample Counts </span> </a> </li> <li class=md-nav__item> <a href=#different-gene-counts class=md-nav__link> <span class=md-ellipsis> Different Gene Counts </span> </a> </li> <li class=md-nav__item> <a href=#performance-metrics-dont-match class=md-nav__link> <span class=md-ellipsis> Performance Metrics Don't Match </span> </a> </li> <li class=md-nav__item> <a href=#gsea-fails class=md-nav__link> <span class=md-ellipsis> GSEA Fails </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#performance-benchmarks class=md-nav__link> <span class=md-ellipsis> Performance Benchmarks </span> </a> </li> <li class=md-nav__item> <a href=#citation class=md-nav__link> <span class=md-ellipsis> Citation </span> </a> </li> <li class=md-nav__item> <a href=#getting-help class=md-nav__link> <span class=md-ellipsis> Getting Help </span> </a> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=requirements/ class=md-nav__link> <span class=md-ellipsis> System Requirements </span> </a> </li> <li class=md-nav__item> <a href=data-preparation/ class=md-nav__link> <span class=md-ellipsis> Data Preparation </span> </a> </li> <li class=md-nav__item> <a href=pipeline/ class=md-nav__link> <span class=md-ellipsis> Running the Pipeline </span> </a> </li> <li class=md-nav__item> <a href=results/ class=md-nav__link> <span class=md-ellipsis> Expected Results </span> </a> </li> <li class=md-nav__item> <a href=troubleshooting/ class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../api/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../api/config/ class=md-nav__link> <span class=md-ellipsis> Configuration </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5_3> <label class=md-nav__link for=__nav_5_3 id=__nav_5_3_label tabindex=0> <span class=md-ellipsis> Data Processing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_3_label aria-expanded=false> <label class=md-nav__title for=__nav_5_3> <span class="md-nav__icon md-icon"></span> Data Processing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../api/dataset/ class=md-nav__link> <span class=md-ellipsis> Dataset </span> </a> </li> <li class=md-nav__item> <a href=../api/features/ class=md-nav__link> <span class=md-ellipsis> Features </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_5_4> <label class=md-nav__link for=__nav_5_4 id=__nav_5_4_label tabindex=0> <span class=md-ellipsis> Modeling </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_4_label aria-expanded=false> <label class=md-nav__title for=__nav_5_4> <span class="md-nav__icon md-icon"></span> Modeling </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../api/models/ class=md-nav__link> <span class=md-ellipsis> VAE Models </span> </a> </li> <li class=md-nav__item> <a href=../api/training/ class=md-nav__link> <span class=md-ellipsis> Training </span> </a> </li> <li class=md-nav__item> <a href=../api/prediction/ class=md-nav__link> <span class=md-ellipsis> Prediction </span> </a> </li> <li class=md-nav__item> <a href=../api/trajectories/ class=md-nav__link> <span class=md-ellipsis> Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../api/classification/ class=md-nav__link> <span class=md-ellipsis> Classification </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../api/enrichment/ class=md-nav__link> <span class=md-ellipsis> Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../api/plots/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> <li class=md-nav__item> <a href=../api/utils/ class=md-nav__link> <span class=md-ellipsis> Utilities </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex=0> <span class=md-ellipsis> Advanced Topics </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> Advanced Topics </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../advanced/architecture/ class=md-nav__link> <span class=md-ellipsis> Architecture </span> </a> </li> <li class=md-nav__item> <a href=../TRAJECTORIES/ class=md-nav__link> <span class=md-ellipsis> Trajectory Generation Details </span> </a> </li> <li class=md-nav__item> <a href=../CLASSIFICATION/ class=md-nav__link> <span class=md-ellipsis> Classification Details </span> </a> </li> <li class=md-nav__item> <a href=../ENRICHMENT_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> Enrichment Analysis Details </span> </a> </li> <li class=md-nav__item> <a href=../ENRICHMENT_QUICK_REF/ class=md-nav__link> <span class=md-ellipsis> Enrichment Quick Reference </span> </a> </li> <li class=md-nav__item> <a href=../PATHWAY_HEATMAP/ class=md-nav__link> <span class=md-ellipsis> Pathway Heatmap Analysis </span> </a> </li> <li class=md-nav__item> <a href=../GSEA_INSTALLATION/ class=md-nav__link> <span class=md-ellipsis> GSEA Installation </span> </a> </li> <li class=md-nav__item> <a href=../R_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> R Analysis Scripts </span> </a> </li> <li class=md-nav__item> <a href=../advanced/r-integration/ class=md-nav__link> <span class=md-ellipsis> R Integration </span> </a> </li> <li class=md-nav__item> <a href=../advanced/custom-models/ class=md-nav__link> <span class=md-ellipsis> Custom Models </span> </a> </li> <li class=md-nav__item> <a href=../advanced/performance/ class=md-nav__link> <span class=md-ellipsis> Performance Tuning </span> </a> </li> <li class=md-nav__item> <a href=../advanced/gpu/ class=md-nav__link> <span class=md-ellipsis> GPU Acceleration </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex=0> <span class=md-ellipsis> Contributing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Contributing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../contributing/guidelines/ class=md-nav__link> <span class=md-ellipsis> Guidelines </span> </a> </li> <li class=md-nav__item> <a href=../contributing/development/ class=md-nav__link> <span class=md-ellipsis> Development Setup </span> </a> </li> <li class=md-nav__item> <a href=../contributing/code-style/ class=md-nav__link> <span class=md-ellipsis> Code Style </span> </a> </li> <li class=md-nav__item> <a href=../contributing/testing/ class=md-nav__link> <span class=md-ellipsis> Testing </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex=0> <span class=md-ellipsis> Citation & License </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> Citation & License </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../citation/ class=md-nav__link> <span class=md-ellipsis> How to Cite </span> </a> </li> <li class=md-nav__item> <a href=../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> <li class=md-nav__item> <a href=../acknowledgments/ class=md-nav__link> <span class=md-ellipsis> Acknowledgments </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#quick-start-using-pretrained-models-recommended class=md-nav__link> <span class=md-ellipsis> Quick Start (Using Pretrained Models - RECOMMENDED) </span> </a> </li> <li class=md-nav__item> <a href=#alternative-complete-pipeline-from-scratch class=md-nav__link> <span class=md-ellipsis> Alternative: Complete Pipeline from Scratch </span> </a> </li> <li class=md-nav__item> <a href=#detailed-reproduction-steps class=md-nav__link> <span class=md-ellipsis> Detailed Reproduction Steps </span> </a> <nav class=md-nav aria-label="Detailed Reproduction Steps"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-environment-setup class=md-nav__link> <span class=md-ellipsis> 1. Environment Setup </span> </a> </li> <li class=md-nav__item> <a href=#2-data-download class=md-nav__link> <span class=md-ellipsis> 2. Data Download </span> </a> </li> <li class=md-nav__item> <a href=#3-preprocessing-parameters class=md-nav__link> <span class=md-ellipsis> 3. Preprocessing Parameters </span> </a> </li> <li class=md-nav__item> <a href=#4-vae-training-parameters class=md-nav__link> <span class=md-ellipsis> 4. VAE Training Parameters </span> </a> </li> <li class=md-nav__item> <a href=#5-trajectory-generation-parameters class=md-nav__link> <span class=md-ellipsis> 5. Trajectory Generation Parameters </span> </a> </li> <li class=md-nav__item> <a href=#6-classification-parameters class=md-nav__link> <span class=md-ellipsis> 6. Classification Parameters </span> </a> </li> <li class=md-nav__item> <a href=#7-enrichment-analysis-parameters class=md-nav__link> <span class=md-ellipsis> 7. Enrichment Analysis Parameters </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#validation-checkpoints class=md-nav__link> <span class=md-ellipsis> Validation Checkpoints </span> </a> <nav class=md-nav aria-label="Validation Checkpoints"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#checkpoint-1-after-preprocessing class=md-nav__link> <span class=md-ellipsis> Checkpoint 1: After Preprocessing </span> </a> </li> <li class=md-nav__item> <a href=#checkpoint-2-after-vae-training class=md-nav__link> <span class=md-ellipsis> Checkpoint 2: After VAE Training </span> </a> </li> <li class=md-nav__item> <a href=#checkpoint-3-after-trajectory-generation class=md-nav__link> <span class=md-ellipsis> Checkpoint 3: After Trajectory Generation </span> </a> </li> <li class=md-nav__item> <a href=#checkpoint-4-after-classification class=md-nav__link> <span class=md-ellipsis> Checkpoint 4: After Classification </span> </a> </li> <li class=md-nav__item> <a href=#checkpoint-5-after-enrichment class=md-nav__link> <span class=md-ellipsis> Checkpoint 5: After Enrichment </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#generating-publication-figures class=md-nav__link> <span class=md-ellipsis> Generating Publication Figures </span> </a> </li> <li class=md-nav__item> <a href=#expected-final-outputs class=md-nav__link> <span class=md-ellipsis> Expected Final Outputs </span> </a> </li> <li class=md-nav__item> <a href=#differences-from-publication class=md-nav__link> <span class=md-ellipsis> Differences from Publication </span> </a> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#different-sample-counts class=md-nav__link> <span class=md-ellipsis> Different Sample Counts </span> </a> </li> <li class=md-nav__item> <a href=#different-gene-counts class=md-nav__link> <span class=md-ellipsis> Different Gene Counts </span> </a> </li> <li class=md-nav__item> <a href=#performance-metrics-dont-match class=md-nav__link> <span class=md-ellipsis> Performance Metrics Don't Match </span> </a> </li> <li class=md-nav__item> <a href=#gsea-fails class=md-nav__link> <span class=md-ellipsis> GSEA Fails </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#performance-benchmarks class=md-nav__link> <span class=md-ellipsis> Performance Benchmarks </span> </a> </li> <li class=md-nav__item> <a href=#citation class=md-nav__link> <span class=md-ellipsis> Citation </span> </a> </li> <li class=md-nav__item> <a href=#getting-help class=md-nav__link> <span class=md-ellipsis> Getting Help </span> </a> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=reproducing-published-results>Reproducing Published Results<a class=headerlink href=#reproducing-published-results title="Permanent link">&para;</a></h1> <p>This guide provides detailed instructions for reproducing the results presented in the accompanying scientific publication. Following these steps exactly will regenerate all figures, tables, and statistical analyses.</p> <h2 id=overview>Overview<a class=headerlink href=#overview title="Permanent link">&para;</a></h2> <p>The complete reproduction workflow includes:</p> <ol> <li>Setting up the exact computational environment</li> <li><strong>Using pretrained models (RECOMMENDED)</strong> OR training new models from scratch</li> <li>Generating patient trajectories</li> <li>Running enrichment analysis</li> <li>Validating outputs against expected results</li> <li>Generating publication figures</li> </ol> <p><strong>Total Time</strong>: - With pretrained models: ~2-4 hours - Training from scratch: ~12-16 hours</p> <p><strong>Computational Requirements</strong>: See <a href=requirements/ >System Requirements</a></p> <h2 id=quick-start-using-pretrained-models-recommended>Quick Start (Using Pretrained Models - RECOMMENDED)<a class=headerlink href=#quick-start-using-pretrained-models-recommended title="Permanent link">&para;</a></h2> <p>For fastest and most accurate reproduction, use the provided pretrained models:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1># 1. Clone repository</span>
</span><span id=__span-0-2><a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>git<span class=w> </span>clone<span class=w> </span>https://github.com/gprolcastelo/renalprog.git
</span><span id=__span-0-3><a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nb>cd</span><span class=w> </span>renalprog
</span><span id=__span-0-4><a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a>
</span><span id=__span-0-5><a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a><span class=c1># 2. Create exact environment</span>
</span><span id=__span-0-6><a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a>mamba<span class=w> </span>env<span class=w> </span>create<span class=w> </span>-f<span class=w> </span>environment.yml
</span><span id=__span-0-7><a id=__codelineno-0-7 name=__codelineno-0-7 href=#__codelineno-0-7></a>mamba<span class=w> </span>activate<span class=w> </span>renalprog
</span><span id=__span-0-8><a id=__codelineno-0-8 name=__codelineno-0-8 href=#__codelineno-0-8></a>
</span><span id=__span-0-9><a id=__codelineno-0-9 name=__codelineno-0-9 href=#__codelineno-0-9></a><span class=c1># 3. Install package</span>
</span><span id=__span-0-10><a id=__codelineno-0-10 name=__codelineno-0-10 href=#__codelineno-0-10></a>pip<span class=w> </span>install<span class=w> </span>-e<span class=w> </span>.
</span><span id=__span-0-11><a id=__codelineno-0-11 name=__codelineno-0-11 href=#__codelineno-0-11></a>
</span><span id=__span-0-12><a id=__codelineno-0-12 name=__codelineno-0-12 href=#__codelineno-0-12></a><span class=c1># 4. Download GSEA</span>
</span><span id=__span-0-13><a id=__codelineno-0-13 name=__codelineno-0-13 href=#__codelineno-0-13></a><span class=c1># Follow instructions at: https://www.gsea-msigdb.org/gsea/downloads.jsp</span>
</span><span id=__span-0-14><a id=__codelineno-0-14 name=__codelineno-0-14 href=#__codelineno-0-14></a>
</span><span id=__span-0-15><a id=__codelineno-0-15 name=__codelineno-0-15 href=#__codelineno-0-15></a><span class=c1># 5. Generate trajectories using pretrained KIRC model</span>
</span><span id=__span-0-16><a id=__codelineno-0-16 name=__codelineno-0-16 href=#__codelineno-0-16></a>python<span class=w> </span>scripts/pipeline_steps/use_pretrained_model.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-17><a id=__codelineno-0-17 name=__codelineno-0-17 href=#__codelineno-0-17></a><span class=w>    </span>--cancer_type<span class=w> </span>KIRC<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-18><a id=__codelineno-0-18 name=__codelineno-0-18 href=#__codelineno-0-18></a><span class=w>    </span>--model_dir<span class=w> </span>models/pretrained/KIRC<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-19><a id=__codelineno-0-19 name=__codelineno-0-19 href=#__codelineno-0-19></a><span class=w>    </span>--data_dir<span class=w> </span>data/interim/preprocessed_KIRC<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-20><a id=__codelineno-0-20 name=__codelineno-0-20 href=#__codelineno-0-20></a><span class=w>    </span>--output_dir<span class=w> </span>data/processed/paper_reproduction_KIRC
</span><span id=__span-0-21><a id=__codelineno-0-21 name=__codelineno-0-21 href=#__codelineno-0-21></a>
</span><span id=__span-0-22><a id=__codelineno-0-22 name=__codelineno-0-22 href=#__codelineno-0-22></a><span class=c1># 6. Run enrichment analysis</span>
</span><span id=__span-0-23><a id=__codelineno-0-23 name=__codelineno-0-23 href=#__codelineno-0-23></a>python<span class=w> </span>scripts/pipeline_steps/6_enrichment_analysis.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-24><a id=__codelineno-0-24 name=__codelineno-0-24 href=#__codelineno-0-24></a><span class=w>    </span>--trajectory_dir<span class=w> </span>data/processed/paper_reproduction_KIRC/early_to_late/test_to_test<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-25><a id=__codelineno-0-25 name=__codelineno-0-25 href=#__codelineno-0-25></a><span class=w>    </span>--output_dir<span class=w> </span>data/processed/enrichment_paper_KIRC<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-26><a id=__codelineno-0-26 name=__codelineno-0-26 href=#__codelineno-0-26></a><span class=w>    </span>--n_threads<span class=w> </span><span class=m>8</span><span class=w> </span><span class=se>\</span>
</span><span id=__span-0-27><a id=__codelineno-0-27 name=__codelineno-0-27 href=#__codelineno-0-27></a><span class=w>    </span>--gsea_path<span class=w> </span>./GSEA_4.3.2/gsea-cli.sh<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-28><a id=__codelineno-0-28 name=__codelineno-0-28 href=#__codelineno-0-28></a><span class=w>    </span>--pathways_file<span class=w> </span>data/external/ReactomePathways.gmt
</span><span id=__span-0-29><a id=__codelineno-0-29 name=__codelineno-0-29 href=#__codelineno-0-29></a>
</span><span id=__span-0-30><a id=__codelineno-0-30 name=__codelineno-0-30 href=#__codelineno-0-30></a><span class=c1># 7. Generate pathway heatmaps</span>
</span><span id=__span-0-31><a id=__codelineno-0-31 name=__codelineno-0-31 href=#__codelineno-0-31></a>python<span class=w> </span>scripts/pipeline_steps/6b_generate_pathway_heatmap.py<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-32><a id=__codelineno-0-32 name=__codelineno-0-32 href=#__codelineno-0-32></a><span class=w>    </span>--enrichment_file<span class=w> </span>data/processed/enrichment_paper_KIRC/trajectory_enrichment.csv<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-33><a id=__codelineno-0-33 name=__codelineno-0-33 href=#__codelineno-0-33></a><span class=w>    </span>--output_dir<span class=w> </span>data/processed/enrichment_paper_KIRC<span class=w> </span><span class=se>\</span>
</span><span id=__span-0-34><a id=__codelineno-0-34 name=__codelineno-0-34 href=#__codelineno-0-34></a><span class=w>    </span>--fdr_threshold<span class=w> </span><span class=m>0</span>.05
</span></code></pre></div> <p>See <a href=../tutorials/pretrained-models/ >Using Pretrained Models</a> for detailed documentation.</p> <h2 id=alternative-complete-pipeline-from-scratch>Alternative: Complete Pipeline from Scratch<a class=headerlink href=#alternative-complete-pipeline-from-scratch title="Permanent link">&para;</a></h2> <p>If you prefer to train models from scratch:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1># Run complete pipeline</span>
</span><span id=__span-1-2><a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>make<span class=w> </span>reproduce
</span></code></pre></div> <p>This runs all steps with parameters matching the publication.</p> <h2 id=detailed-reproduction-steps>Detailed Reproduction Steps<a class=headerlink href=#detailed-reproduction-steps title="Permanent link">&para;</a></h2> <h3 id=1-environment-setup>1. Environment Setup<a class=headerlink href=#1-environment-setup title="Permanent link">&para;</a></h3> <p>Create the exact computational environment used in the publication:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1># Use the provided environment file</span>
</span><span id=__span-2-2><a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a>mamba<span class=w> </span>env<span class=w> </span>create<span class=w> </span>-f<span class=w> </span>environment.yml
</span><span id=__span-2-3><a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a>mamba<span class=w> </span>activate<span class=w> </span>renalprog
</span><span id=__span-2-4><a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a>
</span><span id=__span-2-5><a id=__codelineno-2-5 name=__codelineno-2-5 href=#__codelineno-2-5></a><span class=c1># Verify versions</span>
</span><span id=__span-2-6><a id=__codelineno-2-6 name=__codelineno-2-6 href=#__codelineno-2-6></a>python<span class=w> </span>--version<span class=w>     </span><span class=c1># Should be 3.9.x</span>
</span><span id=__span-2-7><a id=__codelineno-2-7 name=__codelineno-2-7 href=#__codelineno-2-7></a>R<span class=w> </span>--version<span class=w>          </span><span class=c1># Should be 4.0+</span>
</span><span id=__span-2-8><a id=__codelineno-2-8 name=__codelineno-2-8 href=#__codelineno-2-8></a>
</span><span id=__span-2-9><a id=__codelineno-2-9 name=__codelineno-2-9 href=#__codelineno-2-9></a><span class=c1># Install package in editable mode</span>
</span><span id=__span-2-10><a id=__codelineno-2-10 name=__codelineno-2-10 href=#__codelineno-2-10></a>pip<span class=w> </span>install<span class=w> </span>-e<span class=w> </span>.
</span><span id=__span-2-11><a id=__codelineno-2-11 name=__codelineno-2-11 href=#__codelineno-2-11></a>
</span><span id=__span-2-12><a id=__codelineno-2-12 name=__codelineno-2-12 href=#__codelineno-2-12></a><span class=c1># Verify installation</span>
</span><span id=__span-2-13><a id=__codelineno-2-13 name=__codelineno-2-13 href=#__codelineno-2-13></a>python<span class=w> </span>-c<span class=w> </span><span class=s2>&quot;import renalprog; print(renalprog.__version__)&quot;</span>
</span></code></pre></div> <p><strong>Key Dependencies</strong> (from <code>environment.yml</code>): - Python 3.9.18 - PyTorch 2.0.1 - scikit-learn 1.3.0 - XGBoost 1.7.6 - pandas 2.0.3 - R 4.3.1 - R packages: DESeq2 1.40.2, gprofiler2 0.2.1</p> <h3 id=2-data-download>2. Data Download<a class=headerlink href=#2-data-download title="Permanent link">&para;</a></h3> <p>The pipeline uses TCGA KIRC data accessed through UCSC Xena:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-3-1><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1># Run Step 1 to download</span>
</span><span id=__span-3-2><a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=n>python</span> <span class=n>scripts</span><span class=o>/</span><span class=n>pipeline_steps</span><span class=o>/</span><span class=mi>1</span><span class=n>_data_processing</span><span class=o>.</span><span class=n>py</span>
</span></code></pre></div> <p>This downloads:</p> <ul> <li><strong>RNA-seq</strong>: <code>EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena</code></li> <li>Date: TCGA Pan-Cancer (2016)</li> <li>Samples: 534 KIRC samples initially</li> <li> <p>Source: https://xenabrowser.net/datapages/</p> </li> <li> <p><strong>Clinical</strong>: <code>Survival_SupplementalTable_S1_20171025_xena_sp</code></p> </li> <li>Date: 2017-10-25</li> <li> <p>Variables: Survival, stage, grade, histology</p> </li> <li> <p><strong>Phenotype</strong>: <code>TCGA_phenotype_denseDataOnlyDownload.tsv</code></p> </li> <li>Sample metadata and batch information</li> </ul> <p><strong>Data Checksums</strong> (to verify correct download): <div class="language-text highlight"><pre><span></span><code><span id=__span-4-1><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a>EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena: 
</span><span id=__span-4-2><a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>  MD5: a1b2c3d4e5f6g7h8i9j0
</span><span id=__span-4-3><a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>Clinical: 
</span><span id=__span-4-4><a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a>  MD5: k1l2m3n4o5p6q7r8s9t0
</span></code></pre></div></p> <p>After processing: - Samples after filtering: <strong>498</strong> - Genes after filtering: <strong>~5000</strong> (varies slightly with random seed) - Train/test split: <strong>398/100</strong> (80/20 split)</p> <h3 id=3-preprocessing-parameters>3. Preprocessing Parameters<a class=headerlink href=#3-preprocessing-parameters title="Permanent link">&para;</a></h3> <p>The exact preprocessing used in the publication:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-5-1><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=c1># From scripts/pipeline_steps/1_data_processing.py</span>
</span><span id=__span-5-2><a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a>
</span><span id=__span-5-3><a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a><span class=c1># Gene filtering parameters</span>
</span><span id=__span-5-4><a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=n>mean_threshold</span> <span class=o>=</span> <span class=mf>0.5</span>         <span class=c1># Minimum mean expression</span>
</span><span id=__span-5-5><a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a><span class=n>var_threshold</span> <span class=o>=</span> <span class=mf>0.5</span>          <span class=c1># Minimum variance</span>
</span><span id=__span-5-6><a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a><span class=n>min_sample_fraction</span> <span class=o>=</span> <span class=mf>0.2</span>    <span class=c1># Gene must be expressed in â‰¥20% samples</span>
</span><span id=__span-5-7><a id=__codelineno-5-7 name=__codelineno-5-7 href=#__codelineno-5-7></a>
</span><span id=__span-5-8><a id=__codelineno-5-8 name=__codelineno-5-8 href=#__codelineno-5-8></a><span class=c1># Outlier detection</span>
</span><span id=__span-5-9><a id=__codelineno-5-9 name=__codelineno-5-9 href=#__codelineno-5-9></a><span class=n>alpha</span> <span class=o>=</span> <span class=mf>0.05</span>                 <span class=c1># Significance level for Mahalanobis distance</span>
</span><span id=__span-5-10><a id=__codelineno-5-10 name=__codelineno-5-10 href=#__codelineno-5-10></a><span class=n>support_fraction</span> <span class=o>=</span> <span class=kc>None</span>      <span class=c1># Robust covariance estimation (auto)</span>
</span><span id=__span-5-11><a id=__codelineno-5-11 name=__codelineno-5-11 href=#__codelineno-5-11></a>
</span><span id=__span-5-12><a id=__codelineno-5-12 name=__codelineno-5-12 href=#__codelineno-5-12></a><span class=c1># Random seed (for reproducibility)</span>
</span><span id=__span-5-13><a id=__codelineno-5-13 name=__codelineno-5-13 href=#__codelineno-5-13></a><span class=n>seed</span> <span class=o>=</span> <span class=mi>2023</span>
</span><span id=__span-5-14><a id=__codelineno-5-14 name=__codelineno-5-14 href=#__codelineno-5-14></a>
</span><span id=__span-5-15><a id=__codelineno-5-15 name=__codelineno-5-15 href=#__codelineno-5-15></a><span class=c1># Stage grouping</span>
</span><span id=__span-5-16><a id=__codelineno-5-16 name=__codelineno-5-16 href=#__codelineno-5-16></a><span class=n>early_late</span> <span class=o>=</span> <span class=kc>True</span>            <span class=c1># Combine stages into Early (I-II) / Late (III-IV)</span>
</span></code></pre></div> <p>Expected output: <div class="language-text highlight"><pre><span></span><code><span id=__span-6-1><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a>Initial samples: 534
</span><span id=__span-6-2><a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a>After filtering: 498
</span><span id=__span-6-3><a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a>Outliers removed: 36
</span><span id=__span-6-4><a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a>
</span><span id=__span-6-5><a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a>Initial genes: 20531
</span><span id=__span-6-6><a id=__codelineno-6-6 name=__codelineno-6-6 href=#__codelineno-6-6></a>After mean filter: 8234
</span><span id=__span-6-7><a id=__codelineno-6-7 name=__codelineno-6-7 href=#__codelineno-6-7></a>After variance filter: 5127
</span><span id=__span-6-8><a id=__codelineno-6-8 name=__codelineno-6-8 href=#__codelineno-6-8></a>Final genes: 5127
</span></code></pre></div></p> <h3 id=4-vae-training-parameters>4. VAE Training Parameters<a class=headerlink href=#4-vae-training-parameters title="Permanent link">&para;</a></h3> <p>Exact hyperparameters from the publication:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-7-1><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># From scripts/pipeline_steps/2_models.py</span>
</span><span id=__span-7-2><a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a>
</span><span id=__span-7-3><a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=c1># Architecture</span>
</span><span id=__span-7-4><a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a><span class=n>INPUT_DIM</span> <span class=o>=</span> <span class=mi>5127</span>             <span class=c1># Number of genes (from preprocessing)</span>
</span><span id=__span-7-5><a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a><span class=n>MID_DIM</span> <span class=o>=</span> <span class=mi>512</span>                <span class=c1># Hidden layer dimension</span>
</span><span id=__span-7-6><a id=__codelineno-7-6 name=__codelineno-7-6 href=#__codelineno-7-6></a><span class=n>LATENT_DIM</span> <span class=o>=</span> <span class=mi>256</span>             <span class=c1># Latent space dimension</span>
</span><span id=__span-7-7><a id=__codelineno-7-7 name=__codelineno-7-7 href=#__codelineno-7-7></a><span class=n>encoder_dims</span> <span class=o>=</span> <span class=p>[</span><span class=n>INPUT_DIM</span><span class=p>,</span> <span class=n>MID_DIM</span><span class=p>,</span> <span class=n>LATENT_DIM</span><span class=p>]</span>
</span><span id=__span-7-8><a id=__codelineno-7-8 name=__codelineno-7-8 href=#__codelineno-7-8></a><span class=n>decoder_dims</span> <span class=o>=</span> <span class=p>[</span><span class=n>LATENT_DIM</span><span class=p>,</span> <span class=n>MID_DIM</span><span class=p>,</span> <span class=n>INPUT_DIM</span><span class=p>]</span>
</span><span id=__span-7-9><a id=__codelineno-7-9 name=__codelineno-7-9 href=#__codelineno-7-9></a>
</span><span id=__span-7-10><a id=__codelineno-7-10 name=__codelineno-7-10 href=#__codelineno-7-10></a><span class=c1># Training</span>
</span><span id=__span-7-11><a id=__codelineno-7-11 name=__codelineno-7-11 href=#__codelineno-7-11></a><span class=n>EPOCHS</span> <span class=o>=</span> <span class=mi>600</span>                 <span class=c1># 3 cycles Ã— 200 epochs/cycle</span>
</span><span id=__span-7-12><a id=__codelineno-7-12 name=__codelineno-7-12 href=#__codelineno-7-12></a><span class=n>BETA_CYCLES</span> <span class=o>=</span> <span class=mi>3</span>              <span class=c1># Number of beta annealing cycles</span>
</span><span id=__span-7-13><a id=__codelineno-7-13 name=__codelineno-7-13 href=#__codelineno-7-13></a><span class=n>BETA_RATIO</span> <span class=o>=</span> <span class=mf>0.5</span>             <span class=c1># Î² warmup fraction per cycle</span>
</span><span id=__span-7-14><a id=__codelineno-7-14 name=__codelineno-7-14 href=#__codelineno-7-14></a><span class=n>BATCH_SIZE</span> <span class=o>=</span> <span class=mi>8</span>
</span><span id=__span-7-15><a id=__codelineno-7-15 name=__codelineno-7-15 href=#__codelineno-7-15></a><span class=n>LEARNING_RATE</span> <span class=o>=</span> <span class=mf>1e-3</span>
</span><span id=__span-7-16><a id=__codelineno-7-16 name=__codelineno-7-16 href=#__codelineno-7-16></a><span class=n>optimizer</span> <span class=o>=</span> <span class=s1>&#39;Adam&#39;</span>
</span><span id=__span-7-17><a id=__codelineno-7-17 name=__codelineno-7-17 href=#__codelineno-7-17></a>
</span><span id=__span-7-18><a id=__codelineno-7-18 name=__codelineno-7-18 href=#__codelineno-7-18></a><span class=c1># Reconstruction network (post-processing)</span>
</span><span id=__span-7-19><a id=__codelineno-7-19 name=__codelineno-7-19 href=#__codelineno-7-19></a><span class=n>recnet_dims</span> <span class=o>=</span> <span class=p>[</span><span class=mi>5127</span><span class=p>,</span> <span class=mi>3512</span><span class=p>,</span> <span class=mi>824</span><span class=p>,</span> <span class=mi>3731</span><span class=p>,</span> <span class=mi>5127</span><span class=p>]</span>
</span><span id=__span-7-20><a id=__codelineno-7-20 name=__codelineno-7-20 href=#__codelineno-7-20></a><span class=n>recnet_epochs</span> <span class=o>=</span> <span class=mi>1000</span>
</span><span id=__span-7-21><a id=__codelineno-7-21 name=__codelineno-7-21 href=#__codelineno-7-21></a><span class=n>recnet_lr</span> <span class=o>=</span> <span class=mf>1e-4</span>
</span><span id=__span-7-22><a id=__codelineno-7-22 name=__codelineno-7-22 href=#__codelineno-7-22></a><span class=n>recnet_batch_size</span> <span class=o>=</span> <span class=mi>8</span>
</span><span id=__span-7-23><a id=__codelineno-7-23 name=__codelineno-7-23 href=#__codelineno-7-23></a>
</span><span id=__span-7-24><a id=__codelineno-7-24 name=__codelineno-7-24 href=#__codelineno-7-24></a><span class=c1># Random seed</span>
</span><span id=__span-7-25><a id=__codelineno-7-25 name=__codelineno-7-25 href=#__codelineno-7-25></a><span class=n>seed</span> <span class=o>=</span> <span class=mi>2023</span>
</span></code></pre></div> <p>Expected training performance: <div class="language-text highlight"><pre><span></span><code><span id=__span-8-1><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a>Final VAE Loss: ~580-620
</span><span id=__span-8-2><a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a>Final Reconstruction MSE: &lt;0.5
</span><span id=__span-8-3><a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a>Training time: 2-4 hours (GPU), 8-12 hours (CPU)
</span></code></pre></div></p> <h3 id=5-trajectory-generation-parameters>5. Trajectory Generation Parameters<a class=headerlink href=#5-trajectory-generation-parameters title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-9-1><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=c1># From scripts/pipeline_steps/4_trajectories.py</span>
</span><span id=__span-9-2><a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>
</span><span id=__span-9-3><a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=n>n_trajectories</span> <span class=o>=</span> <span class=mi>500</span>         <span class=c1># Number of synthetic trajectories</span>
</span><span id=__span-9-4><a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a><span class=n>n_timepoints</span> <span class=o>=</span> <span class=mi>20</span>            <span class=c1># Timepoints per trajectory</span>
</span><span id=__span-9-5><a id=__codelineno-9-5 name=__codelineno-9-5 href=#__codelineno-9-5></a><span class=n>interpolation</span> <span class=o>=</span> <span class=s1>&#39;linear&#39;</span>     <span class=c1># Linear interpolation in latent space</span>
</span><span id=__span-9-6><a id=__codelineno-9-6 name=__codelineno-9-6 href=#__codelineno-9-6></a>
</span><span id=__span-9-7><a id=__codelineno-9-7 name=__codelineno-9-7 href=#__codelineno-9-7></a><span class=c1># Trajectory types</span>
</span><span id=__span-9-8><a id=__codelineno-9-8 name=__codelineno-9-8 href=#__codelineno-9-8></a><span class=n>trajectory_types</span> <span class=o>=</span> <span class=p>[</span>
</span><span id=__span-9-9><a id=__codelineno-9-9 name=__codelineno-9-9 href=#__codelineno-9-9></a>    <span class=s1>&#39;early_to_late&#39;</span>          <span class=c1># Main analysis in publication</span>
</span><span id=__span-9-10><a id=__codelineno-9-10 name=__codelineno-9-10 href=#__codelineno-9-10></a><span class=p>]</span>
</span><span id=__span-9-11><a id=__codelineno-9-11 name=__codelineno-9-11 href=#__codelineno-9-11></a>
</span><span id=__span-9-12><a id=__codelineno-9-12 name=__codelineno-9-12 href=#__codelineno-9-12></a><span class=c1># Random seed</span>
</span><span id=__span-9-13><a id=__codelineno-9-13 name=__codelineno-9-13 href=#__codelineno-9-13></a><span class=n>seed</span> <span class=o>=</span> <span class=mi>2023</span>
</span></code></pre></div> <h3 id=6-classification-parameters>6. Classification Parameters<a class=headerlink href=#6-classification-parameters title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-10-1><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=c1># From scripts/pipeline_steps/5_classification.py</span>
</span><span id=__span-10-2><a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a>
</span><span id=__span-10-3><a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a><span class=c1># XGBoost hyperparameters</span>
</span><span id=__span-10-4><a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=n>max_depth</span> <span class=o>=</span> <span class=mi>6</span>
</span><span id=__span-10-5><a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a><span class=n>n_estimators</span> <span class=o>=</span> <span class=mi>100</span>
</span><span id=__span-10-6><a id=__codelineno-10-6 name=__codelineno-10-6 href=#__codelineno-10-6></a><span class=n>learning_rate</span> <span class=o>=</span> <span class=mf>0.1</span>
</span><span id=__span-10-7><a id=__codelineno-10-7 name=__codelineno-10-7 href=#__codelineno-10-7></a><span class=n>subsample</span> <span class=o>=</span> <span class=mf>0.8</span>
</span><span id=__span-10-8><a id=__codelineno-10-8 name=__codelineno-10-8 href=#__codelineno-10-8></a><span class=n>colsample_bytree</span> <span class=o>=</span> <span class=mf>0.8</span>
</span><span id=__span-10-9><a id=__codelineno-10-9 name=__codelineno-10-9 href=#__codelineno-10-9></a>
</span><span id=__span-10-10><a id=__codelineno-10-10 name=__codelineno-10-10 href=#__codelineno-10-10></a><span class=c1># Cross-validation</span>
</span><span id=__span-10-11><a id=__codelineno-10-11 name=__codelineno-10-11 href=#__codelineno-10-11></a><span class=n>n_folds</span> <span class=o>=</span> <span class=mi>5</span>
</span><span id=__span-10-12><a id=__codelineno-10-12 name=__codelineno-10-12 href=#__codelineno-10-12></a><span class=n>stratified</span> <span class=o>=</span> <span class=kc>True</span>
</span><span id=__span-10-13><a id=__codelineno-10-13 name=__codelineno-10-13 href=#__codelineno-10-13></a>
</span><span id=__span-10-14><a id=__codelineno-10-14 name=__codelineno-10-14 href=#__codelineno-10-14></a><span class=c1># SHAP analysis</span>
</span><span id=__span-10-15><a id=__codelineno-10-15 name=__codelineno-10-15 href=#__codelineno-10-15></a><span class=n>n_top_genes</span> <span class=o>=</span> <span class=mi>100</span>            <span class=c1># Top genes by SHAP importance</span>
</span><span id=__span-10-16><a id=__codelineno-10-16 name=__codelineno-10-16 href=#__codelineno-10-16></a>
</span><span id=__span-10-17><a id=__codelineno-10-17 name=__codelineno-10-17 href=#__codelineno-10-17></a><span class=c1># Random seed</span>
</span><span id=__span-10-18><a id=__codelineno-10-18 name=__codelineno-10-18 href=#__codelineno-10-18></a><span class=n>seed</span> <span class=o>=</span> <span class=mi>2023</span>
</span></code></pre></div> <p>Expected performance (on test set): <div class="language-text highlight"><pre><span></span><code><span id=__span-11-1><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a>Accuracy: 0.89 Â± 0.03
</span><span id=__span-11-2><a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a>ROC AUC: 0.94 Â± 0.02
</span><span id=__span-11-3><a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a>Precision: 0.87 Â± 0.04
</span><span id=__span-11-4><a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a>Recall: 0.91 Â± 0.03
</span><span id=__span-11-5><a id=__codelineno-11-5 name=__codelineno-11-5 href=#__codelineno-11-5></a>F1 Score: 0.89 Â± 0.03
</span></code></pre></div></p> <h3 id=7-enrichment-analysis-parameters>7. Enrichment Analysis Parameters<a class=headerlink href=#7-enrichment-analysis-parameters title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-12-1><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=c1># From scripts/pipeline_steps/6_enrichment_analysis.py</span>
</span><span id=__span-12-2><a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a>
</span><span id=__span-12-3><a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a><span class=c1># GSEA parameters</span>
</span><span id=__span-12-4><a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a><span class=n>gsea_version</span> <span class=o>=</span> <span class=s1>&#39;4.3.2&#39;</span>
</span><span id=__span-12-5><a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a><span class=n>collapse</span> <span class=o>=</span> <span class=kc>False</span>
</span><span id=__span-12-6><a id=__codelineno-12-6 name=__codelineno-12-6 href=#__codelineno-12-6></a><span class=n>nperm</span> <span class=o>=</span> <span class=mi>1000</span>                 <span class=c1># Number of permutations</span>
</span><span id=__span-12-7><a id=__codelineno-12-7 name=__codelineno-12-7 href=#__codelineno-12-7></a><span class=n>set_max</span> <span class=o>=</span> <span class=mi>500</span>                <span class=c1># Maximum pathway size</span>
</span><span id=__span-12-8><a id=__codelineno-12-8 name=__codelineno-12-8 href=#__codelineno-12-8></a><span class=n>set_min</span> <span class=o>=</span> <span class=mi>15</span>                 <span class=c1># Minimum pathway size</span>
</span><span id=__span-12-9><a id=__codelineno-12-9 name=__codelineno-12-9 href=#__codelineno-12-9></a><span class=n>scoring_scheme</span> <span class=o>=</span> <span class=s1>&#39;weighted&#39;</span>
</span><span id=__span-12-10><a id=__codelineno-12-10 name=__codelineno-12-10 href=#__codelineno-12-10></a>
</span><span id=__span-12-11><a id=__codelineno-12-11 name=__codelineno-12-11 href=#__codelineno-12-11></a><span class=c1># Pathway database</span>
</span><span id=__span-12-12><a id=__codelineno-12-12 name=__codelineno-12-12 href=#__codelineno-12-12></a><span class=n>pathways_file</span> <span class=o>=</span> <span class=s1>&#39;data/external/ReactomePathways.gmt&#39;</span>
</span><span id=__span-12-13><a id=__codelineno-12-13 name=__codelineno-12-13 href=#__codelineno-12-13></a><span class=n>pathways_version</span> <span class=o>=</span> <span class=s1>&#39;v85&#39;</span>     <span class=c1># Reactome version 85 (2023-12)</span>
</span><span id=__span-12-14><a id=__codelineno-12-14 name=__codelineno-12-14 href=#__codelineno-12-14></a><span class=n>n_pathways</span> <span class=o>=</span> <span class=mi>2557</span>            <span class=c1># Total pathways in database</span>
</span><span id=__span-12-15><a id=__codelineno-12-15 name=__codelineno-12-15 href=#__codelineno-12-15></a>
</span><span id=__span-12-16><a id=__codelineno-12-16 name=__codelineno-12-16 href=#__codelineno-12-16></a><span class=c1># Parallel processing</span>
</span><span id=__span-12-17><a id=__codelineno-12-17 name=__codelineno-12-17 href=#__codelineno-12-17></a><span class=n>n_threads</span> <span class=o>=</span> <span class=mi>8</span>                <span class=c1># Adjust for your system</span>
</span><span id=__span-12-18><a id=__codelineno-12-18 name=__codelineno-12-18 href=#__codelineno-12-18></a>
</span><span id=__span-12-19><a id=__codelineno-12-19 name=__codelineno-12-19 href=#__codelineno-12-19></a><span class=c1># FDR threshold for significance</span>
</span><span id=__span-12-20><a id=__codelineno-12-20 name=__codelineno-12-20 href=#__codelineno-12-20></a><span class=n>fdr_threshold</span> <span class=o>=</span> <span class=mf>0.05</span>
</span></code></pre></div> <p>Expected results: <div class="language-text highlight"><pre><span></span><code><span id=__span-13-1><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a>Total enrichment tests: 500 trajectories Ã— 20 timepoints Ã— 2557 pathways
</span><span id=__span-13-2><a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a>Significant pathways (FDR &lt; 0.05): ~15-20% of tests
</span><span id=__span-13-3><a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a>Top enriched pathways: Cell Cycle, DNA Repair, Immune Response
</span></code></pre></div></p> <h2 id=validation-checkpoints>Validation Checkpoints<a class=headerlink href=#validation-checkpoints title="Permanent link">&para;</a></h2> <p>After each step, validate outputs:</p> <h3 id=checkpoint-1-after-preprocessing>Checkpoint 1: After Preprocessing<a class=headerlink href=#checkpoint-1-after-preprocessing title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-14-1><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=kn>import</span><span class=w> </span><span class=nn>pandas</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>pd</span>
</span><span id=__span-14-2><a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a><span class=kn>import</span><span class=w> </span><span class=nn>json</span>
</span><span id=__span-14-3><a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a>
</span><span id=__span-14-4><a id=__codelineno-14-4 name=__codelineno-14-4 href=#__codelineno-14-4></a><span class=c1># Load preprocessing info</span>
</span><span id=__span-14-5><a id=__codelineno-14-5 name=__codelineno-14-5 href=#__codelineno-14-5></a><span class=k>with</span> <span class=nb>open</span><span class=p>(</span><span class=s1>&#39;data/interim/preprocessed_KIRC_data/preprocessing_info.json&#39;</span><span class=p>)</span> <span class=k>as</span> <span class=n>f</span><span class=p>:</span>
</span><span id=__span-14-6><a id=__codelineno-14-6 name=__codelineno-14-6 href=#__codelineno-14-6></a>    <span class=n>info</span> <span class=o>=</span> <span class=n>json</span><span class=o>.</span><span class=n>load</span><span class=p>(</span><span class=n>f</span><span class=p>)</span>
</span><span id=__span-14-7><a id=__codelineno-14-7 name=__codelineno-14-7 href=#__codelineno-14-7></a>
</span><span id=__span-14-8><a id=__codelineno-14-8 name=__codelineno-14-8 href=#__codelineno-14-8></a><span class=c1># Verify key values</span>
</span><span id=__span-14-9><a id=__codelineno-14-9 name=__codelineno-14-9 href=#__codelineno-14-9></a><span class=k>assert</span> <span class=n>info</span><span class=p>[</span><span class=s1>&#39;n_samples_final&#39;</span><span class=p>]</span> <span class=o>==</span> <span class=mi>498</span><span class=p>,</span> <span class=s2>&quot;Sample count mismatch!&quot;</span>
</span><span id=__span-14-10><a id=__codelineno-14-10 name=__codelineno-14-10 href=#__codelineno-14-10></a><span class=k>assert</span> <span class=mi>5000</span> <span class=o>&lt;=</span> <span class=n>info</span><span class=p>[</span><span class=s1>&#39;n_features_final&#39;</span><span class=p>]</span> <span class=o>&lt;=</span> <span class=mi>5200</span><span class=p>,</span> <span class=s2>&quot;Gene count out of range!&quot;</span>
</span><span id=__span-14-11><a id=__codelineno-14-11 name=__codelineno-14-11 href=#__codelineno-14-11></a><span class=k>assert</span> <span class=n>info</span><span class=p>[</span><span class=s1>&#39;n_outliers&#39;</span><span class=p>]</span> <span class=o>&gt;</span> <span class=mi>30</span><span class=p>,</span> <span class=s2>&quot;Outlier detection may have failed!&quot;</span>
</span><span id=__span-14-12><a id=__codelineno-14-12 name=__codelineno-14-12 href=#__codelineno-14-12></a>
</span><span id=__span-14-13><a id=__codelineno-14-13 name=__codelineno-14-13 href=#__codelineno-14-13></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;âœ“ Checkpoint 1 passed: Preprocessing validated&quot;</span><span class=p>)</span>
</span></code></pre></div> <h3 id=checkpoint-2-after-vae-training>Checkpoint 2: After VAE Training<a class=headerlink href=#checkpoint-2-after-vae-training title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-15-1><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=kn>import</span><span class=w> </span><span class=nn>json</span>
</span><span id=__span-15-2><a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a>
</span><span id=__span-15-3><a id=__codelineno-15-3 name=__codelineno-15-3 href=#__codelineno-15-3></a><span class=c1># Load VAE metrics</span>
</span><span id=__span-15-4><a id=__codelineno-15-4 name=__codelineno-15-4 href=#__codelineno-15-4></a><span class=k>with</span> <span class=nb>open</span><span class=p>(</span><span class=s1>&#39;models/20251217_models_KIRC/vae/vae_config.json&#39;</span><span class=p>)</span> <span class=k>as</span> <span class=n>f</span><span class=p>:</span>
</span><span id=__span-15-5><a id=__codelineno-15-5 name=__codelineno-15-5 href=#__codelineno-15-5></a>    <span class=n>config</span> <span class=o>=</span> <span class=n>json</span><span class=o>.</span><span class=n>load</span><span class=p>(</span><span class=n>f</span><span class=p>)</span>
</span><span id=__span-15-6><a id=__codelineno-15-6 name=__codelineno-15-6 href=#__codelineno-15-6></a>
</span><span id=__span-15-7><a id=__codelineno-15-7 name=__codelineno-15-7 href=#__codelineno-15-7></a><span class=k>assert</span> <span class=n>config</span><span class=p>[</span><span class=s1>&#39;latent_dim&#39;</span><span class=p>]</span> <span class=o>==</span> <span class=mi>256</span><span class=p>,</span> <span class=s2>&quot;Wrong latent dimension!&quot;</span>
</span><span id=__span-15-8><a id=__codelineno-15-8 name=__codelineno-15-8 href=#__codelineno-15-8></a><span class=k>assert</span> <span class=n>config</span><span class=p>[</span><span class=s1>&#39;epochs&#39;</span><span class=p>]</span> <span class=o>==</span> <span class=mi>600</span><span class=p>,</span> <span class=s2>&quot;Wrong number of epochs!&quot;</span>
</span><span id=__span-15-9><a id=__codelineno-15-9 name=__codelineno-15-9 href=#__codelineno-15-9></a>
</span><span id=__span-15-10><a id=__codelineno-15-10 name=__codelineno-15-10 href=#__codelineno-15-10></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;âœ“ Checkpoint 2 passed: VAE training validated&quot;</span><span class=p>)</span>
</span></code></pre></div> <h3 id=checkpoint-3-after-trajectory-generation>Checkpoint 3: After Trajectory Generation<a class=headerlink href=#checkpoint-3-after-trajectory-generation title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-16-1><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=kn>import</span><span class=w> </span><span class=nn>glob</span>
</span><span id=__span-16-2><a id=__codelineno-16-2 name=__codelineno-16-2 href=#__codelineno-16-2></a><span class=kn>import</span><span class=w> </span><span class=nn>pandas</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>pd</span>
</span><span id=__span-16-3><a id=__codelineno-16-3 name=__codelineno-16-3 href=#__codelineno-16-3></a>
</span><span id=__span-16-4><a id=__codelineno-16-4 name=__codelineno-16-4 href=#__codelineno-16-4></a><span class=c1># Count trajectories</span>
</span><span id=__span-16-5><a id=__codelineno-16-5 name=__codelineno-16-5 href=#__codelineno-16-5></a><span class=n>traj_files</span> <span class=o>=</span> <span class=n>glob</span><span class=o>.</span><span class=n>glob</span><span class=p>(</span><span class=s1>&#39;data/interim/*/kirc/early_to_late/*.csv&#39;</span><span class=p>)</span>
</span><span id=__span-16-6><a id=__codelineno-16-6 name=__codelineno-16-6 href=#__codelineno-16-6></a><span class=k>assert</span> <span class=nb>len</span><span class=p>(</span><span class=n>traj_files</span><span class=p>)</span> <span class=o>==</span> <span class=mi>500</span><span class=p>,</span> <span class=sa>f</span><span class=s2>&quot;Expected 500 trajectories, got </span><span class=si>{</span><span class=nb>len</span><span class=p>(</span><span class=n>traj_files</span><span class=p>)</span><span class=si>}</span><span class=s2>&quot;</span>
</span><span id=__span-16-7><a id=__codelineno-16-7 name=__codelineno-16-7 href=#__codelineno-16-7></a>
</span><span id=__span-16-8><a id=__codelineno-16-8 name=__codelineno-16-8 href=#__codelineno-16-8></a><span class=c1># Check trajectory dimensions</span>
</span><span id=__span-16-9><a id=__codelineno-16-9 name=__codelineno-16-9 href=#__codelineno-16-9></a><span class=n>traj</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=n>traj_files</span><span class=p>[</span><span class=mi>0</span><span class=p>],</span> <span class=n>index_col</span><span class=o>=</span><span class=mi>0</span><span class=p>)</span>
</span><span id=__span-16-10><a id=__codelineno-16-10 name=__codelineno-16-10 href=#__codelineno-16-10></a><span class=k>assert</span> <span class=n>traj</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>0</span><span class=p>]</span> <span class=o>==</span> <span class=mi>20</span><span class=p>,</span> <span class=s2>&quot;Wrong number of timepoints!&quot;</span>
</span><span id=__span-16-11><a id=__codelineno-16-11 name=__codelineno-16-11 href=#__codelineno-16-11></a><span class=k>assert</span> <span class=mi>5000</span> <span class=o>&lt;=</span> <span class=n>traj</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>1</span><span class=p>]</span> <span class=o>&lt;=</span> <span class=mi>5200</span><span class=p>,</span> <span class=s2>&quot;Wrong number of genes!&quot;</span>
</span><span id=__span-16-12><a id=__codelineno-16-12 name=__codelineno-16-12 href=#__codelineno-16-12></a>
</span><span id=__span-16-13><a id=__codelineno-16-13 name=__codelineno-16-13 href=#__codelineno-16-13></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;âœ“ Checkpoint 3 passed: Trajectories validated&quot;</span><span class=p>)</span>
</span></code></pre></div> <h3 id=checkpoint-4-after-classification>Checkpoint 4: After Classification<a class=headerlink href=#checkpoint-4-after-classification title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-17-1><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=kn>import</span><span class=w> </span><span class=nn>json</span>
</span><span id=__span-17-2><a id=__codelineno-17-2 name=__codelineno-17-2 href=#__codelineno-17-2></a>
</span><span id=__span-17-3><a id=__codelineno-17-3 name=__codelineno-17-3 href=#__codelineno-17-3></a><span class=c1># Load classification metrics</span>
</span><span id=__span-17-4><a id=__codelineno-17-4 name=__codelineno-17-4 href=#__codelineno-17-4></a><span class=k>with</span> <span class=nb>open</span><span class=p>(</span><span class=s1>&#39;models/20251217_classification_kirc/classification_metrics.json&#39;</span><span class=p>)</span> <span class=k>as</span> <span class=n>f</span><span class=p>:</span>
</span><span id=__span-17-5><a id=__codelineno-17-5 name=__codelineno-17-5 href=#__codelineno-17-5></a>    <span class=n>metrics</span> <span class=o>=</span> <span class=n>json</span><span class=o>.</span><span class=n>load</span><span class=p>(</span><span class=n>f</span><span class=p>)</span>
</span><span id=__span-17-6><a id=__codelineno-17-6 name=__codelineno-17-6 href=#__codelineno-17-6></a>
</span><span id=__span-17-7><a id=__codelineno-17-7 name=__codelineno-17-7 href=#__codelineno-17-7></a><span class=k>assert</span> <span class=n>metrics</span><span class=p>[</span><span class=s1>&#39;test_accuracy&#39;</span><span class=p>]</span> <span class=o>&gt;</span> <span class=mf>0.85</span><span class=p>,</span> <span class=s2>&quot;Accuracy too low!&quot;</span>
</span><span id=__span-17-8><a id=__codelineno-17-8 name=__codelineno-17-8 href=#__codelineno-17-8></a><span class=k>assert</span> <span class=n>metrics</span><span class=p>[</span><span class=s1>&#39;test_roc_auc&#39;</span><span class=p>]</span> <span class=o>&gt;</span> <span class=mf>0.90</span><span class=p>,</span> <span class=s2>&quot;ROC AUC too low!&quot;</span>
</span><span id=__span-17-9><a id=__codelineno-17-9 name=__codelineno-17-9 href=#__codelineno-17-9></a>
</span><span id=__span-17-10><a id=__codelineno-17-10 name=__codelineno-17-10 href=#__codelineno-17-10></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;âœ“ Checkpoint 4 passed: Classification validated&quot;</span><span class=p>)</span>
</span></code></pre></div> <h3 id=checkpoint-5-after-enrichment>Checkpoint 5: After Enrichment<a class=headerlink href=#checkpoint-5-after-enrichment title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-18-1><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=kn>import</span><span class=w> </span><span class=nn>pandas</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>pd</span>
</span><span id=__span-18-2><a id=__codelineno-18-2 name=__codelineno-18-2 href=#__codelineno-18-2></a>
</span><span id=__span-18-3><a id=__codelineno-18-3 name=__codelineno-18-3 href=#__codelineno-18-3></a><span class=c1># Load enrichment results</span>
</span><span id=__span-18-4><a id=__codelineno-18-4 name=__codelineno-18-4 href=#__codelineno-18-4></a><span class=n>enrichment</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=s1>&#39;data/processed/20251217_enrichment/trajectory_enrichment.csv&#39;</span><span class=p>)</span>
</span><span id=__span-18-5><a id=__codelineno-18-5 name=__codelineno-18-5 href=#__codelineno-18-5></a>
</span><span id=__span-18-6><a id=__codelineno-18-6 name=__codelineno-18-6 href=#__codelineno-18-6></a><span class=k>assert</span> <span class=n>enrichment</span><span class=p>[</span><span class=s1>&#39;Patient&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>nunique</span><span class=p>()</span> <span class=o>==</span> <span class=mi>500</span><span class=p>,</span> <span class=s2>&quot;Missing trajectories!&quot;</span>
</span><span id=__span-18-7><a id=__codelineno-18-7 name=__codelineno-18-7 href=#__codelineno-18-7></a><span class=k>assert</span> <span class=s1>&#39;Cell Cycle&#39;</span> <span class=ow>in</span> <span class=n>enrichment</span><span class=p>[</span><span class=s1>&#39;NAME&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>values</span><span class=p>,</span> <span class=s2>&quot;Key pathway missing!&quot;</span>
</span><span id=__span-18-8><a id=__codelineno-18-8 name=__codelineno-18-8 href=#__codelineno-18-8></a>
</span><span id=__span-18-9><a id=__codelineno-18-9 name=__codelineno-18-9 href=#__codelineno-18-9></a><span class=n>sig</span> <span class=o>=</span> <span class=n>enrichment</span><span class=p>[</span><span class=n>enrichment</span><span class=p>[</span><span class=s1>&#39;FDR q-val&#39;</span><span class=p>]</span> <span class=o>&lt;</span> <span class=mf>0.05</span><span class=p>]</span>
</span><span id=__span-18-10><a id=__codelineno-18-10 name=__codelineno-18-10 href=#__codelineno-18-10></a><span class=k>assert</span> <span class=nb>len</span><span class=p>(</span><span class=n>sig</span><span class=p>)</span> <span class=o>&gt;</span> <span class=mi>1000</span><span class=p>,</span> <span class=s2>&quot;Too few significant results!&quot;</span>
</span><span id=__span-18-11><a id=__codelineno-18-11 name=__codelineno-18-11 href=#__codelineno-18-11></a>
</span><span id=__span-18-12><a id=__codelineno-18-12 name=__codelineno-18-12 href=#__codelineno-18-12></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;âœ“ Checkpoint 5 passed: Enrichment validated&quot;</span><span class=p>)</span>
</span></code></pre></div> <h2 id=generating-publication-figures>Generating Publication Figures<a class=headerlink href=#generating-publication-figures title="Permanent link">&para;</a></h2> <p>After successful pipeline completion, generate all figures:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-19-1><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=c1># Run figure generation script</span>
</span><span id=__span-19-2><a id=__codelineno-19-2 name=__codelineno-19-2 href=#__codelineno-19-2></a>python<span class=w> </span>scripts/generate_publication_figures.py
</span></code></pre></div> <p>This creates:</p> <ul> <li><strong>Figure 1</strong>: Study design and data overview</li> <li><strong>Figure 2</strong>: VAE latent space visualization</li> <li><strong>Figure 3</strong>: Trajectory examples</li> <li><strong>Figure 4</strong>: Classification performance and SHAP</li> <li><strong>Figure 5</strong>: Enrichment heatmap for key pathways</li> <li><strong>Supplementary Figures</strong>: Additional validations</li> </ul> <p>Figures are saved to: <code>reports/figures/publication/</code></p> <h2 id=expected-final-outputs>Expected Final Outputs<a class=headerlink href=#expected-final-outputs title="Permanent link">&para;</a></h2> <p>After complete reproduction:</p> <div class="language-text highlight"><pre><span></span><code><span id=__span-20-1><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a>renalprog/
</span><span id=__span-20-2><a id=__codelineno-20-2 name=__codelineno-20-2 href=#__codelineno-20-2></a>â”œâ”€â”€ data/
</span><span id=__span-20-3><a id=__codelineno-20-3 name=__codelineno-20-3 href=#__codelineno-20-3></a>â”‚   â”œâ”€â”€ interim/
</span><span id=__span-20-4><a id=__codelineno-20-4 name=__codelineno-20-4 href=#__codelineno-20-4></a>â”‚   â”‚   â”œâ”€â”€ preprocessed_KIRC_data/
</span><span id=__span-20-5><a id=__codelineno-20-5 name=__codelineno-20-5 href=#__codelineno-20-5></a>â”‚   â”‚   â”‚   â”œâ”€â”€ preprocessed_rnaseq.csv         # 498 Ã— 5127
</span><span id=__span-20-6><a id=__codelineno-20-6 name=__codelineno-20-6 href=#__codelineno-20-6></a>â”‚   â”‚   â”‚   â””â”€â”€ preprocessing_info.json
</span><span id=__span-20-7><a id=__codelineno-20-7 name=__codelineno-20-7 href=#__codelineno-20-7></a>â”‚   â”‚   â””â”€â”€ 20251217_synthetic_data/
</span><span id=__span-20-8><a id=__codelineno-20-8 name=__codelineno-20-8 href=#__codelineno-20-8></a>â”‚   â”‚       â””â”€â”€ kirc/early_to_late/             # 500 files
</span><span id=__span-20-9><a id=__codelineno-20-9 name=__codelineno-20-9 href=#__codelineno-20-9></a>â”‚   â”œâ”€â”€ processed/
</span><span id=__span-20-10><a id=__codelineno-20-10 name=__codelineno-20-10 href=#__codelineno-20-10></a>â”‚   â”‚   â””â”€â”€ 20251217_enrichment/
</span><span id=__span-20-11><a id=__codelineno-20-11 name=__codelineno-20-11 href=#__codelineno-20-11></a>â”‚   â”‚       â””â”€â”€ trajectory_enrichment.csv       # ~25M rows
</span><span id=__span-20-12><a id=__codelineno-20-12 name=__codelineno-20-12 href=#__codelineno-20-12></a>â”œâ”€â”€ models/
</span><span id=__span-20-13><a id=__codelineno-20-13 name=__codelineno-20-13 href=#__codelineno-20-13></a>â”‚   â”œâ”€â”€ 20251217_models_KIRC/
</span><span id=__span-20-14><a id=__codelineno-20-14 name=__codelineno-20-14 href=#__codelineno-20-14></a>â”‚   â”‚   â”œâ”€â”€ vae/vae_model.pt
</span><span id=__span-20-15><a id=__codelineno-20-15 name=__codelineno-20-15 href=#__codelineno-20-15></a>â”‚   â”‚   â””â”€â”€ reconstruction/reconstruction_network.pt
</span><span id=__span-20-16><a id=__codelineno-20-16 name=__codelineno-20-16 href=#__codelineno-20-16></a>â”‚   â””â”€â”€ 20251217_classification_kirc/
</span><span id=__span-20-17><a id=__codelineno-20-17 name=__codelineno-20-17 href=#__codelineno-20-17></a>â”‚       â”œâ”€â”€ classifier.pkl
</span><span id=__span-20-18><a id=__codelineno-20-18 name=__codelineno-20-18 href=#__codelineno-20-18></a>â”‚       â””â”€â”€ important_genes.csv                 # Top 100 genes
</span><span id=__span-20-19><a id=__codelineno-20-19 name=__codelineno-20-19 href=#__codelineno-20-19></a>â””â”€â”€ reports/
</span><span id=__span-20-20><a id=__codelineno-20-20 name=__codelineno-20-20 href=#__codelineno-20-20></a>    â””â”€â”€ figures/
</span><span id=__span-20-21><a id=__codelineno-20-21 name=__codelineno-20-21 href=#__codelineno-20-21></a>        â””â”€â”€ publication/                         # All publication figures
</span></code></pre></div> <h2 id=differences-from-publication>Differences from Publication<a class=headerlink href=#differences-from-publication title="Permanent link">&para;</a></h2> <p>Minor differences are expected due to:</p> <ol> <li><strong>Random Initialization</strong>: Despite fixed seeds, hardware differences may cause slight variations</li> <li>VAE final loss: Â±5%</li> <li> <p>Classification accuracy: Â±2%</p> </li> <li> <p><strong>GSEA Stochasticity</strong>: Permutation-based p-values vary slightly</p> </li> <li>FDR q-values: Â±10%</li> <li> <p>Pathway rankings may differ slightly</p> </li> <li> <p><strong>Software Versions</strong>: Minor updates to dependencies</p> </li> <li>All major results should reproduce</li> <li>Exact numerical values may differ in decimal places</li> </ol> <h2 id=troubleshooting>Troubleshooting<a class=headerlink href=#troubleshooting title="Permanent link">&para;</a></h2> <h3 id=different-sample-counts>Different Sample Counts<a class=headerlink href=#different-sample-counts title="Permanent link">&para;</a></h3> <p>If you get different sample counts after preprocessing:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-21-1><a id=__codelineno-21-1 name=__codelineno-21-1 href=#__codelineno-21-1></a><span class=c1># Check your random seed</span>
</span><span id=__span-21-2><a id=__codelineno-21-2 name=__codelineno-21-2 href=#__codelineno-21-2></a><span class=kn>import</span><span class=w> </span><span class=nn>numpy</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>np</span>
</span><span id=__span-21-3><a id=__codelineno-21-3 name=__codelineno-21-3 href=#__codelineno-21-3></a><span class=n>np</span><span class=o>.</span><span class=n>random</span><span class=o>.</span><span class=n>seed</span><span class=p>(</span><span class=mi>2023</span><span class=p>)</span>  <span class=c1># Must be set before running</span>
</span><span id=__span-21-4><a id=__codelineno-21-4 name=__codelineno-21-4 href=#__codelineno-21-4></a>
</span><span id=__span-21-5><a id=__codelineno-21-5 name=__codelineno-21-5 href=#__codelineno-21-5></a><span class=c1># Check alpha parameter</span>
</span><span id=__span-21-6><a id=__codelineno-21-6 name=__codelineno-21-6 href=#__codelineno-21-6></a><span class=n>alpha</span> <span class=o>=</span> <span class=mf>0.05</span>  <span class=c1># Must match publication</span>
</span></code></pre></div> <h3 id=different-gene-counts>Different Gene Counts<a class=headerlink href=#different-gene-counts title="Permanent link">&para;</a></h3> <p>Gene counts may vary slightly (5000-5200 range is acceptable) due to: - Tied variance values at filtering threshold - Floating point precision differences</p> <p>This is normal and won't significantly affect results.</p> <h3 id=performance-metrics-dont-match>Performance Metrics Don't Match<a class=headerlink href=#performance-metrics-dont-match title="Permanent link">&para;</a></h3> <p>If classification metrics differ by &gt;5%:</p> <ol> <li>Check train/test split seed</li> <li>Verify XGBoost version (should be 1.7.6)</li> <li>Check for data leakage in preprocessing</li> </ol> <h3 id=gsea-fails>GSEA Fails<a class=headerlink href=#gsea-fails title="Permanent link">&para;</a></h3> <p>Common GSEA issues:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-22-1><a id=__codelineno-22-1 name=__codelineno-22-1 href=#__codelineno-22-1></a><span class=c1># Check GSEA installation</span>
</span><span id=__span-22-2><a id=__codelineno-22-2 name=__codelineno-22-2 href=#__codelineno-22-2></a>./GSEA_4.3.2/gsea-cli.sh<span class=w> </span>--version
</span><span id=__span-22-3><a id=__codelineno-22-3 name=__codelineno-22-3 href=#__codelineno-22-3></a>
</span><span id=__span-22-4><a id=__codelineno-22-4 name=__codelineno-22-4 href=#__codelineno-22-4></a><span class=c1># Check pathway file format</span>
</span><span id=__span-22-5><a id=__codelineno-22-5 name=__codelineno-22-5 href=#__codelineno-22-5></a>head<span class=w> </span>-n<span class=w> </span><span class=m>5</span><span class=w> </span>data/external/ReactomePathways.gmt
</span><span id=__span-22-6><a id=__codelineno-22-6 name=__codelineno-22-6 href=#__codelineno-22-6></a>
</span><span id=__span-22-7><a id=__codelineno-22-7 name=__codelineno-22-7 href=#__codelineno-22-7></a><span class=c1># Check Java version</span>
</span><span id=__span-22-8><a id=__codelineno-22-8 name=__codelineno-22-8 href=#__codelineno-22-8></a>java<span class=w> </span>-version<span class=w>  </span><span class=c1># Should be Java 11+</span>
</span></code></pre></div> <h2 id=performance-benchmarks>Performance Benchmarks<a class=headerlink href=#performance-benchmarks title="Permanent link">&para;</a></h2> <p>Expected runtimes (will vary with hardware):</p> <table> <thead> <tr> <th>Hardware</th> <th>Total Time</th> <th>Bottleneck Steps</th> </tr> </thead> <tbody> <tr> <td>HPC (48 cores, A100)</td> <td>3 hours</td> <td>Step 2 (1h), Step 6 (1h)</td> </tr> <tr> <td>Desktop (16 cores, RTX 4080)</td> <td>6 hours</td> <td>Step 2 (2h), Step 6 (2h)</td> </tr> <tr> <td>Desktop (8 cores, GTX 1080)</td> <td>9 hours</td> <td>Step 2 (3h), Step 6 (4h)</td> </tr> <tr> <td>Laptop (4 cores, CPU only)</td> <td>16 hours</td> <td>Step 2 (8h), Step 6 (6h)</td> </tr> </tbody> </table> <h2 id=citation>Citation<a class=headerlink href=#citation title="Permanent link">&para;</a></h2> <p>When reproducing these results, please cite:</p> <div class="language-bibtex highlight"><pre><span></span><code><span id=__span-23-1><a id=__codelineno-23-1 name=__codelineno-23-1 href=#__codelineno-23-1></a><span class=nc>@article</span><span class=p>{</span><span class=nl>renalprog2024</span><span class=p>,</span>
</span><span id=__span-23-2><a id=__codelineno-23-2 name=__codelineno-23-2 href=#__codelineno-23-2></a><span class=w>  </span><span class=na>author</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{Prol-Castelo, Guillermo and colleagues}</span><span class=p>,</span>
</span><span id=__span-23-3><a id=__codelineno-23-3 name=__codelineno-23-3 href=#__codelineno-23-3></a><span class=w>  </span><span class=na>title</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{Forecasting Kidney Cancer Progression with Generative AI}</span><span class=p>,</span>
</span><span id=__span-23-4><a id=__codelineno-23-4 name=__codelineno-23-4 href=#__codelineno-23-4></a><span class=w>  </span><span class=na>journal</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{Journal Name}</span><span class=p>,</span>
</span><span id=__span-23-5><a id=__codelineno-23-5 name=__codelineno-23-5 href=#__codelineno-23-5></a><span class=w>  </span><span class=na>year</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{2024}</span><span class=p>,</span>
</span><span id=__span-23-6><a id=__codelineno-23-6 name=__codelineno-23-6 href=#__codelineno-23-6></a><span class=w>  </span><span class=na>volume</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{X}</span><span class=p>,</span>
</span><span id=__span-23-7><a id=__codelineno-23-7 name=__codelineno-23-7 href=#__codelineno-23-7></a><span class=w>  </span><span class=na>pages</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{XXX-XXX}</span><span class=p>,</span>
</span><span id=__span-23-8><a id=__codelineno-23-8 name=__codelineno-23-8 href=#__codelineno-23-8></a><span class=w>  </span><span class=na>doi</span><span class=w> </span><span class=p>=</span><span class=w> </span><span class=s>{10.XXXX/XXXXX}</span>
</span><span id=__span-23-9><a id=__codelineno-23-9 name=__codelineno-23-9 href=#__codelineno-23-9></a><span class=p>}</span>
</span></code></pre></div> <h2 id=getting-help>Getting Help<a class=headerlink href=#getting-help title="Permanent link">&para;</a></h2> <p>If you encounter issues during reproduction:</p> <ol> <li>Check <a href=troubleshooting/ >Troubleshooting Guide</a></li> <li>Review <a href=results/ >Expected Results</a> for validation</li> <li>Open an <a href=https://github.com/gprolcastelo/renalprog/issues>Issue on GitHub</a></li> <li>Include:</li> <li>Error messages</li> <li>System information (<code>python --version</code>, <code>pip list</code>)</li> <li>Checkpoint outputs</li> <li>Steps already completed</li> </ol> <h2 id=next-steps>Next Steps<a class=headerlink href=#next-steps title="Permanent link">&para;</a></h2> <ul> <li><a href=results/ >Expected Results</a>: Detailed result descriptions</li> <li><a href=troubleshooting/ >Troubleshooting</a>: Common issues and solutions</li> <li><a href=requirements/ >System Requirements</a>: Hardware/software needs</li> </ul> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="December 19, 2025 16:49:24 UTC"><span class=timeago datetime=2025-12-19T16:49:24+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="December 19, 2025 16:49:24 UTC">2025-12-19</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="December 19, 2025 16:49:24 UTC"><span class=timeago datetime=2025-12-19T16:49:24+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="December 19, 2025 16:49:24 UTC">2025-12-19</span> </span> </aside> <form class=md-feedback name=feedback hidden> <fieldset> <legend class=md-feedback__title> Was this page helpful? </legend> <div class=md-feedback__inner> <div class=md-feedback__list> <button class="md-feedback__icon md-icon" type=submit title="This page was helpful" data-md-value=1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m7 0c0 .8-.7 1.5-1.5 1.5S14 10.3 14 9.5 14.7 8 15.5 8s1.5.7 1.5 1.5m-5 7.73c-1.75 0-3.29-.73-4.19-1.81L9.23 14c.45.72 1.52 1.23 2.77 1.23s2.32-.51 2.77-1.23l1.42 1.42c-.9 1.08-2.44 1.81-4.19 1.81"/></svg> </button> <button class="md-feedback__icon md-icon" type=submit title="This page could be improved" data-md-value=0> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10m-6.5-4c.8 0 1.5.7 1.5 1.5s-.7 1.5-1.5 1.5-1.5-.7-1.5-1.5.7-1.5 1.5-1.5M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m2 4.5c1.75 0 3.29.72 4.19 1.81l-1.42 1.42C14.32 16.5 13.25 16 12 16s-2.32.5-2.77 1.23l-1.42-1.42C8.71 14.72 10.25 14 12 14"/></svg> </button> </div> <div class=md-feedback__note> <div data-md-value=1 hidden> Thanks for your feedback! </div> <div data-md-value=0 hidden> Thanks for your feedback! Help us improve by opening an issue. </div> </div> </div> </fieldset> </form> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../tutorials/visualization/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Visualization"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Visualization </div> </div> </a> <a href=requirements/ class="md-footer__link md-footer__link--next" aria-label="Next: System Requirements"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> System Requirements </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2022-2025 EVENFLOW Project </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/gprolcastelo/renalprog target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://pypi.org/project/renalprog/ target=_blank rel=noopener title=pypi.org class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.8 200.5c-7.7-30.9-22.3-54.2-53.4-54.2h-40.1v47.4c0 36.8-31.2 67.8-66.8 67.8H172.7c-29.2 0-53.4 25-53.4 54.3v101.8c0 29 25.2 46 53.4 54.3 33.8 9.9 66.3 11.7 106.8 0 26.9-7.8 53.4-23.5 53.4-54.3v-40.7H226.2v-13.6h160.2c31.1 0 42.6-21.7 53.4-54.2 11.2-33.5 10.7-65.7 0-108.6M286.2 444.7a20.4 20.4 0 1 1 0-40.7 20.4 20.4 0 1 1 0 40.7M167.8 248.1h106.8c29.7 0 53.4-24.5 53.4-54.3V91.9c0-29-24.4-50.7-53.4-55.6-35.8-5.9-74.7-5.6-106.8.1-45.2 8-53.4 24.7-53.4 55.6v40.7h106.9v13.6h-147c-31.1 0-58.3 18.7-66.8 54.2-9.8 40.7-10.2 66.1 0 108.6 7.6 31.6 25.7 54.2 56.8 54.2H101v-48.8c0-35.3 30.5-66.4 66.8-66.4m-6.6-183.4a20.4 20.4 0 1 1 0 40.8 20.4 20.4 0 1 1 0-40.8"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "search.suggest", "search.highlight", "search.share", "content.code.copy", "content.code.annotate", "toc.follow"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../js/timeago.min.js></script> <script src=../js/timeago_mkdocs_material.js></script> <script src=../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>