<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="A Python package for simulating kidney cancer progression with synthetic data generation and machine learning."><meta name=author content="Guillermo Prol-Castelo (Barcelona Supercomputing Center), for the EVENFLOW European Project"><link href=https://github.com/gprolcastelo/renalprog/advanced/R_ANALYSIS/ rel=canonical><link href=../GSEA_INSTALLATION/ rel=prev><link href=../performance/ rel=next><link rel=icon href=../../assets/images/kidneys.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>R Analysis Scripts - renalprog</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../css/timeago.css><link rel=stylesheet href=../../assets/_mkdocstrings.css><link rel=stylesheet href=../../stylesheets/extra.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#r-analysis-pipeline class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=renalprog class="md-header__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../../assets/images/kidneys.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> renalprog </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> R Analysis Scripts </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../../tutorials/installation/ class=md-tabs__link> Getting Started </a> </li> <li class=md-tabs__item> <a href=../../tutorials/ class=md-tabs__link> Tutorials </a> </li> <li class=md-tabs__item> <a href=../../api/ class=md-tabs__link> API Reference </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=../ENRICHMENT_ANALYSIS/ class=md-tabs__link> Advanced Topics </a> </li> <li class=md-tabs__item> <a href=../../contributing/guidelines/ class=md-tabs__link> Contributing </a> </li> <li class=md-tabs__item> <a href=../../citation/ class=md-tabs__link> Citation & License </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=renalprog class="md-nav__button md-logo" aria-label=renalprog data-md-component=logo> <img src=../../assets/images/kidneys.png alt=logo> </a> renalprog </label> <div class=md-nav__source> <a href=https://github.com/gprolcastelo/renalprog title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </div> <div class=md-source__repository> gprolcastelo/renalprog </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> <span class=md-ellipsis> Getting Started </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Getting Started </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../tutorials/installation/ class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/data-requirements/ class=md-nav__link> <span class=md-ellipsis> Data Requirements </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> <span class=md-ellipsis> Tutorials </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../tutorials/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/pretrained-models/ class=md-nav__link> <span class=md-ellipsis> Using Pretrained Models </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/complete-pipeline/ class=md-nav__link> <span class=md-ellipsis> Complete Pipeline </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/step1-data-processing/ class=md-nav__link> <span class=md-ellipsis> Step 1 - Data Processing </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/step2-vae-training/ class=md-nav__link> <span class=md-ellipsis> Step 2 - VAE Training </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/step3-reconstruction/ class=md-nav__link> <span class=md-ellipsis> Step 3 - Reconstruction </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/step4-trajectories/ class=md-nav__link> <span class=md-ellipsis> Step 4 - Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/step5-classification/ class=md-nav__link> <span class=md-ellipsis> Step 5 - Classification </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/step6-enrichment/ class=md-nav__link> <span class=md-ellipsis> Step 6 - Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../../tutorials/visualization/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../api/config/ class=md-nav__link> <span class=md-ellipsis> Configuration </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4_3> <label class=md-nav__link for=__nav_4_3 id=__nav_4_3_label tabindex=0> <span class=md-ellipsis> Data Processing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_3_label aria-expanded=false> <label class=md-nav__title for=__nav_4_3> <span class="md-nav__icon md-icon"></span> Data Processing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/dataset/ class=md-nav__link> <span class=md-ellipsis> Dataset </span> </a> </li> <li class=md-nav__item> <a href=../../api/features/ class=md-nav__link> <span class=md-ellipsis> Features </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_4_4> <label class=md-nav__link for=__nav_4_4 id=__nav_4_4_label tabindex=0> <span class=md-ellipsis> Modeling </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4_4> <span class="md-nav__icon md-icon"></span> Modeling </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/models/ class=md-nav__link> <span class=md-ellipsis> VAE Models </span> </a> </li> <li class=md-nav__item> <a href=../../api/training/ class=md-nav__link> <span class=md-ellipsis> Training </span> </a> </li> <li class=md-nav__item> <a href=../../api/prediction/ class=md-nav__link> <span class=md-ellipsis> Prediction </span> </a> </li> <li class=md-nav__item> <a href=../../api/trajectories/ class=md-nav__link> <span class=md-ellipsis> Trajectories </span> </a> </li> <li class=md-nav__item> <a href=../../api/classification/ class=md-nav__link> <span class=md-ellipsis> Classification </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../api/enrichment/ class=md-nav__link> <span class=md-ellipsis> Enrichment </span> </a> </li> <li class=md-nav__item> <a href=../../api/plots/ class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> <li class=md-nav__item> <a href=../../api/utils/ class=md-nav__link> <span class=md-ellipsis> Utilities </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex> <span class=md-ellipsis> Advanced Topics </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=true> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Advanced Topics </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../ENRICHMENT_ANALYSIS/ class=md-nav__link> <span class=md-ellipsis> Enrichment Analysis Details </span> </a> </li> <li class=md-nav__item> <a href=../PATHWAY_HEATMAP/ class=md-nav__link> <span class=md-ellipsis> Pathway Heatmap Analysis </span> </a> </li> <li class=md-nav__item> <a href=../GSEA_INSTALLATION/ class=md-nav__link> <span class=md-ellipsis> GSEA Installation </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> R Analysis Scripts </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> R Analysis Scripts </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> <span class=md-ellipsis> Prerequisites </span> </a> <nav class=md-nav aria-label=Prerequisites> <ul class=md-nav__list> <li class=md-nav__item> <a href=#r-installation class=md-nav__link> <span class=md-ellipsis> R Installation </span> </a> </li> <li class=md-nav__item> <a href=#installation-methods class=md-nav__link> <span class=md-ellipsis> Installation Methods </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#script-1-differential-expression-analysis class=md-nav__link> <span class=md-ellipsis> Script 1: Differential Expression Analysis </span> </a> <nav class=md-nav aria-label="Script 1: Differential Expression Analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#purpose class=md-nav__link> <span class=md-ellipsis> Purpose </span> </a> </li> <li class=md-nav__item> <a href=#quick-start class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class=md-nav__item> <a href=#command-line-arguments class=md-nav__link> <span class=md-ellipsis> Command-Line Arguments </span> </a> </li> <li class=md-nav__item> <a href=#input-data-format class=md-nav__link> <span class=md-ellipsis> Input Data Format </span> </a> </li> <li class=md-nav__item> <a href=#output-files class=md-nav__link> <span class=md-ellipsis> Output Files </span> </a> </li> <li class=md-nav__item> <a href=#output-file-descriptions class=md-nav__link> <span class=md-ellipsis> Output File Descriptions </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#script-2-gene-enrichment-analysis class=md-nav__link> <span class=md-ellipsis> Script 2: Gene Enrichment Analysis </span> </a> <nav class=md-nav aria-label="Script 2: Gene Enrichment Analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#purpose_1 class=md-nav__link> <span class=md-ellipsis> Purpose </span> </a> </li> <li class=md-nav__item> <a href=#quick-start_1 class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class=md-nav__item> <a href=#command-line-arguments_1 class=md-nav__link> <span class=md-ellipsis> Command-Line Arguments </span> </a> </li> <li class=md-nav__item> <a href=#data-sources class=md-nav__link> <span class=md-ellipsis> Data Sources </span> </a> </li> <li class=md-nav__item> <a href=#input-data-format_1 class=md-nav__link> <span class=md-ellipsis> Input Data Format </span> </a> </li> <li class=md-nav__item> <a href=#output-files_1 class=md-nav__link> <span class=md-ellipsis> Output Files </span> </a> </li> <li class=md-nav__item> <a href=#output-file-descriptions_1 class=md-nav__link> <span class=md-ellipsis> Output File Descriptions </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#complete-pipeline-workflow class=md-nav__link> <span class=md-ellipsis> Complete Pipeline Workflow </span> </a> <nav class=md-nav aria-label="Complete Pipeline Workflow"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-by-step-example class=md-nav__link> <span class=md-ellipsis> Step-by-Step Example </span> </a> </li> <li class=md-nav__item> <a href=#integration-with-python-pipeline class=md-nav__link> <span class=md-ellipsis> Integration with Python Pipeline </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#package-installation-issues class=md-nav__link> <span class=md-ellipsis> Package Installation Issues </span> </a> </li> <li class=md-nav__item> <a href=#data-format-issues class=md-nav__link> <span class=md-ellipsis> Data Format Issues </span> </a> </li> <li class=md-nav__item> <a href=#output-issues class=md-nav__link> <span class=md-ellipsis> Output Issues </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#citation class=md-nav__link> <span class=md-ellipsis> Citation </span> </a> </li> <li class=md-nav__item> <a href=#additional-resources class=md-nav__link> <span class=md-ellipsis> Additional Resources </span> </a> <nav class=md-nav aria-label="Additional Resources"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#documentation class=md-nav__link> <span class=md-ellipsis> Documentation </span> </a> </li> <li class=md-nav__item> <a href=#related-tutorials class=md-nav__link> <span class=md-ellipsis> Related Tutorials </span> </a> </li> <li class=md-nav__item> <a href=#external-links class=md-nav__link> <span class=md-ellipsis> External Links </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../performance/ class=md-nav__link> <span class=md-ellipsis> Performance Tuning </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex=0> <span class=md-ellipsis> Contributing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> Contributing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../contributing/guidelines/ class=md-nav__link> <span class=md-ellipsis> Guidelines </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex=0> <span class=md-ellipsis> Citation & License </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Citation & License </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../citation/ class=md-nav__link> <span class=md-ellipsis> How to Cite </span> </a> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> <li class=md-nav__item> <a href=../../acknowledgments/ class=md-nav__link> <span class=md-ellipsis> Acknowledgments </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> <span class=md-ellipsis> Prerequisites </span> </a> <nav class=md-nav aria-label=Prerequisites> <ul class=md-nav__list> <li class=md-nav__item> <a href=#r-installation class=md-nav__link> <span class=md-ellipsis> R Installation </span> </a> </li> <li class=md-nav__item> <a href=#installation-methods class=md-nav__link> <span class=md-ellipsis> Installation Methods </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#script-1-differential-expression-analysis class=md-nav__link> <span class=md-ellipsis> Script 1: Differential Expression Analysis </span> </a> <nav class=md-nav aria-label="Script 1: Differential Expression Analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#purpose class=md-nav__link> <span class=md-ellipsis> Purpose </span> </a> </li> <li class=md-nav__item> <a href=#quick-start class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class=md-nav__item> <a href=#command-line-arguments class=md-nav__link> <span class=md-ellipsis> Command-Line Arguments </span> </a> </li> <li class=md-nav__item> <a href=#input-data-format class=md-nav__link> <span class=md-ellipsis> Input Data Format </span> </a> </li> <li class=md-nav__item> <a href=#output-files class=md-nav__link> <span class=md-ellipsis> Output Files </span> </a> </li> <li class=md-nav__item> <a href=#output-file-descriptions class=md-nav__link> <span class=md-ellipsis> Output File Descriptions </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#script-2-gene-enrichment-analysis class=md-nav__link> <span class=md-ellipsis> Script 2: Gene Enrichment Analysis </span> </a> <nav class=md-nav aria-label="Script 2: Gene Enrichment Analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#purpose_1 class=md-nav__link> <span class=md-ellipsis> Purpose </span> </a> </li> <li class=md-nav__item> <a href=#quick-start_1 class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class=md-nav__item> <a href=#command-line-arguments_1 class=md-nav__link> <span class=md-ellipsis> Command-Line Arguments </span> </a> </li> <li class=md-nav__item> <a href=#data-sources class=md-nav__link> <span class=md-ellipsis> Data Sources </span> </a> </li> <li class=md-nav__item> <a href=#input-data-format_1 class=md-nav__link> <span class=md-ellipsis> Input Data Format </span> </a> </li> <li class=md-nav__item> <a href=#output-files_1 class=md-nav__link> <span class=md-ellipsis> Output Files </span> </a> </li> <li class=md-nav__item> <a href=#output-file-descriptions_1 class=md-nav__link> <span class=md-ellipsis> Output File Descriptions </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#complete-pipeline-workflow class=md-nav__link> <span class=md-ellipsis> Complete Pipeline Workflow </span> </a> <nav class=md-nav aria-label="Complete Pipeline Workflow"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-by-step-example class=md-nav__link> <span class=md-ellipsis> Step-by-Step Example </span> </a> </li> <li class=md-nav__item> <a href=#integration-with-python-pipeline class=md-nav__link> <span class=md-ellipsis> Integration with Python Pipeline </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> <nav class=md-nav aria-label=Troubleshooting> <ul class=md-nav__list> <li class=md-nav__item> <a href=#package-installation-issues class=md-nav__link> <span class=md-ellipsis> Package Installation Issues </span> </a> </li> <li class=md-nav__item> <a href=#data-format-issues class=md-nav__link> <span class=md-ellipsis> Data Format Issues </span> </a> </li> <li class=md-nav__item> <a href=#output-issues class=md-nav__link> <span class=md-ellipsis> Output Issues </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#citation class=md-nav__link> <span class=md-ellipsis> Citation </span> </a> </li> <li class=md-nav__item> <a href=#additional-resources class=md-nav__link> <span class=md-ellipsis> Additional Resources </span> </a> <nav class=md-nav aria-label="Additional Resources"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#documentation class=md-nav__link> <span class=md-ellipsis> Documentation </span> </a> </li> <li class=md-nav__item> <a href=#related-tutorials class=md-nav__link> <span class=md-ellipsis> Related Tutorials </span> </a> </li> <li class=md-nav__item> <a href=#external-links class=md-nav__link> <span class=md-ellipsis> External Links </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=r-analysis-pipeline>R Analysis Pipeline<a class=headerlink href=#r-analysis-pipeline title="Permanent link">&para;</a></h1> <p>This guide provides comprehensive documentation for the R analysis scripts in the renalprog package, covering differential expression analysis and gene enrichment analysis.</p> <h2 id=overview>Overview<a class=headerlink href=#overview title="Permanent link">&para;</a></h2> <p>The R analysis pipeline provides two main capabilities:</p> <ol> <li><strong>Differential Expression Analysis</strong> - Identify genes that are differentially expressed between disease stages</li> <li><strong>Gene Enrichment Analysis</strong> - Identify biological pathways and processes enriched in gene lists</li> </ol> <p>Both scripts are include a command-line interface with argument parsing, comprehensive logging, and error handling.</p> <h2 id=prerequisites>Prerequisites<a class=headerlink href=#prerequisites title="Permanent link">&para;</a></h2> <h3 id=r-installation>R Installation<a class=headerlink href=#r-installation title="Permanent link">&para;</a></h3> <p><strong>Required:</strong> R version 4.0 or higher</p> <p>Check your R version: <div class="language-bash highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>R<span class=w> </span>--version
</span></code></pre></div></p> <h3 id=installation-methods>Installation Methods<a class=headerlink href=#installation-methods title="Permanent link">&para;</a></h3> <p>This section summarizes different methods to install the required R packages. Check the <a href=../../tutorials/installation/ >detailed installation instructions</a>.</p> <h4 id=method-1-condamamba-recommended>Method 1: Conda/Mamba (Recommended)<a class=headerlink href=#method-1-condamamba-recommended title="Permanent link">&para;</a></h4> <p>If using the renalprog conda environment:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1># Activate environment</span>
</span><span id=__span-1-2><a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>mamba<span class=w> </span>activate<span class=w> </span>renalprog
</span><span id=__span-1-3><a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>
</span><span id=__span-1-4><a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=c1># Install R packages via conda</span>
</span><span id=__span-1-5><a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a>mamba<span class=w> </span>install<span class=w> </span>-c<span class=w> </span>conda-forge<span class=w> </span>-c<span class=w> </span>bioconda<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-6><a id=__codelineno-1-6 name=__codelineno-1-6 href=#__codelineno-1-6></a><span class=w>    </span>r-base<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-7><a id=__codelineno-1-7 name=__codelineno-1-7 href=#__codelineno-1-7></a><span class=w>    </span>r-gprofiler2<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-8><a id=__codelineno-1-8 name=__codelineno-1-8 href=#__codelineno-1-8></a><span class=w>    </span>r-ggplot2<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-9><a id=__codelineno-1-9 name=__codelineno-1-9 href=#__codelineno-1-9></a><span class=w>    </span>r-optparse<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-10><a id=__codelineno-1-10 name=__codelineno-1-10 href=#__codelineno-1-10></a><span class=w>    </span>r-dplyr<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-11><a id=__codelineno-1-11 name=__codelineno-1-11 href=#__codelineno-1-11></a><span class=w>    </span>r-effsize<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-12><a id=__codelineno-1-12 name=__codelineno-1-12 href=#__codelineno-1-12></a><span class=w>    </span>bioconductor-limma<span class=w> </span><span class=se>\</span>
</span><span id=__span-1-13><a id=__codelineno-1-13 name=__codelineno-1-13 href=#__codelineno-1-13></a><span class=w>    </span>bioconductor-edger
</span></code></pre></div> <p>Advantages:</p> <ul> <li>All dependencies in one environment</li> <li>No admin privileges needed</li> <li>Reproducible across systems</li> <li>Automatic dependency resolution</li> </ul> <h4 id=method-2-automated-script>Method 2: Automated Script<a class=headerlink href=#method-2-automated-script title="Permanent link">&para;</a></h4> <p>Use the provided installation script:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1># From repository root</span>
</span><span id=__span-2-2><a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a>Rscript<span class=w> </span>scripts/r_analysis/install_r_packages.R
</span></code></pre></div> <p>This installs:</p> <ul> <li><strong>CRAN packages</strong>: gprofiler2, ggplot2, optparse, dplyr, effsize</li> <li><strong>Bioconductor packages</strong>: limma, edgeR</li> </ul> <h4 id=method-3-manual-installation>Method 3: Manual Installation<a class=headerlink href=#method-3-manual-installation title="Permanent link">&para;</a></h4> <p>In an R console:</p> <div class="language-r highlight"><pre><span></span><code><span id=__span-3-1><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1># CRAN packages</span>
</span><span id=__span-3-2><a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=nf>install.packages</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;gprofiler2&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;ggplot2&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;optparse&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;dplyr&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;effsize&quot;</span><span class=p>))</span>
</span><span id=__span-3-3><a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a>
</span><span id=__span-3-4><a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a><span class=c1># Bioconductor packages</span>
</span><span id=__span-3-5><a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a><span class=kr>if</span><span class=w> </span><span class=p>(</span><span class=o>!</span><span class=nf>requireNamespace</span><span class=p>(</span><span class=s>&quot;BiocManager&quot;</span><span class=p>,</span><span class=w> </span><span class=n>quietly</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>))</span>
</span><span id=__span-3-6><a id=__codelineno-3-6 name=__codelineno-3-6 href=#__codelineno-3-6></a><span class=w>    </span><span class=nf>install.packages</span><span class=p>(</span><span class=s>&quot;BiocManager&quot;</span><span class=p>)</span>
</span><span id=__span-3-7><a id=__codelineno-3-7 name=__codelineno-3-7 href=#__codelineno-3-7></a><span class=n>BiocManager</span><span class=o>::</span><span class=nf>install</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;limma&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;edgeR&quot;</span><span class=p>))</span>
</span></code></pre></div> <hr> <h2 id=script-1-differential-expression-analysis>Script 1: Differential Expression Analysis<a class=headerlink href=#script-1-differential-expression-analysis title="Permanent link">&para;</a></h2> <h3 id=purpose>Purpose<a class=headerlink href=#purpose title="Permanent link">&para;</a></h3> <p>Identifies genes that are differentially expressed between two groups (e.g., early vs. late stage cancer) using two complementary statistical methods:</p> <ol> <li><strong>Limma</strong> - Linear Models for Microarray Data (parametric, empirical Bayes)</li> <li><strong>Wilcoxon</strong> - Non-parametric rank-sum test with Cohen's <em>d</em> effect sizes</li> </ol> <h3 id=quick-start>Quick Start<a class=headerlink href=#quick-start title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-4-1><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1># Basic usage with default parameters</span>
</span><span id=__span-4-2><a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>Rscript<span class=w> </span>scripts/r_analysis/differential_expression.R
</span><span id=__span-4-3><a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>
</span><span id=__span-4-4><a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a><span class=c1># Custom analysis</span>
</span><span id=__span-4-5><a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a>Rscript<span class=w> </span>scripts/r_analysis/differential_expression.R<span class=w> </span><span class=se>\</span>
</span><span id=__span-4-6><a id=__codelineno-4-6 name=__codelineno-4-6 href=#__codelineno-4-6></a><span class=w>    </span>--data-path<span class=w> </span>data/interim/preprocessed_KIRC_data/preprocessed_rnaseq.csv<span class=w> </span><span class=se>\</span>
</span><span id=__span-4-7><a id=__codelineno-4-7 name=__codelineno-4-7 href=#__codelineno-4-7></a><span class=w>    </span>--clinical-path<span class=w> </span>data/interim/preprocessed_KIRC_data/clinical_data.csv<span class=w> </span><span class=se>\</span>
</span><span id=__span-4-8><a id=__codelineno-4-8 name=__codelineno-4-8 href=#__codelineno-4-8></a><span class=w>    </span>--output-dir<span class=w> </span>results/differential_expression<span class=w> </span><span class=se>\</span>
</span><span id=__span-4-9><a id=__codelineno-4-9 name=__codelineno-4-9 href=#__codelineno-4-9></a><span class=w>    </span>--alpha<span class=w> </span><span class=m>0</span>.01<span class=w> </span><span class=se>\</span>
</span><span id=__span-4-10><a id=__codelineno-4-10 name=__codelineno-4-10 href=#__codelineno-4-10></a><span class=w>    </span>--description<span class=w> </span>KIRC_analysis
</span><span id=__span-4-11><a id=__codelineno-4-11 name=__codelineno-4-11 href=#__codelineno-4-11></a>
</span><span id=__span-4-12><a id=__codelineno-4-12 name=__codelineno-4-12 href=#__codelineno-4-12></a><span class=c1># View help</span>
</span><span id=__span-4-13><a id=__codelineno-4-13 name=__codelineno-4-13 href=#__codelineno-4-13></a>Rscript<span class=w> </span>scripts/r_analysis/differential_expression.R<span class=w> </span>--help
</span></code></pre></div> <h3 id=command-line-arguments>Command-Line Arguments<a class=headerlink href=#command-line-arguments title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>Argument</th> <th>Short</th> <th>Type</th> <th>Default</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>--data-path</code></td> <td><code>-d</code></td> <td>string</td> <td><code>preprocessed_rnaseq.csv</code></td> <td>RNA-seq expression matrix file</td> </tr> <tr> <td><code>--clinical-path</code></td> <td><code>-c</code></td> <td>string</td> <td><code>clinical_data.csv</code></td> <td>Clinical metadata file</td> </tr> <tr> <td><code>--output-dir</code></td> <td><code>-o</code></td> <td>string</td> <td><code>results/differential_expression</code></td> <td>Base output directory</td> </tr> <tr> <td><code>--alpha</code></td> <td><code>-a</code></td> <td>numeric</td> <td>0.01</td> <td>Significance threshold (FDR/adjusted p-value)</td> </tr> <tr> <td><code>--group1</code></td> <td><code>-g1</code></td> <td>string</td> <td>"early"</td> <td>First comparison group name</td> </tr> <tr> <td><code>--group2</code></td> <td><code>-g2</code></td> <td>string</td> <td>"late"</td> <td>Second comparison group name</td> </tr> <tr> <td><code>--description</code></td> <td><code>-x</code></td> <td>string</td> <td>(timestamp)</td> <td>Output subdirectory suffix</td> </tr> <tr> <td><code>--help</code></td> <td><code>-h</code></td> <td>flag</td> <td>-</td> <td>Display help message</td> </tr> </tbody> </table> <h3 id=input-data-format>Input Data Format<a class=headerlink href=#input-data-format title="Permanent link">&para;</a></h3> <h4 id=rna-seq-expression-matrix>RNA-seq Expression Matrix<a class=headerlink href=#rna-seq-expression-matrix title="Permanent link">&para;</a></h4> <p>Can be in either orientation (auto-detected):</p> <ul> <li><strong>Genes × Samples</strong> (preferred)</li> <li><strong>Samples × Genes</strong> (auto-transposed)</li> </ul> <p><strong>CSV format:</strong> <div class="language-text highlight"><pre><span></span><code><span id=__span-5-1><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a>,Sample1,Sample2,Sample3,...
</span><span id=__span-5-2><a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a>Gene1,5.23,4.89,6.12,...
</span><span id=__span-5-3><a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a>Gene2,2.45,2.78,2.63,...
</span><span id=__span-5-4><a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a>Gene3,8.91,9.23,8.75,...
</span></code></pre></div></p> <p><strong>Requirements:</strong></p> <ul> <li>First column: Gene names/IDs</li> <li>Remaining columns: Sample expression values</li> <li>Values should be normalized (log2-transformed recommended)</li> </ul> <h4 id=clinical-metadata>Clinical Metadata<a class=headerlink href=#clinical-metadata title="Permanent link">&para;</a></h4> <p><strong>CSV format:</strong> <div class="language-text highlight"><pre><span></span><code><span id=__span-6-1><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a>sample,stage,other_columns,...
</span><span id=__span-6-2><a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a>Sample1,early,...
</span><span id=__span-6-3><a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a>Sample2,late,...
</span><span id=__span-6-4><a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a>Sample3,early,...
</span></code></pre></div></p> <p><strong>Requirements:</strong></p> <ul> <li>Must have a <code>stage</code> column containing group labels</li> <li>Accepts: "early", "late", "Early", "Late", "Stage I", "Stage II", etc.</li> <li>Automatic mapping: Stage I/II → early, Stage III/IV → late</li> </ul> <h3 id=output-files>Output Files<a class=headerlink href=#output-files title="Permanent link">&para;</a></h3> <p>The script creates a timestamped directory structure:</p> <div class="language-text highlight"><pre><span></span><code><span id=__span-7-1><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a>results/differential_expression/YYYYMMDD_HHMMSS_description/
</span><span id=__span-7-2><a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a>├── limma/
</span><span id=__span-7-3><a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a>│   ├── sorted_table.csv              # All genes with limma statistics
</span><span id=__span-7-4><a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a>│   ├── significant_genes.csv         # Significant genes (FDR &lt; alpha)
</span><span id=__span-7-5><a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a>│   └── log.txt                       # Limma analysis log
</span><span id=__span-7-6><a id=__codelineno-7-6 name=__codelineno-7-6 href=#__codelineno-7-6></a>├── wilcox/
</span><span id=__span-7-7><a id=__codelineno-7-7 name=__codelineno-7-7 href=#__codelineno-7-7></a>│   ├── wilcox_results.csv            # All genes with Wilcoxon statistics
</span><span id=__span-7-8><a id=__codelineno-7-8 name=__codelineno-7-8 href=#__codelineno-7-8></a>│   ├── significant_genes.csv         # Significant genes (FDR &lt; alpha)
</span><span id=__span-7-9><a id=__codelineno-7-9 name=__codelineno-7-9 href=#__codelineno-7-9></a>│   └── log.txt                       # Wilcoxon analysis log
</span><span id=__span-7-10><a id=__codelineno-7-10 name=__codelineno-7-10 href=#__codelineno-7-10></a>├── combined_results_all_genes.csv            # All genes from both methods
</span><span id=__span-7-11><a id=__codelineno-7-11 name=__codelineno-7-11 href=#__codelineno-7-11></a>├── combined_results_both_significant.csv     # Consensus genes (both methods)
</span><span id=__span-7-12><a id=__codelineno-7-12 name=__codelineno-7-12 href=#__codelineno-7-12></a>├── summary_statistics.csv                    # Summary of results
</span><span id=__span-7-13><a id=__codelineno-7-13 name=__codelineno-7-13 href=#__codelineno-7-13></a>└── differential_expression_analysis.log      # Complete analysis log
</span></code></pre></div> <h3 id=output-file-descriptions>Output File Descriptions<a class=headerlink href=#output-file-descriptions title="Permanent link">&para;</a></h3> <h4 id=limma-results-limmasorted_tablecsv>Limma Results (<code>limma/sorted_table.csv</code>)<a class=headerlink href=#limma-results-limmasorted_tablecsv title="Permanent link">&para;</a></h4> <table> <thead> <tr> <th>Column</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>gene</code></td> <td>Gene identifier</td> </tr> <tr> <td><code>logFC</code></td> <td>Log2 fold change (group2 vs group1)</td> </tr> <tr> <td><code>AveExpr</code></td> <td>Average expression across all samples</td> </tr> <tr> <td><code>t</code></td> <td>t-statistic</td> </tr> <tr> <td><code>P.Value</code></td> <td>Raw p-value</td> </tr> <tr> <td><code>adj.P.Val</code></td> <td>FDR-adjusted p-value (Benjamini-Hochberg)</td> </tr> <tr> <td><code>B</code></td> <td>Log-odds of differential expression</td> </tr> </tbody> </table> <h4 id=wilcoxon-results-wilcoxwilcox_resultscsv>Wilcoxon Results (<code>wilcox/wilcox_results.csv</code>)<a class=headerlink href=#wilcoxon-results-wilcoxwilcox_resultscsv title="Permanent link">&para;</a></h4> <table> <thead> <tr> <th>Column</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>gene</code></td> <td>Gene identifier</td> </tr> <tr> <td><code>p_value</code></td> <td>Raw p-value from Wilcoxon test</td> </tr> <tr> <td><code>fdr</code></td> <td>FDR-adjusted p-value (Benjamini-Hochberg)</td> </tr> <tr> <td><code>cohens_d</code></td> <td>Cohen's d effect size</td> </tr> <tr> <td><code>mean_group1</code></td> <td>Mean expression in group1</td> </tr> <tr> <td><code>mean_group2</code></td> <td>Mean expression in group2</td> </tr> <tr> <td><code>mean_diff</code></td> <td>Difference in means (group2 - group1)</td> </tr> </tbody> </table> <h4 id=combined-results-combined_results_both_significantcsv>Combined Results (<code>combined_results_both_significant.csv</code>)<a class=headerlink href=#combined-results-combined_results_both_significantcsv title="Permanent link">&para;</a></h4> <p>Consensus genes significant in <strong>both</strong> Limma and Wilcoxon tests.</p> <table> <thead> <tr> <th>Column</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>gene</code></td> <td>Gene identifier</td> </tr> <tr> <td><code>limma_logFC</code></td> <td>Log2 fold change from Limma</td> </tr> <tr> <td><code>limma_adj.P.Val</code></td> <td>FDR from Limma</td> </tr> <tr> <td><code>wilcox_fdr</code></td> <td>FDR from Wilcoxon</td> </tr> <tr> <td><code>wilcox_cohens_d</code></td> <td>Effect size from Wilcoxon</td> </tr> <tr> <td><code>avg_rank</code></td> <td>Average rank across both methods</td> </tr> </tbody> </table> <hr> <h2 id=script-2-gene-enrichment-analysis>Script 2: Gene Enrichment Analysis<a class=headerlink href=#script-2-gene-enrichment-analysis title="Permanent link">&para;</a></h2> <h3 id=purpose_1>Purpose<a class=headerlink href=#purpose_1 title="Permanent link">&para;</a></h3> <p>Performs functional enrichment analysis to identify biological pathways, processes, and functions that are over-represented in a gene list using the g:Profiler web service.</p> <h3 id=quick-start_1>Quick Start<a class=headerlink href=#quick-start_1 title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-8-1><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># Basic usage with default gene list</span>
</span><span id=__span-8-2><a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a>Rscript<span class=w> </span>scripts/r_analysis/gene_enrichment.R
</span><span id=__span-8-3><a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a>
</span><span id=__span-8-4><a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=c1># Custom gene list and parameters</span>
</span><span id=__span-8-5><a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a>Rscript<span class=w> </span>scripts/r_analysis/gene_enrichment.R<span class=w> </span><span class=se>\</span>
</span><span id=__span-8-6><a id=__codelineno-8-6 name=__codelineno-8-6 href=#__codelineno-8-6></a><span class=w>    </span>--input<span class=w> </span>data/external/my_genes.csv<span class=w> </span><span class=se>\</span>
</span><span id=__span-8-7><a id=__codelineno-8-7 name=__codelineno-8-7 href=#__codelineno-8-7></a><span class=w>    </span>--output<span class=w> </span>reports/enrichment_analysis<span class=w> </span><span class=se>\</span>
</span><span id=__span-8-8><a id=__codelineno-8-8 name=__codelineno-8-8 href=#__codelineno-8-8></a><span class=w>    </span>--sources<span class=w> </span>GO,KEGG,REAC,WP<span class=w> </span><span class=se>\</span>
</span><span id=__span-8-9><a id=__codelineno-8-9 name=__codelineno-8-9 href=#__codelineno-8-9></a><span class=w>    </span>--top-n<span class=w> </span><span class=m>20</span><span class=w> </span><span class=se>\</span>
</span><span id=__span-8-10><a id=__codelineno-8-10 name=__codelineno-8-10 href=#__codelineno-8-10></a><span class=w>    </span>--alpha<span class=w> </span><span class=m>0</span>.05
</span><span id=__span-8-11><a id=__codelineno-8-11 name=__codelineno-8-11 href=#__codelineno-8-11></a>
</span><span id=__span-8-12><a id=__codelineno-8-12 name=__codelineno-8-12 href=#__codelineno-8-12></a><span class=c1># View help</span>
</span><span id=__span-8-13><a id=__codelineno-8-13 name=__codelineno-8-13 href=#__codelineno-8-13></a>Rscript<span class=w> </span>scripts/r_analysis/gene_enrichment.R<span class=w> </span>--help
</span></code></pre></div> <h3 id=command-line-arguments_1>Command-Line Arguments<a class=headerlink href=#command-line-arguments_1 title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>Argument</th> <th>Short</th> <th>Type</th> <th>Default</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>--input</code></td> <td><code>-i</code></td> <td>string</td> <td><code>genes.csv</code></td> <td>Input gene list (single-column CSV)</td> </tr> <tr> <td><code>--output</code></td> <td><code>-o</code></td> <td>string</td> <td>(auto-generated)</td> <td>Output directory</td> </tr> <tr> <td><code>--sources</code></td> <td><code>-s</code></td> <td>string</td> <td>"GO,REAC,KEGG,WP"</td> <td>Databases to query (comma-separated)</td> </tr> <tr> <td><code>--top-n</code></td> <td><code>-n</code></td> <td>integer</td> <td>20</td> <td>Number of top terms per database</td> </tr> <tr> <td><code>--organism</code></td> <td>-</td> <td>string</td> <td>"hsapiens"</td> <td>Organism identifier</td> </tr> <tr> <td><code>--alpha</code></td> <td><code>-a</code></td> <td>numeric</td> <td>0.05</td> <td>FDR significance threshold</td> </tr> <tr> <td><code>--width</code></td> <td><code>-w</code></td> <td>numeric</td> <td>25</td> <td>Plot width (inches)</td> </tr> <tr> <td><code>--height</code></td> <td>-</td> <td>numeric</td> <td>15</td> <td>Plot height (inches)</td> </tr> <tr> <td><code>--dpi</code></td> <td>-</td> <td>integer</td> <td>600</td> <td>Plot resolution</td> </tr> </tbody> </table> <h3 id=data-sources>Data Sources<a class=headerlink href=#data-sources title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>Source</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>GO</code></td> <td>Gene Ontology (BP, MF, CC)</td> </tr> <tr> <td><code>KEGG</code></td> <td>KEGG Pathways</td> </tr> <tr> <td><code>REAC</code></td> <td>Reactome Pathways</td> </tr> <tr> <td><code>WP</code></td> <td>WikiPathways</td> </tr> </tbody> </table> <h3 id=input-data-format_1>Input Data Format<a class=headerlink href=#input-data-format_1 title="Permanent link">&para;</a></h3> <p><strong>Gene list CSV:</strong> <div class="language-text highlight"><pre><span></span><code><span id=__span-9-1><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a>gene
</span><span id=__span-9-2><a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>TP53
</span><span id=__span-9-3><a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a>VEGFA
</span><span id=__span-9-4><a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a>HIF1A
</span><span id=__span-9-5><a id=__codelineno-9-5 name=__codelineno-9-5 href=#__codelineno-9-5></a>VHL
</span><span id=__span-9-6><a id=__codelineno-9-6 name=__codelineno-9-6 href=#__codelineno-9-6></a>PTEN
</span><span id=__span-9-7><a id=__codelineno-9-7 name=__codelineno-9-7 href=#__codelineno-9-7></a>...
</span></code></pre></div></p> <p><strong>Requirements:</strong> - CSV file with a column named "gene" (or first column if no header) - Gene symbols (e.g., "TP53") or Ensembl IDs - One gene per row - No duplicates</p> <p><strong>Alternative formats accepted:</strong> <div class="language-text highlight"><pre><span></span><code><span id=__span-10-1><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a># Single column, no header
</span><span id=__span-10-2><a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a>TP53
</span><span id=__span-10-3><a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a>VEGFA
</span><span id=__span-10-4><a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a>HIF1A
</span></code></pre></div></p> <h3 id=output-files_1>Output Files<a class=headerlink href=#output-files_1 title="Permanent link">&para;</a></h3> <div class="language-text highlight"><pre><span></span><code><span id=__span-11-1><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a>reports/enrichment_analysis/YYYYMMDD_HHMMSS/
</span><span id=__span-11-2><a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a>├── gostplot.pdf                  # Publication-quality plot (vector)
</span><span id=__span-11-3><a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a>├── gostplot.png                  # High-resolution raster plot
</span><span id=__span-11-4><a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a>├── gostres.csv                   # Full enrichment results (CSV)
</span><span id=__span-11-5><a id=__codelineno-11-5 name=__codelineno-11-5 href=#__codelineno-11-5></a>├── gostres.tsv                   # Full enrichment results (TSV)
</span><span id=__span-11-6><a id=__codelineno-11-6 name=__codelineno-11-6 href=#__codelineno-11-6></a>├── plot_data_gprofiler.csv       # Plot-ready data (CSV)
</span><span id=__span-11-7><a id=__codelineno-11-7 name=__codelineno-11-7 href=#__codelineno-11-7></a>├── plot_data_gprofiler.tsv       # Plot-ready data (TSV)
</span><span id=__span-11-8><a id=__codelineno-11-8 name=__codelineno-11-8 href=#__codelineno-11-8></a>└── gene_enrichment.log           # Analysis log
</span></code></pre></div> <h3 id=output-file-descriptions_1>Output File Descriptions<a class=headerlink href=#output-file-descriptions_1 title="Permanent link">&para;</a></h3> <h4 id=enrichment-results-gostrescsv>Enrichment Results (<code>gostres.csv</code>)<a class=headerlink href=#enrichment-results-gostrescsv title="Permanent link">&para;</a></h4> <table> <thead> <tr> <th>Column</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>term_id</code></td> <td>Database term identifier (e.g., GO:0007049)</td> </tr> <tr> <td><code>term_name</code></td> <td>Human-readable term name</td> </tr> <tr> <td><code>p_value</code></td> <td>Raw p-value</td> </tr> <tr> <td><code>adjusted_p_value</code></td> <td>FDR-adjusted p-value</td> </tr> <tr> <td><code>source</code></td> <td>Database source (GO, KEGG, etc.)</td> </tr> <tr> <td><code>term_size</code></td> <td>Total genes in term</td> </tr> <tr> <td><code>query_size</code></td> <td>Genes queried</td> </tr> <tr> <td><code>intersection_size</code></td> <td>Genes found in term</td> </tr> <tr> <td><code>intersection</code></td> <td>List of genes in term</td> </tr> <tr> <td><code>precision</code></td> <td>Intersection / query size</td> </tr> <tr> <td><code>recall</code></td> <td>Intersection / term size</td> </tr> </tbody> </table> <h4 id=plot-data-plot_data_gprofilercsv>Plot Data (<code>plot_data_gprofiler.csv</code>)<a class=headerlink href=#plot-data-plot_data_gprofilercsv title="Permanent link">&para;</a></h4> <p>Pre-processed data ready for custom plotting: - Filtered to significant terms (FDR &lt; threshold) - Sorted by significance - Top N terms per database</p> <hr> <h2 id=complete-pipeline-workflow>Complete Pipeline Workflow<a class=headerlink href=#complete-pipeline-workflow title="Permanent link">&para;</a></h2> <h3 id=step-by-step-example>Step-by-Step Example<a class=headerlink href=#step-by-step-example title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-12-1><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=c1># Step 1: Install R packages (one time)</span>
</span><span id=__span-12-2><a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a>Rscript<span class=w> </span>scripts/r_analysis/install_r_packages.R
</span><span id=__span-12-3><a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a>
</span><span id=__span-12-4><a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a><span class=c1># Step 2: Run differential expression analysis</span>
</span><span id=__span-12-5><a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a>Rscript<span class=w> </span>scripts/r_analysis/differential_expression.R<span class=w> </span><span class=se>\</span>
</span><span id=__span-12-6><a id=__codelineno-12-6 name=__codelineno-12-6 href=#__codelineno-12-6></a><span class=w>    </span>--data-path<span class=w> </span>data/interim/preprocessed_KIRC_data/preprocessed_rnaseq.csv<span class=w> </span><span class=se>\</span>
</span><span id=__span-12-7><a id=__codelineno-12-7 name=__codelineno-12-7 href=#__codelineno-12-7></a><span class=w>    </span>--clinical-path<span class=w> </span>data/interim/preprocessed_KIRC_data/clinical_data.csv<span class=w> </span><span class=se>\</span>
</span><span id=__span-12-8><a id=__codelineno-12-8 name=__codelineno-12-8 href=#__codelineno-12-8></a><span class=w>    </span>--alpha<span class=w> </span><span class=m>0</span>.01<span class=w> </span><span class=se>\</span>
</span><span id=__span-12-9><a id=__codelineno-12-9 name=__codelineno-12-9 href=#__codelineno-12-9></a><span class=w>    </span>--description<span class=w> </span>KIRC_early_vs_late
</span><span id=__span-12-10><a id=__codelineno-12-10 name=__codelineno-12-10 href=#__codelineno-12-10></a>
</span><span id=__span-12-11><a id=__codelineno-12-11 name=__codelineno-12-11 href=#__codelineno-12-11></a><span class=c1># Step 3: Extract significant genes for enrichment</span>
</span><span id=__span-12-12><a id=__codelineno-12-12 name=__codelineno-12-12 href=#__codelineno-12-12></a><span class=c1># Option A: Use limma results</span>
</span><span id=__span-12-13><a id=__codelineno-12-13 name=__codelineno-12-13 href=#__codelineno-12-13></a>head<span class=w> </span>-n<span class=w> </span><span class=m>101</span><span class=w> </span>results/differential_expression/*/limma/significant_genes.csv<span class=w> </span>&gt;<span class=w> </span>my_genes.csv
</span><span id=__span-12-14><a id=__codelineno-12-14 name=__codelineno-12-14 href=#__codelineno-12-14></a>
</span><span id=__span-12-15><a id=__codelineno-12-15 name=__codelineno-12-15 href=#__codelineno-12-15></a><span class=c1># Option B: Use consensus genes (both methods)</span>
</span><span id=__span-12-16><a id=__codelineno-12-16 name=__codelineno-12-16 href=#__codelineno-12-16></a>head<span class=w> </span>-n<span class=w> </span><span class=m>101</span><span class=w> </span>results/differential_expression/*/combined_results_both_significant.csv<span class=w> </span>&gt;<span class=w> </span>my_genes.csv
</span><span id=__span-12-17><a id=__codelineno-12-17 name=__codelineno-12-17 href=#__codelineno-12-17></a>
</span><span id=__span-12-18><a id=__codelineno-12-18 name=__codelineno-12-18 href=#__codelineno-12-18></a><span class=c1># Step 4: Run enrichment analysis</span>
</span><span id=__span-12-19><a id=__codelineno-12-19 name=__codelineno-12-19 href=#__codelineno-12-19></a>Rscript<span class=w> </span>scripts/r_analysis/gene_enrichment.R<span class=w> </span><span class=se>\</span>
</span><span id=__span-12-20><a id=__codelineno-12-20 name=__codelineno-12-20 href=#__codelineno-12-20></a><span class=w>    </span>--input<span class=w> </span>my_genes.csv<span class=w> </span><span class=se>\</span>
</span><span id=__span-12-21><a id=__codelineno-12-21 name=__codelineno-12-21 href=#__codelineno-12-21></a><span class=w>    </span>--sources<span class=w> </span>GO,KEGG,REAC<span class=w> </span><span class=se>\</span>
</span><span id=__span-12-22><a id=__codelineno-12-22 name=__codelineno-12-22 href=#__codelineno-12-22></a><span class=w>    </span>--top-n<span class=w> </span><span class=m>15</span><span class=w> </span><span class=se>\</span>
</span><span id=__span-12-23><a id=__codelineno-12-23 name=__codelineno-12-23 href=#__codelineno-12-23></a><span class=w>    </span>--alpha<span class=w> </span><span class=m>0</span>.05
</span></code></pre></div> <h3 id=integration-with-python-pipeline>Integration with Python Pipeline<a class=headerlink href=#integration-with-python-pipeline title="Permanent link">&para;</a></h3> <p>The R scripts integrate seamlessly with the Python renalprog pipeline if R is installed in the same conda/mamba environment.</p> <hr> <h2 id=troubleshooting>Troubleshooting<a class=headerlink href=#troubleshooting title="Permanent link">&para;</a></h2> <h3 id=package-installation-issues>Package Installation Issues<a class=headerlink href=#package-installation-issues title="Permanent link">&para;</a></h3> <h4 id=problem-biocmanager-not-found>Problem: BiocManager not found<a class=headerlink href=#problem-biocmanager-not-found title="Permanent link">&para;</a></h4> <div class="language-r highlight"><pre><span></span><code><span id=__span-13-1><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=c1># Solution: Install BiocManager first</span>
</span><span id=__span-13-2><a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=nf>install.packages</span><span class=p>(</span><span class=s>&quot;BiocManager&quot;</span><span class=p>)</span>
</span><span id=__span-13-3><a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a><span class=n>BiocManager</span><span class=o>::</span><span class=nf>install</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;limma&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;edgeR&quot;</span><span class=p>))</span>
</span></code></pre></div> <h3 id=data-format-issues>Data Format Issues<a class=headerlink href=#data-format-issues title="Permanent link">&para;</a></h3> <h4 id=problem-no-stage-column-found>Problem: "No stage column found"<a class=headerlink href=#problem-no-stage-column-found title="Permanent link">&para;</a></h4> <p><strong>Solution:</strong> Ensure clinical data has a <code>stage</code> column: <div class="language-r highlight"><pre><span></span><code><span id=__span-14-1><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=c1># Check column names</span>
</span><span id=__span-14-2><a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a><span class=n>clinical</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read.csv</span><span class=p>(</span><span class=s>&quot;clinical_data.csv&quot;</span><span class=p>)</span>
</span><span id=__span-14-3><a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a><span class=nf>colnames</span><span class=p>(</span><span class=n>clinical</span><span class=p>)</span>
</span></code></pre></div></p> <h3 id=output-issues>Output Issues<a class=headerlink href=#output-issues title="Permanent link">&para;</a></h3> <h4 id=problem-plots-not-generated>Problem: Plots not generated<a class=headerlink href=#problem-plots-not-generated title="Permanent link">&para;</a></h4> <p><strong>Solutions:</strong> <div class="language-r highlight"><pre><span></span><code><span id=__span-15-1><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=c1># Check ggplot2 installation</span>
</span><span id=__span-15-2><a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a><span class=nf>install.packages</span><span class=p>(</span><span class=s>&quot;ggplot2&quot;</span><span class=p>)</span>
</span><span id=__span-15-3><a id=__codelineno-15-3 name=__codelineno-15-3 href=#__codelineno-15-3></a>
</span><span id=__span-15-4><a id=__codelineno-15-4 name=__codelineno-15-4 href=#__codelineno-15-4></a><span class=c1># Reduce plot dimensions</span>
</span><span id=__span-15-5><a id=__codelineno-15-5 name=__codelineno-15-5 href=#__codelineno-15-5></a><span class=n>Rscript</span><span class=w> </span><span class=n>scripts</span><span class=o>/</span><span class=n>r_analysis</span><span class=o>/</span><span class=n>gene_enrichment.R</span><span class=w> </span><span class=o>--</span><span class=n>width</span><span class=w> </span><span class=m>15</span><span class=w> </span><span class=o>--</span><span class=n>height</span><span class=w> </span><span class=m>10</span>
</span><span id=__span-15-6><a id=__codelineno-15-6 name=__codelineno-15-6 href=#__codelineno-15-6></a>
</span><span id=__span-15-7><a id=__codelineno-15-7 name=__codelineno-15-7 href=#__codelineno-15-7></a><span class=c1># Check write permissions</span>
</span><span id=__span-15-8><a id=__codelineno-15-8 name=__codelineno-15-8 href=#__codelineno-15-8></a><span class=nf>dir.create</span><span class=p>(</span><span class=s>&quot;output_directory&quot;</span><span class=p>,</span><span class=w> </span><span class=n>recursive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
</span></code></pre></div></p> <hr> <h2 id=citation>Citation<a class=headerlink href=#citation title="Permanent link">&para;</a></h2> <p>If you use these R scripts in your research, please cite:</p> <p><strong>For differential expression (Limma):</strong></p> <blockquote> <p>Ritchie, M. E., Phipson, B., Wu, D. I., Hu, Y., Law, C. W., Shi, W., &amp; Smyth, G. K. (2015). "limma powers differential expression analyses for RNA-sequencing and microarray studies." Nucleic Acids Research, 43(7), e47.</p> </blockquote> <p><strong>For enrichment analysis (g:Profiler):</strong></p> <blockquote> <p>Kolberg, L., Raudvere, U., Kuzmin, I., Vilo, J., &amp; Peterson, H. (2020). "gprofiler2--an R package for gene list functional enrichment analysis and namespace conversion toolset g:Profiler." F1000Research, 9, ELIXIR-709.</p> </blockquote> <hr> <h2 id=additional-resources>Additional Resources<a class=headerlink href=#additional-resources title="Permanent link">&para;</a></h2> <h3 id=documentation>Documentation<a class=headerlink href=#documentation title="Permanent link">&para;</a></h3> <ul> <li><a href=https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf>Limma User's Guide</a></li> <li><a href=https://biit.cs.ut.ee/gprofiler/ >g:Profiler</a></li> <li><a href=https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf>edgeR User's Guide</a></li> </ul> <h3 id=related-tutorials>Related Tutorials<a class=headerlink href=#related-tutorials title="Permanent link">&para;</a></h3> <ul> <li><a href=../../tutorials/complete-pipeline/ >Full Pipeline</a></li> <li><a href=../../tutorials/step6-enrichment/ >Dynamic Enrichment Analysis Tutorial</a></li> <li><a href=../../tutorials/step5-classification/ >Classification Tutorial</a></li> </ul> <h3 id=external-links>External Links<a class=headerlink href=#external-links title="Permanent link">&para;</a></h3> <ul> <li><a href=https://bioconductor.org/ >Bioconductor</a></li> <li><a href=https://cran.r-project.org/ >CRAN</a></li> </ul> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="January 28, 2026 15:39:21 UTC"><span class=timeago datetime=2026-01-28T15:39:21+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="January 28, 2026 15:39:21 UTC">2026-01-28</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="January 28, 2026 15:30:37 UTC"><span class=timeago datetime=2026-01-28T15:30:37+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="January 28, 2026 15:30:37 UTC">2026-01-28</span> </span> </aside> <form class=md-feedback name=feedback hidden> <fieldset> <legend class=md-feedback__title> Was this page helpful? </legend> <div class=md-feedback__inner> <div class=md-feedback__list> <button class="md-feedback__icon md-icon" type=submit title="This page was helpful" data-md-value=1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m7 0c0 .8-.7 1.5-1.5 1.5S14 10.3 14 9.5 14.7 8 15.5 8s1.5.7 1.5 1.5m-5 7.73c-1.75 0-3.29-.73-4.19-1.81L9.23 14c.45.72 1.52 1.23 2.77 1.23s2.32-.51 2.77-1.23l1.42 1.42c-.9 1.08-2.44 1.81-4.19 1.81"/></svg> </button> <button class="md-feedback__icon md-icon" type=submit title="This page could be improved" data-md-value=0> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 12a8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8 8 8 0 0 0 8-8m2 0a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10m-6.5-4c.8 0 1.5.7 1.5 1.5s-.7 1.5-1.5 1.5-1.5-.7-1.5-1.5.7-1.5 1.5-1.5M10 9.5c0 .8-.7 1.5-1.5 1.5S7 10.3 7 9.5 7.7 8 8.5 8s1.5.7 1.5 1.5m2 4.5c1.75 0 3.29.72 4.19 1.81l-1.42 1.42C14.32 16.5 13.25 16 12 16s-2.32.5-2.77 1.23l-1.42-1.42C8.71 14.72 10.25 14 12 14"/></svg> </button> </div> <div class=md-feedback__note> <div data-md-value=1 hidden> Thanks for your feedback! </div> <div data-md-value=0 hidden> Thanks for your feedback! Help us improve by opening an issue. </div> </div> </div> </fieldset> </form> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../GSEA_INSTALLATION/ class="md-footer__link md-footer__link--prev" aria-label="Previous: GSEA Installation"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> GSEA Installation </div> </div> </a> <a href=../performance/ class="md-footer__link md-footer__link--next" aria-label="Next: Performance Tuning"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Performance Tuning </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2022-2025 EVENFLOW Project </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/gprolcastelo/renalprog target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://pypi.org/project/renalprog/ target=_blank rel=noopener title=pypi.org class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.8 200.5c-7.7-30.9-22.3-54.2-53.4-54.2h-40.1v47.4c0 36.8-31.2 67.8-66.8 67.8H172.7c-29.2 0-53.4 25-53.4 54.3v101.8c0 29 25.2 46 53.4 54.3 33.8 9.9 66.3 11.7 106.8 0 26.9-7.8 53.4-23.5 53.4-54.3v-40.7H226.2v-13.6h160.2c31.1 0 42.6-21.7 53.4-54.2 11.2-33.5 10.7-65.7 0-108.6M286.2 444.7a20.4 20.4 0 1 1 0-40.7 20.4 20.4 0 1 1 0 40.7M167.8 248.1h106.8c29.7 0 53.4-24.5 53.4-54.3V91.9c0-29-24.4-50.7-53.4-55.6-35.8-5.9-74.7-5.6-106.8.1-45.2 8-53.4 24.7-53.4 55.6v40.7h106.9v13.6h-147c-31.1 0-58.3 18.7-66.8 54.2-9.8 40.7-10.2 66.1 0 108.6 7.6 31.6 25.7 54.2 56.8 54.2H101v-48.8c0-35.3 30.5-66.4 66.8-66.4m-6.6-183.4a20.4 20.4 0 1 1 0 40.8 20.4 20.4 0 1 1 0-40.8"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "search.suggest", "search.highlight", "search.share", "content.code.copy", "content.code.annotate", "toc.follow"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../../js/timeago.min.js></script> <script src=../../js/timeago_mkdocs_material.js></script> <script src=../../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>